// arikajs/dist/http/Handler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Handler = void 0;
const http_1 = require("@arikajs/http");
const logging_1 = require("@arikajs/logging");
class Handler {
    constructor() {
        /**
         * A list of the exception types that should not be reported.
         */
        this.dontReport = [];
        /**
         * Custom renderers for specific exception types.
         */
        this.renderers = new Map();
    }
    /**
     * Report or log an exception.
     */
    report(error) {
        if (this.shouldReport(error)) {
            logging_1.Log.error(error.message || 'Error', {
                stack: error.stack,
                name: error.name || 'Error',
                originalError: error.originalError
            });
        }
    }
    /**
     * Render an exception into an HTTP response.
     */
    async render(request, error, response) {
        // 1. Check if the error has a custom renderer
        for (const [type, renderer] of this.renderers.entries()) {
            if (error instanceof type) {
                return renderer(request, error, response);
            }
        }
        // 2. Check if the error is "renderable" (has a render method)
        if (typeof error.render === 'function') {
            return error.render(request, response);
        }
        // 3. Handle HttpException specifically
        if (error instanceof http_1.HttpException) {
            const status = error.getStatusCode();
            const isBrowserRequest = this.isBrowserRequest(request);
            if (isBrowserRequest) {
                return this.renderForBrowser(request, status, error, response);
            }
            return response.status(status).json({
                error: true,
                message: error.message,
                ...(this.shouldDisplayStackTrace() ? { trace: error.stack } : {})
            });
        }
        // 4. Default error handling
        const status = error.statusCode || error.status || 500;
        const message = status === 500 && !this.shouldDisplayStackTrace()
            ? 'Internal Server Error'
            : error.message || 'Unknown Error';
        const isBrowserRequest = this.isBrowserRequest(request);
        if (isBrowserRequest) {
            return this.renderForBrowser(request, status, error, response);
        }
        return response.status(status).json({
            error: true,
            message: message,
            ...(this.shouldDisplayStackTrace() ? {
                name: error.name,
                trace: error.stack
            } : {})
        });
    }
    /**
     * Determine if the exception should be reported.
     */
    shouldReport(error) {
        return !this.dontReport.some(type => error instanceof type);
    }
    /**
     * Determine if the stack trace should be displayed.
     */
    shouldDisplayStackTrace() {
        return process.env.NODE_ENV === 'development' || process.env.APP_DEBUG === 'true';
    }
    /**
     * Determine if the incoming request is a browser (non-API, non-JSON) request.
     */
    isBrowserRequest(request) {
        if (!request || typeof request.path !== 'function')
            return false;
        if (request.path().startsWith('/api'))
            return false;
        const accept = request.header('accept') || '';
        if (accept.includes('application/json') && !accept.includes('text/html'))
            return false;
        return true;
    }
    /**
     * Render an error response for browser (HTML) clients.
     * Override in the application Handler to show custom error views.
     */
    async renderForBrowser(request, status, error, response) {
        const supportedErrors = [401, 403, 404, 419, 429, 500, 503];
        if (supportedErrors.includes(status)) {
            const fs = await Promise.resolve().then(() => __importStar(require('fs')));
            const path = await Promise.resolve().then(() => __importStar(require('path')));
            const appName = process.env.APP_NAME || 'ArikaJS';
            const renderFile = async (filePath) => {
                try {
                    let html = await fs.promises.readFile(filePath, 'utf8');
                    html = html.replace(/\{\{app_name\}\}/g, appName);
                    // Also replace ArkJS template config() calls
                    html = html.replace(/\{\{\s*config\('app\.name'[^)]*\)\s*\}\}/g, appName);
                    return html;
                }
                catch {
                    return null;
                }
            };
            // 1. App override: resources/views/errors/{status}.ark.html
            const root = process.env.PROJECT_ROOT || process.cwd();
            const appView = path.join(root, 'resources', 'views', 'errors', `${status}.ark.html`);
            const appHtml = await renderFile(appView);
            if (appHtml)
                return response.status(status).send(appHtml);
            // 2. Framework bundled views (packages/arikajs/src/http/views/errors/)
            const frameworkView = path.join(__dirname, 'views', 'errors', `${status}.ark.html`);
            const frameworkHtml = await renderFile(frameworkView);
            if (frameworkHtml)
                return response.status(status).send(frameworkHtml);
            // 3. Also try dist path (when running from compiled JS)
            const frameworkViewDist = path.join(__dirname, '..', '..', 'src', 'http', 'views', 'errors', `${status}.ark.html`);
            const frameworkHtmlDist = await renderFile(frameworkViewDist);
            if (frameworkHtmlDist)
                return response.status(status).send(frameworkHtmlDist);
        }
        return response.status(status).send(`<!DOCTYPE html><html><head><title>${status} Error</title></head><body style="font-family:sans-serif;text-align:center;padding:4rem"><h1 style="font-size:4rem;color:#8b5cf6">${status}</h1><p style="color:#64748b">${error.message || 'An error occurred'}</p><a href="/" style="color:#8b5cf6">Return Home</a></body></html>`);
    }
    /**
     * Set the exceptions that should not be reported.
     */
    dontReportExceptions(exceptions) {
        this.dontReport = [...this.dontReport, ...exceptions];
        return this;
    }
    /**
     * Map an exception to a custom renderer.
     */
    map(type, renderer) {
        this.renderers.set(type, renderer);
        return this;
    }
    /**
     * Register a custom renderer for an exception type.
     */
    renderable(type, renderer) {
        return this.map(type, renderer);
    }
    /**
     * Add an exception type to the dontReport list.
     */
    ignore(type) {
        if (!this.dontReport.includes(type)) {
            this.dontReport.push(type);
        }
        return this;
    }
}
exports.Handler = Handler;
