// arikajs/dist/http/Kernel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Kernel = void 0;
const http_1 = require("@arikajs/http");
const middleware_1 = require("@arikajs/middleware");
const RequestLoggingMiddleware_1 = require("./Middleware/RequestLoggingMiddleware");
const http_2 = require("@arikajs/http");
const Handler_1 = require("./Handler");
const ViewMiddleware_1 = require("./Middleware/ViewMiddleware");
const VerifyCsrfToken_1 = require("./Middleware/VerifyCsrfToken");
const ServeStaticMiddleware_1 = require("./Middleware/ServeStaticMiddleware");
const session_1 = require("@arikajs/session");
const auth_1 = require("@arikajs/auth");
class Kernel {
    constructor(app) {
        this.app = app;
        /**
         * The application's global HTTP middleware stack.
         */
        this.middleware = [
            new http_2.CorsMiddleware(),
            new http_2.SecurityHeaders(),
            new RequestLoggingMiddleware_1.RequestLoggingMiddleware(),
            new http_2.BodyParserMiddleware(),
            new http_2.TrimStrings(),
            new http_2.ConvertEmptyStringsToNull(),
            new ServeStaticMiddleware_1.ServeStaticMiddleware(),
        ];
        /**
         * The application's route middleware groups.
         */
        this.middlewareGroups = {
            web: [
                session_1.StartSession,
                ViewMiddleware_1.ViewMiddleware,
                VerifyCsrfToken_1.VerifyCsrfToken,
            ],
            api: [],
        };
        /**
         * The application's route middleware.
         */
        this.routeMiddleware = {
            'auth': auth_1.Authenticate,
            'verified': auth_1.EnsureEmailIsVerified,
            'throttle': http_2.Throttle,
        };
        try {
            this.handler = this.app.make(Handler_1.Handler);
        }
        catch (e) {
            this.handler = new Handler_1.Handler();
        }
        const router = this.app.getRouter();
        if (router.setMiddlewareGroups) {
            router.setMiddlewareGroups(this.middlewareGroups);
        }
        if (router.setRouteMiddleware) {
            router.setRouteMiddleware(this.routeMiddleware);
        }
    }
    /**
     * Handle an incoming HTTP request.
     */
    async handle(request, response) {
        try {
            const pipeline = new middleware_1.Pipeline(this.app.getContainer());
            pipeline.setMiddlewareGroups(this.middlewareGroups);
            pipeline.setAliases(this.routeMiddleware);
            return await pipeline.pipe(this.middleware)
                .handle(request, async (req) => {
                return this.dispatchToRouter(req, response);
            }, response);
        }
        catch (error) {
            this.handler.report(error);
            return await this.handler.render(request, error, response);
        }
    }
    /**
     * Dispatch the request to the router.
     */
    async dispatchToRouter(request, response) {
        const router = this.app.getRouter();
        const result = await router.dispatch(request, response);
        if (result === null) {
            throw new http_1.NotFoundHttpException(`Route not found: [${request.method()}] ${request.path()}`);
        }
        return result;
    }
    /**
     * Send the response back to the client.
     */
    terminate(request, response) {
        response.terminate();
    }
}
exports.Kernel = Kernel;
