// @arikajs/database/dist/Schema/SchemaBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaBuilder = void 0;
const Schema_1 = require("./Schema");
/**
 * Schema builder for creating and modifying tables
 */
class SchemaBuilder {
    constructor(connection) {
        this.connection = connection;
        this.resolvedConnection = null;
    }
    async getResolvedConnection() {
        if (!this.resolvedConnection) {
            this.resolvedConnection = await this.connection;
        }
        return this.resolvedConnection;
    }
    /**
     * Create a new table on the schema
     */
    async create(tableName, callback) {
        const blueprint = Schema_1.Schema.create(tableName, callback);
        const connection = await this.getResolvedConnection();
        const grammar = connection.getSchemaGrammar();
        const sql = grammar.compileCreate(blueprint);
        await connection.query(sql);
    }
    /**
     * Modify an existing table on the schema
     */
    async table(tableName, callback) {
        const blueprint = Schema_1.Schema.table(tableName, callback);
        const connection = await this.getResolvedConnection();
        const grammar = connection.getSchemaGrammar();
        if (typeof grammar.compileAlter !== 'function') {
            throw new Error(`The active schema grammar does not support table alterations.`);
        }
        const statements = grammar.compileAlter(blueprint);
        for (const sql of statements) {
            await connection.query(sql);
        }
    }
    /**
     * Drop a table from the schema
     */
    async drop(tableName) {
        const connection = await this.getResolvedConnection();
        const grammar = connection.getSchemaGrammar();
        const sql = grammar.compileDrop(tableName);
        await connection.query(sql);
    }
    /**
     * Drop a table from the schema if it exists
     */
    async dropIfExists(tableName) {
        const connection = await this.getResolvedConnection();
        const grammar = connection.getSchemaGrammar();
        const sql = grammar.compileDropIfExists(tableName);
        await connection.query(sql);
    }
}
exports.SchemaBuilder = SchemaBuilder;
