// @arikajs/database/dist/Schema/Schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schema = exports.TableBlueprint = exports.ForeignKeyDefinition = exports.ColumnDefinition = void 0;
/**
 * Column definition implementation
 */
class ColumnDefinition {
    constructor(name, type) {
        this.nullable_ = false;
        this.unique_ = false;
        this.primary = false;
        this.autoIncrement = false;
        this.unsigned_ = false;
        this.name = name;
        this.type = type;
    }
    nullable() {
        this.nullable_ = true;
        return this;
    }
    default(value) {
        this.default_ = value;
        return this;
    }
    unique() {
        this.unique_ = true;
        return this;
    }
    unsigned() {
        this.unsigned_ = true;
        return this;
    }
    comment(comment) {
        this.comment_ = comment;
        return this;
    }
}
exports.ColumnDefinition = ColumnDefinition;
/**
 * Foreign key definition implementation
 */
class ForeignKeyDefinition {
    constructor(column) {
        this.column = column;
    }
    references(column) {
        this.referencedColumn = column;
        return this;
    }
    on(table) {
        this.referencedTable = table;
        return this;
    }
    onDelete(action) {
        this.onDeleteAction = action;
        return this;
    }
    onUpdate(action) {
        this.onUpdateAction = action;
        return this;
    }
}
exports.ForeignKeyDefinition = ForeignKeyDefinition;
/**
 * Table blueprint implementation
 */
class TableBlueprint {
    constructor(tableName) {
        this.tableName = tableName;
        this.columns = [];
        this.foreignKeys = [];
        this.indexes = [];
        this.dropColumns = [];
        this.dropIndexes = [];
        this.dropForeignKeys = [];
    }
    id(name = 'id') {
        const column = new ColumnDefinition(name, 'bigInteger');
        column.primary = true;
        column.autoIncrement = true;
        column.unsigned_ = true;
        this.columns.push(column);
        return column;
    }
    string(name, length = 255) {
        const column = new ColumnDefinition(name, 'string');
        column.length = length;
        this.columns.push(column);
        return column;
    }
    integer(name) {
        const column = new ColumnDefinition(name, 'integer');
        this.columns.push(column);
        return column;
    }
    bigInteger(name) {
        const column = new ColumnDefinition(name, 'bigInteger');
        this.columns.push(column);
        return column;
    }
    boolean(name) {
        const column = new ColumnDefinition(name, 'boolean');
        this.columns.push(column);
        return column;
    }
    text(name) {
        const column = new ColumnDefinition(name, 'text');
        this.columns.push(column);
        return column;
    }
    timestamp(name) {
        const column = new ColumnDefinition(name, 'timestamp');
        this.columns.push(column);
        return column;
    }
    timestamps() {
        this.timestamp('created_at').nullable();
        this.timestamp('updated_at').nullable();
    }
    decimal(name, precision = 8, scale = 2) {
        const column = new ColumnDefinition(name, 'decimal');
        column.precision = precision;
        column.scale = scale;
        this.columns.push(column);
        return column;
    }
    date(name) {
        const column = new ColumnDefinition(name, 'date');
        this.columns.push(column);
        return column;
    }
    datetime(name) {
        const column = new ColumnDefinition(name, 'datetime');
        this.columns.push(column);
        return column;
    }
    json(name) {
        const column = new ColumnDefinition(name, 'json');
        this.columns.push(column);
        return column;
    }
    foreign(column) {
        const foreignKey = new ForeignKeyDefinition(column);
        this.foreignKeys.push(foreignKey);
        return foreignKey;
    }
    index(columns, name) {
        const columnArray = Array.isArray(columns) ? columns : [columns];
        this.indexes.push({ columns: columnArray, name, unique: false });
    }
    unique(columns, name) {
        const columnArray = Array.isArray(columns) ? columns : [columns];
        this.indexes.push({ columns: columnArray, name, unique: true });
    }
    dropColumn(name) {
        this.dropColumns.push(name);
    }
    dropIndex(name) {
        this.dropIndexes.push(name);
    }
    dropForeignKey(name) {
        this.dropForeignKeys.push(name);
    }
    /**
     * Get all columns
     */
    getColumns() {
        return this.columns;
    }
    /**
     * Get all foreign keys
     */
    getForeignKeys() {
        return this.foreignKeys;
    }
    /**
     * Get all indexes
     */
    getIndexes() {
        return this.indexes;
    }
    /**
     * Get columns to drop
     */
    getDropColumns() {
        return this.dropColumns;
    }
    /**
     * Get indexes to drop
     */
    getDropIndexes() {
        return this.dropIndexes;
    }
    /**
     * Get foreign keys to drop
     */
    getDropForeignKeys() {
        return this.dropForeignKeys;
    }
}
exports.TableBlueprint = TableBlueprint;
/**
 * Schema builder
 */
class Schema {
    /**
     * Create a new table
     */
    static create(tableName, callback) {
        const blueprint = new TableBlueprint(tableName);
        callback(blueprint);
        return blueprint;
    }
    /**
     * Modify an existing table
     */
    static table(tableName, callback) {
        const blueprint = new TableBlueprint(tableName);
        callback(blueprint);
        return blueprint;
    }
    /**
     * Drop a table
     */
    static drop(tableName) {
        return { tableName, action: 'drop' };
    }
    /**
     * Drop a table if it exists
     */
    static dropIfExists(tableName) {
        return { tableName, action: 'dropIfExists' };
    }
}
exports.Schema = Schema;
