// @arikajs/database/dist/Migrations/Migrator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Migrator = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
/**
 * Migrator class for handling database migrations
 */
class Migrator {
    constructor(db, migrationsPath) {
        this.db = db;
        this.migrationsPath = migrationsPath;
        this.tableName = 'migrations';
    }
    /**
     * Run all pending migrations
     */
    async migrate() {
        await this.ensureMigrationsTable();
        const ran = await this.getRan();
        const files = this.getMigrationFiles();
        const pending = files.filter(file => !ran.includes(file));
        if (pending.length === 0) {
            console.log('\x1b[32mNothing to migrate.\x1b[0m');
            return;
        }
        const batch = await this.getNextBatchNumber();
        for (const file of pending) {
            console.log(`\x1b[33mMigrating: ${file}\x1b[0m`);
            try {
                const migration = await this.resolve(file);
                await migration.up(this.db.schema());
                await this.logMigration(file, batch);
                console.log(`\x1b[32mMigrated:  ${file}\x1b[0m`);
            }
            catch (error) {
                console.error(`\x1b[31mMigration failed: ${file}\x1b[0m`);
                console.error(error.message);
                throw error;
            }
        }
    }
    /**
     * Rollback the last batch of migrations
     */
    async rollback() {
        await this.ensureMigrationsTable();
        const lastBatch = await this.getLastBatch();
        if (lastBatch.length === 0) {
            console.log('\x1b[32mNothing to rollback.\x1b[0m');
            return;
        }
        for (const migrationData of lastBatch) {
            console.log(`\x1b[33mRolling back: ${migrationData.migration}\x1b[0m`);
            try {
                const migration = await this.resolve(migrationData.migration);
                await migration.down(this.db.schema());
                await this.deleteMigration(migrationData.migration);
                console.log(`\x1b[32mRolled back:  ${migrationData.migration}\x1b[0m`);
            }
            catch (error) {
                console.error(`\x1b[31mRollback failed: ${migrationData.migration}\x1b[0m`);
                console.error(error.message);
                throw error;
            }
        }
    }
    /**
     * Ensure the migrations table exists
     */
    async ensureMigrationsTable() {
        const connection = await this.db.connection();
        const grammar = connection.getSchemaGrammar().constructor.name;
        let createSql = '';
        if (grammar === 'MySQLGrammar') {
            createSql = `CREATE TABLE IF NOT EXISTS ${this.tableName} (id INT AUTO_INCREMENT PRIMARY KEY, migration VARCHAR(255), batch INT) ENGINE=InnoDB`;
        }
        else if (grammar === 'PostgreSQLGrammar') {
            createSql = `CREATE TABLE IF NOT EXISTS ${this.tableName} (id SERIAL PRIMARY KEY, migration VARCHAR(255), batch INT)`;
        }
        else {
            createSql = `CREATE TABLE IF NOT EXISTS ${this.tableName} (id INTEGER PRIMARY KEY AUTOINCREMENT, migration TEXT, batch INTEGER)`;
        }
        await connection.query(createSql);
    }
    /**
     * Get all ran migrations
     */
    async getRan() {
        const rows = await this.db.table(this.tableName).select('migration').get();
        return rows.map((row) => row.migration);
    }
    /**
     * Get all migration files in the directory
     */
    getMigrationFiles() {
        if (!fs_1.default.existsSync(this.migrationsPath))
            return [];
        return fs_1.default.readdirSync(this.migrationsPath)
            .filter(file => (file.endsWith('.ts') || file.endsWith('.js')) && !file.endsWith('.d.ts'))
            .sort();
    }
    /**
     * Resolve a migration file to a Migration instance
     */
    async resolve(file) {
        const fullPath = path_1.default.resolve(this.migrationsPath, file);
        // Use dynamic import so tsx can handle .ts migration files
        const module = await Promise.resolve(`${fullPath}`).then(s => __importStar(require(s)));
        // Check for default export or named export
        let MigrationClass = module.default || module;
        // If it's still an object and has multiple exports, look for one that looks like a migration
        if (typeof MigrationClass !== 'function' && typeof MigrationClass === 'object') {
            MigrationClass = Object.values(MigrationClass).find(v => typeof v === 'function');
        }
        if (typeof MigrationClass !== 'function') {
            // Handle case where it's an exported function up/down (Knex style)
            if (module.up && module.down) {
                return {
                    up: () => module.up(this.db.schema()),
                    down: () => module.down(this.db.schema())
                };
            }
            throw new Error(`Migration ${file} does not export a valid Migration class or up/down functions.`);
        }
        return new MigrationClass();
    }
    /**
     * Log a migration as ran
     */
    async logMigration(file, batch) {
        await this.db.table(this.tableName).insert({
            migration: file,
            batch
        });
    }
    /**
     * Delete a migration from the log
     */
    async deleteMigration(file) {
        await this.db.table(this.tableName).where('migration', file).delete();
    }
    /**
     * Get the next batch number
     */
    async getNextBatchNumber() {
        const rows = await this.db.table(this.tableName).select('batch').orderBy('batch', 'desc').limit(1).get();
        return rows.length > 0 ? rows[0].batch + 1 : 1;
    }
    /**
     * Get all migrations from the last batch
     */
    async getLastBatch() {
        const rows = await this.db.table(this.tableName).select('batch').orderBy('batch', 'desc').limit(1).get();
        if (rows.length === 0)
            return [];
        const batch = rows[0].batch;
        return await this.db.table(this.tableName).where('batch', batch).orderBy('migration', 'desc').get();
    }
}
exports.Migrator = Migrator;
