// @arikajs/logging/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DailyDriver = exports.FileDriver = exports.ConsoleDriver = exports.LogManager = exports.Log = void 0;
const LogManager_1 = require("./LogManager");
Object.defineProperty(exports, "LogManager", { enumerable: true, get: function () { return LogManager_1.LogManager; } });
let logManager;
class Log {
    static setManager(manager) {
        logManager = manager;
    }
    static getManager() {
        if (!logManager) {
            throw new Error('Log system not configured. Please use Log.setManager() to configure.');
        }
        return logManager;
    }
    static channel(name) {
        return this.getManager().channel(name);
    }
    static emergency(message, context) { this.getManager().emergency(message, context); }
    static alert(message, context) { this.getManager().alert(message, context); }
    static critical(message, context) { this.getManager().critical(message, context); }
    static error(message, context) { this.getManager().error(message, context); }
    static warning(message, context) { this.getManager().warning(message, context); }
    static notice(message, context) { this.getManager().notice(message, context); }
    static info(message, context) { this.getManager().info(message, context); }
    static debug(message, context) { this.getManager().debug(message, context); }
    static log(level, message, context) { this.getManager().log(level, message, context); }
}
exports.Log = Log;
var ConsoleDriver_1 = require("./Drivers/ConsoleDriver");
Object.defineProperty(exports, "ConsoleDriver", { enumerable: true, get: function () { return ConsoleDriver_1.ConsoleDriver; } });
var FileDriver_1 = require("./Drivers/FileDriver");
Object.defineProperty(exports, "FileDriver", { enumerable: true, get: function () { return FileDriver_1.FileDriver; } });
var DailyDriver_1 = require("./Drivers/DailyDriver");
Object.defineProperty(exports, "DailyDriver", { enumerable: true, get: function () { return DailyDriver_1.DailyDriver; } });
