// @arikajs/logging/dist/LogManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogManager = void 0;
const Logger_1 = require("./Logger");
const ConsoleDriver_1 = require("./Drivers/ConsoleDriver");
const FileDriver_1 = require("./Drivers/FileDriver");
const DailyDriver_1 = require("./Drivers/DailyDriver");
const SlackDriver_1 = require("./Drivers/SlackDriver");
const JsonFormatter_1 = require("./Formatters/JsonFormatter");
const LineFormatter_1 = require("./Formatters/LineFormatter");
class LogManager {
    constructor(config) {
        this.config = config;
        this.channels = new Map();
        this.customCreators = new Map();
    }
    channel(name) {
        const channelName = name || this.config.default;
        if (!this.channels.has(channelName)) {
            this.channels.set(channelName, this.resolve(channelName));
        }
        return this.channels.get(channelName);
    }
    resolve(name) {
        const config = this.config.channels[name];
        if (!config) {
            throw new Error(`Log channel [${name}] is not defined.`);
        }
        const formatter = this.resolveFormatter(config.formatter);
        if (this.customCreators.has(config.driver)) {
            return new Logger_1.Logger(this.customCreators.get(config.driver)({ ...config, formatter, _rawFormatter: config.formatter }));
        }
        const driverMethod = `create${config.driver.charAt(0).toUpperCase() + config.driver.slice(1)}Driver`;
        if (typeof this[driverMethod] === 'function') {
            return new Logger_1.Logger(this[driverMethod]({ ...config, formatter, _rawFormatter: config.formatter }));
        }
        throw new Error(`Driver [${config.driver}] is not supported.`);
    }
    resolveFormatter(formatter) {
        if (typeof formatter === 'string') {
            switch (formatter) {
                case 'json':
                    return new JsonFormatter_1.JsonFormatter();
                case 'line':
                case 'text':
                    return new LineFormatter_1.LineFormatter();
                case 'pretty':
                    // We'll let the driver handle the 'pretty' default if it wants colors
                    // or return a standard LineFormatter here.
                    return new LineFormatter_1.LineFormatter();
                default:
                    throw new Error(`Formatter [${formatter}] not supported.`);
            }
        }
        if (typeof formatter === 'object' && 'format' in formatter) {
            return formatter;
        }
        return new LineFormatter_1.LineFormatter();
    }
    createDailyDriver(config) {
        return new DailyDriver_1.DailyDriver(config);
    }
    createConsoleDriver(config) {
        return new ConsoleDriver_1.ConsoleDriver(config);
    }
    createFileDriver(config) {
        return new FileDriver_1.FileDriver(config);
    }
    createSlackDriver(config) {
        return new SlackDriver_1.SlackDriver(config);
    }
    createStackDriver(config) {
        const channels = config.channels.map((channel) => this.channel(channel));
        return new class {
            log(level, message, context) {
                channels.forEach((channel) => channel.log(level, message, context));
            }
            emergency(message, context) { this.log('emergency', message, context); }
            alert(message, context) { this.log('alert', message, context); }
            critical(message, context) { this.log('critical', message, context); }
            error(message, context) { this.log('error', message, context); }
            warning(message, context) { this.log('warning', message, context); }
            notice(message, context) { this.log('notice', message, context); }
            info(message, context) { this.log('info', message, context); }
            debug(message, context) { this.log('debug', message, context); }
        };
    }
    extend(driver, callback) {
        this.customCreators.set(driver, callback);
        return this;
    }
    // Proxy standard levels to the default channel
    emergency(message, context) { this.channel().emergency(message, context); }
    alert(message, context) { this.channel().alert(message, context); }
    critical(message, context) { this.channel().critical(message, context); }
    error(message, context) { this.channel().error(message, context); }
    warning(message, context) { this.channel().warning(message, context); }
    notice(message, context) { this.channel().notice(message, context); }
    info(message, context) { this.channel().info(message, context); }
    debug(message, context) { this.channel().debug(message, context); }
    log(level, message, context) { this.channel().log(level, message, context); }
}
exports.LogManager = LogManager;
