// @arikajs/authorization/dist/AuthorizationContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthorizationContext = void 0;
const Gate_1 = require("./Gate");
const RolePermission_1 = require("./RolePermission");
/**
 * Per-request authorization context.
 * Avoids the global mutable `Gate.currentUser` problem under concurrency.
 * Each request gets its own `AuthorizationContext` with an isolated user reference.
 */
class AuthorizationContext {
    constructor(user) {
        this.cache = new Map();
        this.user = user;
    }
    // ── Single ability checks ───────────────────────────────────────
    async can(ability, ...args) {
        const cacheKey = this.buildCacheKey(ability, args);
        if (this.cache.has(cacheKey)) {
            return this.cache.get(cacheKey);
        }
        const result = await Gate_1.Gate.forUser(this.user).allows(ability, ...args);
        this.cache.set(cacheKey, result);
        return result;
    }
    async cannot(ability, ...args) {
        return !(await this.can(ability, ...args));
    }
    async authorize(ability, ...args) {
        await Gate_1.Gate.forUser(this.user).authorize(ability, ...args);
    }
    async inspect(ability, ...args) {
        return await Gate_1.Gate.forUser(this.user).inspect(ability, ...args);
    }
    // ── Bulk ability checks ─────────────────────────────────────────
    async any(abilities, ...args) {
        return await Gate_1.Gate.forUser(this.user).any(abilities, ...args);
    }
    async every(abilities, ...args) {
        return await Gate_1.Gate.forUser(this.user).every(abilities, ...args);
    }
    async none(abilities, ...args) {
        return await Gate_1.Gate.forUser(this.user).none(abilities, ...args);
    }
    // ── Role & Permission checks (via mixin) ────────────────────────
    hasRole(role) {
        return RolePermission_1.RolePermissionMixin.hasRole(this.user, role);
    }
    hasAnyRole(roles) {
        return RolePermission_1.RolePermissionMixin.hasAnyRole(this.user, roles);
    }
    hasAllRoles(roles) {
        return RolePermission_1.RolePermissionMixin.hasAllRoles(this.user, roles);
    }
    hasPermission(permission) {
        return RolePermission_1.RolePermissionMixin.hasPermission(this.user, permission);
    }
    hasAnyPermission(permissions) {
        return RolePermission_1.RolePermissionMixin.hasAnyPermission(this.user, permissions);
    }
    hasAllPermissions(permissions) {
        return RolePermission_1.RolePermissionMixin.hasAllPermissions(this.user, permissions);
    }
    // ── Internals ───────────────────────────────────────────────────
    getUser() {
        return this.user;
    }
    /**
     * Clear the authorization cache (e.g. after role change mid-request).
     */
    flushCache() {
        this.cache.clear();
    }
    buildCacheKey(ability, args) {
        // Build a simple cache key from ability + resource IDs
        const argIds = args.map(a => {
            if (a && typeof a === 'object' && a.id !== undefined)
                return `${a.constructor?.name || 'obj'}:${a.id}`;
            if (a && typeof a === 'object' && a.constructor)
                return a.constructor.name;
            return String(a);
        });
        return `${ability}:${argIds.join(',')}`;
    }
}
exports.AuthorizationContext = AuthorizationContext;
