// @arikajs/authorization/dist/AuthResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthResponse = void 0;
/**
 * Represents the result of an authorization check.
 * Allows gates/policies to return custom denial messages instead of just true/false.
 */
class AuthResponse {
    constructor(allowed, message = null, code = null) {
        this._allowed = allowed;
        this._message = message;
        this._code = code;
    }
    allowed() {
        return this._allowed;
    }
    denied() {
        return !this._allowed;
    }
    message() {
        return this._message;
    }
    code() {
        return this._code;
    }
    /**
     * Create an "allow" response.
     */
    static allow(message = null) {
        return new AuthResponse(true, message);
    }
    /**
     * Create a "deny" response with an optional custom message.
     */
    static deny(message = 'This action is unauthorized.', code = null) {
        return new AuthResponse(false, message, code);
    }
}
exports.AuthResponse = AuthResponse;
