// @arikajs/authorization/dist/AuthorizationManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthorizationManager = void 0;
const AuthorizationContext_1 = require("./AuthorizationContext");
class AuthorizationManager {
    constructor(user) {
        this.user = user;
        this.context = new AuthorizationContext_1.AuthorizationContext(user);
    }
    /**
     * Create a request-scoped authorization context and bind it to the request.
     */
    static createContext(request) {
        const user = request.user || (request.auth ? null : null);
        const context = new AuthorizationContext_1.AuthorizationContext(user);
        request.can = context.can.bind(context);
        request.cannot = context.cannot.bind(context);
        request.authorize = context.authorize.bind(context);
        return context;
    }
    // ── Delegated checks ────────────────────────────────────────────
    async can(ability, ...args) {
        return await this.context.can(ability, ...args);
    }
    async cannot(ability, ...args) {
        return await this.context.cannot(ability, ...args);
    }
    async authorize(ability, ...args) {
        return await this.context.authorize(ability, ...args);
    }
    async inspect(ability, ...args) {
        return await this.context.inspect(ability, ...args);
    }
    async any(abilities, ...args) {
        return await this.context.any(abilities, ...args);
    }
    async every(abilities, ...args) {
        return await this.context.every(abilities, ...args);
    }
    async none(abilities, ...args) {
        return await this.context.none(abilities, ...args);
    }
    // ── Role & Permission ───────────────────────────────────────────
    hasRole(role) {
        return this.context.hasRole(role);
    }
    hasAnyRole(roles) {
        return this.context.hasAnyRole(roles);
    }
    hasAllRoles(roles) {
        return this.context.hasAllRoles(roles);
    }
    hasPermission(permission) {
        return this.context.hasPermission(permission);
    }
    hasAnyPermission(permissions) {
        return this.context.hasAnyPermission(permissions);
    }
    hasAllPermissions(permissions) {
        return this.context.hasAllPermissions(permissions);
    }
}
exports.AuthorizationManager = AuthorizationManager;
