// @arikajs/authorization/dist/PolicyResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolicyResolver = void 0;
class PolicyResolver {
    constructor() {
        this.policies = new Map();
        this.instances = new Map();
    }
    /**
     * Register a policy for a given model/class.
     */
    register(model, policy) {
        this.policies.set(model, policy);
    }
    /**
     * Resolve the policy class/constructor for a given resource.
     */
    resolvePolicy(resource) {
        if (!resource)
            return null;
        // Check by constructor
        const constructor = resource.constructor;
        if (constructor && this.policies.has(constructor)) {
            return this.policies.get(constructor);
        }
        // Direct lookup
        if (this.policies.has(resource)) {
            return this.policies.get(resource);
        }
        // Auto-discovery by naming convention
        if (constructor && constructor.name) {
            const policyName = `${constructor.name}Policy`;
            for (const [, policy] of this.policies) {
                const pName = typeof policy === 'function' ? policy.name : policy.constructor?.name;
                if (pName === policyName) {
                    return policy;
                }
            }
        }
        return null;
    }
    /**
     * Get or create a cached instance for a policy class.
     */
    getInstance(policy) {
        if (typeof policy !== 'function')
            return policy;
        if (!this.instances.has(policy)) {
            this.instances.set(policy, new policy());
        }
        return this.instances.get(policy);
    }
    /**
     * Get bound policy method for ability.
     */
    getPolicyMethod(policy, ability) {
        if (!policy)
            return null;
        const instance = this.getInstance(policy);
        if (typeof instance[ability] === 'function') {
            return instance[ability].bind(instance);
        }
        return null;
    }
}
exports.PolicyResolver = PolicyResolver;
