// @arikajs/database/dist/Query/QueryLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryLogger = void 0;
/**
 * Query Logger — records all executed queries for debugging
 */
class QueryLogger {
    /**
     * Enable query logging
     */
    static enable() {
        this.enabled = true;
    }
    /**
     * Disable query logging
     */
    static disable() {
        this.enabled = false;
    }
    /**
     * Check if logging is active
     */
    static isEnabled() {
        return this.enabled;
    }
    /**
     * Log a query
     */
    static log(sql, bindings, time, connection = 'default') {
        if (!this.enabled)
            return;
        const entry = {
            sql,
            bindings,
            time,
            connection,
            timestamp: new Date(),
        };
        this.logs.push(entry);
        // Notify live listeners (e.g. debug console output)
        for (const listener of this.listeners) {
            listener(entry);
        }
    }
    /**
     * Register a real-time listener on each query
     */
    static listen(callback) {
        this.listeners.push(callback);
    }
    /**
     * Get all logged queries
     */
    static getLog() {
        return [...this.logs];
    }
    /**
     * Get the last executed query
     */
    static getLastQuery() {
        return this.logs[this.logs.length - 1];
    }
    /**
     * Clear the query log
     */
    static flush() {
        this.logs = [];
    }
    /**
     * Print all queries to console for debugging
     */
    static dump() {
        for (const entry of this.logs) {
            console.log(`[${entry.connection}] ${entry.sql}`, entry.bindings.length ? `-- bindings: ${JSON.stringify(entry.bindings)}` : '', `(${entry.time.toFixed(2)}ms)`);
        }
    }
}
exports.QueryLogger = QueryLogger;
QueryLogger.logs = [];
QueryLogger.enabled = false;
QueryLogger.listeners = [];
