// @arikajs/cli/dist/src/TemplateManager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemplateManager = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const crypto_1 = require("crypto");
class TemplateManager {
    constructor() {
        // Templates are located relative to the dist/src directory when running
        this.templatesDir = path_1.default.resolve(__dirname, '../templates');
    }
    /**
     * Resolve the template directory path.
     */
    resolveTemplatePath() {
        const templatePath = path_1.default.join(this.templatesDir, 'app');
        if (!fs_1.default.existsSync(templatePath)) {
            throw new Error(`Template "app" not found in ${this.templatesDir}`);
        }
        return templatePath;
    }
    /**
     * Copy template files to the target directory.
     */
    copyTemplate(templatePath, targetDir, appName) {
        this.copyRecursiveSync(templatePath, targetDir);
        // Rename migrations to today's date
        this.renameMigrations(targetDir);
        // Post-process files if needed (e.g., replacing placeholders)
        this.processPlaceholders(targetDir, appName);
    }
    /**
     * Rename migration files in the target directory to use the current date.
     */
    renameMigrations(targetDir) {
        const migrationsDir = path_1.default.join(targetDir, 'database/migrations');
        if (!fs_1.default.existsSync(migrationsDir))
            return;
        const files = fs_1.default.readdirSync(migrationsDir);
        const date = new Date();
        const timestamp = date.getFullYear().toString() + '_' +
            (date.getMonth() + 1).toString().padStart(2, '0') + '_' +
            date.getDate().toString().padStart(2, '0') + '_' +
            date.getHours().toString().padStart(2, '0') +
            date.getMinutes().toString().padStart(2, '0');
        files.forEach((file, index) => {
            // Check if file looks like a migration (starts with a date pattern)
            if (/^\d{4}_\d{2}_\d{2}_\d{6}_/.test(file) || file.startsWith('2024_01_01')) {
                const suffix = file.split('_').slice(4).join('_');
                const newTimestamp = timestamp + (date.getSeconds() + index).toString().padStart(2, '0');
                const newName = `${newTimestamp}_${suffix}`;
                fs_1.default.renameSync(path_1.default.join(migrationsDir, file), path_1.default.join(migrationsDir, newName));
            }
        });
    }
    /**
     * Create .env and .env.example if they don't exist.
     */
    setupEnvironmentFiles(targetDir, appName) {
        const examplePath = path_1.default.join(targetDir, '.env.example');
        const envPath = path_1.default.join(targetDir, '.env');
        if (fs_1.default.existsSync(examplePath) && !fs_1.default.existsSync(envPath)) {
            let content = fs_1.default.readFileSync(examplePath, 'utf8');
            content = content.replace(/APP_NAME=.*/, `APP_NAME=ArikaJs`);
            // Automatically generate app key
            const appKey = 'base64:' + (0, crypto_1.randomBytes)(32).toString('base64');
            if (content.includes('APP_KEY=')) {
                content = content.replace(/APP_KEY=.*/, `APP_KEY=${appKey}`);
            }
            else {
                content += `\nAPP_KEY=${appKey}\n`;
            }
            fs_1.default.writeFileSync(envPath, content);
        }
    }
    copyRecursiveSync(src, dest) {
        const exists = fs_1.default.existsSync(src);
        if (!exists)
            return;
        const stats = fs_1.default.statSync(src);
        const isDirectory = stats.isDirectory();
        if (isDirectory) {
            if (!fs_1.default.existsSync(dest)) {
                fs_1.default.mkdirSync(dest, { recursive: true });
            }
            fs_1.default.readdirSync(src).forEach((childItemName) => {
                this.copyRecursiveSync(path_1.default.join(src, childItemName), path_1.default.join(dest, childItemName));
            });
        }
        else {
            fs_1.default.copyFileSync(src, dest);
        }
    }
    processPlaceholders(targetDir, appName) {
        const packageJsonPath = path_1.default.join(targetDir, 'package.json');
        if (fs_1.default.existsSync(packageJsonPath)) {
            let pkg = JSON.parse(fs_1.default.readFileSync(packageJsonPath, 'utf8'));
            // Replace app name
            pkg.name = appName;
            // Sync framework version with CLI version
            const cliPkgPath = path_1.default.resolve(__dirname, '../../package.json');
            if (fs_1.default.existsSync(cliPkgPath)) {
                const cliPkg = JSON.parse(fs_1.default.readFileSync(cliPkgPath, 'utf8'));
                const targetVersion = `^${cliPkg.version}`;
                if (pkg.dependencies && pkg.dependencies['arikajs']) {
                    pkg.dependencies['arikajs'] = targetVersion;
                }
                if (pkg.devDependencies && pkg.devDependencies['@arikajs/cli']) {
                    pkg.devDependencies['@arikajs/cli'] = targetVersion;
                }
            }
            // Handle local @arikajs dependencies for development
            // templatesDir is /path/to/cli/dist/templates
            // ArikaJs root is /path/to/
            const devWorkspaceRoot = path_1.default.resolve(this.templatesDir, '../../../');
            const handleDeps = (deps) => {
                if (!deps)
                    return;
                Object.keys(deps).forEach(dep => {
                    let depName = '';
                    if (dep.startsWith('@arikajs/')) {
                        depName = dep.split('/')[1];
                        // Map 'core' package to 'arikajs' directory if 'core' directory doesn't exist
                        if (depName === 'core') {
                            const coreDir = path_1.default.join(devWorkspaceRoot, 'core');
                            if (!fs_1.default.existsSync(coreDir)) {
                                depName = 'arikajs';
                            }
                        }
                    }
                    else if (dep === 'arikajs') {
                        depName = 'arikajs';
                    }
                    if (depName && process.env.DEV_MODE === 'true') {
                        let localPath = path_1.default.join(devWorkspaceRoot, depName);
                        if (!fs_1.default.existsSync(localPath)) {
                            localPath = path_1.default.join(devWorkspaceRoot, 'packages', depName);
                        }
                        if (fs_1.default.existsSync(localPath)) {
                            // Calculate relative path from target directory to local package
                            let relativePath = path_1.default.relative(targetDir, localPath);
                            // Ensure it starts with ./ or ../
                            if (!relativePath.startsWith('.')) {
                                relativePath = './' + relativePath;
                            }
                            deps[dep] = `file:${relativePath}`;
                        }
                    }
                });
            };
            handleDeps(pkg.dependencies);
            handleDeps(pkg.devDependencies);
            fs_1.default.writeFileSync(packageJsonPath, JSON.stringify(pkg, null, 2));
        }
        // Replace placeholders in view files
        const viewsDir = path_1.default.join(targetDir, 'resources/views');
        if (fs_1.default.existsSync(viewsDir)) {
            const files = fs_1.default.readdirSync(viewsDir);
            files.forEach(file => {
                if (file.endsWith('.html') || file.endsWith('.ark.html')) {
                    const filePath = path_1.default.join(viewsDir, file);
                    let content = fs_1.default.readFileSync(filePath, 'utf8');
                    content = content.replace(/{{name}}/g, appName);
                    // Replace version
                    const cliPkgPath = path_1.default.resolve(__dirname, '../../package.json');
                    if (fs_1.default.existsSync(cliPkgPath)) {
                        const cliPkg = JSON.parse(fs_1.default.readFileSync(cliPkgPath, 'utf8'));
                        content = content.replace(/{{version}}/g, cliPkg.version);
                    }
                    fs_1.default.writeFileSync(filePath, content);
                }
            });
        }
    }
}
exports.TemplateManager = TemplateManager;
