// @arikajs/cli/dist/src/Commands/AuthInstallCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthInstallCommand = void 0;
const console_1 = require("@arikajs/console");
const inquirer_1 = __importDefault(require("inquirer"));
class AuthInstallCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'auth:install {--force}';
        this.description = 'Interactive menu to scaffold authentication system';
    }
    async handle() {
        this.writeln('');
        this.info(' 🔐 ArikaJS Authentication Scaffolding');
        this.writeln('');
        const { type } = await inquirer_1.default.prompt([
            {
                type: 'list',
                name: 'type',
                message: 'Which type of authentication would you like to scaffold?',
                choices: [
                    { name: 'Web (Session-based, includes views)', value: 'web' },
                    { name: 'API (JWT-based, for mobile/SPAs)', value: 'api' },
                    new inquirer_1.default.Separator(),
                    { name: 'Cancel', value: 'cancel' }
                ]
            }
        ]);
        if (type === 'cancel') {
            this.comment('Operation cancelled.');
            return;
        }
        const force = this.option('force') ? ' --force' : '';
        if (type === 'web') {
            await this.registry.run(['auth:install:web' + force]);
        }
        else {
            await this.registry.run(['auth:install:api' + force]);
        }
    }
}
exports.AuthInstallCommand = AuthInstallCommand;
