// @arikajs/cli/dist/src/Commands/CacheTableCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheTableCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class CacheTableCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'cache:table';
        this.description = 'Create a migration for the cache database table';
    }
    async handle() {
        const migrationsDir = path_1.default.join(process.cwd(), 'database/migrations');
        if (!fs_1.default.existsSync(migrationsDir)) {
            fs_1.default.mkdirSync(migrationsDir, { recursive: true });
        }
        const date = new Date();
        const timestamp = date.getFullYear().toString() + '_' +
            (date.getMonth() + 1).toString().padStart(2, '0') + '_' +
            date.getDate().toString().padStart(2, '0') + '_' +
            date.getHours().toString().padStart(2, '0') +
            date.getMinutes().toString().padStart(2, '0') +
            date.getSeconds().toString().padStart(2, '0');
        const fileName = `${timestamp}_create_cache_table.ts`;
        const filePath = path_1.default.join(migrationsDir, fileName);
        const content = `import { Migration, SchemaBuilder } from 'arikajs';

export default class CreateCacheTable extends Migration {
    /**
     * Run the migrations.
     */
    public async up(schema: SchemaBuilder): Promise<void> {
        await schema.create('cache', (table: any) => {
            table.string('key').unique();
            table.text('value');
            table.integer('expiration').nullable();
        });
    }

    /**
     * Reverse the migrations.
     */
    public async down(schema: SchemaBuilder): Promise<void> {
        await schema.dropIfExists('cache');
    }
}
`;
        fs_1.default.writeFileSync(filePath, content);
        this.info(`Migration created: ${filePath} `);
    }
}
exports.CacheTableCommand = CacheTableCommand;
