// @arikajs/cli/dist/src/Commands/KeyGenerateCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyGenerateCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const crypto_1 = require("crypto");
class KeyGenerateCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'key:generate';
        this.description = 'Set the application key';
    }
    async handle() {
        const envPath = path_1.default.join(process.cwd(), '.env');
        if (!fs_1.default.existsSync(envPath)) {
            this.error('.env file not found.');
            return;
        }
        const key = this.generateRandomKey();
        if (this.setKeyInEnvironmentFile(envPath, key)) {
            this.info('Application key set successfully.');
        }
    }
    generateRandomKey() {
        return 'base64:' + (0, crypto_1.randomBytes)(32).toString('base64');
    }
    setKeyInEnvironmentFile(path, key) {
        let content = fs_1.default.readFileSync(path, 'utf8');
        if (content.includes('APP_KEY=')) {
            content = content.replace(/APP_KEY=.*/, `APP_KEY=${key}`);
        }
        else {
            content += `\nAPP_KEY=${key}\n`;
        }
        fs_1.default.writeFileSync(path, content);
        return true;
    }
}
exports.KeyGenerateCommand = KeyGenerateCommand;
