// @arikajs/cli/dist/src/Commands/MakeCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakeCommand = void 0;
const console_1 = require("@arikajs/console");
const inquirer_1 = __importDefault(require("inquirer"));
class MakeCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'make';
        this.description = 'Interactive menu to generate application boilerplate';
    }
    async handle() {
        this.writeln('');
        this.info(' 🪄 ArikaJS Interactive Make');
        this.writeln('');
        const { type } = await inquirer_1.default.prompt([
            {
                type: 'list',
                name: 'type',
                message: 'What would you like to generate?',
                choices: [
                    'Controller',
                    'Model',
                    'Migration',
                    'Seeder',
                    'Middleware',
                    'Service Provider',
                    'Console Command',
                    'Event',
                    'Listener',
                    'Job',
                    'View',
                    new inquirer_1.default.Separator(),
                    'Cancel'
                ]
            }
        ]);
        if (type === 'Cancel') {
            this.comment('Operation cancelled.');
            return;
        }
        const { name } = await inquirer_1.default.prompt([
            {
                type: 'input',
                name: 'name',
                message: `What should we name the ${type}?`,
                validate: (input) => input.trim() !== '' ? true : 'Please enter a name.'
            }
        ]);
        let commandToRun = '';
        const mockRawArgs = [];
        switch (type) {
            case 'Controller':
                commandToRun = 'make:controller';
                break;
            case 'Migration':
                commandToRun = 'make:migration';
                break;
            case 'Seeder':
                commandToRun = 'make:seeder';
                break;
            case 'Middleware':
                commandToRun = 'make:middleware';
                break;
            case 'Service Provider':
                commandToRun = 'make:provider';
                break;
            case 'Console Command':
                commandToRun = 'make:command';
                break;
            case 'Event':
                commandToRun = 'make:event';
                break;
            case 'Listener':
                commandToRun = 'make:listener';
                break;
            case 'Job':
                commandToRun = 'make:job';
                break;
            case 'View':
                commandToRun = 'make:view';
                break;
            case 'Model': {
                commandToRun = 'make:model';
                const { options } = await inquirer_1.default.prompt([
                    {
                        type: 'checkbox',
                        name: 'options',
                        message: 'Include additional files?',
                        choices: [
                            { name: 'Migration', value: '--migration' },
                            { name: 'Controller', value: '--controller' }
                        ]
                    }
                ]);
                mockRawArgs.push(...options);
                break;
            }
        }
        if (commandToRun) {
            this.writeln('');
            await this.registry.run([commandToRun, name, ...mockRawArgs]);
        }
    }
}
exports.MakeCommand = MakeCommand;
