// @arikajs/cli/dist/src/Commands/MakeListenerCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakeListenerCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class MakeListenerCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'make:listener {name}';
        this.description = 'Create a new event listener class';
    }
    async handle() {
        const name = this.argument('name');
        if (!name) {
            this.error('Listener name is required.');
            return;
        }
        const dir = path_1.default.join(process.cwd(), 'app/Listeners');
        if (!fs_1.default.existsSync(dir)) {
            fs_1.default.mkdirSync(dir, { recursive: true });
        }
        const fileName = `${name}.ts`;
        const filePath = path_1.default.join(dir, fileName);
        if (fs_1.default.existsSync(filePath)) {
            this.error(`Listener already exists: app/Listeners/${fileName}`);
            return;
        }
        fs_1.default.writeFileSync(filePath, this.getStub(name));
        this.success(`Listener created: app/Listeners/${fileName}`);
    }
    getStub(name) {
        return `export class ${name} {
    /**
     * Create the event listener.
     */
    constructor() {
        //
    }

    /**
     * Handle the event.
     */
    public async handle(event: any): Promise<void> {
        // Handle the event
    }
}
`;
    }
}
exports.MakeListenerCommand = MakeListenerCommand;
