// @arikajs/cli/dist/src/Commands/MakeSeederCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakeSeederCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
/**
 * Command to create a new seeder file
 */
class MakeSeederCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'make:seeder {name}';
        this.description = 'Create a new seeder file';
    }
    async handle() {
        const name = this.argument('name');
        if (!name) {
            this.error('Seeder name is required.');
            return;
        }
        const fileName = `${name}.ts`;
        const seedersDir = path_1.default.join(process.cwd(), 'database/seeders');
        if (!fs_1.default.existsSync(seedersDir)) {
            fs_1.default.mkdirSync(seedersDir, { recursive: true });
        }
        const filePath = path_1.default.join(seedersDir, fileName);
        const stub = this.getStub(name);
        fs_1.default.writeFileSync(filePath, stub);
        this.info(`Seeder created: database/seeders/${fileName}`);
    }
    /**
     * Get the seeder stub content
     */
    getStub(name) {
        return `import { Seeder, DatabaseManager } from '@arikajs/database';

export default class ${name} extends Seeder {
    /**
     * Run the database seeds.
     */
    public async run(db: DatabaseManager): Promise<void> {
        // await db.table('users').insert({
        //     name: 'John Doe',
        //     email: 'john@example.com'
        // });
    }
}
`;
    }
}
exports.MakeSeederCommand = MakeSeederCommand;
