// @arikajs/cli/dist/src/Commands/MakeCommandCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakeCommandCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
/**
 * Command to generate a new console command file.
 *
 * Usage:
 *   arika make:command SendReport
 *
 * Generates:
 *   app/Console/Commands/SendReport.ts
 */
class MakeCommandCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'make:command {name}';
        this.description = 'Create a new Artisan command';
    }
    async handle() {
        const name = this.argument('name');
        if (!name) {
            this.error('Command name is required.');
            return;
        }
        const commandsDir = path_1.default.join(process.cwd(), 'app/Console/Commands');
        if (!fs_1.default.existsSync(commandsDir)) {
            fs_1.default.mkdirSync(commandsDir, { recursive: true });
        }
        const fileName = `${name}.ts`;
        const filePath = path_1.default.join(commandsDir, fileName);
        if (fs_1.default.existsSync(filePath)) {
            this.error(`Command already exists: app/Console/Commands/${fileName}`);
            return;
        }
        const stub = this.getStub(name);
        fs_1.default.writeFileSync(filePath, stub);
        this.success(`Command created: app/Console/Commands/${fileName}`);
    }
    /**
     * Get the command stub content.
     */
    getStub(name) {
        // Convert PascalCase to kebab-case for the signature
        // e.g. SendReport → send:report, CleanupOldData → cleanup:old-data
        const signature = name
            .replace(/([a-z])([A-Z])/g, '$1-$2')
            .replace(/([A-Z]+)([A-Z][a-z])/g, '$1-$2')
            .toLowerCase()
            .replace('-', ':');
        return `import { Command } from '@arikajs/console';

export class ${name} extends Command {
    /**
     * The name and signature of the command.
     */
    public signature = '${signature}';

    /**
     * The command description.
     */
    public description = 'Description of the ${signature} command';

    /**
     * Execute the command.
     */
    public async handle() {
        this.info('${name} executed successfully.');

        // Your command logic here...
    }
}
`;
    }
}
exports.MakeCommandCommand = MakeCommandCommand;
