// @arikajs/cli/dist/src/Commands/MakeMiddlewareCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakeMiddlewareCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
/**
 * Command to generate a new middleware.
 *
 * Usage:
 *   arika make:middleware AuthMiddleware
 *
 * Generates:
 *   app/Http/Middleware/AuthMiddleware.ts
 */
class MakeMiddlewareCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'make:middleware {name}';
        this.description = 'Create a new middleware class';
    }
    async handle() {
        const name = this.argument('name');
        if (!name) {
            this.error('Middleware name is required.');
            return;
        }
        const dir = path_1.default.join(process.cwd(), 'app/Http/Middleware');
        if (!fs_1.default.existsSync(dir)) {
            fs_1.default.mkdirSync(dir, { recursive: true });
        }
        const fileName = `${name}.ts`;
        const filePath = path_1.default.join(dir, fileName);
        if (fs_1.default.existsSync(filePath)) {
            this.error(`Middleware already exists: app/Http/Middleware/${fileName}`);
            return;
        }
        fs_1.default.writeFileSync(filePath, this.getStub(name));
        this.success(`Middleware created: app/Http/Middleware/${fileName}`);
    }
    getStub(name) {
        return `import { Request, Response, NextFunction } from '@arikajs/http';

/**
 * ${name}
 *
 * This middleware is executed for every matching route.
 * Call next() to pass control to the next middleware or route handler.
 */
export class ${name} {
    /**
     * Handle an incoming request.
     */
    public async handle(req: Request, res: Response, next: NextFunction): Promise<void> {
        // Perform actions before the request reaches the controller...

        await next();

        // Perform actions after the response is generated...
    }
}
`;
    }
}
exports.MakeMiddlewareCommand = MakeMiddlewareCommand;
