// @arikajs/cli/dist/src/Commands/MakeMigrationCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakeMigrationCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
/**
 * Command to create a new migration file
 */
class MakeMigrationCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'make:migration {name}';
        this.description = 'Create a new migration file';
    }
    async handle() {
        const name = this.argument('name');
        if (!name) {
            this.error('Migration name is required.');
            return;
        }
        const date = new Date();
        const timestamp = date.getFullYear().toString() + '_' +
            (date.getMonth() + 1).toString().padStart(2, '0') + '_' +
            date.getDate().toString().padStart(2, '0') + '_' +
            date.getHours().toString().padStart(2, '0') +
            date.getMinutes().toString().padStart(2, '0') +
            date.getSeconds().toString().padStart(2, '0');
        const fileName = `${timestamp}_${name}.ts`;
        const migrationsDir = path_1.default.join(process.cwd(), 'database/migrations');
        if (!fs_1.default.existsSync(migrationsDir)) {
            fs_1.default.mkdirSync(migrationsDir, { recursive: true });
        }
        const filePath = path_1.default.join(migrationsDir, fileName);
        const stub = this.getStub(name);
        fs_1.default.writeFileSync(filePath, stub);
        this.info(`Migration created: database/migrations/${fileName}`);
    }
    /**
     * Get the migration stub content
     */
    getStub(name) {
        const className = name.split('_').map(s => s.charAt(0).toUpperCase() + s.slice(1)).join('');
        return `import { Migration, SchemaBuilder } from 'arikajs';

export default class ${className} extends Migration {
    /**
     * Run the migrations.
     */
    public async up(schema: SchemaBuilder): Promise<void> {
        // await schema.create('table_name', (table) => {
        //     table.id();
        //     table.string('name');
        //     table.timestamps();
        // });
    }

    /**
     * Reverse the migrations.
     */
    public async down(schema: SchemaBuilder): Promise<void> {
        // await schema.dropIfExists('table_name');
    }
}
`;
    }
}
exports.MakeMigrationCommand = MakeMigrationCommand;
