// @arikajs/cli/dist/src/Commands/MakeControllerCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakeControllerCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
/**
 * Command to generate a new controller.
 *
 * Usage:
 *   arika make:controller UserController
 *
 * Generates:
 *   app/Http/Controllers/UserController.ts
 */
class MakeControllerCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'make:controller {name}';
        this.description = 'Create a new controller class';
    }
    async handle() {
        const name = this.argument('name');
        if (!name) {
            this.error('Controller name is required.');
            return;
        }
        const dir = path_1.default.join(process.cwd(), 'app/Http/Controllers');
        if (!fs_1.default.existsSync(dir)) {
            fs_1.default.mkdirSync(dir, { recursive: true });
        }
        const fileName = `${name}.ts`;
        const filePath = path_1.default.join(dir, fileName);
        if (fs_1.default.existsSync(filePath)) {
            this.error(`Controller already exists: app/Http/Controllers/${fileName}`);
            return;
        }
        fs_1.default.writeFileSync(filePath, this.getStub(name));
        this.success(`Controller created: app/Http/Controllers/${fileName}`);
    }
    getStub(name) {
        return `import { Request, Response } from '@arikajs/http';

export class ${name} {
    /**
     * Display a listing of the resource.
     */
    public async index(req: Request, res: Response) {
        return res.json({ message: 'List of resources' });
    }

    /**
     * Store a newly created resource in storage.
     */
    public async store(req: Request, res: Response) {
        return res.json({ message: 'Resource created' }, 201);
    }

    /**
     * Display the specified resource.
     */
    public async show(req: Request, res: Response) {
        const id = req.param('id');
        return res.json({ message: \`Showing resource \${id}\` });
    }

    /**
     * Update the specified resource in storage.
     */
    public async update(req: Request, res: Response) {
        const id = req.param('id');
        return res.json({ message: \`Resource \${id} updated\` });
    }

    /**
     * Remove the specified resource from storage.
     */
    public async destroy(req: Request, res: Response) {
        const id = req.param('id');
        return res.json({ message: \`Resource \${id} deleted\` });
    }
}
`;
    }
}
exports.MakeControllerCommand = MakeControllerCommand;
