// @arikajs/cli/dist/src/Commands/MakeEventCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakeEventCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class MakeEventCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'make:event {name}';
        this.description = 'Create a new event class';
    }
    async handle() {
        const name = this.argument('name');
        if (!name) {
            this.error('Event name is required.');
            return;
        }
        const dir = path_1.default.join(process.cwd(), 'app/Events');
        if (!fs_1.default.existsSync(dir)) {
            fs_1.default.mkdirSync(dir, { recursive: true });
        }
        const fileName = `${name}.ts`;
        const filePath = path_1.default.join(dir, fileName);
        if (fs_1.default.existsSync(filePath)) {
            this.error(`Event already exists: app/Events/${fileName}`);
            return;
        }
        fs_1.default.writeFileSync(filePath, this.getStub(name));
        this.success(`Event created: app/Events/${fileName}`);
    }
    getStub(name) {
        return `export class ${name} {
    /**
     * Create a new event instance.
     */
    constructor() {
        //
    }
}
`;
    }
}
exports.MakeEventCommand = MakeEventCommand;
