// @arikajs/cli/dist/src/Commands/MakeProviderCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakeProviderCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
/**
 * Command to generate a new service provider.
 *
 * Usage:
 *   arika make:provider PaymentServiceProvider
 *
 * Generates:
 *   app/Providers/PaymentServiceProvider.ts
 */
class MakeProviderCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'make:provider {name}';
        this.description = 'Create a new service provider';
    }
    async handle() {
        const name = this.argument('name');
        if (!name) {
            this.error('Provider name is required.');
            return;
        }
        const dir = path_1.default.join(process.cwd(), 'app/Providers');
        if (!fs_1.default.existsSync(dir)) {
            fs_1.default.mkdirSync(dir, { recursive: true });
        }
        const fileName = `${name}.ts`;
        const filePath = path_1.default.join(dir, fileName);
        if (fs_1.default.existsSync(filePath)) {
            this.error(`Provider already exists: app/Providers/${fileName}`);
            return;
        }
        fs_1.default.writeFileSync(filePath, this.getStub(name));
        this.success(`Provider created: app/Providers/${fileName}`);
    }
    getStub(name) {
        return `import { ServiceProvider } from '@arikajs/foundation';

export class ${name} extends ServiceProvider {
    /**
     * Register any application services.
     *
     * This is called before any boot() methods run.
     * Use this to bind services into the container.
     */
    public async register(): Promise<void> {
        //
    }

    /**
     * Bootstrap any application services.
     *
     * This is called after all providers have been registered.
     * You can safely resolve services from the container here.
     */
    public async boot(): Promise<void> {
        //
    }
}
`;
    }
}
exports.MakeProviderCommand = MakeProviderCommand;
