// @arikajs/cli/dist/src/Commands/MakeJobCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakeJobCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class MakeJobCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'make:job {name}';
        this.description = 'Create a new job class';
    }
    async handle() {
        const name = this.argument('name');
        if (!name) {
            this.error('Job name is required.');
            return;
        }
        const dir = path_1.default.join(process.cwd(), 'app/Jobs');
        if (!fs_1.default.existsSync(dir)) {
            fs_1.default.mkdirSync(dir, { recursive: true });
        }
        const fileName = `${name}.ts`;
        const filePath = path_1.default.join(dir, fileName);
        if (fs_1.default.existsSync(filePath)) {
            this.error(`Job already exists: app/Jobs/${fileName}`);
            return;
        }
        fs_1.default.writeFileSync(filePath, this.getStub(name));
        this.success(`Job created: app/Jobs/${fileName}`);
    }
    getStub(name) {
        return `export class ${name} {
    /**
     * Create a new job instance.
     */
    constructor() {
        //
    }

    /**
     * Execute the job.
     */
    public async handle(): Promise<void> {
        // Job logic goes here
    }
}
`;
    }
}
exports.MakeJobCommand = MakeJobCommand;
