// @arikajs/cli/dist/src/Commands/NewCommand.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const child_process_1 = require("child_process");
class NewCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'new {name}';
        this.description = 'Create a new ArikaJS application';
    }
    async handle() {
        const name = this.argument('name');
        const targetDir = path_1.default.resolve(process.cwd(), name);
        const relativePath = path_1.default.relative(process.cwd(), targetDir) || '.';
        if (fs_1.default.existsSync(targetDir)) {
            this.error(`Directory "${name}" already exists.`);
            return;
        }
        const { TemplateManager } = await Promise.resolve().then(() => __importStar(require('../TemplateManager')));
        const templateManager = new TemplateManager();
        this.writeln('');
        this.info(` 🚀 Creating a new ArikaJS application: ${name}`);
        this.writeln('');
        const shouldInstall = await this.confirm('Would you like to install dependencies automatically?', true);
        try {
            const templatePath = templateManager.resolveTemplatePath();
            await this.task(`Scaffolding project in ${targetDir}`, async () => {
                templateManager.copyTemplate(templatePath, targetDir, name);
                templateManager.setupEnvironmentFiles(targetDir, name);
            });
            if (shouldInstall) {
                this.writeln('');
                this.info(' 📦 Installing dependencies...');
                this.writeln('');
                await this.installDependencies(targetDir);
            }
            this.writeln('');
            this.success('Project created successfully!');
            this.writeln('');
            this.comment(' Next steps:');
            this.comment(`  - cd ${relativePath}`);
            if (!shouldInstall) {
                this.comment('  - npm install');
            }
            this.comment('  - npm run dev');
            this.writeln('');
            this.info(' Happy coding with ArikaJS!');
            this.writeln('');
        }
        catch (error) {
            this.writeln('');
            this.error(`Failed to create project: ${error.message}`);
        }
    }
    async installDependencies(targetDir) {
        return new Promise((resolve, reject) => {
            const pkgPath = path_1.default.join(targetDir, 'package.json');
            let pkgObj = null;
            try {
                if (fs_1.default.existsSync(pkgPath)) {
                    const pkgStr = fs_1.default.readFileSync(pkgPath, 'utf8');
                    pkgObj = JSON.parse(pkgStr);
                    const strippedPkg = JSON.parse(pkgStr);
                    // Temporarily strip file: references so npm doesn't freeze traversing local deps
                    ['dependencies', 'devDependencies'].forEach(section => {
                        if (strippedPkg[section]) {
                            for (const key of Object.keys(strippedPkg[section])) {
                                if (strippedPkg[section][key].startsWith('file:')) {
                                    delete strippedPkg[section][key];
                                }
                            }
                        }
                    });
                    fs_1.default.writeFileSync(pkgPath, JSON.stringify(strippedPkg, null, 2));
                }
            }
            catch (e) {
                // Ignore error
            }
            const child = (0, child_process_1.spawn)('npm', [
                'install',
                '--no-package-lock',
                '--legacy-peer-deps',
                '--ignore-scripts',
                '--prefer-offline',
                '--no-audit',
                '--no-fund'
            ], {
                cwd: targetDir,
                stdio: 'inherit',
                shell: true
            });
            child.on('close', (code) => {
                if (pkgObj) {
                    // Restore original package.json after install completes
                    fs_1.default.writeFileSync(pkgPath, JSON.stringify(pkgObj, null, 2));
                    // Manually symlink local dependencies into node_modules
                    const nodeModulesPath = path_1.default.join(targetDir, 'node_modules');
                    if (!fs_1.default.existsSync(nodeModulesPath)) {
                        fs_1.default.mkdirSync(nodeModulesPath);
                    }
                    ['dependencies', 'devDependencies'].forEach(section => {
                        if (pkgObj[section]) {
                            for (const key of Object.keys(pkgObj[section])) {
                                const val = pkgObj[section][key];
                                if (val.startsWith('file:')) {
                                    const sourcePath = path_1.default.resolve(targetDir, val.replace('file:', ''));
                                    const targetLinkPath = path_1.default.join(nodeModulesPath, key);
                                    // Handle @scoped/packages directories
                                    const scopeMatch = key.match(/^(@[^/]+)\/(.+)$/);
                                    if (scopeMatch) {
                                        const scopeDir = path_1.default.join(nodeModulesPath, scopeMatch[1]);
                                        if (!fs_1.default.existsSync(scopeDir)) {
                                            fs_1.default.mkdirSync(scopeDir, { recursive: true });
                                        }
                                    }
                                    try {
                                        if (fs_1.default.existsSync(targetLinkPath))
                                            fs_1.default.unlinkSync(targetLinkPath);
                                        fs_1.default.symlinkSync(sourcePath, targetLinkPath, 'dir');
                                    }
                                    catch (err) {
                                        // Ignore symlink failure
                                    }
                                }
                            }
                        }
                    });
                }
                if (code === 0) {
                    resolve();
                }
                else {
                    reject(new Error(`npm install failed with exit code ${code}`));
                }
            });
        });
    }
}
exports.NewCommand = NewCommand;
