// @arikajs/cli/dist/src/Commands/QueueRetryCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueRetryCommand = void 0;
const console_1 = require("@arikajs/console");
const arikajs_1 = require("arikajs");
class QueueRetryCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'queue:retry {id}';
        this.description = 'Retry a failed queue job';
    }
    async handle() {
        const id = this.argument('id');
        const db = (0, arikajs_1.app)().make('db');
        const config = (0, arikajs_1.app)().make('config');
        const failedJob = await db.table('failed_jobs').where('id', id).first();
        if (!failedJob) {
            this.error(`Failed job with ID [${id}] not found.`);
            return;
        }
        const jobsTable = config.get('queue.connections.database.table', 'jobs');
        // Re-insert into active jobs table
        await db.table(jobsTable).insert({
            queue: failedJob.queue,
            payload: failedJob.payload,
            attempts: 0,
            available_at: Math.floor(Date.now() / 1000),
            created_at: Math.floor(Date.now() / 1000),
        });
        // Delete from failed jobs
        await db.table('failed_jobs').where('id', id).delete();
        this.success(`The failed job [${id}] has been pushed back onto the queue.`);
    }
}
exports.QueueRetryCommand = QueueRetryCommand;
