// @arikajs/cli/dist/src/Commands/QueueWorkCommand.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueWorkCommand = void 0;
const console_1 = require("@arikajs/console");
const arikajs_1 = require("arikajs");
class QueueWorkCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'queue:work';
        this.description = 'Start processing jobs on the queue as a background worker';
    }
    async handle() {
        this.info('ArikaJS Queue Worker started...');
        this.info('Press Ctrl+C to stop.');
        const queue = (0, arikajs_1.app)().make('queue');
        const config = (0, arikajs_1.app)().make('config');
        const driverName = this.option('connection') || config.get('queue.default');
        const driver = queue.driver(driverName);
        // Auto-register jobs from app/Jobs directory if possible
        await this.registerJobsFromDirectory();
        const worker = new arikajs_1.Worker(driver, {
            queue: this.option('queue') || 'default',
            sleep: parseInt(this.option('sleep') || '3')
        });
        await worker.run();
    }
    async registerJobsFromDirectory() {
        const path = require('path');
        const fs = require('fs');
        const jobsDir = path.join(process.cwd(), 'app', 'Jobs');
        if (fs.existsSync(jobsDir)) {
            const files = fs.readdirSync(jobsDir);
            for (const file of files) {
                if (file.endsWith('.ts') || file.endsWith('.js')) {
                    const jobName = path.parse(file).name;
                    try {
                        const jobModule = await Promise.resolve(`${path.join(jobsDir, file)}`).then(s => __importStar(require(s)));
                        const JobClass = jobModule[jobName] || jobModule.default;
                        if (JobClass) {
                            arikajs_1.Worker.registerJob(jobName, JobClass);
                        }
                    }
                    catch (e) {
                        // Skip if cannot import
                    }
                }
            }
        }
        // Also register VerifyEmail if it is used (though technically it is a Mailable, 
        // in ArikaJS mail is often sent via queue using the Mailable itself as a job)
        // For standard jobs, this directory sweep is enough.
    }
}
exports.QueueWorkCommand = QueueWorkCommand;
