// @arikajs/cli/dist/src/Commands/QueueFailedTableCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueFailedTableCommand = void 0;
const console_1 = require("@arikajs/console");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class QueueFailedTableCommand extends console_1.Command {
    constructor() {
        super(...arguments);
        this.signature = 'queue:failed-table';
        this.description = 'Create a migration for the failed queue jobs database table';
    }
    async handle() {
        const migrationsDir = path_1.default.join(process.cwd(), 'database/migrations');
        if (!fs_1.default.existsSync(migrationsDir)) {
            fs_1.default.mkdirSync(migrationsDir, { recursive: true });
        }
        const date = new Date();
        const timestamp = date.getFullYear().toString() + '_' +
            (date.getMonth() + 1).toString().padStart(2, '0') + '_' +
            date.getDate().toString().padStart(2, '0') + '_' +
            date.getHours().toString().padStart(2, '0') +
            date.getMinutes().toString().padStart(2, '0') +
            date.getSeconds().toString().padStart(2, '0');
        const fileName = `${timestamp}_create_failed_jobs_table.ts`;
        const filePath = path_1.default.join(migrationsDir, fileName);
        const content = `import { Migration, SchemaBuilder } from 'arikajs';

export default class CreateFailedJobsTable extends Migration {
    public async up(schema: SchemaBuilder) {
        await schema.create('failed_jobs', (table: any) => {
            table.id();
            table.string('uuid').unique();
            table.string('connection');
            table.string('queue');
            table.text('payload');
            table.text('exception');
            table.timestamp('failed_at').useCurrent();
        });
    }

    public async down(schema: SchemaBuilder) {
        await schema.dropIfExists('failed_jobs');
    }
}
`;
        fs_1.default.writeFileSync(filePath, content);
        this.info(`Migration created: ${filePath}`);
    }
}
exports.QueueFailedTableCommand = QueueFailedTableCommand;
