// @arikajs/storage/dist/tests/EnterpriseFeatures.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = require("node:test");
const assert = __importStar(require("node:assert"));
const fs = __importStar(require("fs/promises"));
const path = __importStar(require("path"));
const src_1 = require("../src");
const stream_1 = require("stream");
class DummyMemoryDriver {
    constructor() {
        this.files = new Map();
    }
    async put(path, contents) {
        this.files.set(path, contents);
    }
    async putStream(path, stream) { }
    async get(path) {
        if (!this.files.has(path))
            throw new Error('Not found');
        const content = this.files.get(path);
        return Buffer.isBuffer(content) ? content : Buffer.from(content);
    }
    readStream(path) { return new stream_1.Readable(); }
    async exists(path) {
        return this.files.has(path);
    }
    async delete(path) {
        this.files.delete(path);
    }
    url(path) { return `memory://${path}`; }
    async temporaryUrl(path, expiresAt) { return ''; }
    async size(path) { return 0; }
    async lastModified(path) { return 0; }
    async mimeType(path) { return 'text/plain'; }
}
(0, node_test_1.describe)('Enterprise Storage Features', () => {
    const testRoot = path.join(process.cwd(), 'test-storage-enterprise');
    let storage;
    (0, node_test_1.beforeEach)(async () => {
        await fs.mkdir(testRoot, { recursive: true });
        const config = {
            default: 'local',
            disks: {
                local: {
                    driver: 'local',
                    root: testRoot
                },
                memory: {
                    driver: 'memory'
                }
            }
        };
        storage = new src_1.StorageManager(config);
    });
    (0, node_test_1.afterEach)(async () => {
        try {
            await fs.rm(testRoot, { recursive: true, force: true });
        }
        catch (error) { }
    });
    (0, node_test_1.it)('can extend storage manager with custom driver', async () => {
        const dummyDriver = new DummyMemoryDriver();
        storage.extend('memory', (config) => dummyDriver);
        await storage.disk('memory').put('test.txt', 'Memory Content');
        assert.strictEqual(dummyDriver.files.has('test.txt'), true);
        const retrieved = await storage.disk('memory').get('test.txt');
        assert.strictEqual(retrieved.toString(), 'Memory Content');
    });
    (0, node_test_1.it)('can use middleware to intercept pipeline operations', async () => {
        let loggedPath = '';
        const loggingMiddleware = async (data, next) => {
            if (data.operation === 'put') {
                loggedPath = data.path;
                data.contents = data.contents + ' - Intercepted';
            }
            return await next(data);
        };
        storage.disk('local').middleware(loggingMiddleware);
        await storage.put('middleware.txt', 'Original');
        assert.strictEqual(loggedPath, 'middleware.txt');
        const contents = await storage.get('middleware.txt');
        assert.strictEqual(contents.toString(), 'Original - Intercepted');
    });
    (0, node_test_1.it)('can copy files with fallback', async () => {
        await storage.put('original.txt', 'Copy Me');
        await storage.copy('original.txt', 'copied.txt');
        assert.strictEqual(await storage.exists('copied.txt'), true);
        assert.strictEqual((await storage.get('copied.txt')).toString(), 'Copy Me');
    });
    (0, node_test_1.it)('can move files with fallback', async () => {
        await storage.put('source.txt', 'Move Me');
        await storage.move('source.txt', 'destination.txt');
        assert.strictEqual(await storage.exists('source.txt'), false);
        assert.strictEqual(await storage.exists('destination.txt'), true);
        assert.strictEqual((await storage.get('destination.txt')).toString(), 'Move Me');
    });
    (0, node_test_1.it)('can append to files', async () => {
        await storage.put('append.txt', 'Line 1');
        await storage.append('append.txt', 'Line 2');
        const contents = await storage.get('append.txt');
        assert.strictEqual(contents.toString(), 'Line 1\nLine 2');
    });
    (0, node_test_1.it)('can prepend to files', async () => {
        await storage.put('prepend.txt', 'Line 2');
        await storage.prepend('prepend.txt', 'Line 1');
        const contents = await storage.get('prepend.txt');
        assert.strictEqual(contents.toString(), 'Line 1\nLine 2');
    });
});
