// @arikajs/storage/dist/tests/Storage.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = require("node:test");
const assert = __importStar(require("node:assert"));
const fs = __importStar(require("fs/promises"));
const path = __importStar(require("path"));
const src_1 = require("../src");
(0, node_test_1.describe)('Arika Storage', () => {
    const testRoot = path.join(process.cwd(), 'test-storage');
    let storage;
    (0, node_test_1.beforeEach)(async () => {
        // Create test storage directory
        await fs.mkdir(testRoot, { recursive: true });
        // Configure storage
        const config = {
            default: 'local',
            disks: {
                local: {
                    driver: 'local',
                    root: testRoot
                },
                public: {
                    driver: 'local',
                    root: path.join(testRoot, 'public'),
                    url: '/storage'
                }
            }
        };
        storage = new src_1.StorageManager(config);
    });
    (0, node_test_1.afterEach)(async () => {
        // Clean up test storage directory
        try {
            await fs.rm(testRoot, { recursive: true, force: true });
        }
        catch (error) {
            // Ignore cleanup errors
        }
    });
    (0, node_test_1.it)('puts and gets file content', async () => {
        await storage.put('test.txt', 'Hello Arika');
        const content = await storage.get('test.txt');
        assert.strictEqual(content.toString(), 'Hello Arika');
    });
    (0, node_test_1.it)('checks if file exists', async () => {
        await storage.put('exists.txt', 'content');
        const exists = await storage.exists('exists.txt');
        assert.strictEqual(exists, true);
        const notExists = await storage.exists('not-exists.txt');
        assert.strictEqual(notExists, false);
    });
    (0, node_test_1.it)('deletes files', async () => {
        await storage.put('delete-me.txt', 'content');
        assert.strictEqual(await storage.exists('delete-me.txt'), true);
        await storage.delete('delete-me.txt');
        assert.strictEqual(await storage.exists('delete-me.txt'), false);
    });
    (0, node_test_1.it)('throws FileNotFoundException for missing files', async () => {
        await assert.rejects(async () => await storage.get('missing.txt'), src_1.FileNotFoundException);
    });
    (0, node_test_1.it)('works with different disks', async () => {
        await storage.disk('local').put('local.txt', 'local content');
        await storage.disk('public').put('public.txt', 'public content');
        const localContent = await storage.disk('local').get('local.txt');
        const publicContent = await storage.disk('public').get('public.txt');
        assert.strictEqual(localContent.toString(), 'local content');
        assert.strictEqual(publicContent.toString(), 'public content');
    });
    (0, node_test_1.it)('generates URLs for files', async () => {
        await storage.disk('public').put('image.png', 'fake image');
        const url = storage.disk('public').url('image.png');
        assert.strictEqual(url, '/storage/image.png');
    });
    (0, node_test_1.it)('handles nested directories', async () => {
        await storage.put('nested/deep/file.txt', 'nested content');
        const exists = await storage.exists('nested/deep/file.txt');
        assert.strictEqual(exists, true);
        const content = await storage.get('nested/deep/file.txt');
        assert.strictEqual(content.toString(), 'nested content');
    });
    (0, node_test_1.it)('handles Buffer content', async () => {
        const buffer = Buffer.from('binary content');
        await storage.put('binary.dat', buffer);
        const retrieved = await storage.get('binary.dat');
        assert.deepStrictEqual(retrieved, buffer);
    });
});
