// @arikajs/queue/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisDriver = exports.BaseJob = exports.Worker = exports.QueueManager = exports.Queue = exports.queueManager = void 0;
const QueueManager_1 = require("./QueueManager");
Object.defineProperty(exports, "QueueManager", { enumerable: true, get: function () { return QueueManager_1.QueueManager; } });
const Worker_1 = require("./Worker");
Object.defineProperty(exports, "Worker", { enumerable: true, get: function () { return Worker_1.Worker; } });
class Queue {
    static setManager(manager) {
        exports.queueManager = manager;
    }
    static connection(name) {
        return {
            push: (job) => exports.queueManager.push(job, {}, name),
            later: (delay, job) => exports.queueManager.later(delay, job, undefined, name),
        };
    }
    static async dispatch(job) {
        if (!exports.queueManager) {
            throw new Error('Queue not configured. Please use Queue.setManager().');
        }
        return exports.queueManager.dispatch(job);
    }
    static async push(job) {
        return exports.queueManager.push(job);
    }
    static async later(delay, job) {
        return exports.queueManager.later(delay, job);
    }
    static async bulk(jobs) {
        return exports.queueManager.bulk(jobs);
    }
}
exports.Queue = Queue;
var Job_1 = require("./Job");
Object.defineProperty(exports, "BaseJob", { enumerable: true, get: function () { return Job_1.BaseJob; } });
var RedisDriver_1 = require("./Drivers/RedisDriver");
Object.defineProperty(exports, "RedisDriver", { enumerable: true, get: function () { return RedisDriver_1.RedisDriver; } });
