// @arikajs/queue/dist/QueueManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueManager = void 0;
const SyncDriver_1 = require("./Drivers/SyncDriver");
const DatabaseDriver_1 = require("./Drivers/DatabaseDriver");
const RedisDriver_1 = require("./Drivers/RedisDriver");
class QueueManager {
    constructor(config, database) {
        this.drivers = new Map();
        this.config = config;
        this.database = database;
    }
    driver(name) {
        const driverName = name || this.config.default;
        if (!this.drivers.has(driverName)) {
            this.drivers.set(driverName, this.resolve(driverName));
        }
        return this.drivers.get(driverName);
    }
    resolve(name) {
        const config = this.config.connections[name];
        if (!config) {
            throw new Error(`Queue connection [${name}] not configured.`);
        }
        switch (config.driver) {
            case 'sync':
                return new SyncDriver_1.SyncDriver();
            case 'database':
                return new DatabaseDriver_1.DatabaseDriver(this.database, config);
            case 'redis':
                return new RedisDriver_1.RedisDriver(config);
            default:
                throw new Error(`Unsupported queue driver [${config.driver}].`);
        }
    }
    async push(job, options, connection) {
        return this.driver(connection || job.connection).push(job, options);
    }
    async later(delay, job, queue, connection) {
        return this.push(job, { delay, queue }, connection);
    }
    async dispatch(job, connection) {
        return this.push(job, {
            queue: job.queue,
            delay: job.delay
        }, connection || job.connection);
    }
    async bulk(jobs, connection) {
        return Promise.all(jobs.map(job => this.dispatch(job, connection)));
    }
}
exports.QueueManager = QueueManager;
