// @arikajs/logging/dist/Drivers/ConsoleDriver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleDriver = void 0;
const LineFormatter_1 = require("../Formatters/LineFormatter");
class ConsoleDriver {
    constructor(config = {}) {
        this.config = config;
        this.colors = {
            emergency: '\x1b[41m\x1b[37m', // White on Red
            alert: '\x1b[41m\x1b[37m', // White on Red
            critical: '\x1b[41m\x1b[37m', // White on Red
            error: '\x1b[31m', // Red
            warning: '\x1b[33m', // Yellow
            notice: '\x1b[34m', // Blue
            info: '\x1b[32m', // Green
            debug: '\x1b[90m', // Gray
        };
        this.levels = {
            emergency: 0,
            alert: 1,
            critical: 2,
            error: 3,
            warning: 4,
            notice: 5,
            info: 6,
            debug: 7,
        };
        // If config.formatter was 'pretty', LogManager might have returned a default LineFormatter.
        // We want to ensure ConsoleDriver uses colors for 'pretty' or default.
        if (!config.formatter || (typeof config._rawFormatter === 'string' && config._rawFormatter === 'pretty')) {
            this.formatter = new LineFormatter_1.LineFormatter(this.colors);
        }
        else {
            this.formatter = config.formatter;
        }
    }
    log(level, message, context) {
        const configLevel = this.config.level || 'debug';
        if (this.levels[level] > this.levels[configLevel]) {
            return;
        }
        const output = this.formatter.format(level, message, context);
        switch (level) {
            case 'emergency':
            case 'alert':
            case 'critical':
            case 'error':
                console.error(output);
                break;
            case 'warning':
                console.warn(output);
                break;
            case 'info':
            case 'notice':
                console.info(output);
                break;
            case 'debug':
                console.debug(output);
                break;
            default:
                console.log(output);
        }
    }
    emergency(message, context) { this.log('emergency', message, context); }
    alert(message, context) { this.log('alert', message, context); }
    critical(message, context) { this.log('critical', message, context); }
    error(message, context) { this.log('error', message, context); }
    warning(message, context) { this.log('warning', message, context); }
    notice(message, context) { this.log('notice', message, context); }
    info(message, context) { this.log('info', message, context); }
    debug(message, context) { this.log('debug', message, context); }
}
exports.ConsoleDriver = ConsoleDriver;
