// @arikajs/logging/dist/Drivers/FileDriver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileDriver = void 0;
const LineFormatter_1 = require("../Formatters/LineFormatter");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class FileDriver {
    constructor(config) {
        this.config = config;
        this.levels = {
            emergency: 0,
            alert: 1,
            critical: 2,
            error: 3,
            warning: 4,
            notice: 5,
            info: 6,
            debug: 7,
        };
        this.filePath = config.path || 'app.log';
        this.ensureFileExists();
        this.formatter = config.formatter || new LineFormatter_1.LineFormatter();
    }
    ensureFileExists() {
        const dir = path_1.default.dirname(this.filePath);
        if (!fs_1.default.existsSync(dir)) {
            fs_1.default.mkdirSync(dir, { recursive: true });
        }
    }
    log(level, message, context) {
        const configLevel = this.config.level || 'debug';
        if (this.levels[level] > this.levels[configLevel]) {
            return;
        }
        const line = this.formatter.format(level, message, context) + '\n';
        fs_1.default.appendFileSync(this.filePath, line);
    }
    emergency(message, context) { this.log('emergency', message, context); }
    alert(message, context) { this.log('alert', message, context); }
    critical(message, context) { this.log('critical', message, context); }
    error(message, context) { this.log('error', message, context); }
    warning(message, context) { this.log('warning', message, context); }
    notice(message, context) { this.log('notice', message, context); }
    info(message, context) { this.log('info', message, context); }
    debug(message, context) { this.log('debug', message, context); }
}
exports.FileDriver = FileDriver;
