// @arikajs/logging/dist/Drivers/SlackDriver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlackDriver = void 0;
const https_1 = __importDefault(require("https"));
class SlackDriver {
    constructor(config) {
        this.config = config;
        this.levels = {
            emergency: 0,
            alert: 1,
            critical: 2,
            error: 3,
            warning: 4,
            notice: 5,
            info: 6,
            debug: 7,
        };
        if (!config.url) {
            throw new Error('Slack webhook URL is required for Slack driver.');
        }
    }
    log(level, message, context) {
        const configLevel = this.config.level || 'critical';
        if (this.levels[level] > this.levels[configLevel]) {
            return;
        }
        const payload = {
            text: `*[${level.toUpperCase()}]* ${message}`,
            attachments: context ? [
                {
                    color: this.getAttachmentColor(level),
                    fields: Object.keys(context).map(key => ({
                        title: key,
                        value: typeof context[key] === 'object' ? JSON.stringify(context[key]) : String(context[key]),
                        short: true
                    }))
                }
            ] : []
        };
        const data = JSON.stringify(payload);
        const req = https_1.default.request(this.config.url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': data.length,
            },
        });
        req.on('error', (error) => {
            console.error('Failed to send log to Slack:', error);
        });
        req.write(data);
        req.end();
    }
    getAttachmentColor(level) {
        switch (level) {
            case 'emergency':
            case 'alert':
            case 'critical':
                return '#FF0000';
            case 'error':
                return '#E01E5A';
            case 'warning':
                return '#ECB22E';
            case 'info':
                return '#36C5F0';
            case 'debug':
                return '#909090';
            default:
                return '#D0D0D0';
        }
    }
    emergency(message, context) { this.log('emergency', message, context); }
    alert(message, context) { this.log('alert', message, context); }
    critical(message, context) { this.log('critical', message, context); }
    error(message, context) { this.log('error', message, context); }
    warning(message, context) { this.log('warning', message, context); }
    notice(message, context) { this.log('notice', message, context); }
    info(message, context) { this.log('info', message, context); }
    debug(message, context) { this.log('debug', message, context); }
}
exports.SlackDriver = SlackDriver;
