// @arikajs/cache/dist/Drivers/DatabaseDriver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatabaseDriver = void 0;
class DatabaseDriver {
    constructor(database, table, prefix = '', connection) {
        this.database = database;
        this.table = table;
        this.prefix = prefix;
        this.connection = connection;
    }
    async get(key) {
        const cache = await this.database.table(this.table, this.connection)
            .where('key', this.prefix + key)
            .first();
        if (!cache) {
            return null;
        }
        if (cache.expiration && cache.expiration <= Math.floor(Date.now() / 1000)) {
            await this.forget(key);
            return null;
        }
        return JSON.parse(cache.value);
    }
    async put(key, value, seconds) {
        const expiration = Math.floor(Date.now() / 1000) + seconds;
        const jsonValue = JSON.stringify(value);
        const prefixedKey = this.prefix + key;
        const exists = await this.database.table(this.table, this.connection).where('key', prefixedKey).exists();
        if (exists) {
            await this.database.table(this.table, this.connection)
                .where('key', prefixedKey)
                .update({
                value: jsonValue,
                expiration: expiration
            });
        }
        else {
            await this.database.table(this.table, this.connection).insert({
                key: prefixedKey,
                value: jsonValue,
                expiration: expiration
            });
        }
    }
    async increment(key, value = 1) {
        const current = await this.get(key) || 0;
        const newValue = current + value;
        await this.forever(key, newValue);
        return newValue;
    }
    async decrement(key, value = 1) {
        return this.increment(key, -value);
    }
    async forever(key, value) {
        const jsonValue = JSON.stringify(value);
        const prefixedKey = this.prefix + key;
        const exists = await this.database.table(this.table, this.connection).where('key', prefixedKey).exists();
        if (exists) {
            await this.database.table(this.table, this.connection)
                .where('key', prefixedKey)
                .update({
                value: jsonValue,
                expiration: null
            });
        }
        else {
            await this.database.table(this.table, this.connection).insert({
                key: prefixedKey,
                value: jsonValue,
                expiration: null
            });
        }
    }
    async forget(key) {
        await this.database.table(this.table, this.connection)
            .where('key', this.prefix + key)
            .delete();
    }
    async flush() {
        await this.database.table(this.table, this.connection).delete();
    }
    async add(key, value, seconds) {
        const prefixedKey = this.prefix + key;
        const jsonValue = JSON.stringify(value);
        const expiration = Math.floor(Date.now() / 1000) + seconds;
        // Try updating an already expired key first
        const updated = await this.database.table(this.table, this.connection)
            .where('key', prefixedKey)
            .where('expiration', '<=', Math.floor(Date.now() / 1000))
            .update({
            value: jsonValue,
            expiration: expiration
        });
        if (updated > 0)
            return true;
        try {
            await this.database.table(this.table, this.connection).insert({
                key: prefixedKey,
                value: jsonValue,
                expiration: expiration
            });
            return true;
        }
        catch (e) {
            return false;
        }
    }
    async getMultiple(keys) {
        const results = {};
        for (const k of keys)
            results[k] = null;
        if (keys.length === 0)
            return results;
        const prefixedKeys = keys.map(k => this.prefix + k);
        const rows = await this.database.table(this.table, this.connection).whereIn('key', prefixedKeys);
        const now = Math.floor(Date.now() / 1000);
        const expiredKeys = [];
        for (const row of rows) {
            const originalKey = row.key.substring(this.prefix.length);
            if (row.expiration && row.expiration <= now) {
                expiredKeys.push(originalKey);
            }
            else {
                results[originalKey] = JSON.parse(row.value);
            }
        }
        if (expiredKeys.length > 0) {
            await this.forgetMultiple(expiredKeys);
        }
        return results;
    }
    async putMultiple(values, seconds) {
        await Promise.all(Object.entries(values).map(([key, value]) => this.put(key, value, seconds)));
    }
    async forgetMultiple(keys) {
        if (keys.length === 0)
            return;
        const prefixedKeys = keys.map(k => this.prefix + k);
        await this.database.table(this.table, this.connection).whereIn('key', prefixedKeys).delete();
    }
    getPrefix() {
        return this.prefix;
    }
}
exports.DatabaseDriver = DatabaseDriver;
