// @arikajs/database/dist/Database.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Database = void 0;
const Expression_1 = require("./Query/Expression");
const QueryLogger_1 = require("./Query/QueryLogger");
/**
 * Database facade - static interface to the database manager
 */
class Database {
    /**
     * Set the database manager instance
     */
    static setManager(manager) {
        Database.manager = manager;
    }
    /**
     * Get the database manager instance
     */
    static getManager() {
        if (!Database.manager) {
            throw new Error('DatabaseManager not initialized. Call Database.setManager() first.');
        }
        return Database.manager;
    }
    /**
     * Create a new query builder for a table
     */
    static table(tableName, connectionName) {
        return Database.getManager().table(tableName, connectionName);
    }
    /**
     * Get a schema builder instance
     */
    static schema(connectionName) {
        return Database.getManager().schema(connectionName);
    }
    /**
     * Execute a callback within a transaction (nested savepoints supported)
     */
    static async transaction(callback, connectionName) {
        return await Database.getManager().transaction(callback, connectionName);
    }
    /**
     * Get a TransactionManager for manual begin/commit/rollback control
     */
    static async getTransactionManager(connectionName) {
        return Database.getManager().getTransactionManager(connectionName);
    }
    /**
     * Enable query logging
     */
    static enableQueryLog() {
        QueryLogger_1.QueryLogger.enable();
    }
    /**
     * Disable query logging
     */
    static disableQueryLog() {
        QueryLogger_1.QueryLogger.disable();
    }
    /**
     * Get all logged queries
     */
    static getQueryLog() {
        return QueryLogger_1.QueryLogger.getLog();
    }
    /**
     * Flush (clear) the query log
     */
    static flushQueryLog() {
        QueryLogger_1.QueryLogger.flush();
    }
    /**
     * Get a connection by name
     */
    static connection(name) {
        return Database.getManager().connection(name);
    }
    /**
     * Create a raw SQL expression
     */
    static raw(value) {
        return new Expression_1.Expression(value);
    }
    /**
     * Close all connections
     */
    static async closeAll() {
        return Database.getManager().closeAll();
    }
}
exports.Database = Database;
