// @arikajs/database/dist/DatabaseManager.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatabaseManager = void 0;
const QueryBuilder_1 = require("./Query/QueryBuilder");
const SchemaBuilder_1 = require("./Schema/SchemaBuilder");
const TransactionManager_1 = require("./Transactions/TransactionManager");
/**
 * Database manager - handles connections and query builders
 */
class DatabaseManager {
    constructor(config) {
        this.connections = new Map();
        this.config = config;
    }
    /**
     * Get a connection by name
     */
    connection(name) {
        const connectionName = name || this.config.default;
        if (this.connections.has(connectionName)) {
            return this.connections.get(connectionName);
        }
        const connectionConfig = this.config.connections[connectionName];
        if (!connectionConfig) {
            throw new Error(`Connection "${connectionName}" not configured`);
        }
        const connectionPromise = this.createConnection(connectionConfig).then(connection => {
            this.connections.set(connectionName, connection);
            return connection;
        });
        return connectionPromise;
    }
    /**
     * Create a new query builder
     */
    table(tableName, connectionName) {
        const connection = this.connection(connectionName);
        return new QueryBuilder_1.QueryBuilder(connection).table(tableName);
    }
    /**
     * Get a schema builder instance
     */
    schema(connectionName) {
        const connection = this.connection(connectionName);
        return new SchemaBuilder_1.SchemaBuilder(connection);
    }
    /**
     * Execute a callback within a transaction (nested savepoints supported)
     */
    async transaction(callback, connectionName) {
        const connection = await this.connection(connectionName);
        const txManager = new TransactionManager_1.TransactionManager(connection);
        return txManager.transaction(callback);
    }
    /**
     * Get a TransactionManager for manual begin/commit/rollback control
     */
    async getTransactionManager(connectionName) {
        const connection = await this.connection(connectionName);
        return new TransactionManager_1.TransactionManager(connection);
    }
    /**
     * Set a query cache implementation
     */
    setCache(cache) {
        this.cacheStore = cache;
    }
    /**
     * Get the configured query cache store
     */
    getCache() {
        return this.cacheStore;
    }
    /**
     * Close all connections
     */
    async closeAll() {
        const promises = Array.from(this.connections.values()).map(conn => conn.close());
        await Promise.all(promises);
        this.connections.clear();
    }
    /**
     * Create a connection based on driver type
     */
    async createConnection(config) {
        switch (config.driver) {
            case 'mysql': {
                const { MySQLConnection } = await Promise.resolve().then(() => __importStar(require('./Connections/MySQLConnection')));
                return new MySQLConnection(config);
            }
            case 'pgsql': {
                const { PostgreSQLConnection } = await Promise.resolve().then(() => __importStar(require('./Connections/PostgreSQLConnection')));
                return new PostgreSQLConnection(config);
            }
            case 'sqlite': {
                const { SQLiteConnection } = await Promise.resolve().then(() => __importStar(require('./Connections/SQLiteConnection')));
                return new SQLiteConnection(config);
            }
            case 'mongodb': {
                const { MongoDBConnection } = await Promise.resolve().then(() => __importStar(require('./Connections/MongoDBConnection')));
                return new MongoDBConnection(config);
            }
            default:
                throw new Error(`Unsupported database driver: ${config.driver}`);
        }
    }
}
exports.DatabaseManager = DatabaseManager;
