// @arikajs/database/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SeedRunner = exports.Seeder = exports.Migrator = exports.Migration = exports.SchemaBuilder = exports.ForeignKeyDefinition = exports.ColumnDefinition = exports.TableBlueprint = exports.Schema = exports.QueryLogger = exports.TransactionManager = exports.CallbackGlobalScope = exports.GlobalScopeRegistry = exports.ObserverRegistry = exports.withSoftDeletes = exports.Relation = exports.MorphTo = exports.MorphMany = exports.MorphOne = exports.HasManyThrough = exports.HasOneThrough = exports.BelongsToMany = exports.BelongsTo = exports.HasMany = exports.HasOne = exports.ModelQueryBuilder = exports.Model = exports.Expression = exports.QueryBuilder = exports.MongoDBConnection = exports.SQLiteConnection = exports.PostgreSQLConnection = exports.MySQLConnection = exports.DB = exports.DatabaseManager = exports.Database = void 0;
// Core exports
var Database_1 = require("./Database");
Object.defineProperty(exports, "Database", { enumerable: true, get: function () { return Database_1.Database; } });
var DatabaseManager_1 = require("./DatabaseManager");
Object.defineProperty(exports, "DatabaseManager", { enumerable: true, get: function () { return DatabaseManager_1.DatabaseManager; } });
// DB Facade alias (for convenience)
var Database_2 = require("./Database");
Object.defineProperty(exports, "DB", { enumerable: true, get: function () { return Database_2.Database; } });
// Connections
var MySQLConnection_1 = require("./Connections/MySQLConnection");
Object.defineProperty(exports, "MySQLConnection", { enumerable: true, get: function () { return MySQLConnection_1.MySQLConnection; } });
var PostgreSQLConnection_1 = require("./Connections/PostgreSQLConnection");
Object.defineProperty(exports, "PostgreSQLConnection", { enumerable: true, get: function () { return PostgreSQLConnection_1.PostgreSQLConnection; } });
var SQLiteConnection_1 = require("./Connections/SQLiteConnection");
Object.defineProperty(exports, "SQLiteConnection", { enumerable: true, get: function () { return SQLiteConnection_1.SQLiteConnection; } });
var MongoDBConnection_1 = require("./Connections/MongoDBConnection");
Object.defineProperty(exports, "MongoDBConnection", { enumerable: true, get: function () { return MongoDBConnection_1.MongoDBConnection; } });
// Query Builder
var QueryBuilder_1 = require("./Query/QueryBuilder");
Object.defineProperty(exports, "QueryBuilder", { enumerable: true, get: function () { return QueryBuilder_1.QueryBuilder; } });
var Expression_1 = require("./Query/Expression");
Object.defineProperty(exports, "Expression", { enumerable: true, get: function () { return Expression_1.Expression; } });
// Model
var Model_1 = require("./Model/Model");
Object.defineProperty(exports, "Model", { enumerable: true, get: function () { return Model_1.Model; } });
Object.defineProperty(exports, "ModelQueryBuilder", { enumerable: true, get: function () { return Model_1.ModelQueryBuilder; } });
var Relations_1 = require("./Model/Relations");
Object.defineProperty(exports, "HasOne", { enumerable: true, get: function () { return Relations_1.HasOne; } });
Object.defineProperty(exports, "HasMany", { enumerable: true, get: function () { return Relations_1.HasMany; } });
Object.defineProperty(exports, "BelongsTo", { enumerable: true, get: function () { return Relations_1.BelongsTo; } });
Object.defineProperty(exports, "BelongsToMany", { enumerable: true, get: function () { return Relations_1.BelongsToMany; } });
Object.defineProperty(exports, "HasOneThrough", { enumerable: true, get: function () { return Relations_1.HasOneThrough; } });
Object.defineProperty(exports, "HasManyThrough", { enumerable: true, get: function () { return Relations_1.HasManyThrough; } });
Object.defineProperty(exports, "MorphOne", { enumerable: true, get: function () { return Relations_1.MorphOne; } });
Object.defineProperty(exports, "MorphMany", { enumerable: true, get: function () { return Relations_1.MorphMany; } });
Object.defineProperty(exports, "MorphTo", { enumerable: true, get: function () { return Relations_1.MorphTo; } });
Object.defineProperty(exports, "Relation", { enumerable: true, get: function () { return Relations_1.Relation; } });
var SoftDeletes_1 = require("./Model/SoftDeletes");
Object.defineProperty(exports, "withSoftDeletes", { enumerable: true, get: function () { return SoftDeletes_1.withSoftDeletes; } });
var Observer_1 = require("./Model/Observer");
Object.defineProperty(exports, "ObserverRegistry", { enumerable: true, get: function () { return Observer_1.ObserverRegistry; } });
var GlobalScope_1 = require("./Model/GlobalScope");
Object.defineProperty(exports, "GlobalScopeRegistry", { enumerable: true, get: function () { return GlobalScope_1.GlobalScopeRegistry; } });
Object.defineProperty(exports, "CallbackGlobalScope", { enumerable: true, get: function () { return GlobalScope_1.CallbackGlobalScope; } });
// Transactions
var TransactionManager_1 = require("./Transactions/TransactionManager");
Object.defineProperty(exports, "TransactionManager", { enumerable: true, get: function () { return TransactionManager_1.TransactionManager; } });
// Query Logging
var QueryLogger_1 = require("./Query/QueryLogger");
Object.defineProperty(exports, "QueryLogger", { enumerable: true, get: function () { return QueryLogger_1.QueryLogger; } });
// Schema
var Schema_1 = require("./Schema/Schema");
Object.defineProperty(exports, "Schema", { enumerable: true, get: function () { return Schema_1.Schema; } });
Object.defineProperty(exports, "TableBlueprint", { enumerable: true, get: function () { return Schema_1.TableBlueprint; } });
Object.defineProperty(exports, "ColumnDefinition", { enumerable: true, get: function () { return Schema_1.ColumnDefinition; } });
Object.defineProperty(exports, "ForeignKeyDefinition", { enumerable: true, get: function () { return Schema_1.ForeignKeyDefinition; } });
var SchemaBuilder_1 = require("./Schema/SchemaBuilder");
Object.defineProperty(exports, "SchemaBuilder", { enumerable: true, get: function () { return SchemaBuilder_1.SchemaBuilder; } });
// Migrations
var Migration_1 = require("./Migrations/Migration");
Object.defineProperty(exports, "Migration", { enumerable: true, get: function () { return Migration_1.Migration; } });
var Migrator_1 = require("./Migrations/Migrator");
Object.defineProperty(exports, "Migrator", { enumerable: true, get: function () { return Migrator_1.Migrator; } });
// Seeders
var Seeder_1 = require("./Seeders/Seeder");
Object.defineProperty(exports, "Seeder", { enumerable: true, get: function () { return Seeder_1.Seeder; } });
var SeedRunner_1 = require("./Seeders/SeedRunner");
Object.defineProperty(exports, "SeedRunner", { enumerable: true, get: function () { return SeedRunner_1.SeedRunner; } });
