// @arikajs/middleware/dist/tests/Pipeline.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Pipeline_1 = require("../src/Pipeline");
const http_1 = require("@arikajs/http");
const node_assert_1 = __importDefault(require("node:assert"));
const node_test_1 = require("node:test");
const mockServerResponse = () => ({
    setHeader: () => { },
    end: () => { },
});
(0, node_test_1.test)('Pipeline executes middleware in order', async () => {
    const pipeline = new Pipeline_1.Pipeline();
    const calls = [];
    pipeline.pipe(async (req, next) => {
        calls.push('start 1');
        const res = await next(req);
        calls.push('end 1');
        return res;
    });
    pipeline.pipe(async (req, next) => {
        calls.push('start 2');
        const res = await next(req);
        calls.push('end 2');
        return res;
    });
    const request = {};
    const destination = async (req) => {
        calls.push('destination');
        return new http_1.Response(mockServerResponse()).send('OK');
    };
    const response = await pipeline.handle(request, destination);
    node_assert_1.default.strictEqual(response.getContent(), 'OK');
    node_assert_1.default.deepStrictEqual(calls, [
        'start 1',
        'start 2',
        'destination',
        'end 2',
        'end 1'
    ]);
});
(0, node_test_1.test)('Middleware can short-circuit', async () => {
    const pipeline = new Pipeline_1.Pipeline();
    const calls = [];
    pipeline.pipe(async (req, next) => {
        calls.push('m1');
        return new http_1.Response(mockServerResponse()).send('Short-circuit');
    });
    pipeline.pipe(async (req, next) => {
        calls.push('m2');
        return next(req);
    });
    const request = {};
    const destination = async (req) => {
        calls.push('destination');
        return new http_1.Response(mockServerResponse()).send('OK');
    };
    const response = await pipeline.handle(request, destination);
    node_assert_1.default.strictEqual(response.getContent(), 'Short-circuit');
    node_assert_1.default.deepStrictEqual(calls, ['m1']);
});
(0, node_test_1.test)('Middleware can be a class', async () => {
    const pipeline = new Pipeline_1.Pipeline();
    const calls = [];
    class Logger {
        async handle(req, next) {
            calls.push('logger');
            return next(req);
        }
    }
    pipeline.pipe(new Logger());
    const request = {};
    const destination = async (req) => {
        calls.push('destination');
        return new http_1.Response(mockServerResponse()).send('OK');
    };
    await pipeline.handle(request, destination);
    node_assert_1.default.deepStrictEqual(calls, ['logger', 'destination']);
});
(0, node_test_1.test)('Middleware can be resolved from container', async () => {
    const container = {
        has: (token) => token === 'auth',
        make: (token) => {
            if (token === 'auth')
                return new AuthMiddleware();
            return null;
        }
    };
    const calls = [];
    class AuthMiddleware {
        async handle(req, next) {
            calls.push('auth');
            return next(req);
        }
    }
    const pipeline = new Pipeline_1.Pipeline(container);
    pipeline.pipe('auth');
    const request = {};
    const destination = async (req) => {
        calls.push('destination');
        return new http_1.Response(mockServerResponse()).send('OK');
    };
    await pipeline.handle(request, destination);
    node_assert_1.default.deepStrictEqual(calls, ['auth', 'destination']);
});
