// @arikajs/auth/dist/Providers/EloquentUserProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EloquentUserProvider = void 0;
const Hasher_1 = require("../Hasher");
class EloquentUserProvider {
    constructor(model) {
        this.model = model;
    }
    /**
     * Retrieve a user by their unique identifier.
     */
    async retrieveById(identifier) {
        return await this.model.where('id', identifier).first();
    }
    /**
     * Retrieve a user by their unique identifier and "remember me" token.
     */
    async retrieveByToken(identifier, token) {
        // Implement remember token logic if needed
        return null;
    }
    /**
     * Update the "remember me" token for the given user in storage.
     */
    async updateRememberToken(user, token) {
        // Implement remember token logic if needed
    }
    /**
     * Retrieve a user by the given credentials.
     */
    async retrieveByCredentials(credentials) {
        if (Object.keys(credentials).length === 0 ||
            (Object.keys(credentials).length === 1 && 'password' in credentials)) {
            return null;
        }
        let query = this.model;
        for (const key in credentials) {
            if (key === 'password') {
                continue;
            }
            query = query.where(key, credentials[key]);
        }
        return await query.first();
    }
    /**
     * Validate a user against the given credentials.
     */
    async validateCredentials(user, credentials) {
        const plain = credentials.password;
        return await Hasher_1.Hasher.check(plain, user.password);
    }
    /**
     * Rehash the user's password if required.
     */
    async rehashPasswordIfRequired(user, credentials, force = false) {
        if (Hasher_1.Hasher.needsRehash(user.password) || force) {
            user.password = await Hasher_1.Hasher.make(credentials.password);
            await user.save();
        }
    }
}
exports.EloquentUserProvider = EloquentUserProvider;
