// @arikajs/auth/dist/tests/Auth.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = require("node:test");
const assert = __importStar(require("node:assert"));
const src_1 = require("../src/index");
const SessionGuard_1 = require("../src/Guards/SessionGuard");
class MockPoolProvider {
    constructor() {
        this.users = [
            { id: 1, email: 'test@example.com', password: '$2a$10$...' }
        ];
    }
    async retrieveById(id) {
        return this.users.find(u => u.id === Number(id)) || null;
    }
    async retrieveByCredentials(credentials) {
        if (credentials.email === 'test@example.com') {
            return this.users[0];
        }
        return null;
    }
    validateCredentials(user, credentials) {
        return credentials.password === 'secret';
    }
}
(0, node_test_1.describe)('Arika Auth', () => {
    let authManager;
    let config;
    (0, node_test_1.beforeEach)(() => {
        config = {
            default: 'web',
            guards: {
                web: { driver: 'session', provider: 'users' },
                api: { driver: 'token', provider: 'users' }
            }
        };
        authManager = new src_1.AuthManager(config);
        authManager.registerProvider('users', new MockPoolProvider());
    });
    (0, node_test_1.it)('resolves session guard by default via context', async () => {
        const request = { session: { get: () => null, put: () => { } } };
        const ctx = authManager.createContext(request);
        const guard = ctx.guard();
        assert.ok(guard);
        const isSessionGuard = guard instanceof SessionGuard_1.SessionGuard || guard.constructor.name === 'SessionGuard';
        assert.ok(isSessionGuard);
    });
    (0, node_test_1.it)('proxies methods through context', async () => {
        const session = {};
        const request = {
            session: {
                get: (key) => session[key] || null,
                put: (key, val) => { session[key] = val; },
                forget: (key) => { delete session[key]; }
            }
        };
        const ctx = authManager.createContext(request);
        const success = await ctx.attempt({ email: 'test@example.com', password: 'secret' });
        assert.strictEqual(success, true);
        const user = await ctx.user();
        assert.ok(user);
        assert.strictEqual(user.id, 1);
        const check = await ctx.check();
        assert.strictEqual(check, true);
        await ctx.logout();
        const checkAfterLogout = await ctx.check();
        assert.strictEqual(checkAfterLogout, false);
    });
    (0, node_test_1.it)('can validate credentials via context guard', async () => {
        const request = { session: { get: () => null, put: () => { } } };
        const ctx = authManager.createContext(request);
        const guard = ctx.guard('web');
        const success = await guard.validate({ email: 'test@example.com', password: 'secret' });
        assert.strictEqual(success, true);
        const fail = await guard.validate({ email: 'test@example.com', password: 'wrong' });
        assert.strictEqual(fail, false);
    });
    (0, node_test_1.it)('hashes passwords securely', async () => {
        const password = 'my-secret-password';
        const hash = await src_1.Hasher.make(password);
        assert.ok(hash.length > 0);
        assert.notStrictEqual(password, hash);
        const valid = await src_1.Hasher.check(password, hash);
        assert.strictEqual(valid, true);
        const invalid = await src_1.Hasher.check('wrong', hash);
        assert.strictEqual(invalid, false);
    });
    (0, node_test_1.it)('request-scoped auth context isolation', async () => {
        // Simulate two concurrent requests
        const session1 = {};
        const request1 = {
            session: {
                get: (key) => session1[key] || null,
                put: (key, val) => { session1[key] = val; },
                forget: (key) => { delete session1[key]; }
            }
        };
        const session2 = {};
        const request2 = {
            session: {
                get: (key) => session2[key] || null,
                put: (key, val) => { session2[key] = val; },
                forget: (key) => { delete session2[key]; }
            }
        };
        const ctx1 = authManager.createContext(request1);
        const ctx2 = authManager.createContext(request2);
        // Log in user on ctx1 only
        await ctx1.attempt({ email: 'test@example.com', password: 'secret' });
        const user1 = await ctx1.user();
        assert.ok(user1);
        // ctx2 should NOT have a user
        const user2 = await ctx2.user();
        assert.strictEqual(user2, null);
    });
    (0, node_test_1.it)('middleware protects routes with AuthContext', async () => {
        const { Authenticate } = require('../src/Middleware/Authenticate');
        const middleware = new Authenticate(authManager);
        const next = async () => 'success';
        // Mock request with token
        const request = {
            headers: { authorization: 'Bearer token-123' },
            user: null,
            session: { get: () => null, put: () => { } }
        };
        // Override config to default to 'api' for this test
        authManager.shouldUse('api');
        // Hack: mock the provider to accept this token
        const provider = authManager['providers'].get('users');
        provider.retrieveByCredentials = async (creds) => {
            if (creds.api_token === 'token-123')
                return { id: 1 };
            return null;
        };
        // Should pass
        const result = await middleware.handle(request, next);
        assert.strictEqual(result, 'success');
        // req.auth should have been set
        assert.ok(request.auth);
        const user = await request.auth.user();
        assert.strictEqual(user.id, 1);
    });
});
