// arikajs/dist/providers/DatabaseServiceProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatabaseServiceProvider = void 0;
const foundation_1 = require("@arikajs/foundation");
const database_1 = require("@arikajs/database");
class DatabaseServiceProvider extends foundation_1.ServiceProvider {
    /**
     * Register the database services.
     */
    async register() {
        this.app.singleton('db', () => {
            const config = this.app.config().get('database');
            if (!config) {
                // If no config provided, we'll use a default memory sqlite for convenience
                // though usually this should throw in production.
                return new database_1.DatabaseManager({
                    default: 'sqlite',
                    connections: {
                        sqlite: {
                            driver: 'sqlite',
                            database: ':memory:'
                        }
                    }
                });
            }
            return new database_1.DatabaseManager(config);
        });
        this.app.bind(database_1.DatabaseManager, () => this.app.resolve('db'));
    }
    /**
     * Boot the database services.
     */
    async boot() {
        // Initialize the static Database facade
        const dbManager = this.app.resolve(database_1.DatabaseManager);
        database_1.Database.setManager(dbManager);
        // Register caching for query builder
        if (this.app.getContainer().has('cache')) {
            dbManager.setCache(this.app.resolve('cache'));
        }
    }
}
exports.DatabaseServiceProvider = DatabaseServiceProvider;
