// arikajs/dist/providers/SessionServiceProvider.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionServiceProvider = void 0;
const foundation_1 = require("@arikajs/foundation");
const session_1 = require("@arikajs/session");
const session_2 = require("@arikajs/session");
const path = __importStar(require("path"));
class SessionServiceProvider extends foundation_1.ServiceProvider {
    /**
     * Register the session manager as a singleton.
     */
    async register() {
        this.app.singleton(session_1.SessionManager, () => {
            const config = this.app.config();
            const basePath = this.app.getBasePath();
            const sessionConfig = {
                driver: config.get('session.driver', 'file'),
                lifetime: config.get('session.lifetime', 120), // minutes
                cookie: config.get('session.cookie', 'arika_session'),
                path: config.get('session.path', '/'),
                storagePath: config.get('session.storagePath', path.join(basePath, 'storage', 'sessions')),
                secure: config.get('session.secure', false),
                httpOnly: config.get('session.httpOnly', true),
                sameSite: config.get('session.sameSite', 'Lax'),
                locking: config.get('session.locking', false),
                lockTimeout: config.get('session.lockTimeout', 10),
                gcProbability: config.get('session.gcProbability', 0.01),
                secret: config.get('app.key', 'fallback-secret-change-me'),
            };
            if (sessionConfig.driver === 'database') {
                try {
                    // Try to resolve the database manager from the container
                    const dbManager = this.app.make('db');
                    sessionConfig.connection = dbManager; // passing manager or connection directly
                    sessionConfig.table = config.get('session.table', 'sessions');
                }
                catch {
                    // Framework fallback
                }
            }
            if (sessionConfig.driver === 'redis') {
                try {
                    // Try to resolve redis from container or cache store
                    const redis = this.app.make('redis') || this.app.make('cache').store('redis').getRedis();
                    sessionConfig.connection = redis;
                    sessionConfig.prefix = config.get('session.prefix', 'arika_session:');
                }
                catch {
                    // Framework fallback
                }
            }
            return new session_1.SessionManager(sessionConfig);
        });
        this.app.singleton('session', () => this.app.make(session_1.SessionManager));
        this.app.singleton(session_2.StartSession, () => {
            return new session_2.StartSession(this.app.make(session_1.SessionManager));
        });
    }
    /**
     * Boot — nothing required.
     */
    async boot() { }
}
exports.SessionServiceProvider = SessionServiceProvider;
