// @arikajs/foundation/dist/container/Container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Container = void 0;
/**
 * Lightweight dependency injection container for ArikaJS.
 *
 * Responsible for:
 * - Registering bindings and singletons
 * - Resolving services on demand
 * - Holding pre-built instances
 */
class Container {
    constructor() {
        this.bindings = new Map();
        this.aliases = new Map();
        this.tags = new Map();
        this.extenders = new Map();
    }
    /**
     * Bind a token to a factory. Produces a new instance on every resolution.
     */
    bind(token, factory) {
        this.bindings.set(token, { factory, singleton: false });
    }
    /**
     * Bind a token as a singleton. The factory runs once and the
     * same instance is returned for all subsequent resolutions.
     */
    singleton(token, factory) {
        this.bindings.set(token, { factory, singleton: true });
    }
    /**
     * Directly register an existing instance for a token.
     */
    instance(token, value) {
        this.bindings.set(token, {
            factory: () => value,
            singleton: true,
            instance: value,
        });
    }
    /**
     * Create an alias for a token.
     */
    alias(token, alias) {
        this.aliases.set(alias, token);
    }
    /**
     * Assign a tag to a given token.
     */
    tag(token, tag) {
        if (!this.tags.has(tag)) {
            this.tags.set(tag, []);
        }
        this.tags.get(tag)?.push(token);
    }
    /**
     * Resolve all bindings associated with a given tag.
     */
    tagged(tag) {
        const tokens = this.tags.get(tag) || [];
        return tokens.map(token => this.make(token));
    }
    /**
     * Extend a binding in the container.
     */
    extend(token, callback) {
        const originalToken = this.getAlias(token);
        if (!this.extenders.has(originalToken)) {
            this.extenders.set(originalToken, []);
        }
        this.extenders.get(originalToken)?.push(callback);
    }
    /**
     * Resolve a token from the container.
     */
    make(token) {
        const originalToken = this.getAlias(token);
        let binding = this.bindings.get(originalToken);
        if (!binding) {
            if (typeof originalToken === 'function') {
                return this.instantiate(originalToken);
            }
            throw new Error(`Container: no binding found for token "${String(token)}"`);
        }
        let instance;
        if (binding.singleton) {
            if (binding.instance === undefined) {
                binding.instance = binding.factory(this);
                binding.instance = this.applyExtenders(originalToken, binding.instance);
            }
            instance = binding.instance;
        }
        else {
            instance = binding.factory(this);
            instance = this.applyExtenders(originalToken, instance);
        }
        return instance;
    }
    getAlias(token) {
        return this.aliases.get(token) || token;
    }
    applyExtenders(token, instance) {
        const extenders = this.extenders.get(token) || [];
        return extenders.reduce((inst, extender) => extender(inst), instance);
    }
    instantiate(constructor) {
        // Basic automatic injection (assumes constructor only takes container)
        // In the future, this will use Reflect metadata for property/param injection
        return new constructor(this);
    }
    /**
     * Alias for make() - resolves a token from the container.
     */
    resolve(token) {
        return this.make(token);
    }
    /**
     * Check whether a token has been registered.
     */
    has(token) {
        return this.bindings.has(this.getAlias(token));
    }
}
exports.Container = Container;
