// @arikajs/database/dist/Connections/SQLiteConnection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLiteConnection = void 0;
const better_sqlite3_1 = __importDefault(require("better-sqlite3"));
/**
 * SQLite database connection
 */
class SQLiteConnection {
    constructor(config) {
        this.db = null;
        this.writeDb = null;
        this.config = config;
        this.connect();
    }
    /**
     * Connect to the database
     */
    async connect() {
        let readConfig = this.config;
        if (this.config.read) {
            if (Array.isArray(this.config.read)) {
                const randomRead = this.config.read[Math.floor(Math.random() * this.config.read.length)];
                readConfig = { ...this.config, ...randomRead };
            }
            else {
                readConfig = { ...this.config, ...this.config.read };
            }
        }
        if (!this.db) {
            const dbPath = readConfig.database || ':memory:';
            // If config has explicit read/write split, open read DB read-only
            const isReadOnly = !!(this.config.read && this.config.write);
            this.db = new better_sqlite3_1.default(dbPath, { readonly: isReadOnly });
        }
        if (this.config.write && !this.writeDb) {
            const writeConfig = { ...this.config, ...this.config.write };
            this.writeDb = new better_sqlite3_1.default(writeConfig.database || ':memory:');
        }
    }
    /**
     * Disconnect from the database
     */
    async disconnect() {
        if (this.db) {
            this.db.close();
            this.db = null;
        }
        if (this.writeDb) {
            this.writeDb.close();
            this.writeDb = null;
        }
    }
    /**
     * Execute a query
     */
    async query(sql, bindings = []) {
        if (!this.db) {
            await this.connect();
        }
        const preparedBindings = bindings.map(b => {
            if (b instanceof Date)
                return b.toISOString();
            if (typeof b === 'boolean')
                return b ? 1 : 0;
            return b;
        });
        const isWriteOperation = /^\s*(?:insert|update|delete|create|alter|drop|truncate|replace)/i.test(sql);
        const connection = (isWriteOperation && this.writeDb) ? this.writeDb : this.db;
        try {
            const stmt = connection.prepare(sql);
            // Check if it's a SELECT query
            if (!isWriteOperation && sql.trim().toLowerCase().startsWith('select')) {
                return stmt.all(...preparedBindings);
            }
            // For INSERT, UPDATE, DELETE
            const info = stmt.run(...preparedBindings);
            return [{ affectedRows: info.changes, insertId: info.lastInsertRowid }];
        }
        catch (error) {
            throw new Error(`SQLite query failed: ${error.message}`);
        }
    }
    /**
     * Execute a raw query
     */
    async raw(sql) {
        if (!this.db) {
            await this.connect();
        }
        try {
            return this.db.exec(sql);
        }
        catch (error) {
            throw new Error(`SQLite raw query failed: ${error.message}`);
        }
    }
    /**
     * Get the underlying database instance
     */
    getDriver() {
        if (!this.db) {
            throw new Error('Database not connected');
        }
        return this.db;
    }
    /**
     * Close the connection
     */
    async close() {
        await this.disconnect();
    }
    /**
     * Begin a transaction
     */
    async beginTransaction() {
        await this.query('BEGIN TRANSACTION');
    }
    /**
     * Commit a transaction
     */
    async commit() {
        await this.query('COMMIT');
    }
    /**
     * Rollback a transaction
     */
    async rollback() {
        await this.query('ROLLBACK');
    }
    /**
     * Get the schema grammar for this connection
     */
    getSchemaGrammar() {
        const { SQLiteGrammar } = require('../Schema/Grammars/SQLiteGrammar');
        return new SQLiteGrammar();
    }
}
exports.SQLiteConnection = SQLiteConnection;
