// @arikajs/http/dist/Exceptions/HttpException.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnauthorizedHttpException = exports.MethodNotAllowedHttpException = exports.NotFoundHttpException = exports.HttpException = void 0;
class HttpException extends Error {
    constructor(statusCode, message, originalError = null) {
        super(message);
        this.statusCode = statusCode;
        this.originalError = originalError;
        this.name = 'HttpException';
    }
    getStatusCode() {
        return this.statusCode;
    }
}
exports.HttpException = HttpException;
class NotFoundHttpException extends HttpException {
    constructor(message = 'Not Found', originalError = null) {
        super(404, message, originalError);
        this.name = 'NotFoundHttpException';
    }
}
exports.NotFoundHttpException = NotFoundHttpException;
class MethodNotAllowedHttpException extends HttpException {
    constructor(message = 'Method Not Allowed', originalError = null) {
        super(405, message, originalError);
        this.name = 'MethodNotAllowedHttpException';
    }
}
exports.MethodNotAllowedHttpException = MethodNotAllowedHttpException;
class UnauthorizedHttpException extends HttpException {
    constructor(message = 'Unauthorized', originalError = null) {
        super(401, message, originalError);
        this.name = 'UnauthorizedHttpException';
    }
}
exports.UnauthorizedHttpException = UnauthorizedHttpException;
