// arikajs/src/index.ts
export * from './Application';
export { createApp } from './createApp';
export { app, config, env, info, error, warning, debug, route, lang, trans, __, view } from './helpers.js';

// Re-export common foundation items for convenience
export { ServiceProvider } from '@arikajs/foundation';
export { Container } from '@arikajs/foundation';
export { Repository as Config } from '@arikajs/foundation';

// Re-export HTTP items
export { Request, Response, Middleware, HttpException, NotFoundHttpException } from '@arikajs/http';
export { 
    BodyParserMiddleware, 
    CorsMiddleware, 
    TrimStrings, 
    ConvertEmptyStringsToNull, 
    SecurityHeaders,
    Throttle,
    ServeStaticMiddleware as HttpServeStaticMiddleware 
} from '@arikajs/http';
export { Kernel } from './http/Kernel';
export { Handler } from './http/Handler';

// Re-export Routing items
export { Route } from '@arikajs/router';

// Re-export Middleware items
export { Pipeline, MiddlewareHandler } from '@arikajs/middleware';
export { RequestLoggingMiddleware } from './http/Middleware/RequestLoggingMiddleware';
export { ServeStaticMiddleware } from './http/Middleware/ServeStaticMiddleware';
export { ValidateRequestMiddleware } from './http/Middleware/ValidateRequestMiddleware';
export { ViewMiddleware } from './http/Middleware/ViewMiddleware';
export { VerifyCsrfToken } from './http/Middleware/VerifyCsrfToken';

// Re-export Database items
export { Model, Database as DB, Schema, Migration, SchemaBuilder } from '@arikajs/database';

// Re-export Cache items
export { Cache } from '@arikajs/cache';

// Re-export Queue items
export { Queue, BaseJob, Worker } from '@arikajs/queue';

// Re-export Event items
export { Event } from '@arikajs/events';

// Re-export Scheduler items
export { Scheduler, Schedule } from '@arikajs/scheduler';

// Re-export Encryption items
export { Encrypter } from '@arikajs/encryption';

// Re-export Console items
export { Command } from '@arikajs/console';

// Re-export Logging items
export { Log } from '@arikajs/logging';

// Re-export View items
export { View } from '@arikajs/view';

// Re-export Validation items
export { Validator } from '@arikajs/validation';

// Re-export Localization items
export { Translator } from '@arikajs/localization';

// Re-export Mail items
export { Mail, Mailable } from '@arikajs/mail';

// Re-export Auth items
export { AuthManager, AuthContext, Hasher, EloquentUserProvider, Authenticate } from '@arikajs/auth';

// Re-export Session items
export { Session, SessionManager, StartSession, SessionDriver, MemoryDriver as SessionMemoryDriver, FileDriver as SessionFileDriver } from '@arikajs/session';
