// @arikajs/localization/dist/Translator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Translator = void 0;
const MessageSelector_1 = require("./MessageSelector");
class Translator {
    constructor(locale = 'en', fallbackLocale = 'en') {
        this.locale = 'en';
        this.fallbackLocale = 'en';
        this.translations = {};
        this.locale = locale;
        this.fallbackLocale = fallbackLocale;
        this.selector = new MessageSelector_1.MessageSelector();
    }
    /**
     * Get the translation for the given key.
     */
    get(key, replace = {}, locale = null) {
        const currentLocale = locale || this.locale;
        let line = this.getLine(currentLocale, key);
        // Fallback if not found
        if (!line && currentLocale !== this.fallbackLocale) {
            line = this.getLine(this.fallbackLocale, key);
        }
        if (!line) {
            return key;
        }
        return this.makeReplacements(line, replace);
    }
    /**
     * Get a translation according to an integer value.
     */
    choice(key, number, replace = {}, locale = null) {
        const line = this.get(key, replace, locale);
        // If the line is the key itself (not found), return it
        if (line === key) {
            return key;
        }
        const selected = this.selector.choose(line, number, locale || this.locale);
        return this.makeReplacements(selected, { ...replace, count: number });
    }
    /**
     * Set the current locale.
     */
    setLocale(locale) {
        this.locale = locale;
    }
    /**
     * Get the current locale.
     */
    getLocale() {
        return this.locale;
    }
    /**
     * Load translations into the translator.
     */
    load(locale, group, lines) {
        if (!this.translations[locale]) {
            this.translations[locale] = {};
        }
        this.translations[locale][group] = lines;
    }
    /**
     * Retrieve a line from the translation repository.
     */
    getLine(locale, key) {
        const segments = key.split('.');
        const group = segments.shift();
        if (!group || !this.translations[locale] || !this.translations[locale][group]) {
            return null;
        }
        let line = this.translations[locale][group];
        for (const segment of segments) {
            if (line[segment] === undefined) {
                return null;
            }
            line = line[segment];
        }
        return typeof line === 'string' ? line : null;
    }
    /**
     * Make the place-holder replacements in the given line.
     */
    makeReplacements(line, replace) {
        if (Object.keys(replace).length === 0) {
            return line;
        }
        Object.entries(replace).forEach(([key, value]) => {
            line = line.replace(new RegExp(`:${key}`, 'g'), String(value));
        });
        return line;
    }
}
exports.Translator = Translator;
