// @arikajs/router/dist/RouteEntry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouteEntry = void 0;
class RouteEntry {
    constructor(method, path, handler, prefix, groupMiddleware = []) {
        this.middleware = [];
        this.paramKeys = [];
        this.constraints = {};
        this.method = method.toUpperCase();
        this.path = path;
        this.handler = handler;
        this.prefix = prefix;
        this.middleware = [...groupMiddleware];
        const { regex, paramKeys } = this.compilePath(path);
        this.regex = regex;
        this.paramKeys = paramKeys;
    }
    /**
     * Add a regular expression requirement for a route parameter.
     */
    where(param, regex) {
        if (typeof param === 'string') {
            this.constraints[param] = regex;
        }
        else {
            Object.assign(this.constraints, param);
        }
        // Re-compile regex when constraints are updated
        const compiled = this.compilePath(this.path);
        this.regex = compiled.regex;
        return this;
    }
    /**
     * Set a name for the route.
     */
    name(name) {
        this._name = name;
        return this;
    }
    /**
     * Set an alias name for the route.
     */
    as(name) {
        return this.name(name);
    }
    /**
     * Add middleware to the route.
     */
    withMiddleware(middleware) {
        if (Array.isArray(middleware)) {
            this.middleware.push(...middleware);
        }
        else {
            this.middleware.push(middleware);
        }
        return this;
    }
    /**
     * Compile the path into a regex.
     */
    compilePath(path) {
        let paramKeys = [];
        const pattern = path
            .replace(/:([a-zA-Z0-9_]+)|\{([a-zA-Z0-9_]+)\}/g, (_, key1, key2) => {
            const key = key1 || key2;
            paramKeys.push(key);
            const constrain = this.constraints[key] || '[^/]+';
            return `(${constrain})`;
        })
            .replace(/\//g, '\\/');
        return {
            regex: new RegExp(`^${pattern}$`),
            paramKeys
        };
    }
}
exports.RouteEntry = RouteEntry;
