// @arikajs/router/dist/RouteMatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouteMatcher = void 0;
const RouteRegistry_1 = require("./RouteRegistry");
class RadixNode {
    constructor() {
        this.children = new Map();
        this.paramName = null;
        this.paramNode = null;
        this.routes = []; // Support multiple methods on same path
    }
}
class RouteMatcher {
    constructor() {
        this.root = new RadixNode();
        this.isBuilt = false;
    }
    buildTree() {
        const routes = RouteRegistry_1.RouteRegistry.getInstance().getRoutes();
        this.root = new RadixNode(); // Reset root
        for (const route of routes) {
            this.insert(route);
        }
        this.isBuilt = true;
    }
    insert(route) {
        let node = this.root;
        const parts = route.path.split('/').filter(Boolean);
        for (const part of parts) {
            if (part.startsWith(':')) {
                const paramName = part.slice(1);
                if (!node.paramNode) {
                    node.paramNode = new RadixNode();
                    node.paramName = paramName;
                }
                node = node.paramNode;
            }
            else {
                if (!node.children.has(part)) {
                    node.children.set(part, new RadixNode());
                }
                node = node.children.get(part);
            }
        }
        node.routes.push(route);
    }
    match(method, path) {
        if (!this.isBuilt)
            this.buildTree();
        let node = this.root;
        const normalizedPath = path.split('?')[0];
        const parts = normalizedPath.split('/').filter(Boolean);
        const params = {};
        const normalizedMethod = method.toUpperCase();
        for (let i = 0; i < parts.length; i++) {
            const part = parts[i];
            const nextNode = node.children.get(part);
            if (nextNode) {
                node = nextNode;
            }
            else if (node.paramNode) {
                if (node.paramName) {
                    params[node.paramName] = part;
                }
                node = node.paramNode;
            }
            else {
                return null;
            }
        }
        if (node.routes.length > 0) {
            // Find a route that matches the method OR has method 'ANY'
            const route = node.routes.find(r => r.method === normalizedMethod || r.method === 'ANY');
            if (route) {
                const hasParams = Object.keys(params).length > 0;
                // If there are constraints, verify them
                if (route.constraints && Object.keys(route.constraints).length > 0) {
                    for (const [key, pattern] of Object.entries(route.constraints)) {
                        if (params[key]) {
                            const regex = new RegExp(`^${pattern}$`);
                            if (!regex.test(params[key])) {
                                return null;
                            }
                        }
                    }
                }
                return { route, params, hasParams };
            }
        }
        return null;
    }
}
exports.RouteMatcher = RouteMatcher;
