// @arikajs/router/dist/Router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Router = void 0;
const RouteMatcher_1 = require("./RouteMatcher");
const dispatcher_1 = require("@arikajs/dispatcher");
const RouteRegistry_1 = require("./RouteRegistry");
class Router {
    constructor(container) {
        this.cache = new Map();
        this.modelsSynced = false;
        this.isSynced = false;
        this.matcher = new RouteMatcher_1.RouteMatcher();
        this.dispatcher = new dispatcher_1.Dispatcher(container);
    }
    /**
     * Sync models from the static registry.
     */
    syncModels() {
        if (this.modelsSynced)
            return;
        const models = RouteRegistry_1.RouteRegistry.getInstance().getModels();
        for (const [key, resolver] of models.entries()) {
            this.dispatcher.bind(key, resolver);
        }
        this.modelsSynced = true;
    }
    sync() {
        if (this.isSynced)
            return;
        this.syncModels();
        this.dispatcher.precompile(RouteRegistry_1.RouteRegistry.getInstance().getRoutes());
        this.isSynced = true;
    }
    /**
     * Set the container for the dispatcher.
     */
    setContainer(container) {
        this.dispatcher.setContainer(container);
        return this;
    }
    /**
     * Set the middleware groups.
     */
    setMiddlewareGroups(groups) {
        this.dispatcher.setMiddlewareGroups(groups);
        return this;
    }
    /**
     * Set the route middleware aliases.
     */
    setRouteMiddleware(middleware) {
        this.dispatcher.setRouteMiddleware(middleware);
        return this;
    }
    /**
     * Register a route model binding.
     */
    model(key, resolver) {
        this.dispatcher.bind(key, resolver);
        return this;
    }
    /**
     * Match a request to a route.
     */
    match(method, path) {
        if (!this.isSynced)
            this.sync();
        const cacheKey = `${method.toUpperCase()}:${path}`;
        if (this.cache.has(cacheKey)) {
            return this.cache.get(cacheKey);
        }
        const matched = this.matcher.match(method, path);
        this.cache.set(cacheKey, matched);
        return matched;
    }
    /**
     * Generate a URL for a named route.
     */
    route(name, params = {}) {
        const route = RouteRegistry_1.RouteRegistry.getInstance().getRouteByName(name);
        if (!route) {
            throw new Error(`Route [${name}] not found.`);
        }
        let path = route.path;
        for (const [key, value] of Object.entries(params)) {
            path = path.replace(`:${key}`, String(value)).replace(`{${key}}`, String(value));
        }
        // Return without checking for lingering parameters (that could be optional in regex)
        // just a basic replacement implementation.
        return path;
    }
    /**
     * Clear the route cache.
     */
    clearCache() {
        this.cache.clear();
    }
    /**
     * Dispatch the request to the matching route and return the result.
     * Note: This does not wrap the result in a Response object.
     */
    async dispatch(request, response) {
        this.syncModels();
        const matched = this.match(request.method(), request.path());
        if (!matched) {
            return null;
        }
        return await this.dispatcher.dispatch(matched, request, response);
    }
}
exports.Router = Router;
