// @arikajs/router/dist/RouteRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouteRegistry = void 0;
const RouteEntry_1 = require("./RouteEntry");
class RouteRegistry {
    constructor() {
        this.routes = [];
        this.groupStack = [];
        this.middlewareStack = [];
        this.models = new Map();
    }
    static getInstance() {
        if (!RouteRegistry.instance) {
            RouteRegistry.instance = new RouteRegistry();
        }
        return RouteRegistry.instance;
    }
    model(key, resolver) {
        this.models.set(key, resolver);
    }
    getModels() {
        return this.models;
    }
    addRoute(method, path, handler) {
        const prefix = this.groupStack.join('');
        const fullPath = this.normalizePath(prefix + '/' + path);
        const groupMiddleware = this.middlewareStack.flat();
        const route = new RouteEntry_1.RouteEntry(method, fullPath, handler, prefix || undefined, groupMiddleware);
        this.routes.push(route);
        return route;
    }
    group(options, callback) {
        let prefix = typeof options === 'string' ? options : options.prefix || '';
        if (prefix && !prefix.startsWith('/')) {
            prefix = '/' + prefix;
        }
        const middleware = typeof options === 'string' ? [] : (Array.isArray(options.middleware) ? options.middleware : (options.middleware ? [options.middleware] : []));
        this.groupStack.push(prefix);
        this.middlewareStack.push(middleware);
        callback();
        this.middlewareStack.pop();
        this.groupStack.pop();
    }
    getRoutes() {
        return this.routes;
    }
    getRouteByName(name) {
        return this.routes.find(r => r._name === name);
    }
    clear() {
        this.routes = [];
        this.groupStack = [];
        this.middlewareStack = [];
    }
    normalizePath(path) {
        const normalized = path.replace(/\/+/g, '/');
        return normalized.startsWith('/') ? normalized : '/' + normalized;
    }
}
exports.RouteRegistry = RouteRegistry;
