// @arikajs/router/dist/Route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route = exports.RouteGroupBuilder = void 0;
const RouteRegistry_1 = require("./RouteRegistry");
class RouteGroupBuilder {
    constructor() {
        this.options = {};
    }
    prefix(prefix) {
        this.options.prefix = prefix;
        return this;
    }
    middleware(middleware) {
        this.options.middleware = Array.isArray(middleware) ? middleware : [middleware];
        return this;
    }
    group(callback) {
        RouteRegistry_1.RouteRegistry.getInstance().group(this.options, callback);
    }
}
exports.RouteGroupBuilder = RouteGroupBuilder;
class Route {
    static get(path, handler) {
        return RouteRegistry_1.RouteRegistry.getInstance().addRoute('GET', path, handler);
    }
    static post(path, handler) {
        return RouteRegistry_1.RouteRegistry.getInstance().addRoute('POST', path, handler);
    }
    static put(path, handler) {
        return RouteRegistry_1.RouteRegistry.getInstance().addRoute('PUT', path, handler);
    }
    static patch(path, handler) {
        return RouteRegistry_1.RouteRegistry.getInstance().addRoute('PATCH', path, handler);
    }
    static delete(path, handler) {
        return RouteRegistry_1.RouteRegistry.getInstance().addRoute('DELETE', path, handler);
    }
    static any(path, handler) {
        const methods = ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS', 'HEAD'];
        return methods.map(method => RouteRegistry_1.RouteRegistry.getInstance().addRoute(method, path, handler));
    }
    static options(path, handler) {
        return RouteRegistry_1.RouteRegistry.getInstance().addRoute('OPTIONS', path, handler);
    }
    static match(methods, path, handler) {
        return methods.map(method => RouteRegistry_1.RouteRegistry.getInstance().addRoute(method.toUpperCase(), path, handler));
    }
    static redirect(path, destination, status = 302) {
        return RouteRegistry_1.RouteRegistry.getInstance().addRoute('ANY', path, (request, response) => {
            if (response && typeof response.redirect === 'function') {
                return response.redirect(destination, status);
            }
            return { $status: status, $redirect: destination }; // Fallback signaling for dispatcher
        });
    }
    static resource(name, controller) {
        Route.get(`/${name}`, [controller, 'index']).name(`${name}.index`);
        Route.get(`/${name}/create`, [controller, 'create']).name(`${name}.create`);
        Route.post(`/${name}`, [controller, 'store']).name(`${name}.store`);
        Route.get(`/${name}/:id`, [controller, 'show']).name(`${name}.show`);
        Route.get(`/${name}/:id/edit`, [controller, 'edit']).name(`${name}.edit`);
        Route.put(`/${name}/:id`, [controller, 'update']).name(`${name}.update`);
        Route.patch(`/${name}/:id`, [controller, 'update']).name(`${name}.update`);
        Route.delete(`/${name}/:id`, [controller, 'destroy']).name(`${name}.destroy`);
    }
    static group(options, callback) {
        RouteRegistry_1.RouteRegistry.getInstance().group(options, callback);
    }
    static prefix(prefix) {
        return new RouteGroupBuilder().prefix(prefix);
    }
    static middleware(middleware) {
        return new RouteGroupBuilder().middleware(middleware);
    }
    static model(key, resolver) {
        RouteRegistry_1.RouteRegistry.getInstance().model(key, resolver);
    }
}
exports.Route = Route;
