// @arikajs/auth/dist/src/AuthContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthContext = void 0;
class AuthContext {
    constructor(manager, request) {
        this.guards = new Map();
        this.manager = manager;
        this.request = request;
    }
    guard(name) {
        name = name || this.manager.getDefaultGuard();
        if (!name) {
            throw new Error('No auth guard defined.');
        }
        if (!this.guards.has(name)) {
            this.guards.set(name, this.manager.resolveGuard(name, this.request));
        }
        return this.guards.get(name);
    }
    // Proxy methods to the default guard
    async check() {
        return await this.guard().check();
    }
    async guest() {
        return await this.guard().guest();
    }
    async user() {
        return await this.guard().user();
    }
    async id() {
        return await this.guard().id();
    }
    async validate(credentials) {
        return await this.guard().validate(credentials);
    }
    setUser(user) {
        this.guard().setUser(user);
    }
    async attempt(credentials, remember = false) {
        return await this.manager.attemptForContext(this, credentials, remember);
    }
    async login(user, remember = false) {
        return await this.manager.loginForContext(this, user, remember);
    }
    async logout() {
        return await this.manager.logoutForContext(this);
    }
    async sendVerification(user) {
        return await this.manager.sendVerification(this, user);
    }
    async isLocked(credentials) {
        return await this.manager.isLocked(this, credentials);
    }
    async unlockAccount(credentials) {
        return await this.manager.unlockAccount(this, credentials);
    }
    getRequest() {
        return this.request;
    }
    resolve(token) {
        return this.guard(token);
    }
}
exports.AuthContext = AuthContext;
