// @arikajs/auth/dist/src/AuthManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthManager = void 0;
const SessionGuard_1 = require("./Guards/SessionGuard");
const TokenGuard_1 = require("./Guards/TokenGuard");
const JwtGuard_1 = require("./Guards/JwtGuard");
const BasicGuard_1 = require("./Guards/BasicGuard");
const AuthContext_1 = require("./AuthContext");
const async_hooks_1 = require("async_hooks");
const EloquentUserProvider_1 = require("./Providers/EloquentUserProvider");
class AuthManager {
    constructor(config) {
        this.providers = new Map();
        this.eventDispatcher = null;
        this.rateLimiter = null;
        this.als = new async_hooks_1.AsyncLocalStorage();
        this.config = config;
        this.defaultGuard = config.default || 'web';
    }
    setEventDispatcher(dispatcher) {
        this.eventDispatcher = dispatcher;
    }
    setRateLimiter(limiter) {
        this.rateLimiter = limiter;
    }
    registerProvider(name, provider) {
        this.providers.set(name, provider);
    }
    extend(name, callback) {
        // Implementation for custom guards
    }
    getDefaultGuard() {
        return this.defaultGuard;
    }
    createContext(request) {
        const context = new AuthContext_1.AuthContext(this, request);
        // Optionally bind it directly to the request
        if (request) {
            request.auth = context;
        }
        return context;
    }
    resolveGuard(name, request) {
        const config = this.config.guards[name];
        if (!config) {
            throw new Error(`Auth guard [${name}] is not defined.`);
        }
        if (config.driver === 'session') {
            return this.createSessionDriver(name, config, request);
        }
        if (config.driver === 'token') {
            return this.createTokenDriver(name, config, request);
        }
        if (config.driver === 'jwt') {
            return this.createJwtDriver(name, config, request);
        }
        if (config.driver === 'basic') {
            return this.createBasicDriver(name, config, request);
        }
        throw new Error(`Auth driver [${config.driver}] for guard [${name}] is not supported.`);
    }
    createSessionDriver(name, config, request) {
        const provider = this.getProvider(config.provider);
        return new SessionGuard_1.SessionGuard(provider, request.session);
    }
    createTokenDriver(name, config, request) {
        const provider = this.getProvider(config.provider);
        return new TokenGuard_1.TokenGuard(provider, request);
    }
    createJwtDriver(name, config, request) {
        const provider = this.getProvider(config.provider);
        return new JwtGuard_1.JwtGuard(provider, request, config.secret || 'default_secret', config.options || {});
    }
    createBasicDriver(name, config, request) {
        const provider = this.getProvider(config.provider);
        return new BasicGuard_1.BasicGuard(provider, request);
    }
    getProvider(name) {
        // 1. Check if registered manually
        if (this.providers.has(name)) {
            return this.providers.get(name);
        }
        // 2. Resolve from config
        const config = this.config.providers?.[name];
        if (!config) {
            throw new Error(`User provider [${name}] is not defined.`);
        }
        if (config.driver === 'eloquent') {
            const provider = new EloquentUserProvider_1.EloquentUserProvider(config.model);
            this.providers.set(name, provider);
            return provider;
        }
        throw new Error(`User provider driver [${config.driver}] is not supported.`);
    }
    shouldUse(name) {
        this.defaultGuard = name;
    }
    runWithContext(context, fn) {
        return this.als.run(context, fn);
    }
    getContext() {
        const ctx = this.als.getStore();
        if (!ctx) {
            throw new Error('AuthContext not found. Ensure you are running within a request scope or use req.auth instead of the global facade.');
        }
        return ctx;
    }
    // Proxy methods to the context
    guard(name) {
        return this.getContext().guard(name);
    }
    resolve(token) {
        return this.getContext().resolve(token);
    }
    async check() {
        return await this.getContext().check();
    }
    async guest() {
        return await this.getContext().guest();
    }
    async user() {
        return await this.getContext().user();
    }
    async id() {
        return await this.getContext().id();
    }
    async validate(credentials) {
        return await this.getContext().validate(credentials);
    }
    setUser(user) {
        this.getContext().setUser(user);
    }
    async attempt(credentials, remember = false) {
        return await this.getContext().attempt(credentials, remember);
    }
    async login(user, remember = false) {
        return await this.getContext().login(user, remember);
    }
    async logout() {
        return await this.getContext().logout();
    }
    // Called by AuthContext to run attempts
    async attemptForContext(context, credentials, remember = false) {
        this.fireEvent('Auth.Attempting', { credentials, remember, guard: this.defaultGuard });
        const throttleKey = this.getThrottleKey(credentials, context.getRequest());
        if (throttleKey && this.rateLimiter) {
            if (await this.rateLimiter.tooManyAttempts(throttleKey, 5)) {
                this.fireEvent('Auth.Lockout', { credentials });
                throw new Error('Too many login attempts. Please try again later.');
            }
        }
        const guard = context.guard();
        if (typeof guard.attempt === 'function') {
            const successOrToken = await guard.attempt(credentials, remember);
            if (successOrToken) {
                if (throttleKey && this.rateLimiter) {
                    await this.rateLimiter.clear(throttleKey);
                }
                const user = await guard.user();
                this.fireEvent('Auth.Login', { user, guard: this.defaultGuard });
                return successOrToken; // Can return boolean true or JWT string
            }
            if (throttleKey && this.rateLimiter) {
                await this.rateLimiter.hit(throttleKey, 1); // 1 minute decay
            }
            this.fireEvent('Auth.Failed', { credentials, guard: this.defaultGuard });
            return false;
        }
        throw new Error(`Guard [${this.defaultGuard}] does not support login attempts.`);
    }
    // Called by AuthContext to log in
    async loginForContext(context, user, remember = false) {
        const guard = context.guard();
        if (typeof guard.login === 'function') {
            await guard.login(user, remember);
            this.fireEvent('Auth.Login', { user, guard: this.defaultGuard });
            return;
        }
        throw new Error(`Guard [${this.defaultGuard}] does not support login.`);
    }
    // Called by AuthContext to log out
    async logoutForContext(context) {
        const guard = context.guard();
        const user = await guard.user();
        if (typeof guard.logout === 'function') {
            guard.logout();
            this.fireEvent('Auth.Logout', { user, guard: this.defaultGuard });
            return;
        }
        throw new Error(`Guard [${this.defaultGuard}] does not support logout.`);
    }
    // ── Email Verification ──────────────────────────────────────────
    async sendVerification(context, user) {
        const targetUser = user || await context.user();
        if (!targetUser) {
            throw new Error('No authenticated user to verify.');
        }
        if (typeof targetUser.hasVerifiedEmail === 'function' && targetUser.hasVerifiedEmail()) {
            return; // Already verified
        }
        if (typeof targetUser.sendEmailVerificationNotification === 'function') {
            await targetUser.sendEmailVerificationNotification();
            this.fireEvent('Auth.VerificationSent', { user: targetUser });
        }
        else {
            throw new Error('User model does not implement sendEmailVerificationNotification().');
        }
    }
    // ── Account Locking ─────────────────────────────────────────────
    get lockoutThreshold() {
        return this.config.lockout?.maxAttempts ?? 5;
    }
    get lockoutDuration() {
        return this.config.lockout?.decayMinutes ?? 15;
    }
    async isLocked(context, credentials) {
        if (!this.rateLimiter)
            return false;
        const throttleKey = this.getLockKey(credentials, context.getRequest());
        if (!throttleKey)
            return false;
        return await this.rateLimiter.tooManyAttempts(throttleKey, this.lockoutThreshold);
    }
    async unlockAccount(context, credentials) {
        if (!this.rateLimiter)
            return;
        const throttleKey = this.getLockKey(credentials, context.getRequest());
        if (throttleKey) {
            await this.rateLimiter.clear(throttleKey);
            this.fireEvent('Auth.AccountUnlocked', { credentials });
        }
    }
    getLockKey(credentials, request) {
        if (credentials.email) {
            const ip = request?.ip || '127.0.0.1';
            return `account_lock:${credentials.email}:${ip}`;
        }
        return null;
    }
    // ── Internals ───────────────────────────────────────────────────
    fireEvent(name, payload) {
        if (this.eventDispatcher) {
            this.eventDispatcher.dispatch(name, payload);
        }
    }
    getThrottleKey(credentials, request) {
        if (credentials.email) {
            const ip = request?.ip || '127.0.0.1';
            return `login_attempts:${credentials.email}:${ip}`;
        }
        return null;
    }
}
exports.AuthManager = AuthManager;
