// @arikajs/docs/dist/Generator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentationGenerator = void 0;
const PostmanGenerator_1 = require("./PostmanGenerator");
const MarkdownGenerator_1 = require("./MarkdownGenerator");
const HtmlGenerator_1 = require("./HtmlGenerator");
const OpenApiGenerator_1 = require("./OpenApiGenerator");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class DocumentationGenerator {
    /**
     * Generate all documentation formats based on parsed routes.
     */
    generateAll(routes, appName, outputDir, options = {}) {
        const baseUrl = options.baseUrl || 'http://localhost:3000';
        if (!fs_1.default.existsSync(outputDir)) {
            fs_1.default.mkdirSync(outputDir, { recursive: true });
        }
        // Generate Postman Environment
        const envJson = {
            name: `${appName} Env`,
            values: [
                { key: "base_url", value: baseUrl, enabled: true },
                { key: "token", value: "", enabled: true }
            ]
        };
        const envContent = JSON.stringify(envJson, null, 2);
        fs_1.default.writeFileSync(path_1.default.join(outputDir, 'postman_environment.json'), envContent);
        // Generate Postman Collection
        const postmanGenerator = new PostmanGenerator_1.PostmanGenerator();
        const postmanJson = postmanGenerator.generate(routes, appName, baseUrl);
        const postmanContent = JSON.stringify(postmanJson, null, 2);
        fs_1.default.writeFileSync(path_1.default.join(outputDir, 'postman_collection.json'), postmanContent);
        // Generate OpenAPI Spec
        const openApiGenerator = new OpenApiGenerator_1.OpenApiGenerator();
        const openApiJson = openApiGenerator.generate(routes, appName, baseUrl);
        const openApiContent = JSON.stringify(openApiJson, null, 2);
        fs_1.default.writeFileSync(path_1.default.join(outputDir, 'openapi.json'), openApiContent);
        // Generate Markdown
        const markdownGenerator = new MarkdownGenerator_1.MarkdownGenerator();
        const markdownContent = markdownGenerator.generate(routes, appName, baseUrl);
        fs_1.default.writeFileSync(path_1.default.join(outputDir, 'DOCS.md'), markdownContent);
        // Generate HTML with embedded JSONs
        const htmlGenerator = new HtmlGenerator_1.HtmlGenerator();
        const htmlContent = htmlGenerator.generate(routes, appName, baseUrl, {
            postmanContent,
            envContent,
            openApiContent
        });
        fs_1.default.writeFileSync(path_1.default.join(outputDir, 'api_docs.html'), htmlContent);
    }
}
exports.DocumentationGenerator = DocumentationGenerator;
