// @arikajs/docs/dist/RouteParser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouteParser = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class RouteParser {
    parseApplicationRoutes(appDir, routeFileName = null) {
        const routesDir = path_1.default.join(appDir, 'routes');
        if (!fs_1.default.existsSync(routesDir))
            return [];
        const routes = [];
        let filesToParse = [];
        if (routeFileName) {
            const specificFile = path_1.default.join(routesDir, routeFileName.endsWith('.ts') ? routeFileName : `${routeFileName}.ts`);
            if (fs_1.default.existsSync(specificFile)) {
                filesToParse = [specificFile];
            }
            else {
                console.warn(`\n  ⚠ Route file not found: ${specificFile}\n`);
                return [];
            }
        }
        else {
            const allFiles = fs_1.default.readdirSync(routesDir).filter(f => f.endsWith('.ts'));
            if (allFiles.includes('api.ts')) {
                filesToParse = [path_1.default.join(routesDir, 'api.ts')];
            }
            else if (allFiles.length > 0) {
                filesToParse = [path_1.default.join(routesDir, allFiles[0])];
            }
        }
        for (const file of filesToParse) {
            const content = fs_1.default.readFileSync(file, 'utf8');
            this.extractRoutesFromContent(content, routes);
        }
        return routes;
    }
    extractRoutesFromContent(content, routes) {
        const strippedContent = this.extractGroupRoutes(content, routes);
        this.extractFlatRoutes(strippedContent, routes, '', []);
    }
    extractGroupRoutes(content, routes) {
        const groupRegex = /Route\.group\s*\(\s*\{([^}]+)\}\s*,\s*\(\s*\)\s*=>\s*\{([\s\S]*?)\}\s*\)/g;
        let match;
        let contentCopy = content;
        while ((match = groupRegex.exec(content)) !== null) {
            const groupConfig = match[1];
            const groupBody = match[2];
            let prefix = '';
            let middleware = [];
            const prefixMatch = groupConfig.match(/prefix\s*:\s*['"`]([^'"`]+)['"`]/);
            if (prefixMatch)
                prefix = prefixMatch[1];
            const mwMatch = groupConfig.match(/middleware\s*:\s*(['"`][^'"`]+['"`]|\[[^\]]+\])/);
            if (mwMatch) {
                const mwStr = mwMatch[1];
                if (mwStr.startsWith('[')) {
                    const mws = mwStr.match(/['"`]([^'"`]+)['"`]/g);
                    if (mws)
                        middleware = mws.map(m => m.replace(/['"`]/g, ''));
                }
                else {
                    middleware.push(mwStr.replace(/['"`]/g, ''));
                }
            }
            this.extractFlatRoutes(groupBody, routes, prefix, middleware);
        }
        return this.stripGroupBlocks(contentCopy);
    }
    stripGroupBlocks(content) {
        let result = content;
        let changed = true;
        while (changed) {
            changed = false;
            const groupStart = result.indexOf('Route.group(');
            if (groupStart === -1)
                break;
            const arrowIdx = result.indexOf('=>', groupStart);
            if (arrowIdx === -1)
                break;
            const braceStart = result.indexOf('{', arrowIdx);
            if (braceStart === -1)
                break;
            let depth = 0;
            let i = braceStart;
            for (; i < result.length; i++) {
                if (result[i] === '{')
                    depth++;
                else if (result[i] === '}') {
                    depth--;
                    if (depth === 0)
                        break;
                }
            }
            let endIdx = result.indexOf(';', i);
            if (endIdx === -1)
                endIdx = i + 1;
            result = result.substring(0, groupStart) + result.substring(endIdx + 1);
            changed = true;
        }
        return result;
    }
    extractFlatRoutes(content, routes, prefix, middleware) {
        const normalized = content.replace(/\s+/g, ' ');
        const routeRegex = /Route\.(get|post|put|patch|delete|options)\s*\(\s*['"`]([^'"`]*)['"`]\s*,\s*((?:[^\(\)]*|\([^\)]*\))*)\)\s*([^;]*)?;?/g;
        let match;
        while ((match = routeRegex.exec(normalized)) !== null) {
            const method = match[1].toUpperCase();
            const routePath = match[2];
            const handlerStr = match[3]?.trim() || '';
            const modifiersStr = match[4] || '';
            const controllerMatch = handlerStr.match(/\[\s*(\w+)\s*,\s*['"`](\w+)['"`]\s*\]/);
            const instanceMatch = handlerStr.match(/(\w+)\.(\w+)/);
            let handlerLabel;
            if (controllerMatch) {
                handlerLabel = `${controllerMatch[1]}@${controllerMatch[2]}`;
            }
            else if (instanceMatch && !handlerStr.includes('=>')) {
                handlerLabel = `${instanceMatch[1]}.${instanceMatch[2]}`;
            }
            else if (handlerStr.includes('=>')) {
                handlerLabel = '[Closure]';
            }
            else {
                handlerLabel = handlerStr.substring(0, 40);
            }
            const routeMiddleware = [...middleware];
            const withMwRegex = /\.withMiddleware\s*\(\s*(['"`][^'"`]+['"`]|\[[^\]]+\])\s*\)/;
            const withMwMatch = modifiersStr.match(withMwRegex);
            if (withMwMatch) {
                const mwStr = withMwMatch[1];
                if (mwStr.startsWith('[')) {
                    const mws = mwStr.match(/['"`]([^'"`]+)['"`]/g);
                    if (mws)
                        routeMiddleware.push(...mws.map(m => m.replace(/['"`]/g, '')));
                }
                else {
                    routeMiddleware.push(mwStr.replace(/['"`]/g, ''));
                }
            }
            let name;
            const asRegex = /\.as\s*\(\s*['"`]([^'"`]+)['"`]\s*\)/;
            const asMatch = modifiersStr.match(asRegex);
            if (asMatch)
                name = asMatch[1];
            const fullPath = this.joinPaths(prefix, routePath);
            const paramKeys = this.extractParamKeys(fullPath);
            routes.push({
                method,
                path: fullPath,
                handler: handlerLabel !== '[Closure]' && handlerLabel.length > 0 ? handlerLabel : undefined,
                name,
                middleware: routeMiddleware,
                prefix,
                paramKeys
            });
        }
    }
    joinPaths(prefix, path) {
        const cleanPrefix = prefix.replace(/^\/+|\/+$/g, '');
        const cleanPath = path.replace(/^\/+|\/+$/g, '');
        const sep = cleanPrefix && cleanPath ? '/' : '';
        return `/${cleanPrefix}${sep}${cleanPath}`;
    }
    extractParamKeys(pathStr) {
        const keys = [];
        const regex = /:([a-zA-Z0-9_]+)|\{([a-zA-Z0-9_]+)\}/g;
        let match;
        while ((match = regex.exec(pathStr)) !== null) {
            keys.push(match[1] || match[2]);
        }
        return keys;
    }
}
exports.RouteParser = RouteParser;
