// arikajs/src/http/Kernel.ts
import { Application } from '../Contracts/Application';
import { Request, Response, NotFoundHttpException } from '@arikajs/http';
import { Pipeline } from '@arikajs/middleware';
import { Dispatcher } from '@arikajs/dispatcher';
import { RequestLoggingMiddleware } from './Middleware/RequestLoggingMiddleware';
import { BodyParserMiddleware, CorsMiddleware, TrimStrings, ConvertEmptyStringsToNull, SecurityHeaders, Throttle } from '@arikajs/http';
import { Handler } from './Handler';
import { ViewMiddleware } from './Middleware/ViewMiddleware';
import { VerifyCsrfToken } from './Middleware/VerifyCsrfToken';
import { ServeStaticMiddleware } from './Middleware/ServeStaticMiddleware';
import { StartSession } from '@arikajs/session';
import { Authenticate, EnsureEmailIsVerified } from '@arikajs/auth';


export class Kernel {
    /**
     * The application's global HTTP middleware stack.
     */
    protected middleware: any[] = [
        new CorsMiddleware(),
        new SecurityHeaders(),
        new RequestLoggingMiddleware(),
        new BodyParserMiddleware(),
        new TrimStrings(),
        new ConvertEmptyStringsToNull(),
        new ServeStaticMiddleware(),
    ];

    /**
     * The application's route middleware groups.
     */
    protected middlewareGroups: Record<string, any[]> = {
        web: [
            StartSession,
            ViewMiddleware,
            VerifyCsrfToken,
        ],
        api: [],
    };

    /**
     * The application's route middleware.
     */
    protected routeMiddleware: Record<string, any> = {
        'auth': Authenticate,
        'verified': EnsureEmailIsVerified,
        'throttle': Throttle,
    };

    protected handler: Handler;

    constructor(protected app: Application) {
        try {
            this.handler = this.app.make(Handler);
        } catch (e) {
            this.handler = new Handler();
        }

        const router = this.app.getRouter();
        if ((router as any).setMiddlewareGroups) {
            (router as any).setMiddlewareGroups(this.middlewareGroups);
        }
        if ((router as any).setRouteMiddleware) {
            (router as any).setRouteMiddleware(this.routeMiddleware);
        }
    }

    /**
     * Handle an incoming HTTP request.
     */
    public async handle(request: Request, response: Response): Promise<Response> {
        try {
            const pipeline = new Pipeline<Request, Response>(this.app.getContainer());
            pipeline.setMiddlewareGroups(this.middlewareGroups);
            pipeline.setAliases(this.routeMiddleware);

            return await pipeline.pipe(this.middleware)
                .handle(request, async (req: Request) => {
                    return this.dispatchToRouter(req, response);
                }, response);
        } catch (error: any) {
            this.handler.report(error);
            return await this.handler.render(request, error, response);
        }
    }

    /**
     * Dispatch the request to the router.
     */
    protected async dispatchToRouter(request: Request, response: Response): Promise<Response> {
        const router = this.app.getRouter();

        const result = await router.dispatch(request, response);

        if (result === null) {
            throw new NotFoundHttpException(`Route not found: [${request.method()}] ${request.path()}`);
        }

        return result as Response;
    }

    /**
     * Send the response back to the client.
     */
    public terminate(request: Request, response: Response): void {
        response.terminate();
    }
}
