// @arikajs/dispatcher/dist/Dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dispatcher = void 0;
const ControllerResolver_1 = require("./ControllerResolver");
const MethodInvoker_1 = require("./MethodInvoker");
const ResponseResolver_1 = require("./ResponseResolver");
const middleware_1 = require("@arikajs/middleware");
class Dispatcher {
    constructor(container) {
        this.container = container;
        this.middlewareGroups = {};
        this.routeMiddleware = {};
        this.parameterBinders = new Map();
        this.executionPlans = new Map();
        this.invoker = new MethodInvoker_1.MethodInvoker(container);
        this.responseResolver = new ResponseResolver_1.ResponseResolver();
        if (container) {
            this.controllerResolver = new ControllerResolver_1.ControllerResolver(container);
        }
    }
    precompile(routes) {
        for (let i = 0; i < routes.length; i++) {
            const route = routes[i];
            const handler = route.handler;
            let resolvedHandler;
            let controllerMiddleware = [];
            if (Array.isArray(handler)) {
                if (this.controllerResolver) {
                    try {
                        const resolved = this.controllerResolver.resolve(handler);
                        resolvedHandler = resolved;
                        if (typeof resolved.controller.getMiddleware === 'function') {
                            controllerMiddleware = resolved.controller.getMiddleware() || [];
                        }
                        else if (resolved.controller.constructor && resolved.controller.constructor.middleware) {
                            controllerMiddleware = resolved.controller.constructor.middleware;
                        }
                    }
                    catch (e) {
                        // Skip if cannot resolve at boot (e.g. missing container)
                        continue;
                    }
                }
                else {
                    continue;
                }
            }
            else {
                resolvedHandler = handler;
            }
            const hasMiddleware = (route.middleware && route.middleware.length > 0) || (controllerMiddleware && controllerMiddleware.length > 0);
            this.executionPlans.set(route, {
                route,
                hasMiddleware,
                controllerMiddleware,
                resolvedHandler,
                handler: (req, res, params) => this.invoker.invoke(resolvedHandler, req, res, params),
                middleware: hasMiddleware ? [...(route.middleware || []), ...(controllerMiddleware || [])] : []
            });
            // Pre-serialize simple responses if possible (only if handler takes no arguments)
            if (!hasMiddleware && typeof resolvedHandler === 'function' && resolvedHandler.length === 0) {
                try {
                    const result = resolvedHandler();
                    if (result && typeof result === 'object' && !(result instanceof Promise)) {
                        this.responseResolver.bufferCache.set(route, Buffer.from(JSON.stringify(result)));
                    }
                }
                catch (e) { }
            }
        }
    }
    /**
     * Set the container for resolving controllers.
     */
    setContainer(container) {
        this.container = container;
        this.controllerResolver = new ControllerResolver_1.ControllerResolver(container);
        this.invoker.setContainer(container);
        return this;
    }
    /**
     * Set a global exception handler to catch and format errors from route execution.
     */
    setExceptionHandler(handler) {
        this.exceptionHandler = handler;
        return this;
    }
    /**
     * Set the middleware groups mapping.
     */
    setMiddlewareGroups(groups) {
        this.middlewareGroups = groups;
        return this;
    }
    /**
     * Set the route middleware mapping.
     */
    setRouteMiddleware(middleware) {
        this.routeMiddleware = middleware;
        return this;
    }
    /**
     * Register a route parameter binder.
     */
    bind(key, resolver) {
        if (resolver && typeof resolver.findOrFail === 'function') {
            this.parameterBinders.set(key, (value) => resolver.findOrFail(value));
        }
        else {
            this.parameterBinders.set(key, resolver);
        }
        return this;
    }
    /**
     * Dispatch the matched route to its handler.
     */
    async dispatch(matchedRoute, request, response) {
        const { route, params } = matchedRoute;
        const handler = route.handler;
        // Set parameters on the request object
        if (typeof request.setParams === 'function') {
            request.setParams(params);
        }
        // 0. Resolve Route Parameters (Model Binding) - only if we have binders
        const resolvedParams = this.parameterBinders.size > 0 ? await this.resolveParameters(params) : params;
        // 1. Resolve Handler
        let resolvedHandler;
        let controllerMiddleware = [];
        const plan = this.executionPlans.get(route);
        if (plan) {
            resolvedHandler = plan.resolvedHandler;
            controllerMiddleware = plan.controllerMiddleware;
        }
        else {
            if (Array.isArray(handler)) {
                if (!this.controllerResolver) {
                    throw new Error('Container required for controller resolution.');
                }
                const resolved = this.controllerResolver.resolve(handler);
                resolvedHandler = resolved;
                // Extract controller-level middleware
                if (typeof resolved.controller.getMiddleware === 'function') {
                    controllerMiddleware = resolved.controller.getMiddleware() || [];
                }
                else if (resolved.controller.constructor && resolved.controller.constructor.middleware) {
                    controllerMiddleware = resolved.controller.constructor.middleware;
                }
            }
            else if (typeof handler === 'function') {
                resolvedHandler = handler;
            }
            else {
                throw new Error('Invalid route handler.');
            }
        }
        // 2. Prepare Middleware Pipeline
        const pipeline = new middleware_1.Pipeline(this.container);
        if (plan && plan.hasMiddleware) {
            pipeline.pipe(plan.middleware);
        }
        else if (!plan) {
            // Fallback for non-plan routes
            if (route.middleware)
                pipeline.pipe(route.middleware);
            if (controllerMiddleware)
                pipeline.pipe(controllerMiddleware);
        }
        pipeline.setMiddlewareGroups(this.middlewareGroups);
        pipeline.setAliases(this.routeMiddleware);
        // 3. Execute Pipeline
        try {
            return await pipeline.handle(request, async (req) => {
                // 4. Invoke Handler
                const result = await this.invoker.invoke(resolvedHandler, req, response, resolvedParams);
                // 5. Normalize Response
                return await this.responseResolver.resolve(result, response);
            }, response);
        }
        catch (error) {
            if (this.exceptionHandler) {
                const handledResult = await this.exceptionHandler(error, request, response);
                return await this.responseResolver.resolve(handledResult, response);
            }
            throw error;
        }
    }
    /**
     * Resolve route parameters using registered binders.
     */
    async resolveParameters(params) {
        const resolved = { ...params };
        for (const [key, value] of Object.entries(params)) {
            if (this.parameterBinders.has(key)) {
                const resolver = this.parameterBinders.get(key);
                resolved[key] = await resolver(value);
            }
        }
        return resolved;
    }
}
exports.Dispatcher = Dispatcher;
