// @arikajs/scheduler/dist/Event.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Event = void 0;
const cron_parser_1 = __importDefault(require("cron-parser"));
class Event {
    command;
    cronExpression = '* * * * *';
    timezoneVal;
    withoutOverlappingVal = false;
    expiresAt = 1440; // 24 hours default for overlap lock
    successCallbacks = [];
    failureCallbacks = [];
    timeoutVal = 0;
    retriesVal = 0;
    retryDelayVal = 0;
    descriptionVal = '';
    constructor(command) {
        this.command = command;
    }
    name(description) {
        this.descriptionVal = description;
        return this;
    }
    timeout(seconds) {
        this.timeoutVal = seconds;
        return this;
    }
    retry(times, delay = 0) {
        this.retriesVal = times;
        this.retryDelayVal = delay;
        return this;
    }
    cron(expression) {
        this.cronExpression = expression;
        return this;
    }
    timezone(timezone) {
        this.timezoneVal = timezone;
        return this;
    }
    withoutOverlapping(expiresAt = 1440) {
        this.withoutOverlappingVal = true;
        this.expiresAt = expiresAt;
        return this;
    }
    onSuccess(callback) {
        this.successCallbacks.push(callback);
        return this;
    }
    onFailure(callback) {
        this.failureCallbacks.push(callback);
        return this;
    }
    // Frequencies
    everySecond() { return this.cron('* * * * * *'); }
    everyMinute() { return this.cron('* * * * *'); }
    everyFiveMinutes() { return this.cron('*/5 * * * *'); }
    everyTenMinutes() { return this.cron('*/10 * * * *'); }
    hourly() { return this.cron('0 * * * *'); }
    hourlyAt(minute) { return this.cron(`${minute} * * * *`); }
    daily() { return this.cron('0 0 * * *'); }
    dailyAt(time) {
        const [hour, minute] = time.split(':');
        return this.cron(`${minute} ${hour} * * *`);
    }
    weekly() { return this.cron('0 0 * * 0'); }
    weeklyOn(day, time = '00:00') {
        const [hour, minute] = time.split(':');
        return this.cron(`${minute} ${hour} * * ${day}`);
    }
    monthly() { return this.cron('0 0 1 * *'); }
    monthlyOn(day, time = '00:00') {
        const [hour, minute] = time.split(':');
        return this.cron(`${minute} ${hour} ${day} * *`);
    }
    expression() {
        return this.cronExpression;
    }
    isDue(date, timezone) {
        const options = {
            currentDate: new Date(date.getTime() - 1000),
            tz: this.timezoneVal || timezone
        };
        try {
            const interval = cron_parser_1.default.parseExpression(this.cronExpression, options);
            const next = interval.next().toDate();
            return next.getMinutes() === date.getMinutes() &&
                next.getHours() === date.getHours() &&
                next.getDate() === date.getDate() &&
                next.getMonth() === date.getMonth() &&
                next.getFullYear() === date.getFullYear();
        }
        catch (e) {
            return false;
        }
    }
    async run() {
        try {
            if (typeof this.command === 'function') {
                await this.command();
            }
            else {
                // For commands, we would usually use the CommandRegistry
                // But this will be handled by the Scheduler runner
            }
            for (const callback of this.successCallbacks) {
                await callback();
            }
        }
        catch (e) {
            for (const callback of this.failureCallbacks) {
                await callback(e);
            }
            throw e;
        }
    }
    shouldSkipOverlapping() {
        return this.withoutOverlappingVal;
    }
    mutexExpiration() {
        return this.expiresAt;
    }
    getTimeout() {
        return this.timeoutVal;
    }
    getRetries() {
        return this.retriesVal;
    }
    getRetryDelay() {
        return this.retryDelayVal;
    }
    getDescription() {
        return this.descriptionVal;
    }
}
exports.Event = Event;
