// @arikajs/scheduler/dist/Schedule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schedule = void 0;
const Event_1 = require("./Event");
class Schedule {
    events = [];
    /**
     * Add a new callback event to the schedule.
     */
    call(callback) {
        const event = new Event_1.Event(callback);
        this.events.push(event);
        return event;
    }
    /**
     * Add a new command event to the schedule.
     */
    command(command) {
        const event = new Event_1.Event(command);
        this.events.push(event);
        return event;
    }
    /**
     * Add a new job event to the schedule.
     */
    job(job) {
        const event = new Event_1.Event(async () => {
            const { Queue } = await import('@arikajs/queue');
            await Queue.dispatch(job);
        });
        this.events.push(event);
        return event;
    }
    /**
     * Get all events.
     */
    allEvents() {
        return this.events;
    }
    /**
     * Get due events.
     */
    dueEvents(date, timezone) {
        return this.events.filter(event => event.isDue(date, timezone));
    }
}
exports.Schedule = Schedule;
