// @arikajs/scheduler/dist/Scheduler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scheduler = void 0;
const logging_1 = require("@arikajs/logging");
const Schedule_1 = require("./Schedule");
class Scheduler {
    container;
    schedule;
    constructor(container) {
        this.container = container;
        this.schedule = new Schedule_1.Schedule();
    }
    /**
     * Define the schedule.
     */
    define(callback) {
        callback(this.schedule);
        return this;
    }
    /**
     * Run the scheduled tasks.
     */
    async run(date = new Date()) {
        const config = this.container.make('config');
        const timezone = config.get('app.timezone', 'UTC');
        const dueEvents = this.schedule.dueEvents(date, timezone);
        if (dueEvents.length === 0) {
            return;
        }
        // Leader Election (Prevent multiple instances from running the same schedule)
        if (await this.shouldSkipBecauseAnotherInstanceIsRunning()) {
            return;
        }
        logging_1.Log.info(`Running ${dueEvents.length} scheduled tasks in parallel...`);
        // Run events in parallel
        await Promise.allSettled(dueEvents.map(event => this.runEvent(event)));
    }
    async shouldSkipBecauseAnotherInstanceIsRunning() {
        if (!this.container.has('cache'))
            return false;
        const cache = this.container.make('cache');
        const lockKey = 'framework/scheduler-leader-lock';
        const isLeader = await cache.add(lockKey, true, 55); // Lock for 55 seconds
        if (!isLeader) {
            logging_1.Log.debug('Another instance is already running the scheduler. Skipping...');
            return true;
        }
        return false;
    }
    async runEvent(event) {
        const name = this.getEventName(event);
        try {
            await this.dispatchLifecycleEvent('TaskStarting', event);
            // Check for overlapping
            if (event.shouldSkipOverlapping()) {
                const locked = await this.isLocked(event);
                if (locked) {
                    logging_1.Log.debug(`Skipping task [${name}] as it is still running.`);
                    return;
                }
                await this.lock(event);
            }
            logging_1.Log.info(`Running scheduled task: [${name}]`);
            // Handle execution with Retries and Timeout
            await this.executeWithRetries(event);
            logging_1.Log.info(`Task [${name}] completed successfully.`);
            await this.dispatchLifecycleEvent('TaskFinished', event);
        }
        catch (e) {
            logging_1.Log.error(`Task [${name}] failed: ${e.message}`);
            await this.dispatchLifecycleEvent('TaskFailed', event, { error: e });
        }
        finally {
            if (event.shouldSkipOverlapping()) {
                await this.unlock(event);
            }
        }
    }
    async executeWithRetries(event) {
        const maxRetries = event.getRetries();
        let attempt = 0;
        while (attempt <= maxRetries) {
            try {
                await this.executeWithTimeout(event);
                return;
            }
            catch (e) {
                attempt++;
                if (attempt > maxRetries)
                    throw e;
                const delay = event.getRetryDelay();
                if (delay > 0) {
                    await new Promise(resolve => setTimeout(resolve, delay * 1000));
                }
                logging_1.Log.warning(`Retrying task [${this.getEventName(event)}] (Attempt ${attempt}/${maxRetries})...`);
            }
        }
    }
    async executeWithTimeout(event) {
        const timeoutSeconds = event.getTimeout();
        if (timeoutSeconds <= 0) {
            return this.actuallyRun(event);
        }
        return new Promise(async (resolve, reject) => {
            const timer = setTimeout(() => {
                reject(new Error(`Task timed out after ${timeoutSeconds} seconds.`));
            }, timeoutSeconds * 1000);
            try {
                await this.actuallyRun(event);
                clearTimeout(timer);
                resolve();
            }
            catch (e) {
                clearTimeout(timer);
                reject(e);
            }
        });
    }
    async actuallyRun(event) {
        if (typeof event.command === 'string') {
            const { CommandRegistry } = await import('@arikajs/console');
            const registry = this.container.make(CommandRegistry);
            await registry.run([event.command]);
        }
        else {
            await event.run();
        }
    }
    async dispatchLifecycleEvent(type, event, extra = {}) {
        // Only if @arikajs/events is available
        try {
            const { Event: Emitter } = await import('@arikajs/events');
            await Emitter.dispatch({
                type: `scheduler.${type}`,
                task: this.getEventName(event),
                expression: event.expression(),
                ...extra
            });
        }
        catch (e) {
            // Events package might not be installed, ignore silently
        }
    }
    getEventName(event) {
        return event.getDescription() || (typeof event.command === 'string' ? event.command : 'closure');
    }
    async isLocked(event) {
        if (!this.container.has('cache'))
            return false;
        const cache = this.container.make('cache');
        return await cache.has(this.getMutexName(event));
    }
    async lock(event) {
        if (!this.container.has('cache'))
            return;
        const cache = this.container.make('cache');
        await cache.put(this.getMutexName(event), true, event.mutexExpiration());
    }
    async unlock(event) {
        if (!this.container.has('cache'))
            return;
        const cache = this.container.make('cache');
        await cache.forget(this.getMutexName(event));
    }
    getMutexName(event) {
        return `framework/schedule-${Buffer.from(this.getEventName(event) + event.expression()).toString('base64')}`;
    }
}
exports.Scheduler = Scheduler;
