// @arikajs/storage/dist/src/StorageManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageManager = void 0;
const Disk_1 = require("./Disk");
const LocalDriver_1 = require("./Drivers/LocalDriver");
const S3Driver_1 = require("./Drivers/S3Driver");
const GCSDriver_1 = require("./Drivers/GCSDriver");
const AzureDriver_1 = require("./Drivers/AzureDriver");
class StorageManager {
    constructor(config) {
        this.disks = new Map();
        this.customCreators = new Map();
        this.config = config;
    }
    /**
     * Register a custom driver creator.
     */
    extend(driver, callback) {
        this.customCreators.set(driver, callback);
        return this;
    }
    /**
     * Get a disk instance.
     */
    disk(name) {
        name = name || this.config.default;
        if (!name) {
            throw new Error('No default disk configured.');
        }
        if (!this.disks.has(name)) {
            this.disks.set(name, this.resolveDisk(name));
        }
        return this.disks.get(name);
    }
    /**
     * Resolve a disk from configuration.
     */
    resolveDisk(name) {
        const diskConfig = this.config.disks[name];
        if (!diskConfig) {
            throw new Error(`Disk [${name}] is not configured.`);
        }
        const driver = this.createDriver(diskConfig);
        return new Disk_1.Disk(driver);
    }
    /**
     * Create a driver instance based on configuration.
     */
    createDriver(config) {
        if (this.customCreators.has(config.driver)) {
            return this.customCreators.get(config.driver)(config);
        }
        switch (config.driver) {
            case 'local':
                return new LocalDriver_1.LocalDriver(config);
            case 's3':
                return new S3Driver_1.S3Driver(config);
            case 'gcs':
                return new GCSDriver_1.GCSDriver(config);
            case 'azure':
                return new AzureDriver_1.AzureDriver(config);
            default:
                throw new Error(`Driver [${config.driver}] is not supported.`);
        }
    }
    /**
     * Proxy methods to the default disk.
     */
    async put(path, contents) {
        return await this.disk().put(path, contents);
    }
    async putStream(path, stream) {
        return await this.disk().putStream(path, stream);
    }
    async get(path) {
        return await this.disk().get(path);
    }
    readStream(path) {
        return this.disk().readStream(path);
    }
    async exists(path) {
        return await this.disk().exists(path);
    }
    async delete(path) {
        return await this.disk().delete(path);
    }
    url(path) {
        return this.disk().url(path);
    }
    async temporaryUrl(path, expiresAt) {
        return await this.disk().temporaryUrl(path, expiresAt);
    }
    async size(path) {
        return await this.disk().size(path);
    }
    async lastModified(path) {
        return await this.disk().lastModified(path);
    }
    async mimeType(path) {
        return await this.disk().mimeType(path);
    }
    async copy(path, newPath) {
        return await this.disk().copy(path, newPath);
    }
    async move(path, newPath) {
        return await this.disk().move(path, newPath);
    }
    async append(path, contents) {
        return await this.disk().append(path, contents);
    }
    async prepend(path, contents) {
        return await this.disk().prepend(path, contents);
    }
}
exports.StorageManager = StorageManager;
