// @arikajs/console/dist/Command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Command = void 0;
class Command {
    setInput(input) {
        this.input = input;
    }
    setOutput(output) {
        this.output = output;
    }
    setRegistry(registry) {
        this.registry = registry;
    }
    setContainer(container) {
        this.container = container;
    }
    /**
     * Optional initialization hook that runs before handle().
     */
    initialize() {
        // can be overridden
    }
    argument(key, defaultValue = null) {
        return this.input.argument(key, defaultValue);
    }
    option(key, defaultValue = null) {
        return this.input.option(key, defaultValue);
    }
    info(message) {
        this.output.info(message);
    }
    error(message) {
        this.output.error(message);
    }
    warning(message) {
        this.output.warning(message);
    }
    comment(message) {
        this.output.comment(message);
    }
    success(message) {
        this.output.success(message);
    }
    table(headers, rows) {
        this.output.table(headers, rows);
    }
    async confirm(question, defaultAction = true) {
        return await this.output.confirm(question, defaultAction);
    }
    async ask(question, defaultValue = null) {
        return await this.output.ask(question, defaultValue);
    }
    progressStart(total) {
        this.output.progressStart(total);
    }
    progressAdvance(step = 1) {
        this.output.progressAdvance(step);
    }
    progressFinish() {
        this.output.progressFinish();
    }
    async task(message, task) {
        return await this.output.task(message, task);
    }
    write(message) {
        this.output.write(message);
    }
    writeln(message) {
        this.output.writeln(message);
    }
}
exports.Command = Command;
