// @arikajs/console/dist/CommandRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandRegistry = void 0;
const Parser_1 = require("./Parser");
const Input_1 = require("./Input");
const Output_1 = require("./Output");
class CommandRegistry {
    constructor(resolver) {
        this.commands = new Map();
        this.resolver = resolver || ((cls) => new cls());
    }
    setContainer(container) {
        this.container = container;
    }
    /**
     * Register a command class.
     */
    register(command) {
        const instance = this.resolver(command);
        const { name } = Parser_1.Parser.parseSignature(instance.signature);
        this.commands.set(name, {
            class: command,
            signature: instance.signature,
            description: instance.description
        });
    }
    /**
     * Register a command lazily.
     */
    registerLazy(signature, description, importer) {
        const { name } = Parser_1.Parser.parseSignature(signature);
        this.commands.set(name, {
            isLazy: true,
            signature,
            description,
            importer
        });
    }
    async run(rawArgs) {
        const commandName = rawArgs[0];
        const commandData = this.commands.get(commandName);
        if (!commandData) {
            throw new Error(`Command "${commandName}" not found.`);
        }
        let commandClass = commandData.class;
        if (commandData.isLazy) {
            const module = await commandData.importer();
            // Try to find the class in the module (default export or first exported class)
            commandClass = module.default || Object.values(module).find(v => typeof v === 'function');
        }
        if (!commandClass) {
            throw new Error(`Could not resolve command class for "${commandName}".`);
        }
        const command = this.resolver(commandClass);
        const signature = Parser_1.Parser.parseSignature(command.signature);
        const { arguments: args, options } = Parser_1.Parser.parseArguments(rawArgs.slice(1), signature);
        const input = new Input_1.Input(args, options);
        const output = new Output_1.Output();
        command.setInput(input);
        command.setOutput(output);
        command.setRegistry(this);
        if (this.container) {
            command.setContainer(this.container);
        }
        await command.initialize();
        return await command.handle();
    }
    all() {
        return this.commands;
    }
}
exports.CommandRegistry = CommandRegistry;
