// @arikajs/console/dist/Output.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Output = void 0;
const readline_1 = __importDefault(require("readline"));
class Output {
    constructor() {
        this.colors = {
            reset: '\x1b[0m',
            red: '\x1b[31m',
            green: '\x1b[32m',
            yellow: '\x1b[33m',
            blue: '\x1b[34m',
            magenta: '\x1b[35m',
            cyan: '\x1b[36m',
            white: '\x1b[37m',
            gray: '\x1b[90m',
            bold: '\x1b[1m'
        };
        /**
         * Progress bar support.
         */
        this.progressTotal = 0;
        this.progressCurrent = 0;
    }
    write(message) {
        process.stdout.write(message);
    }
    writeln(message) {
        process.stdout.write(message + '\n');
    }
    info(message) {
        this.writeln(`${this.colors.green}${message}${this.colors.reset}`);
    }
    error(message) {
        process.stderr.write(`${this.colors.red}${message}${this.colors.reset}\n`);
    }
    warning(message) {
        this.writeln(`${this.colors.yellow}${message}${this.colors.reset}`);
    }
    comment(message) {
        this.writeln(`${this.colors.gray}${message}${this.colors.reset}`);
    }
    success(message) {
        this.writeln(`${this.colors.green}✔ ${this.colors.reset}${message}`);
    }
    /**
     * Display a table.
     */
    table(headers, rows) {
        if (rows.length === 0)
            return;
        // Calculate column widths
        const colWidths = headers.map((header, i) => {
            const rowWidths = rows.map(row => String(row[i]).length);
            return Math.max(header.length, ...rowWidths);
        });
        const horizontalLine = ' +' + colWidths.map(w => '-'.repeat(w + 2)).join('+') + '+';
        // Print header
        this.writeln(horizontalLine);
        this.writeln(' | ' + headers.map((h, i) => h.padEnd(colWidths[i])).join(' | ') + ' |');
        this.writeln(horizontalLine);
        // Print rows
        rows.forEach(row => {
            this.writeln(' | ' + row.map((cell, i) => String(cell).padEnd(colWidths[i])).join(' | ') + ' |');
        });
        this.writeln(horizontalLine);
    }
    /**
     * Interactive Confirmation.
     */
    async confirm(question, defaultAction = true) {
        const rl = readline_1.default.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        const suffix = defaultAction ? '[Y/n]' : '[y/N]';
        return new Promise(resolve => {
            rl.question(`${this.colors.bold}${question}${this.colors.reset} ${this.colors.gray}${suffix}${this.colors.reset} `, (answer) => {
                rl.close();
                if (!answer)
                    return resolve(defaultAction);
                resolve(answer.toLowerCase().startsWith('y'));
            });
        });
    }
    /**
     * Ask for input.
     */
    async ask(question, defaultValue = null) {
        const rl = readline_1.default.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        const suffix = defaultValue ? ` (${defaultValue})` : '';
        return new Promise(resolve => {
            rl.question(`${this.colors.bold}${question}${this.colors.reset}${suffix}: `, (answer) => {
                rl.close();
                resolve(answer || defaultValue);
            });
        });
    }
    progressStart(total) {
        this.progressTotal = total;
        this.progressCurrent = 0;
        this.renderProgressBar();
    }
    progressAdvance(step = 1) {
        this.progressCurrent += step;
        this.renderProgressBar();
    }
    progressFinish() {
        this.progressCurrent = this.progressTotal;
        this.renderProgressBar();
        this.writeln('');
    }
    renderProgressBar() {
        const width = 40;
        const progress = Math.min(1, this.progressCurrent / this.progressTotal);
        const filledWidth = Math.round(width * progress);
        const emptyWidth = width - filledWidth;
        const bar = '█'.repeat(filledWidth) + '░'.repeat(emptyWidth);
        const percentage = Math.round(progress * 100);
        readline_1.default.cursorTo(process.stdout, 0);
        process.stdout.write(` ${bar} ${percentage}% (${this.progressCurrent}/${this.progressTotal})`);
    }
    /**
     * Run a task with a status indicator.
     */
    async task(message, task) {
        this.write(`${message} ... `);
        try {
            const result = await task();
            readline_1.default.cursorTo(process.stdout, message.length + 5);
            this.writeln(`${this.colors.green}DONE${this.colors.reset}`);
            return result;
        }
        catch (error) {
            readline_1.default.cursorTo(process.stdout, message.length + 5);
            this.writeln(`${this.colors.red}FAIL${this.colors.reset}`);
            this.error(error.message);
            throw error;
        }
    }
}
exports.Output = Output;
