// @arikajs/auth/dist/src/Guards/JwtGuard.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.JwtGuard = void 0;
const jwt = __importStar(require("jsonwebtoken"));
const crypto = __importStar(require("crypto"));
class JwtGuard {
    constructor(provider, request, secret, options = {}) {
        this.loggedUser = null;
        this.provider = provider;
        this.request = request;
        this.secret = secret;
        this.options = options;
    }
    async check() {
        return !!(await this.user());
    }
    async guest() {
        return !(await this.check());
    }
    async user() {
        if (this.loggedUser)
            return this.loggedUser;
        const token = this.getTokenForRequest();
        if (!token)
            return null;
        try {
            const decoded = jwt.verify(token, this.secret, this.options);
            if (decoded && decoded.sub) {
                // If the provider supports retrieving by ID, use it for stateless verification payload validation
                // In a purely stateless app, you might just return the decoded payload or a Model proxy instead
                // of querying the DB. But to ensure user is active/exists, we retrieve it:
                this.loggedUser = await this.provider.retrieveById(decoded.sub);
                return this.loggedUser;
            }
        }
        catch (e) {
            return null; // Invalid token
        }
        return null;
    }
    async id() {
        if (this.loggedUser)
            return this.loggedUser.id;
        const token = this.getTokenForRequest();
        if (!token)
            return null;
        try {
            const decoded = jwt.verify(token, this.secret, this.options);
            return decoded ? decoded.sub : null;
        }
        catch (e) {
            return null;
        }
    }
    async validate(credentials) {
        const user = await this.provider.retrieveByCredentials(credentials);
        if (!user)
            return false;
        return await this.provider.validateCredentials(user, credentials);
    }
    /**
     * Authenticate a user and return a JWT token and refresh token
     */
    async attempt(credentials) {
        const user = await this.provider.retrieveByCredentials(credentials);
        if (user && await this.provider.validateCredentials(user, credentials)) {
            this.login(user);
            return await this.issueTokens(user);
        }
        return false;
    }
    async issueTokens(user, additionalPayload = {}) {
        const payload = { sub: user.id, ...additionalPayload };
        const access_token = jwt.sign(payload, this.secret, this.options);
        const response = { access_token };
        if (this.provider.updateRefreshToken) {
            const refresh_token = crypto.randomBytes(40).toString('hex');
            await this.provider.updateRefreshToken(user, refresh_token);
            response.refresh_token = refresh_token;
        }
        return response;
    }
    async refresh(refreshToken) {
        if (!this.provider.retrieveByRefreshToken) {
            throw new Error('UserProvider does not support retrieveByRefreshToken');
        }
        const user = await this.provider.retrieveByRefreshToken(refreshToken);
        if (!user) {
            throw new Error('Invalid or expired refresh token');
        }
        return await this.issueTokens(user);
    }
    login(user) {
        this.setUser(user);
    }
    logout() {
        this.loggedUser = null;
    }
    setUser(user) {
        this.loggedUser = user;
        if (this.request) {
            this.request.user = user;
        }
    }
    setRequest(request) {
        this.request = request;
    }
    getTokenForRequest() {
        let authHeader;
        // Support ArikaJS Request with header() method
        if (typeof this.request?.header === 'function') {
            authHeader = this.request.header('authorization');
        }
        // Fallback to raw headers object
        if (!authHeader && this.request?.headers?.['authorization']) {
            authHeader = this.request.headers['authorization'];
        }
        if (authHeader && authHeader.startsWith('Bearer ')) {
            return authHeader.substring(7);
        }
        return null;
    }
}
exports.JwtGuard = JwtGuard;
