// @arikajs/auth/dist/Guards/BasicGuard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicGuard = void 0;
class BasicGuard {
    constructor(provider, request) {
        this.loggedUser = null;
        this.provider = provider;
        this.request = request;
    }
    async check() {
        return !!(await this.user());
    }
    async guest() {
        return !(await this.check());
    }
    async user() {
        if (this.loggedUser)
            return this.loggedUser;
        const credentials = this.getCredentialsFromRequest();
        if (!credentials)
            return null;
        const user = await this.provider.retrieveByCredentials(credentials);
        if (user && await this.provider.validateCredentials(user, credentials)) {
            this.loggedUser = user;
            return user;
        }
        return null;
    }
    async id() {
        const user = await this.user();
        return user ? user.id : null;
    }
    async validate(credentials) {
        const user = await this.provider.retrieveByCredentials(credentials);
        if (!user)
            return false;
        return await this.provider.validateCredentials(user, credentials);
    }
    setRequest(request) {
        this.request = request;
    }
    setUser(user) {
        this.loggedUser = user;
    }
    getCredentialsFromRequest() {
        if (this.request?.headers?.['authorization']) {
            const authHeader = this.request.headers['authorization'];
            if (authHeader.startsWith('Basic ')) {
                const b64 = authHeader.substring(6);
                const decoded = Buffer.from(b64, 'base64').toString('ascii');
                const [email, password] = decoded.split(':');
                return { email, password };
            }
        }
        return null;
    }
}
exports.BasicGuard = BasicGuard;
