// @arikajs/auth/dist/Guards/SessionGuard.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionGuard = void 0;
const crypto = __importStar(require("crypto"));
class SessionGuard {
    constructor(provider, session) {
        this.loggedUser = null;
        this.provider = provider;
        // Accept any session-like object (sync or async API).
        // The @arikajs/session Session class has async get/put;
        // the legacy fallback below is sync — both work because we await all calls.
        if (session && (typeof session.get === 'function' || typeof session.put === 'function')) {
            this.session = session;
        }
        else {
            const store = {};
            this.session = {
                get(key) { return store[key] ?? null; },
                put(key, value) { store[key] = value; },
                forget(key) { delete store[key]; },
            };
        }
    }
    async check() {
        return !!(await this.user());
    }
    async guest() {
        return !(await this.check());
    }
    async user() {
        if (this.loggedUser) {
            return this.loggedUser;
        }
        // Await to support both sync (legacy) and async (Session class) APIs
        const id = this.session.get ? await Promise.resolve(this.session.get('auth_user_id')) : null;
        if (id) {
            this.loggedUser = await this.provider.retrieveById(id);
        }
        else {
            // Check for remember me cookie
            const rememberTokenString = this.getRememberCookie();
            if (rememberTokenString) {
                const [userId, token] = rememberTokenString.split('|');
                if (userId && token && this.provider.retrieveByToken) {
                    const user = await this.provider.retrieveByToken(userId, token);
                    if (user) {
                        this.login(user, true); // re-authenticate
                        this.loggedUser = user;
                    }
                }
            }
        }
        return this.loggedUser;
    }
    async id() {
        const user = await this.user();
        return user ? user.id : null;
    }
    async validate(credentials) {
        const user = await this.provider.retrieveByCredentials(credentials);
        if (!user) {
            return false;
        }
        return await this.provider.validateCredentials(user, credentials);
    }
    async attempt(credentials, remember = false) {
        if (await this.validate(credentials)) {
            const user = await this.provider.retrieveByCredentials(credentials);
            await this.login(user, remember);
            return true;
        }
        return false;
    }
    async login(user, remember = false) {
        this.loggedUser = user;
        if (this.session.put) {
            // Await to support both sync and async session APIs
            await Promise.resolve(this.session.put('auth_user_id', user.id));
        }
        if (remember) {
            const token = crypto.randomBytes(32).toString('hex');
            if (this.provider.updateRememberToken) {
                await this.provider.updateRememberToken(user, token);
            }
            this.queueRememberCookie(user.id, token);
        }
    }
    async logout() {
        const userId = this.loggedUser?.id;
        this.loggedUser = null;
        if (this.session.forget) {
            await Promise.resolve(this.session.forget('auth_user_id'));
        }
        this.clearRememberCookie();
        if (userId && this.provider.updateRememberToken) {
            // Invalidate token in provider asynchronously 
            this.provider.updateRememberToken({ id: userId }, null).catch(() => { });
        }
    }
    setUser(user) {
        this.loggedUser = user;
    }
    setRequest(request) {
        this.request = request;
        if (request && request.session) {
            this.session = request.session;
        }
    }
    getRememberCookie() {
        if (this.request?.cookies && typeof this.request.cookies === 'function') {
            return this.request.cookies('remember_web');
        }
        return this.request?.cookies?.['remember_web'] || null;
    }
    queueRememberCookie(userId, token) {
        const val = `${userId}|${token}`;
        if (this.request?.cookie && typeof this.request.cookie === 'function') {
            // Expires in 5 years essentially "forever" in internet time
            this.request.cookie('remember_web', val, { maxAge: 5 * 365 * 24 * 60 * 60 * 1000, httpOnly: true });
        }
    }
    clearRememberCookie() {
        if (this.request?.clearCookie && typeof this.request.clearCookie === 'function') {
            this.request.clearCookie('remember_web');
        }
    }
}
exports.SessionGuard = SessionGuard;
