// @arikajs/auth/dist/Guards/TokenGuard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenGuard = void 0;
class TokenGuard {
    constructor(provider, request, inputKey = 'api_token', storageKey = 'api_token') {
        this.storageKey = 'api_token';
        this.inputKey = 'api_token';
        this.provider = provider;
        this.request = request;
        this.inputKey = inputKey;
        this.storageKey = storageKey;
    }
    async user() {
        if (!this.request)
            return null;
        let token = this.getTokenForRequest();
        if (!token)
            return null;
        return await this.provider.retrieveByCredentials({ [this.storageKey]: token });
    }
    async check() {
        return !!(await this.user());
    }
    async guest() {
        return !(await this.check());
    }
    async id() {
        const user = await this.user();
        return user ? user.id : null;
    }
    async validate(credentials) {
        return !!(await this.provider.retrieveByCredentials(credentials));
    }
    setUser(user) {
        if (this.request) {
            this.request.user = user;
        }
    }
    setRequest(request) {
        this.request = request;
    }
    getTokenForRequest() {
        // Simple implementation: check query param, input body, or Bearer token
        if (this.request.query && this.request.query[this.inputKey]) {
            return this.request.query[this.inputKey];
        }
        if (this.request.body && this.request.body[this.inputKey]) {
            return this.request.body[this.inputKey];
        }
        if (this.request.headers && this.request.headers['authorization']) {
            const authHeader = this.request.headers['authorization'];
            if (authHeader.startsWith('Bearer ')) {
                return authHeader.substring(7);
            }
        }
        return null;
    }
}
exports.TokenGuard = TokenGuard;
