// @arikajs/database/dist/Transactions/TransactionManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionManager = void 0;
/**
 * Manages database transactions with automatic rollback on failure
 */
class TransactionManager {
    constructor(connection) {
        this.depth = 0;
        this.connection = connection;
    }
    /**
     * Run a callback within a transaction.
     * Automatically commits on success and rolls back on failure.
     * Supports nested transactions via savepoints.
     */
    async transaction(callback) {
        if (this.depth === 0) {
            await this.connection.beginTransaction();
        }
        else {
            // Nested transaction: use savepoints
            await this.connection.query(`SAVEPOINT sp_level${this.depth}`);
        }
        this.depth++;
        try {
            const result = await callback(this.connection);
            this.depth--;
            if (this.depth === 0) {
                await this.connection.commit();
            }
            else {
                await this.connection.query(`RELEASE SAVEPOINT sp_level${this.depth}`);
            }
            return result;
        }
        catch (error) {
            this.depth--;
            if (this.depth === 0) {
                await this.connection.rollback();
            }
            else {
                await this.connection.query(`ROLLBACK TO SAVEPOINT sp_level${this.depth}`);
            }
            throw error;
        }
    }
    /**
     * Begin a transaction manually
     */
    async begin() {
        await this.connection.beginTransaction();
    }
    /**
     * Commit the current transaction
     */
    async commit() {
        await this.connection.commit();
    }
    /**
     * Rollback the current transaction
     */
    async rollback() {
        await this.connection.rollback();
    }
}
exports.TransactionManager = TransactionManager;
