// @arikajs/authorization/dist/tests/Authorization.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = require("node:test");
const assert = __importStar(require("node:assert"));
const src_1 = require("../src/index");
class Post {
    constructor(id, userId, title) {
        this.id = id;
        this.userId = userId;
        this.title = title;
    }
}
class PostPolicy {
    before(user, ability) {
        if (user.isSuperAdmin)
            return true; // Super admin bypass
        return null; // Continue to normal check
    }
    view(user, post) {
        return true;
    }
    update(user, post) {
        return user.id === post.userId;
    }
    delete(user, post) {
        if (user.id !== post.userId) {
            return src_1.AuthResponse.deny('You do not own this post.', 'POST_NOT_OWNED');
        }
        return src_1.AuthResponse.allow();
    }
}
(0, node_test_1.describe)('Arika Authorization', () => {
    const user = { id: 1, name: 'John', isAdmin: false, isSuperAdmin: false, roles: ['editor'], permissions: ['view-posts'] };
    const adminUser = { id: 2, name: 'Admin', isAdmin: true, isSuperAdmin: false, roles: ['admin', 'editor'], permissions: ['view-posts', 'edit-posts', 'delete-posts'] };
    const superAdmin = { id: 3, name: 'SuperAdmin', isSuperAdmin: true, roles: ['super-admin'], permissions: [] };
    const post = new Post(1, 1, 'Test Post');
    const otherPost = new Post(2, 2, 'Other Post');
    (0, node_test_1.beforeEach)(() => {
        src_1.Gate.reset();
    });
    (0, node_test_1.afterEach)(() => {
        src_1.Gate.reset();
    });
    // ── Basic Gate Tests ────────────────────────────────────────────
    (0, node_test_1.it)('defines and checks gates', async () => {
        src_1.Gate.define('edit-post', (user, post) => user.id === post.userId);
        const canEdit = await src_1.Gate.forUser(user).allows('edit-post', post);
        assert.strictEqual(canEdit, true);
        const cannotEdit = await src_1.Gate.forUser(user).allows('edit-post', otherPost);
        assert.strictEqual(cannotEdit, false);
    });
    (0, node_test_1.it)('denies unauthorized actions', async () => {
        src_1.Gate.define('delete-post', (user) => user.isAdmin);
        const denied = await src_1.Gate.forUser(user).denies('delete-post', post);
        assert.strictEqual(denied, true);
        const allowed = await src_1.Gate.forUser(adminUser).denies('delete-post', post);
        assert.strictEqual(allowed, false);
    });
    // ── Policy Tests ────────────────────────────────────────────────
    (0, node_test_1.it)('works with policies', async () => {
        src_1.Gate.policy(Post, PostPolicy);
        const canView = await src_1.Gate.forUser(user).allows('view', post);
        assert.strictEqual(canView, true);
        const canUpdate = await src_1.Gate.forUser(user).allows('update', post);
        assert.strictEqual(canUpdate, true);
        const cannotUpdate = await src_1.Gate.forUser(user).allows('update', otherPost);
        assert.strictEqual(cannotUpdate, false);
    });
    // ── Feature 2: before() / after() Hooks ─────────────────────────
    (0, node_test_1.it)('runs before() hook on policy for super admin bypass', async () => {
        src_1.Gate.policy(Post, PostPolicy);
        // SuperAdmin can update ANY post via before() returning true
        const canUpdate = await src_1.Gate.forUser(superAdmin).allows('update', otherPost);
        assert.strictEqual(canUpdate, true);
        // Regular user cannot
        const cannotUpdate = await src_1.Gate.forUser(user).allows('update', otherPost);
        assert.strictEqual(cannotUpdate, false);
    });
    (0, node_test_1.it)('runs global before() hook', async () => {
        src_1.Gate.before((user) => {
            if (user.isSuperAdmin)
                return true;
            return null;
        });
        src_1.Gate.define('admin-only', (user) => user.isAdmin);
        // Super admin bypasses everything
        const allowed = await src_1.Gate.forUser(superAdmin).allows('admin-only');
        assert.strictEqual(allowed, true);
        // Regular user denied
        const denied = await src_1.Gate.forUser(user).allows('admin-only');
        assert.strictEqual(denied, false);
    });
    (0, node_test_1.it)('runs after() hook for auditing', async () => {
        let auditLog = [];
        src_1.Gate.after((user, ability, result) => {
            auditLog.push({ ability, result });
        });
        src_1.Gate.define('edit-post', (user, post) => user.id === post.userId);
        await src_1.Gate.forUser(user).allows('edit-post', post);
        await src_1.Gate.forUser(user).allows('edit-post', otherPost);
        assert.strictEqual(auditLog.length, 2);
        assert.strictEqual(auditLog[0].result, true);
        assert.strictEqual(auditLog[1].result, false);
    });
    // ── Feature 3: Bulk Checks ──────────────────────────────────────
    (0, node_test_1.it)('Gate.any() checks if user has any of the abilities', async () => {
        src_1.Gate.define('edit-post', (user, post) => user.id === post.userId);
        src_1.Gate.define('delete-post', (user) => user.isAdmin);
        const hasAny = await src_1.Gate.forUser(user).any(['edit-post', 'delete-post'], post);
        assert.strictEqual(hasAny, true); // can edit-post
        const hasNone = await src_1.Gate.forUser({ id: 999, isAdmin: false }).any(['edit-post', 'delete-post'], post);
        assert.strictEqual(hasNone, false);
    });
    (0, node_test_1.it)('Gate.every() checks if user has all abilities', async () => {
        src_1.Gate.define('edit-post', (user, post) => user.id === post.userId);
        src_1.Gate.define('publish-post', () => true);
        const hasAll = await src_1.Gate.forUser(user).every(['edit-post', 'publish-post'], post);
        assert.strictEqual(hasAll, true);
        src_1.Gate.define('admin-only', (user) => user.isAdmin);
        const notAll = await src_1.Gate.forUser(user).every(['edit-post', 'admin-only'], post);
        assert.strictEqual(notAll, false);
    });
    (0, node_test_1.it)('Gate.none() checks that user has none of the abilities', async () => {
        src_1.Gate.define('admin-only', (user) => user.isAdmin);
        src_1.Gate.define('super-admin-only', (user) => user.isSuperAdmin);
        const noneForUser = await src_1.Gate.forUser(user).none(['admin-only', 'super-admin-only']);
        assert.strictEqual(noneForUser, true);
        const noneForAdmin = await src_1.Gate.forUser(adminUser).none(['admin-only', 'super-admin-only']);
        assert.strictEqual(noneForAdmin, false);
    });
    // ── Feature 4: Response-Based Authorization ─────────────────────
    (0, node_test_1.it)('returns custom deny message via AuthResponse', async () => {
        src_1.Gate.policy(Post, PostPolicy);
        const response = await src_1.Gate.forUser(user).inspect('delete', otherPost);
        assert.strictEqual(response.denied(), true);
        assert.strictEqual(response.message(), 'You do not own this post.');
        assert.strictEqual(response.code(), 'POST_NOT_OWNED');
        const allowResponse = await src_1.Gate.forUser(user).inspect('delete', post);
        assert.strictEqual(allowResponse.allowed(), true);
    });
    (0, node_test_1.it)('throws exception with custom message from AuthResponse', async () => {
        src_1.Gate.policy(Post, PostPolicy);
        await assert.rejects(async () => await src_1.Gate.forUser(user).authorize('delete', otherPost), (err) => {
            assert.strictEqual(err.message, 'You do not own this post.');
            assert.strictEqual(err.code, 'POST_NOT_OWNED');
            return true;
        });
    });
    // ── Feature 5: Request-Scoped Context ───────────────────────────
    (0, node_test_1.it)('AuthorizationContext provides isolated per-request checks', async () => {
        src_1.Gate.define('edit-post', (user, post) => user.id === post.userId);
        const ctx1 = new src_1.AuthorizationContext(user);
        const ctx2 = new src_1.AuthorizationContext(adminUser);
        const can1 = await ctx1.can('edit-post', post);
        assert.strictEqual(can1, true);
        const can2 = await ctx2.can('edit-post', post);
        assert.strictEqual(can2, false); // adminUser.id !== post.userId
    });
    (0, node_test_1.it)('AuthorizationContext caches results within a request', async () => {
        let callCount = 0;
        src_1.Gate.define('counted-gate', (user) => {
            callCount++;
            return true;
        });
        const ctx = new src_1.AuthorizationContext(user);
        await ctx.can('counted-gate');
        await ctx.can('counted-gate'); // should be cached
        assert.strictEqual(callCount, 1); // Only called once
    });
    // ── Feature 1: Role & Permission ────────────────────────────────
    (0, node_test_1.it)('checks roles via RolePermissionMixin', () => {
        assert.strictEqual(src_1.RolePermissionMixin.hasRole(user, 'editor'), true);
        assert.strictEqual(src_1.RolePermissionMixin.hasRole(user, 'admin'), false);
        assert.strictEqual(src_1.RolePermissionMixin.hasAnyRole(adminUser, ['admin', 'super-admin']), true);
        assert.strictEqual(src_1.RolePermissionMixin.hasAllRoles(adminUser, ['admin', 'editor']), true);
        assert.strictEqual(src_1.RolePermissionMixin.hasAllRoles(adminUser, ['admin', 'super-admin']), false);
    });
    (0, node_test_1.it)('checks permissions via RolePermissionMixin', () => {
        assert.strictEqual(src_1.RolePermissionMixin.hasPermission(user, 'view-posts'), true);
        assert.strictEqual(src_1.RolePermissionMixin.hasPermission(user, 'delete-posts'), false);
        assert.strictEqual(src_1.RolePermissionMixin.hasAnyPermission(adminUser, ['edit-posts', 'nuke-server']), true);
        assert.strictEqual(src_1.RolePermissionMixin.hasAllPermissions(adminUser, ['view-posts', 'edit-posts', 'delete-posts']), true);
    });
    (0, node_test_1.it)('checks roles/permissions via AuthorizationContext', () => {
        const ctx = new src_1.AuthorizationContext(adminUser);
        assert.strictEqual(ctx.hasRole('admin'), true);
        assert.strictEqual(ctx.hasAnyRole(['admin', 'super-admin']), true);
        assert.strictEqual(ctx.hasPermission('edit-posts'), true);
        assert.strictEqual(ctx.hasAllPermissions(['view-posts', 'edit-posts']), true);
    });
    // ── AuthorizationManager ────────────────────────────────────────
    (0, node_test_1.it)('AuthorizationManager delegates to context', async () => {
        src_1.Gate.define('edit-post', (user, post) => user.id === post.userId);
        const authz = new src_1.AuthorizationManager(user);
        const can = await authz.can('edit-post', post);
        assert.strictEqual(can, true);
        const cannot = await authz.cannot('edit-post', otherPost);
        assert.strictEqual(cannot, true);
        assert.strictEqual(authz.hasRole('editor'), true);
        assert.strictEqual(authz.hasPermission('view-posts'), true);
    });
});
