// @arikajs/events/dist/EventManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventManager = void 0;
const Dispatcher_1 = require("./Dispatcher");
const ListenerResolver_1 = require("./ListenerResolver");
class EventManager {
    constructor(resolver) {
        this.listeners = new Map();
        this.wildcards = new Map();
        this.dispatched = [];
        this.isFaking = false;
        this.resolver = resolver || new ListenerResolver_1.ListenerResolver();
        this.dispatcher = new Dispatcher_1.Dispatcher((ln) => this.resolver.resolve(ln));
    }
    listen(event, listener, priority = 0) {
        if (typeof event === 'string' && event.includes('*')) {
            this.setupWildcard(event, listener, priority);
        }
        else {
            if (!this.listeners.has(event)) {
                this.listeners.set(event, []);
            }
            this.listeners.get(event)?.push({ listener, priority });
        }
        return this;
    }
    setupWildcard(event, listener, priority) {
        if (!this.wildcards.has(event)) {
            this.wildcards.set(event, []);
        }
        this.wildcards.get(event)?.push({ listener, priority });
    }
    subscribe(subscriber) {
        const subscriberInstance = typeof subscriber === 'function' ? new subscriber() : subscriber;
        if (typeof subscriberInstance.subscribe === 'function') {
            subscriberInstance.subscribe(this);
        }
    }
    forget(event) {
        this.listeners.delete(event);
        this.wildcards.delete(event);
        return this;
    }
    async dispatch(event) {
        if (this.isFaking) {
            this.dispatched.push(event);
            return;
        }
        const listeners = this.getSortedListeners(event);
        if (listeners.length > 0) {
            await this.dispatcher.dispatch(event, listeners);
        }
    }
    getSortedListeners(event) {
        let listeners = [];
        // 1. Get exact listeners
        const eventKey = typeof event === 'object' ? event.constructor : event;
        if (this.listeners.has(eventKey)) {
            listeners = [...this.listeners.get(eventKey)];
        }
        // 2. Get wildcard listeners
        if (typeof event === 'string' || (typeof event === 'object' && event.constructor.name)) {
            const eventName = typeof event === 'string' ? event : event.constructor.name;
            for (const [pattern, wildcardListeners] of this.wildcards.entries()) {
                if (this.matchWildcard(pattern, eventName)) {
                    listeners = listeners.concat(wildcardListeners);
                }
            }
        }
        // 3. Sort by priority (descending)
        return listeners
            .sort((a, b) => b.priority - a.priority)
            .map(l => l.listener);
    }
    matchWildcard(pattern, eventName) {
        const regex = new RegExp('^' + pattern.replace(/\./g, '\\.').replace(/\*/g, '.*') + '$');
        return regex.test(eventName);
    }
    // Testing helpers
    fake() {
        this.isFaking = true;
    }
    assertDispatched(event, callback) {
        const found = this.dispatched.some(e => {
            const matches = e instanceof event || e.constructor === event || e === event;
            if (matches && callback) {
                return callback(e);
            }
            return matches;
        });
        if (!found) {
            throw new Error(`Event [${event.name || event}] was not dispatched.`);
        }
    }
}
exports.EventManager = EventManager;
