// @arikajs/events/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventException = exports.EventManager = exports.Event = void 0;
const EventManager_1 = require("./EventManager");
Object.defineProperty(exports, "EventManager", { enumerable: true, get: function () { return EventManager_1.EventManager; } });
const EventException_1 = require("./Exceptions/EventException");
Object.defineProperty(exports, "EventException", { enumerable: true, get: function () { return EventException_1.EventException; } });
class Event {
    static dispatch(event) {
        return this.instance.dispatch(event);
    }
    static listen(event, listener, priority = 0) {
        return this.instance.listen(event, listener, priority);
    }
    static forget(event) {
        return this.instance.forget(event);
    }
    static fake() {
        return this.instance.fake();
    }
    static assertDispatched(event, callback) {
        return this.instance.assertDispatched(event, callback);
    }
    static subscribe(subscriber) {
        return this.instance.subscribe(subscriber);
    }
    // Facade helper to set a custom manager
    static setManager(manager) {
        this.instance = manager;
    }
    static getManager() {
        return this.instance;
    }
}
exports.Event = Event;
Event.instance = new EventManager_1.EventManager();
