// @arikajs/auth/dist/Passwords/PasswordResetBroker.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PasswordResetBroker = exports.InMemoryTokenRepository = void 0;
const crypto = __importStar(require("crypto"));
const PasswordBroker_1 = require("../Contracts/PasswordBroker");
/**
 * In-memory token repository (production apps should use a database-backed one)
 */
class InMemoryTokenRepository {
    constructor(expiryMinutes = 60) {
        this.tokens = new Map();
        this.expiryMinutes = expiryMinutes;
    }
    async create(user) {
        const rawToken = crypto.randomBytes(32).toString('hex');
        const hashedToken = crypto.createHash('sha256').update(rawToken).digest('hex');
        this.tokens.set(String(user.id), {
            token: hashedToken,
            createdAt: new Date(),
        });
        return rawToken; // Return un-hashed version to be sent via email
    }
    async exists(user, token) {
        const record = this.tokens.get(String(user.id));
        if (!record)
            return false;
        const hashedToken = crypto.createHash('sha256').update(token).digest('hex');
        // Check expiry
        const now = new Date();
        const diffMs = now.getTime() - record.createdAt.getTime();
        const diffMins = diffMs / (1000 * 60);
        if (diffMins > this.expiryMinutes) {
            this.tokens.delete(String(user.id));
            return false;
        }
        return record.token === hashedToken;
    }
    async delete(user) {
        this.tokens.delete(String(user.id));
    }
    async deleteExpired() {
        const now = new Date();
        for (const [key, record] of this.tokens.entries()) {
            const diffMs = now.getTime() - record.createdAt.getTime();
            const diffMins = diffMs / (1000 * 60);
            if (diffMins > this.expiryMinutes) {
                this.tokens.delete(key);
            }
        }
    }
}
exports.InMemoryTokenRepository = InMemoryTokenRepository;
class PasswordResetBroker {
    constructor(provider, tokens) {
        this.provider = provider;
        this.tokens = tokens || new InMemoryTokenRepository(60);
    }
    /**
     * Send a password reset link to a user.
     */
    async sendResetLink(credentials) {
        const user = await this.provider.retrieveByCredentials(credentials);
        if (!user) {
            return PasswordBroker_1.PasswordResetStatus.INVALID_USER;
        }
        const token = await this.tokens.create(user);
        if (typeof user.sendPasswordResetNotification === 'function') {
            await user.sendPasswordResetNotification(token);
        }
        return PasswordBroker_1.PasswordResetStatus.RESET_LINK_SENT;
    }
    /**
     * Reset the password for the given token.
     */
    async reset(credentials, callback) {
        const user = await this.provider.retrieveByCredentials(credentials);
        if (!user) {
            return PasswordBroker_1.PasswordResetStatus.INVALID_USER;
        }
        if (!credentials.token || !(await this.tokens.exists(user, credentials.token))) {
            return PasswordBroker_1.PasswordResetStatus.INVALID_TOKEN;
        }
        await callback(user, credentials.password);
        await this.tokens.delete(user);
        return PasswordBroker_1.PasswordResetStatus.PASSWORD_RESET;
    }
    /**
     * Clean up expired tokens
     */
    async deleteExpiredTokens() {
        await this.tokens.deleteExpired();
    }
}
exports.PasswordResetBroker = PasswordResetBroker;
