// @arikajs/storage/dist/src/Drivers/S3Driver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3Driver = void 0;
const client_s3_1 = require("@aws-sdk/client-s3");
const s3_request_presigner_1 = require("@aws-sdk/s3-request-presigner");
const mime = __importStar(require("mime-types"));
class S3Driver {
    constructor(config) {
        const s3Config = {
            region: config.region,
            credentials: {
                accessKeyId: config.key,
                secretAccessKey: config.secret,
            },
            endpoint: config.endpoint,
            forcePathStyle: config.forcePathStyle,
        };
        this.client = new client_s3_1.S3Client(s3Config);
        this.bucket = config.bucket;
        this.region = config.region;
        this.urlEndpoint = config.url;
    }
    /**
     * Write contents to a file.
     */
    async put(path, contents) {
        const command = new client_s3_1.PutObjectCommand({
            Bucket: this.bucket,
            Key: path,
            Body: contents,
            ContentType: mime.lookup(path) || 'application/octet-stream'
        });
        await this.client.send(command);
    }
    /**
     * Write a stream to a file.
     */
    async putStream(path, stream) {
        // Note: For large streams, you might want to use Upload from @aws-sdk/lib-storage
        // but for basic implementation, this works.
        const command = new client_s3_1.PutObjectCommand({
            Bucket: this.bucket,
            Key: path,
            Body: stream,
            ContentType: mime.lookup(path) || 'application/octet-stream'
        });
        await this.client.send(command);
    }
    /**
     * Read file contents.
     */
    async get(path) {
        const command = new client_s3_1.GetObjectCommand({
            Bucket: this.bucket,
            Key: path,
        });
        const response = await this.client.send(command);
        if (!response.Body) {
            throw new Error(`File not found: ${path}`);
        }
        // Convert stream to buffer
        return this.streamToBuffer(response.Body);
    }
    /**
     * Read file as a stream.
     */
    readStream(path) {
        // We use a custom stream that fetches from S3 on demand
        // Alternatively, we can use a pass-through stream but it's more complex.
        // For simplicity, we return the S3 body which is a Readable.
        // However, this needs to be an async call in some contexts.
        // A better way is to provide a method that returns a Promise of a Readable.
        // But to match the interface sync signature:
        const { Readable } = require('stream');
        const passThrough = new Readable({
            read() { }
        });
        this.client.send(new client_s3_1.GetObjectCommand({
            Bucket: this.bucket,
            Key: path
        })).then(response => {
            if (response.Body) {
                response.Body.pipe(passThrough);
            }
        }).catch(err => {
            passThrough.emit('error', err);
        });
        return passThrough;
    }
    /**
     * Check if a file exists.
     */
    async exists(path) {
        try {
            const command = new client_s3_1.HeadObjectCommand({
                Bucket: this.bucket,
                Key: path,
            });
            await this.client.send(command);
            return true;
        }
        catch (error) {
            if (error.name === 'NotFound' || error.$metadata?.httpStatusCode === 404) {
                return false;
            }
            throw error;
        }
    }
    /**
     * Delete a file.
     */
    async delete(path) {
        const command = new client_s3_1.DeleteObjectCommand({
            Bucket: this.bucket,
            Key: path,
        });
        await this.client.send(command);
    }
    /**
     * Get the URL for a file.
     */
    url(path) {
        if (this.urlEndpoint) {
            return `${this.urlEndpoint}/${path}`;
        }
        return `https://${this.bucket}.s3.${this.region}.amazonaws.com/${path}`;
    }
    /**
     * Get a temporary URL for a file.
     */
    async temporaryUrl(path, expiresAt) {
        const command = new client_s3_1.GetObjectCommand({
            Bucket: this.bucket,
            Key: path,
        });
        const expiresIn = Math.floor((expiresAt.getTime() - Date.now()) / 1000);
        return await (0, s3_request_presigner_1.getSignedUrl)(this.client, command, { expiresIn });
    }
    /**
     * Get the file size in bytes.
     */
    async size(path) {
        const command = new client_s3_1.HeadObjectCommand({
            Bucket: this.bucket,
            Key: path,
        });
        const response = await this.client.send(command);
        return response.ContentLength || 0;
    }
    /**
     * Get the last modified time of the file.
     */
    async lastModified(path) {
        const command = new client_s3_1.HeadObjectCommand({
            Bucket: this.bucket,
            Key: path,
        });
        const response = await this.client.send(command);
        return response.LastModified?.getTime() || 0;
    }
    /**
     * Get the mime type of the file.
     */
    async mimeType(path) {
        const command = new client_s3_1.HeadObjectCommand({
            Bucket: this.bucket,
            Key: path,
        });
        const response = await this.client.send(command);
        return response.ContentType || mime.lookup(path) || 'application/octet-stream';
    }
    /**
     * Helper to convert stream to buffer.
     */
    async streamToBuffer(stream) {
        return new Promise((resolve, reject) => {
            const chunks = [];
            stream.on('data', (chunk) => chunks.push(chunk));
            stream.on('error', reject);
            stream.on('end', () => resolve(Buffer.concat(chunks)));
        });
    }
}
exports.S3Driver = S3Driver;
