// @arikajs/benchmark/dist/reporter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.printReport = printReport;
const chalk_1 = __importDefault(require("chalk"));
const cli_table3_1 = __importDefault(require("cli-table3"));
// ─── Colour palette ────────────────────────────────────────────────────────
const FRAMEWORK_COLORS = {
    'ArikaJS': chalk_1.default.hex('#7C3AED'), // violet
    'Fastify': chalk_1.default.hex('#00B7A3'), // teal
    'Express': chalk_1.default.hex('#E09000'), // amber
};
function colorFor(name) {
    return FRAMEWORK_COLORS[name] ?? chalk_1.default.white;
}
function medal(rank) {
    if (rank === 1)
        return '🥇';
    if (rank === 2)
        return '🥈';
    if (rank === 3)
        return '🥉';
    return `  ${rank}.`;
}
function fmtRps(n) {
    return n >= 1000 ? `${(n / 1000).toFixed(2)}k` : n.toFixed(0);
}
function fmtBytes(n) {
    if (n >= 1024 * 1024)
        return `${(n / (1024 * 1024)).toFixed(2)} MB/s`;
    if (n >= 1024)
        return `${(n / 1024).toFixed(2)} KB/s`;
    return `${n.toFixed(0)} B/s`;
}
function fmtMs(n) {
    return `${n.toFixed(2)} ms`;
}
// ─── Header ────────────────────────────────────────────────────────────────
function printHeader(suite) {
    console.log('');
    console.log(chalk_1.default.hex('#7C3AED').bold('╔══════════════════════════════════════════════════════════════╗'));
    console.log(chalk_1.default.hex('#7C3AED').bold('║') + chalk_1.default.white.bold('         🚀  ArikaJS Performance Benchmark Report           ') + chalk_1.default.hex('#7C3AED').bold('║'));
    console.log(chalk_1.default.hex('#7C3AED').bold('╚══════════════════════════════════════════════════════════════╝'));
    console.log('');
    console.log(chalk_1.default.gray(`  📅  Timestamp : ${suite.timestamp}`));
    console.log(chalk_1.default.gray(`  ⬡  Node.js   : ${suite.nodeVersion}`));
    console.log(chalk_1.default.gray(`  💻  Platform  : ${suite.platform}`));
    console.log('');
}
// ─── Per-scenario table ─────────────────────────────────────────────────────
function printScenarioTable(scenario, rows) {
    // Sort by req/s descending
    const sorted = [...rows].sort((a, b) => b.requestsPerSecond - a.requestsPerSecond);
    const topRps = sorted[0]?.requestsPerSecond ?? 1;
    console.log(chalk_1.default.white.bold(`  📌  Scenario: `) + chalk_1.default.cyan.bold(scenario));
    console.log(chalk_1.default.gray(`     Connections: ${sorted[0]?.connections}   Duration: ${sorted[0]?.duration}s`));
    console.log('');
    const table = new cli_table3_1.default({
        head: [
            chalk_1.default.white.bold('Rank'),
            chalk_1.default.white.bold('Framework'),
            chalk_1.default.white.bold('Req/s'),
            chalk_1.default.white.bold('Latency avg'),
            chalk_1.default.white.bold('Latency p99'),
            chalk_1.default.white.bold('Throughput'),
            chalk_1.default.white.bold('Errors'),
            chalk_1.default.white.bold('vs Best'),
        ],
        colAligns: ['center', 'left', 'right', 'right', 'right', 'right', 'right', 'right'],
        style: {
            head: [],
            border: ['gray'],
        },
        chars: {
            top: '─', 'top-mid': '┬', 'top-left': '╭', 'top-right': '╮',
            bottom: '─', 'bottom-mid': '┴', 'bottom-left': '╰', 'bottom-right': '╯',
            left: '│', 'left-mid': '├', mid: '─', 'mid-mid': '┼',
            right: '│', 'right-mid': '┤', middle: '│',
        },
    });
    sorted.forEach((r, idx) => {
        const rank = idx + 1;
        const color = colorFor(r.framework);
        const ratio = (r.requestsPerSecond / topRps);
        const vsLabel = rank === 1
            ? chalk_1.default.green.bold('  ★ Best')
            : chalk_1.default.red(`  −${((1 - ratio) * 100).toFixed(1)}%`);
        const errText = r.errors > 0
            ? chalk_1.default.red(r.errors.toLocaleString())
            : chalk_1.default.green('0');
        table.push([
            medal(rank),
            color.bold(r.framework),
            color(fmtRps(r.requestsPerSecond)),
            chalk_1.default.white(fmtMs(r.latencyAvg)),
            chalk_1.default.white(fmtMs(r.latencyP99)),
            chalk_1.default.gray(fmtBytes(r.throughput)),
            errText,
            vsLabel,
        ]);
    });
    console.log(table.toString());
    console.log('');
}
// ─── Summary comparison table ───────────────────────────────────────────────
function printSummary(results) {
    const frameworks = [...new Set(results.map(r => r.framework))];
    const scenarios = [...new Set(results.map(r => r.scenario))];
    // Compute average req/s per framework across all scenarios
    const avgMap = {};
    for (const fw of frameworks) {
        const fwResults = results.filter(r => r.framework === fw);
        avgMap[fw] = fwResults.reduce((s, r) => s + r.requestsPerSecond, 0) / fwResults.length;
    }
    const sorted = frameworks.sort((a, b) => avgMap[b] - avgMap[a]);
    const bestRps = avgMap[sorted[0]];
    console.log(chalk_1.default.white.bold('  📊  Overall Summary  ') + chalk_1.default.gray('(avg req/s across all scenarios)'));
    console.log('');
    const table = new cli_table3_1.default({
        head: [
            chalk_1.default.white.bold('Rank'),
            chalk_1.default.white.bold('Framework'),
            chalk_1.default.white.bold('Avg Req/s'),
            chalk_1.default.white.bold('vs Best'),
            ...scenarios.map(s => chalk_1.default.white.bold(s)),
        ],
        colAligns: ['center', 'left', 'right', 'right', ...scenarios.map(() => 'right')],
        style: { head: [], border: ['gray'] },
        chars: {
            top: '─', 'top-mid': '┬', 'top-left': '╭', 'top-right': '╮',
            bottom: '─', 'bottom-mid': '┴', 'bottom-left': '╰', 'bottom-right': '╯',
            left: '│', 'left-mid': '├', mid: '─', 'mid-mid': '┼',
            right: '│', 'right-mid': '┤', middle: '│',
        },
    });
    sorted.forEach((fw, idx) => {
        const rank = idx + 1;
        const color = colorFor(fw);
        const ratio = avgMap[fw] / bestRps;
        const vsLabel = rank === 1
            ? chalk_1.default.green.bold('  ★ Best')
            : chalk_1.default.red(`  −${((1 - ratio) * 100).toFixed(1)}%`);
        const perScenario = scenarios.map(s => {
            const r = results.find(x => x.framework === fw && x.scenario === s);
            return r ? color(fmtRps(r.requestsPerSecond)) : chalk_1.default.gray('N/A');
        });
        table.push([
            medal(rank),
            color.bold(fw),
            color.bold(fmtRps(avgMap[fw])),
            vsLabel,
            ...perScenario,
        ]);
    });
    console.log(table.toString());
    console.log('');
}
// ─── Footer ─────────────────────────────────────────────────────────────────
function printFooter() {
    console.log(chalk_1.default.hex('#7C3AED').bold('─────────────────────────────────────────────────────────────────'));
    console.log(chalk_1.default.gray('  ℹ  Measured with autocannon. Results may vary between runs.'));
    console.log(chalk_1.default.gray('  ℹ  ArikaJS includes full framework boot (routing, middleware, DI).'));
    console.log(chalk_1.default.hex('#7C3AED').bold('─────────────────────────────────────────────────────────────────'));
    console.log('');
}
// ─── Public API ─────────────────────────────────────────────────────────────
function printReport(suite) {
    printHeader(suite);
    const scenarios = [...new Set(suite.results.map(r => r.scenario))];
    for (const scenario of scenarios) {
        const rows = suite.results.filter(r => r.scenario === scenario);
        printScenarioTable(scenario, rows);
    }
    printSummary(suite.results);
    printFooter();
}
