// @arikajs/benchmark/dist/runner.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.startServer = startServer;
exports.stopServer = stopServer;
exports.runScenario = runScenario;
const autocannon_1 = __importDefault(require("autocannon"));
/**
 * Start an HTTP server on a random free port and return the address.
 */
function startServer(server) {
    return new Promise((resolve, reject) => {
        server.on('error', reject);
        server.listen(0, '127.0.0.1', () => {
            const addr = server.address();
            resolve(addr);
        });
    });
}
/**
 * Close an HTTP server gracefully.
 */
function stopServer(server) {
    return new Promise((resolve) => {
        server.close(() => resolve());
        // Force-close any keep-alive connections
        if (typeof server.closeAllConnections === 'function') {
            server.closeAllConnections();
        }
    });
}
/**
 * Run autocannon against a server for a single scenario and return structured results.
 */
async function runScenario(framework, server, scenario, opts) {
    const { port } = await startServer(server);
    const baseUrl = `http://127.0.0.1:${port}`;
    // -- Warmup run (shorter, results discarded) --
    if (opts.warmupDuration && opts.warmupDuration > 0) {
        await (0, autocannon_1.default)({
            url: `${baseUrl}${scenario.url}`,
            connections: opts.connections,
            duration: opts.warmupDuration,
            silent: true,
        });
    }
    // -- Actual benchmark run --
    const result = await (0, autocannon_1.default)({
        url: `${baseUrl}${scenario.url}`,
        method: scenario.method,
        connections: opts.connections,
        duration: opts.duration,
        pipelining: opts.pipelining ?? 1,
        silent: true,
        ...(scenario.body
            ? {
                headers: { 'content-type': 'application/json' },
                body: JSON.stringify(scenario.body),
            }
            : {}),
    });
    await stopServer(server);
    return {
        framework,
        scenario: scenario.name,
        requestsPerSecond: result.requests.average,
        latencyAvg: result.latency.average,
        latencyP99: result.latency.p99,
        throughput: result.throughput.average,
        errors: result.errors,
        timeouts: result.timeouts,
        duration: opts.duration,
        connections: opts.connections,
    };
}
