// arikajs/dist/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__ = exports.trans = void 0;
exports.setApp = setApp;
exports.app = app;
exports.config = config;
exports.env = env;
exports.info = info;
exports.error = error;
exports.warning = warning;
exports.debug = debug;
exports.route = route;
exports.lang = lang;
exports.view = view;
const logging_1 = require("@arikajs/logging");
const localization_1 = require("@arikajs/localization");
const config_1 = require("@arikajs/config");
/**
 * Get the application instance.
 */
let appInstance = null;
function setApp(app) {
    appInstance = app;
}
function app() {
    if (!appInstance) {
        throw new Error('Application instance not set.');
    }
    return appInstance;
}
/**
 * Get a configuration value.
 */
function config(key, defaultValue = null) {
    const repository = app().config();
    if (!key)
        return repository;
    return repository.get(key, defaultValue);
}
/**
 * Get an environment variable.
 */
function env(key, defaultValue) {
    return (0, config_1.env)(key, defaultValue);
}
/**
 * Log an info message.
 */
function info(message, context = {}) {
    logging_1.Log.info(message, context);
}
/**
 * Log an error message.
 */
function error(message, context = {}) {
    logging_1.Log.error(message, context);
}
/**
 * Log a warning message.
 */
function warning(message, context = {}) {
    logging_1.Log.warning(message, context);
}
/**
 * Log a debug message.
 */
function debug(message, context = {}) {
    logging_1.Log.debug(message, context);
}
/**
 * Generate a URL for a named route.
 */
function route(name, params = {}) {
    return app().getRouter().route(name, params);
}
/**
 * Translate the given message.
 */
function lang(key, replace = {}, locale = null) {
    return app().make(localization_1.Translator).get(key, replace, locale);
}
// Alias for common patterns
exports.trans = lang;
exports.__ = lang;
/**
 * Render a view template or get the view engine.
 */
function view(template, data = {}) {
    const engine = app().make('view');
    if (template === undefined)
        return engine;
    return engine.render(template, data);
}
// Add properties to support view.render() and view.share() as seen in the README
view.render = (template, data = {}) => {
    return app().make('view').render(template, data);
};
view.share = (key, value) => {
    return app().make('view').share(key, value);
};
view.composer = (template, callback) => {
    return app().make('view').composer(template, callback);
};
