// arikajs/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Command = exports.Encrypter = exports.Schedule = exports.Scheduler = exports.Event = exports.Worker = exports.BaseJob = exports.Queue = exports.Cache = exports.SchemaBuilder = exports.Migration = exports.Schema = exports.DB = exports.Model = exports.VerifyCsrfToken = exports.ViewMiddleware = exports.ValidateRequestMiddleware = exports.ServeStaticMiddleware = exports.RequestLoggingMiddleware = exports.Pipeline = exports.Route = exports.Handler = exports.Kernel = exports.HttpServeStaticMiddleware = exports.Throttle = exports.SecurityHeaders = exports.ConvertEmptyStringsToNull = exports.TrimStrings = exports.CorsMiddleware = exports.BodyParserMiddleware = exports.NotFoundHttpException = exports.HttpException = exports.Response = exports.Request = exports.Config = exports.Container = exports.ServiceProvider = exports.view = exports.__ = exports.trans = exports.lang = exports.route = exports.debug = exports.warning = exports.error = exports.info = exports.env = exports.config = exports.app = exports.createApp = void 0;
exports.SessionFileDriver = exports.SessionMemoryDriver = exports.StartSession = exports.SessionManager = exports.Session = exports.Authenticate = exports.EloquentUserProvider = exports.Hasher = exports.AuthContext = exports.AuthManager = exports.Mailable = exports.Mail = exports.Translator = exports.Validator = exports.View = exports.Log = void 0;
__exportStar(require("./Application"), exports);
var createApp_1 = require("./createApp");
Object.defineProperty(exports, "createApp", { enumerable: true, get: function () { return createApp_1.createApp; } });
var helpers_js_1 = require("./helpers.js");
Object.defineProperty(exports, "app", { enumerable: true, get: function () { return helpers_js_1.app; } });
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return helpers_js_1.config; } });
Object.defineProperty(exports, "env", { enumerable: true, get: function () { return helpers_js_1.env; } });
Object.defineProperty(exports, "info", { enumerable: true, get: function () { return helpers_js_1.info; } });
Object.defineProperty(exports, "error", { enumerable: true, get: function () { return helpers_js_1.error; } });
Object.defineProperty(exports, "warning", { enumerable: true, get: function () { return helpers_js_1.warning; } });
Object.defineProperty(exports, "debug", { enumerable: true, get: function () { return helpers_js_1.debug; } });
Object.defineProperty(exports, "route", { enumerable: true, get: function () { return helpers_js_1.route; } });
Object.defineProperty(exports, "lang", { enumerable: true, get: function () { return helpers_js_1.lang; } });
Object.defineProperty(exports, "trans", { enumerable: true, get: function () { return helpers_js_1.trans; } });
Object.defineProperty(exports, "__", { enumerable: true, get: function () { return helpers_js_1.__; } });
Object.defineProperty(exports, "view", { enumerable: true, get: function () { return helpers_js_1.view; } });
// Re-export common foundation items for convenience
var foundation_1 = require("@arikajs/foundation");
Object.defineProperty(exports, "ServiceProvider", { enumerable: true, get: function () { return foundation_1.ServiceProvider; } });
var foundation_2 = require("@arikajs/foundation");
Object.defineProperty(exports, "Container", { enumerable: true, get: function () { return foundation_2.Container; } });
var foundation_3 = require("@arikajs/foundation");
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return foundation_3.Repository; } });
// Re-export HTTP items
var http_1 = require("@arikajs/http");
Object.defineProperty(exports, "Request", { enumerable: true, get: function () { return http_1.Request; } });
Object.defineProperty(exports, "Response", { enumerable: true, get: function () { return http_1.Response; } });
Object.defineProperty(exports, "HttpException", { enumerable: true, get: function () { return http_1.HttpException; } });
Object.defineProperty(exports, "NotFoundHttpException", { enumerable: true, get: function () { return http_1.NotFoundHttpException; } });
var http_2 = require("@arikajs/http");
Object.defineProperty(exports, "BodyParserMiddleware", { enumerable: true, get: function () { return http_2.BodyParserMiddleware; } });
Object.defineProperty(exports, "CorsMiddleware", { enumerable: true, get: function () { return http_2.CorsMiddleware; } });
Object.defineProperty(exports, "TrimStrings", { enumerable: true, get: function () { return http_2.TrimStrings; } });
Object.defineProperty(exports, "ConvertEmptyStringsToNull", { enumerable: true, get: function () { return http_2.ConvertEmptyStringsToNull; } });
Object.defineProperty(exports, "SecurityHeaders", { enumerable: true, get: function () { return http_2.SecurityHeaders; } });
Object.defineProperty(exports, "Throttle", { enumerable: true, get: function () { return http_2.Throttle; } });
Object.defineProperty(exports, "HttpServeStaticMiddleware", { enumerable: true, get: function () { return http_2.ServeStaticMiddleware; } });
var Kernel_1 = require("./http/Kernel");
Object.defineProperty(exports, "Kernel", { enumerable: true, get: function () { return Kernel_1.Kernel; } });
var Handler_1 = require("./http/Handler");
Object.defineProperty(exports, "Handler", { enumerable: true, get: function () { return Handler_1.Handler; } });
// Re-export Routing items
var router_1 = require("@arikajs/router");
Object.defineProperty(exports, "Route", { enumerable: true, get: function () { return router_1.Route; } });
// Re-export Middleware items
var middleware_1 = require("@arikajs/middleware");
Object.defineProperty(exports, "Pipeline", { enumerable: true, get: function () { return middleware_1.Pipeline; } });
var RequestLoggingMiddleware_1 = require("./http/Middleware/RequestLoggingMiddleware");
Object.defineProperty(exports, "RequestLoggingMiddleware", { enumerable: true, get: function () { return RequestLoggingMiddleware_1.RequestLoggingMiddleware; } });
var ServeStaticMiddleware_1 = require("./http/Middleware/ServeStaticMiddleware");
Object.defineProperty(exports, "ServeStaticMiddleware", { enumerable: true, get: function () { return ServeStaticMiddleware_1.ServeStaticMiddleware; } });
var ValidateRequestMiddleware_1 = require("./http/Middleware/ValidateRequestMiddleware");
Object.defineProperty(exports, "ValidateRequestMiddleware", { enumerable: true, get: function () { return ValidateRequestMiddleware_1.ValidateRequestMiddleware; } });
var ViewMiddleware_1 = require("./http/Middleware/ViewMiddleware");
Object.defineProperty(exports, "ViewMiddleware", { enumerable: true, get: function () { return ViewMiddleware_1.ViewMiddleware; } });
var VerifyCsrfToken_1 = require("./http/Middleware/VerifyCsrfToken");
Object.defineProperty(exports, "VerifyCsrfToken", { enumerable: true, get: function () { return VerifyCsrfToken_1.VerifyCsrfToken; } });
// Re-export Database items
var database_1 = require("@arikajs/database");
Object.defineProperty(exports, "Model", { enumerable: true, get: function () { return database_1.Model; } });
Object.defineProperty(exports, "DB", { enumerable: true, get: function () { return database_1.Database; } });
Object.defineProperty(exports, "Schema", { enumerable: true, get: function () { return database_1.Schema; } });
Object.defineProperty(exports, "Migration", { enumerable: true, get: function () { return database_1.Migration; } });
Object.defineProperty(exports, "SchemaBuilder", { enumerable: true, get: function () { return database_1.SchemaBuilder; } });
// Re-export Cache items
var cache_1 = require("@arikajs/cache");
Object.defineProperty(exports, "Cache", { enumerable: true, get: function () { return cache_1.Cache; } });
// Re-export Queue items
var queue_1 = require("@arikajs/queue");
Object.defineProperty(exports, "Queue", { enumerable: true, get: function () { return queue_1.Queue; } });
Object.defineProperty(exports, "BaseJob", { enumerable: true, get: function () { return queue_1.BaseJob; } });
Object.defineProperty(exports, "Worker", { enumerable: true, get: function () { return queue_1.Worker; } });
// Re-export Event items
var events_1 = require("@arikajs/events");
Object.defineProperty(exports, "Event", { enumerable: true, get: function () { return events_1.Event; } });
// Re-export Scheduler items
var scheduler_1 = require("@arikajs/scheduler");
Object.defineProperty(exports, "Scheduler", { enumerable: true, get: function () { return scheduler_1.Scheduler; } });
Object.defineProperty(exports, "Schedule", { enumerable: true, get: function () { return scheduler_1.Schedule; } });
// Re-export Encryption items
var encryption_1 = require("@arikajs/encryption");
Object.defineProperty(exports, "Encrypter", { enumerable: true, get: function () { return encryption_1.Encrypter; } });
// Re-export Console items
var console_1 = require("@arikajs/console");
Object.defineProperty(exports, "Command", { enumerable: true, get: function () { return console_1.Command; } });
// Re-export Logging items
var logging_1 = require("@arikajs/logging");
Object.defineProperty(exports, "Log", { enumerable: true, get: function () { return logging_1.Log; } });
// Re-export View items
var view_1 = require("@arikajs/view");
Object.defineProperty(exports, "View", { enumerable: true, get: function () { return view_1.View; } });
// Re-export Validation items
var validation_1 = require("@arikajs/validation");
Object.defineProperty(exports, "Validator", { enumerable: true, get: function () { return validation_1.Validator; } });
// Re-export Localization items
var localization_1 = require("@arikajs/localization");
Object.defineProperty(exports, "Translator", { enumerable: true, get: function () { return localization_1.Translator; } });
// Re-export Mail items
var mail_1 = require("@arikajs/mail");
Object.defineProperty(exports, "Mail", { enumerable: true, get: function () { return mail_1.Mail; } });
Object.defineProperty(exports, "Mailable", { enumerable: true, get: function () { return mail_1.Mailable; } });
// Re-export Auth items
var auth_1 = require("@arikajs/auth");
Object.defineProperty(exports, "AuthManager", { enumerable: true, get: function () { return auth_1.AuthManager; } });
Object.defineProperty(exports, "AuthContext", { enumerable: true, get: function () { return auth_1.AuthContext; } });
Object.defineProperty(exports, "Hasher", { enumerable: true, get: function () { return auth_1.Hasher; } });
Object.defineProperty(exports, "EloquentUserProvider", { enumerable: true, get: function () { return auth_1.EloquentUserProvider; } });
Object.defineProperty(exports, "Authenticate", { enumerable: true, get: function () { return auth_1.Authenticate; } });
// Re-export Session items
var session_1 = require("@arikajs/session");
Object.defineProperty(exports, "Session", { enumerable: true, get: function () { return session_1.Session; } });
Object.defineProperty(exports, "SessionManager", { enumerable: true, get: function () { return session_1.SessionManager; } });
Object.defineProperty(exports, "StartSession", { enumerable: true, get: function () { return session_1.StartSession; } });
Object.defineProperty(exports, "SessionMemoryDriver", { enumerable: true, get: function () { return session_1.MemoryDriver; } });
Object.defineProperty(exports, "SessionFileDriver", { enumerable: true, get: function () { return session_1.FileDriver; } });
