// @arikajs/http/dist/Middleware/Throttle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Throttle = void 0;
const index_1 = require("../index");
/**
 * Simple in-memory rate limiter middleware.
 */
class Throttle {
    constructor(maxAttempts = 60, windowSeconds = 60) {
        this.maxAttempts = maxAttempts;
        this.windowSeconds = windowSeconds;
        this.store = new Map();
    }
    /**
     * Handle the incoming request.
     */
    async handle(request, next, response) {
        const ip = request.ip() || '127.0.0.1';
        const key = `throttle:${ip}:${request.path()}`;
        const now = Date.now();
        const windowMs = this.windowSeconds * 1000;
        const entry = this.store.get(key);
        if (entry && now < entry.resetAt) {
            if (entry.count >= this.maxAttempts) {
                const retryAfter = Math.ceil((entry.resetAt - now) / 1000);
                // If it's a 429, we should return early
                const errorRes = response || new index_1.Response(request.getRaw());
                errorRes.header('Retry-After', String(retryAfter));
                errorRes.header('X-RateLimit-Limit', String(this.maxAttempts));
                errorRes.header('X-RateLimit-Remaining', '0');
                return errorRes.status(429).json({
                    message: `Too many requests. Please try again in ${retryAfter} seconds.`,
                    retry_after: retryAfter
                });
            }
            entry.count++;
        }
        else {
            this.store.set(key, { count: 1, resetAt: now + windowMs });
        }
        const res = await next(request);
        const currentEntry = this.store.get(key);
        res.header('X-RateLimit-Limit', String(this.maxAttempts));
        res.header('X-RateLimit-Remaining', String(Math.max(0, this.maxAttempts - currentEntry.count)));
        // Cleanup expired entries probabilistically
        if (Math.random() < 0.01) {
            this.cleanup();
        }
        return res;
    }
    cleanup() {
        const now = Date.now();
        for (const [key, entry] of this.store.entries()) {
            if (now >= entry.resetAt) {
                this.store.delete(key);
            }
        }
    }
}
exports.Throttle = Throttle;
