// @arikajs/mail/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mail = exports.SendQueuedMailable = exports.SendGridTransport = exports.MailgunTransport = exports.SesTransport = exports.SmtpTransport = exports.Message = exports.Mailable = exports.PendingMail = exports.Mailer = exports.MailManager = void 0;
var MailManager_1 = require("./MailManager");
Object.defineProperty(exports, "MailManager", { enumerable: true, get: function () { return MailManager_1.MailManager; } });
var Mailer_1 = require("./Mailer");
Object.defineProperty(exports, "Mailer", { enumerable: true, get: function () { return Mailer_1.Mailer; } });
Object.defineProperty(exports, "PendingMail", { enumerable: true, get: function () { return Mailer_1.PendingMail; } });
var Mailable_1 = require("./Mailable");
Object.defineProperty(exports, "Mailable", { enumerable: true, get: function () { return Mailable_1.Mailable; } });
var Message_1 = require("./Message");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return Message_1.Message; } });
var SmtpTransport_1 = require("./Transport/SmtpTransport");
Object.defineProperty(exports, "SmtpTransport", { enumerable: true, get: function () { return SmtpTransport_1.SmtpTransport; } });
var SesTransport_1 = require("./Transport/SesTransport");
Object.defineProperty(exports, "SesTransport", { enumerable: true, get: function () { return SesTransport_1.SesTransport; } });
var MailgunTransport_1 = require("./Transport/MailgunTransport");
Object.defineProperty(exports, "MailgunTransport", { enumerable: true, get: function () { return MailgunTransport_1.MailgunTransport; } });
var SendGridTransport_1 = require("./Transport/SendGridTransport");
Object.defineProperty(exports, "SendGridTransport", { enumerable: true, get: function () { return SendGridTransport_1.SendGridTransport; } });
var SendQueuedMailable_1 = require("./Jobs/SendQueuedMailable");
Object.defineProperty(exports, "SendQueuedMailable", { enumerable: true, get: function () { return SendQueuedMailable_1.SendQueuedMailable; } });
let mailManager;
class Mail {
    static setManager(manager) {
        mailManager = manager;
    }
    static setQueue(queue) {
        if (!mailManager) {
            throw new Error('Mail system not configured. Please use Mail.setManager() to configure.');
        }
        mailManager.setQueue(queue);
    }
    static extend(driver, callback) {
        if (!mailManager) {
            throw new Error('Mail system not configured. Please use Mail.setManager() to configure.');
        }
        return mailManager.extend(driver, callback);
    }
    static to(users) {
        if (!mailManager) {
            throw new Error('Mail system not configured. Please use Mail.setManager() to configure.');
        }
        return mailManager.to(users);
    }
    static cc(users) {
        if (!mailManager) {
            throw new Error('Mail system not configured. Please use Mail.setManager() to configure.');
        }
        return mailManager.mailer().cc(users);
    }
    static bcc(users) {
        if (!mailManager) {
            throw new Error('Mail system not configured. Please use Mail.setManager() to configure.');
        }
        return mailManager.mailer().bcc(users);
    }
    static mailer(name) {
        if (!mailManager) {
            throw new Error('Mail system not configured. Please use Mail.setManager() to configure.');
        }
        return mailManager.mailer(name);
    }
}
exports.Mail = Mail;
