// @arikajs/mail/dist/Mailer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PendingMail = exports.Mailer = void 0;
const Message_1 = require("./Message");
class Mailer {
    constructor(name, transport, viewRenderer, manager) {
        this.name = name;
        this.transport = transport;
        this.viewRenderer = viewRenderer;
        this.manager = manager;
    }
    to(users) {
        return new PendingMail(this).to(users);
    }
    cc(users) {
        return new PendingMail(this).cc(users);
    }
    bcc(users) {
        return new PendingMail(this).bcc(users);
    }
    async send(viewPath, data, callback) {
        const message = new Message_1.Message();
        callback(message);
        if (viewPath && this.viewRenderer) {
            const html = await this.viewRenderer.render(viewPath, data);
            message.htmlContent(html);
        }
        return this.transport.send(message.getPayload());
    }
    async sendMailable(mailable) {
        return this.sendRaw(mailable);
    }
    async sendRaw(message) {
        return this.transport.send(message);
    }
    getRenderer() {
        return this.viewRenderer;
    }
    getQueue() {
        return this.manager?.getQueue();
    }
}
exports.Mailer = Mailer;
class PendingMail {
    constructor(mailer) {
        this.mailer = mailer;
        this.viewData = {};
        this.message = new Message_1.Message();
    }
    to(users) {
        this.message.to(users);
        return this;
    }
    cc(users) {
        this.message.cc(users);
        return this;
    }
    bcc(users) {
        this.message.bcc(users);
        return this;
    }
    subject(subject) {
        this.message.subject(subject);
        return this;
    }
    replyTo(address) {
        this.message.replyTo(address);
        return this;
    }
    text(content) {
        this.message.textContent(content);
        return this;
    }
    view(path, data = {}) {
        this.viewPath = path;
        this.viewData = data;
        return this;
    }
    attach(path, options = {}) {
        this.message.attach({ path, ...options });
        return this;
    }
    attachData(data, filename, options = {}) {
        this.message.attachData(data, filename, options);
        return this;
    }
    attachStream(stream, filename, options = {}) {
        this.message.attachStream(stream, filename, options);
        return this;
    }
    async send(mailable) {
        if (mailable && typeof mailable.build === 'function') {
            try {
                mailable.build();
            }
            catch (e) {
                // Ignore build errors if already built or not a builder
            }
        }
        if (mailable) {
            if (mailable.subjectLine)
                this.message.subject(mailable.subjectLine);
            if (mailable.fromAddress)
                this.message.from(mailable.fromAddress);
            if (mailable.replyToAddress)
                this.message.replyTo(mailable.replyToAddress);
            if (mailable.textContent)
                this.message.textContent(mailable.textContent);
            if (mailable.viewPath) {
                this.viewPath = mailable.viewPath;
                this.viewData = mailable.viewData;
            }
            if (mailable.attachments) {
                mailable.attachments.forEach(att => this.message.attach(att));
            }
            if (mailable.ccRecipients) {
                this.message.cc(mailable.ccRecipients);
            }
            if (mailable.bccRecipients) {
                this.message.bcc(mailable.bccRecipients);
            }
        }
        if (this.viewPath) {
            const renderer = this.mailer.getRenderer();
            if (renderer) {
                const html = await renderer.render(this.viewPath, this.viewData);
                this.message.htmlContent(html);
            }
        }
        return this.mailer.sendRaw(this.message.getPayload());
    }
    async queue(mailable) {
        const queue = this.mailer.getQueue();
        if (!queue) {
            throw new Error('Queue not configured. Please use Mail.setQueue() to enable queued emails.');
        }
        if (mailable && typeof mailable.build === 'function') {
            try {
                mailable.build();
            }
            catch (e) { }
        }
        if (mailable && mailable.constructor && mailable.constructor.name !== 'Object') {
            const payload = this.message.getPayload();
            if (payload.to && payload.to.length > 0) {
                mailable.toRecipients = (mailable.toRecipients || []).concat(payload.to);
            }
            if (payload.cc && payload.cc.length > 0) {
                mailable.ccRecipients = (mailable.ccRecipients || []).concat(payload.cc);
            }
            if (payload.bcc && payload.bcc.length > 0) {
                mailable.bccRecipients = (mailable.bccRecipients || []).concat(payload.bcc);
            }
        }
        const { SendQueuedMailable } = require('./Jobs/SendQueuedMailable');
        return queue.dispatch(new SendQueuedMailable(mailable || this.prepareMessage(), this.mailer['name']));
    }
    async later(delay, mailable) {
        const queue = this.mailer.getQueue();
        if (!queue) {
            throw new Error('Queue not configured. Please use Mail.setQueue() to enable queued emails.');
        }
        if (mailable && typeof mailable.build === 'function') {
            try {
                mailable.build();
            }
            catch (e) { }
        }
        if (mailable && mailable.constructor && mailable.constructor.name !== 'Object') {
            const payload = this.message.getPayload();
            if (payload.to && payload.to.length > 0) {
                mailable.toRecipients = (mailable.toRecipients || []).concat(payload.to);
            }
        }
        const { SendQueuedMailable } = require('./Jobs/SendQueuedMailable');
        const job = new SendQueuedMailable(mailable || this.prepareMessage(), this.mailer['name']);
        job.onDelay(delay);
        return queue.dispatch(job);
    }
    prepareMessage() {
        // Logic to finalize the message object before sending/queueing
        return this.message.getPayload();
    }
}
exports.PendingMail = PendingMail;
