// @arikajs/mail/dist/MailManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MailManager = void 0;
const Mailer_1 = require("./Mailer");
const SmtpTransport_1 = require("./Transport/SmtpTransport");
const LogTransport_1 = require("./Transport/LogTransport");
const ArrayTransport_1 = require("./Transport/ArrayTransport");
class MailManager {
    constructor(config, viewRenderer) {
        this.config = config;
        this.viewRenderer = viewRenderer;
        this.mailers = new Map();
        this.customCreators = new Map();
    }
    setQueue(queue) {
        this.queue = queue;
        return this;
    }
    getQueue() {
        return this.queue;
    }
    extend(driver, callback) {
        this.customCreators.set(driver, callback);
        return this;
    }
    mailer(name) {
        const mailerName = name || this.config.default;
        if (!this.mailers.has(mailerName)) {
            this.mailers.set(mailerName, this.resolve(mailerName));
        }
        return this.mailers.get(mailerName);
    }
    resolve(name) {
        const config = this.config.mailers[name];
        if (!config) {
            throw new Error(`Mailer [${name}] is not defined.`);
        }
        let transport;
        if (this.customCreators.has(config.transport)) {
            transport = this.customCreators.get(config.transport)(config);
        }
        else {
            switch (config.transport) {
                case 'smtp':
                    transport = new SmtpTransport_1.SmtpTransport(config);
                    break;
                case 'ses':
                    const { SesTransport } = require('./Transport/SesTransport');
                    transport = new SesTransport(config);
                    break;
                case 'mailgun':
                    const { MailgunTransport } = require('./Transport/MailgunTransport');
                    transport = new MailgunTransport(config);
                    break;
                case 'sendgrid':
                    const { SendGridTransport } = require('./Transport/SendGridTransport');
                    transport = new SendGridTransport(config);
                    break;
                case 'log':
                    transport = new LogTransport_1.LogTransport();
                    break;
                case 'array':
                    transport = new ArrayTransport_1.ArrayTransport();
                    break;
                default:
                    throw new Error(`Unsupported transport driver [${config.transport}].`);
            }
        }
        return new Mailer_1.Mailer(name, transport, this.viewRenderer, this);
    }
    to(users) {
        return this.mailer().to(users);
    }
}
exports.MailManager = MailManager;
