// @arikajs/mail/dist/Mailable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mailable = void 0;
class Mailable {
    constructor() {
        this.viewData = {};
        this.attachments = [];
        this.toRecipients = [];
        this.ccRecipients = [];
        this.bccRecipients = [];
    }
    to(address) {
        if (Array.isArray(address)) {
            this.toRecipients = this.toRecipients.concat(address);
        }
        else {
            this.toRecipients.push(address);
        }
        return this;
    }
    cc(address) {
        if (Array.isArray(address)) {
            this.ccRecipients = this.ccRecipients.concat(address);
        }
        else {
            this.ccRecipients.push(address);
        }
        return this;
    }
    bcc(address) {
        if (Array.isArray(address)) {
            this.bccRecipients = this.bccRecipients.concat(address);
        }
        else {
            this.bccRecipients.push(address);
        }
        return this;
    }
    subject(subject) {
        this.subjectLine = subject;
        return this;
    }
    view(path, data = {}) {
        this.viewPath = path;
        this.viewData = data;
        return this;
    }
    text(content) {
        this.textContent = content;
        return this;
    }
    attach(path, options = {}) {
        this.attachments.push({ path, ...options });
        return this;
    }
    attachData(data, filename, options = {}) {
        this.attachments.push({ content: data, filename, ...options });
        return this;
    }
    attachStream(stream, filename, options = {}) {
        this.attachments.push({ content: stream, filename, ...options });
        return this;
    }
    from(address) {
        this.fromAddress = address;
        return this;
    }
    replyTo(address) {
        this.replyToAddress = address;
        return this;
    }
}
exports.Mailable = Mailable;
