// @arikajs/mail/dist/Message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
class Message {
    constructor() {
        this._to = [];
        this._cc = [];
        this._bcc = [];
        this._attachments = [];
    }
    to(address) {
        if (Array.isArray(address)) {
            this._to = this._to.concat(address);
        }
        else {
            this._to.push(address);
        }
        return this;
    }
    cc(address) {
        if (Array.isArray(address)) {
            this._cc = this._cc.concat(address);
        }
        else {
            this._cc.push(address);
        }
        return this;
    }
    bcc(address) {
        if (Array.isArray(address)) {
            this._bcc = this._bcc.concat(address);
        }
        else {
            this._bcc.push(address);
        }
        return this;
    }
    from(address) {
        this._from = address;
        return this;
    }
    replyTo(address) {
        this._replyTo = address;
        return this;
    }
    subject(subject) {
        this._subject = subject;
        return this;
    }
    textContent(content) {
        this._text = content;
        return this;
    }
    htmlContent(content) {
        this._html = content;
        return this;
    }
    attach(attachment) {
        this._attachments.push(attachment);
        return this;
    }
    attachData(data, filename, options = {}) {
        return this.attach({
            ...options,
            content: data,
            filename
        });
    }
    attachStream(stream, filename, options = {}) {
        return this.attach({
            ...options,
            content: stream,
            filename
        });
    }
    getPayload() {
        return {
            to: this._to,
            cc: this._cc,
            bcc: this._bcc,
            from: this._from,
            replyTo: this._replyTo,
            subject: this._subject,
            text: this._text,
            html: this._html,
            attachments: this._attachments,
        };
    }
}
exports.Message = Message;
