// @arikajs/view/dist/Compiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Compiler = void 0;
const Lexer_1 = require("./Compiler/Lexer");
const Parser_1 = require("./Compiler/Parser");
const CodeGenerator_1 = require("./Compiler/CodeGenerator");
const DirectiveRegistry_1 = require("./Compiler/DirectiveRegistry");
class Compiler {
    constructor() {
        this.registry = new DirectiveRegistry_1.DirectiveRegistry();
        this.generator = new CodeGenerator_1.CodeGenerator(this.registry);
    }
    /**
     * Add a custom directive to the compiler.
     */
    addDirective(name, handler) {
        this.registry.register(name, handler);
    }
    /**
     * Compile a template string into a executable JavaScript function body.
     */
    compile(content) {
        const lexer = new Lexer_1.Lexer(content);
        const tokens = lexer.tokenize();
        const parser = new Parser_1.Parser(tokens);
        const ast = parser.parse();
        return this.generator.generate(ast);
    }
}
exports.Compiler = Compiler;
