// @arikajs/view/dist/Engine.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Engine = void 0;
const fs = __importStar(require("node:fs"));
const path = __importStar(require("node:path"));
const crypto = __importStar(require("node:crypto"));
const Compiler_1 = require("./Compiler");
const Template_1 = require("./Template");
class Engine {
    constructor(config) {
        this.config = config;
        this.sections = {};
        this.sectionStack = [];
        this.parentTemplate = null;
        this.compiledFunctions = new Map();
        this.sharedData = {};
        this.composers = new Map();
        this.helpers = {};
        // For push, prepend & stack
        this.pushes = {};
        this.pushStack = [];
        // For components & slots
        this.componentStack = [];
        this.slotStack = [];
        // For fragments
        this.fragmentMode = null;
        // For once directive
        this.onceKeys = new Set();
        this.config = {
            cache: true,
            dev: false,
            extension: '.ark.html',
            ...config
        };
        this.compiler = new Compiler_1.Compiler();
        this.templateLoader = new Template_1.Template(this.config);
    }
    /**
     * Share data across all renders.
     */
    share(key, value) {
        this.sharedData[key] = value;
    }
    /**
     * Add a view composer.
     */
    composer(template, callback) {
        if (!this.composers.has(template)) {
            this.composers.set(template, []);
        }
        this.composers.get(template).push(callback);
    }
    /**
     * Add a global helper.
     */
    helper(name, callback) {
        this.helpers[name] = callback;
    }
    /**
     * Add a custom directive.
     */
    directive(name, handler) {
        this.compiler.addDirective(name, handler);
    }
    /**
     * Render a template file.
     */
    async render(template, data = {}, isInternal = false) {
        if (!isInternal) {
            this.sections = {};
            this.pushes = {};
            this.onceKeys = new Set();
        }
        const mergedData = { ...this.sharedData, ...this.helpers, ...data };
        // Execute composers
        await this.executeComposers(template, mergedData);
        if (this.composers.has('*')) {
            await this.executeComposers('*', mergedData);
        }
        this.parentTemplate = null;
        const content = await this.renderTemplate(template, mergedData);
        if (this.parentTemplate) {
            const parent = this.parentTemplate;
            this.parentTemplate = null;
            return this.render(parent, data, true);
        }
        return content;
    }
    /**
     * Render a fragment of a template.
     */
    async renderFragment(template, fragment, data = {}) {
        this.fragmentMode = fragment;
        const result = await this.render(template, data);
        this.fragmentMode = null;
        return result;
    }
    async executeComposers(template, data) {
        const callbacks = this.composers.get(template);
        if (callbacks) {
            for (const cb of callbacks) {
                await cb(data);
            }
        }
    }
    async renderTemplate(templateName, data) {
        // 1. Check Memory Cache
        if (this.config.cache && this.compiledFunctions.has(templateName)) {
            return await this.runCompiled(this.compiledFunctions.get(templateName).func, data, templateName);
        }
        // 2. Check Disk Cache (Primary - bypasses file read for speed)
        if (this.config.cache && this.config.cachePath) {
            const cachedFunc = this.loadFromDiskCache(templateName);
            if (cachedFunc) {
                // If we found it on disk, we can skip reading the source!
                // But we don't have the current hash... we'll just use the one from disk.
                this.compiledFunctions.set(templateName, { func: cachedFunc, hash: 'disk-cached' });
                return await this.runCompiled(cachedFunc, data, templateName);
            }
        }
        const rawContent = this.templateLoader.read(templateName);
        const contentHash = crypto.createHash('md5').update(rawContent).digest('hex');
        const jsCode = this.compiler.compile(rawContent);
        const AsyncFunction = Object.getPrototypeOf(async function () { }).constructor;
        // Use 'with' only for data provided to function to allow pure JS expressions
        const wrappedJsCode = `with (_data) {\n${jsCode}\n}`;
        const renderFunc = new AsyncFunction('_engine', '_data', wrappedJsCode);
        if (this.config.cache) {
            this.compiledFunctions.set(templateName, { func: renderFunc, hash: contentHash });
            if (this.config.cachePath) {
                this.saveToDiskCache(templateName, contentHash, jsCode);
            }
        }
        return await this.runCompiled(renderFunc, data, templateName);
    }
    loadFromDiskCache(templateName, hash) {
        try {
            const cacheKey = crypto.createHash('md5').update(templateName).digest('hex');
            const cacheFile = path.join(this.config.cachePath, `${cacheKey}.js`);
            if (fs.existsSync(cacheFile)) {
                const { hash: cachedHash, code } = JSON.parse(fs.readFileSync(cacheFile, 'utf8'));
                if (!hash || cachedHash === hash) {
                    const AsyncFunction = Object.getPrototypeOf(async function () { }).constructor;
                    return new AsyncFunction('_engine', '_data', `with (_data) {\n${code}\n}`);
                }
            }
        }
        catch (e) {
            // Ignore cache errors
        }
        return null;
    }
    saveToDiskCache(templateName, hash, code) {
        try {
            if (!fs.existsSync(this.config.cachePath)) {
                fs.mkdirSync(this.config.cachePath, { recursive: true });
            }
            const cacheKey = crypto.createHash('md5').update(templateName).digest('hex');
            const cacheFile = path.join(this.config.cachePath, `${cacheKey}.js`);
            fs.writeFileSync(cacheFile, JSON.stringify({ hash, code }), 'utf8');
        }
        catch (e) {
            // Ignore cache errors
        }
    }
    async runCompiled(func, data, templateName) {
        try {
            return await func(this, data);
        }
        catch (e) {
            if (this.config.dev) {
                // In dev mode, we could enhance the error further
                throw new Error(`Error in ${templateName}: ${e.message}\nStack: ${e.stack}`);
            }
            throw new Error(`Error rendering template "${templateName}": ${e.message}`);
        }
    }
    // Methods called from compiled templates
    extend(parent) {
        this.parentTemplate = parent;
    }
    startSection(name, previousOutput) {
        this.sectionStack.push({ name, previousOutput });
    }
    popSection(content) {
        const section = this.sectionStack.pop();
        if (section) {
            this.sections[section.name] = content;
            return section.previousOutput;
        }
        return content;
    }
    yield(name) {
        return this.sections[name] || '';
    }
    // --- PUSH, PREPEND & STACK ---
    startPush(name, previousOutput) {
        this.pushStack.push({ name, type: 'push', previousOutput });
    }
    startPrepend(name, previousOutput) {
        this.pushStack.push({ name, type: 'prepend', previousOutput });
    }
    endPush(content) {
        return this.processEndPush(content);
    }
    endPrepend(content) {
        return this.processEndPush(content);
    }
    processEndPush(content) {
        const push = this.pushStack.pop();
        if (push) {
            if (!this.pushes[push.name]) {
                this.pushes[push.name] = [];
            }
            if (push.type === 'prepend') {
                this.pushes[push.name].unshift(content);
            }
            else {
                this.pushes[push.name].push(content);
            }
            return push.previousOutput;
        }
        return content;
    }
    stack(name) {
        return this.pushes[name] ? this.pushes[name].join('\n') : '';
    }
    // --- COMPONENTS & SLOTS ---
    startComponent(name, data, previousOutput) {
        this.componentStack.push({ name, data, previousOutput, slots: {} });
    }
    startSlot(name, previousOutput) {
        this.slotStack.push({ name, previousOutput });
    }
    endSlot(content) {
        const slot = this.slotStack.pop();
        if (slot) {
            if (this.componentStack.length > 0) {
                const component = this.componentStack[this.componentStack.length - 1];
                component.slots[slot.name] = content;
            }
            return slot.previousOutput;
        }
        return content;
    }
    async renderComponent(defaultSlotContent) {
        const component = this.componentStack.pop();
        if (component) {
            const componentData = {
                ...component.data,
                slot: defaultSlotContent,
                ...component.slots
            };
            const rendered = await this.render(component.name, componentData, true);
            return component.previousOutput + rendered;
        }
        return defaultSlotContent;
    }
    // --- ONCE ---
    markOnce(id) {
        this.onceKeys.add(id);
    }
    hasOnce(id) {
        return this.onceKeys.has(id);
    }
    // --- FRAGMENTS ---
    isFragmentMode() {
        return this.fragmentMode !== null;
    }
    getFragment() {
        return this.fragmentMode;
    }
}
exports.Engine = Engine;
