// @arikajs/view/dist/View.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.View = void 0;
const Engine_1 = require("./Engine");
class View {
    constructor(config) {
        this.engine = new Engine_1.Engine(config);
    }
    /**
     * Add a custom directive to the template compiler.
     */
    directive(name, handler) {
        this.engine.directive(name, handler);
        return this;
    }
    /**
     * Add a view composer.
     */
    composer(template, callback) {
        this.engine.composer(template, callback);
        return this;
    }
    /**
     * Add a global helper.
     */
    helper(name, callback) {
        this.engine.helper(name, callback);
        return this;
    }
    /**
     * Share data explicitly across all templates rendered by this view instance.
     */
    share(key, value) {
        this.engine.share(key, value);
        return this;
    }
    /**
     * Render a template with data.
     */
    async render(template, data = {}) {
        return this.engine.render(template, data);
    }
    /**
     * Render a fragment of a template.
     */
    async renderFragment(template, fragment, data = {}) {
        return this.engine.renderFragment(template, fragment, data);
    }
}
exports.View = View;
