// @arikajs/encryption/dist/Encrypter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Encrypter = void 0;
const crypto_1 = require("crypto");
const zlib_1 = require("zlib");
const DecryptionException_1 = require("./Exceptions/DecryptionException");
class Encrypter {
    keys;
    cipher = 'aes-256-gcm';
    constructor(key) {
        const keys = Array.isArray(key) ? key : [key];
        this.keys = keys.map(k => this.parseKey(k));
        if (this.keys.length === 0) {
            throw new Error('At least one encryption key must be provided.');
        }
        for (const k of this.keys) {
            if (k.length !== 32) {
                throw new Error('Each encryption key must be 32 bytes.');
            }
        }
    }
    /**
     * Parse the encryption key.
     */
    parseKey(key) {
        if (key.startsWith('base64:')) {
            return Buffer.from(key.substring(7), 'base64');
        }
        return Buffer.from(key);
    }
    /**
     * Encrypt the given value.
     */
    encrypt(value, options = {}) {
        const iv = (0, crypto_1.randomBytes)(16);
        const cipher = (0, crypto_1.createCipheriv)(this.cipher, this.keys[0], iv);
        if (options.context) {
            cipher.setAAD(Buffer.from(options.context));
        }
        const serialize = options.serialize !== false;
        let data;
        if (serialize) {
            data = JSON.stringify(value);
        }
        else {
            data = Buffer.isBuffer(value) ? value : String(value);
        }
        if (options.compress === true) {
            data = (0, zlib_1.deflateSync)(Buffer.from(data));
        }
        let encrypted = cipher.update(data);
        encrypted = Buffer.concat([encrypted, cipher.final()]);
        const tag = cipher.getAuthTag();
        const payload = {
            iv: iv.toString('base64'),
            value: encrypted.toString('base64'),
            tag: tag.toString('base64'),
            v: 1,
        };
        if (options.compress)
            payload.c = true;
        if (options.ttl) {
            payload.exp = Date.now() + (options.ttl * 1000);
        }
        if (options.context)
            payload.aad = true;
        return Buffer.from(JSON.stringify(payload)).toString('base64');
    }
    /**
     * Decrypt the given value.
     */
    decrypt(payload, options = {}) {
        const jsonPayload = this.getJsonPayload(payload);
        // Check for TTL expiration
        if (jsonPayload.exp && Date.now() > jsonPayload.exp) {
            throw new DecryptionException_1.DecryptionException('The payload has expired.');
        }
        // Verify context requirements
        if (jsonPayload.aad && !options.context) {
            throw new DecryptionException_1.DecryptionException('The payload requires a context for decryption.');
        }
        const iv = Buffer.from(jsonPayload.iv, 'base64');
        const tag = Buffer.from(jsonPayload.tag, 'base64');
        const value = jsonPayload.value;
        const compressed = jsonPayload.c === true;
        // Try decrypting with all available keys (support rotation)
        for (const key of this.keys) {
            try {
                const decipher = (0, crypto_1.createDecipheriv)(this.cipher, key, iv);
                if (options.context) {
                    decipher.setAAD(Buffer.from(options.context));
                }
                decipher.setAuthTag(tag);
                let decrypted = decipher.update(value, 'base64');
                decrypted = Buffer.concat([decrypted, decipher.final()]);
                let result = decrypted;
                if (compressed) {
                    result = (0, zlib_1.inflateSync)(result);
                }
                const finalData = result.toString('utf8');
                return options.unserialize === false ? finalData : JSON.parse(finalData);
            }
            catch (error) {
                // Try next key
                continue;
            }
        }
        throw new DecryptionException_1.DecryptionException();
    }
    /**
     * Sign the given value (non-encrypted but authenticated).
     */
    sign(value) {
        const jsonValue = JSON.stringify(value);
        const signature = this.hash(jsonValue);
        return Buffer.from(JSON.stringify({
            value: jsonValue,
            signature: signature
        })).toString('base64');
    }
    /**
     * Verify the given signed payload.
     */
    verify(payload) {
        try {
            const decoded = JSON.parse(Buffer.from(payload, 'base64').toString('utf8'));
            if (!decoded.value || !decoded.signature) {
                throw new DecryptionException_1.DecryptionException('Invalid signed payload.');
            }
            // Verify with all available keys
            for (const key of this.keys) {
                const signature = (0, crypto_1.createHmac)('sha256', key)
                    .update(decoded.value)
                    .digest('hex');
                if ((0, crypto_1.timingSafeEqual)(Buffer.from(signature), Buffer.from(decoded.signature))) {
                    return JSON.parse(decoded.value);
                }
            }
        }
        catch (e) {
            // Fall through 
        }
        throw new DecryptionException_1.DecryptionException('Signature verification failed.');
    }
    /**
     * Create a hash (HMAC) of the given value.
     */
    hash(value) {
        return (0, crypto_1.createHmac)('sha256', this.keys[0])
            .update(value)
            .digest('hex');
    }
    /**
     * Get the JSON payload from the given string.
     */
    getJsonPayload(payload) {
        try {
            const decoded = Buffer.from(payload, 'base64').toString('utf8');
            const data = JSON.parse(decoded);
            if (!this.validPayload(data)) {
                throw new DecryptionException_1.DecryptionException();
            }
            return data;
        }
        catch (error) {
            throw new DecryptionException_1.DecryptionException();
        }
    }
    /**
     * Verify that the encryption payload is valid.
     */
    validPayload(payload) {
        return typeof payload === 'object' &&
            payload !== null &&
            payload.iv &&
            payload.value &&
            payload.tag;
    }
}
exports.Encrypter = Encrypter;
