// @arikajs/foundation/dist/src/application/Application.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Application = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const Container_1 = require("../container/Container");
const ServiceProvider_1 = require("../providers/ServiceProvider");
const config_1 = require("@arikajs/config");
/**
 * Application is the core runtime of ArikaJS.
 *
 * It manages:
 * - Service container
 * - Service providers
 * - Configuration
 * - Application lifecycle (register → boot → run)
 */
class Application {
    constructor(basePath) {
        this.providers = [];
        this.booted = false;
        this.terminatingCallbacks = [];
        this.basePath = path.resolve(basePath);
        this.container = new Container_1.Container();
        this.configRepository = new config_1.Repository();
        Application.instance = this;
        // Register the config repository in the global support helper
        (0, config_1.setConfigRepository)(this.configRepository);
        // Register the application, container, and config for injection
        this.container.instance(Application, this);
        this.container.instance(Container_1.Container, this.container);
        this.container.instance(config_1.Repository, this.configRepository);
        this.container.alias(config_1.Repository, 'config');
        // Load environment variables from .env if it exists
        config_1.EnvLoader.load(this.basePath);
        // Load configuration
        const cachedConfigPath = path.join(this.basePath, 'bootstrap', 'cache', 'config.json');
        if (fs.existsSync(cachedConfigPath)) {
            try {
                const cachedData = fs.readFileSync(cachedConfigPath, 'utf8');
                this.configRepository = new config_1.Repository(JSON.parse(cachedData));
            }
            catch (e) {
                console.error('Failed to load cached config, falling back to directory loading...', e);
                const configPath = path.join(this.basePath, 'config');
                this.configRepository.loadConfigDirectory(configPath);
            }
        }
        else {
            const configPath = path.join(this.basePath, 'config');
            this.configRepository.loadConfigDirectory(configPath);
        }
        // Apply timezone from config to the Node.js process
        const timezone = this.configRepository.get('app.timezone');
        if (timezone) {
            process.env.TZ = timezone;
        }
    }
    /**
     * Get the globally available application instance.
     */
    static getInstance() {
        if (!this.instance) {
            throw new Error('Application instance has not been created yet.');
        }
        return this.instance;
    }
    /**
     * Get the base path of the application.
     */
    getBasePath() {
        return this.basePath;
    }
    /**
     * Access the underlying container.
     */
    getContainer() {
        return this.container;
    }
    /**
     * Access the configuration repository.
     */
    config() {
        return this.configRepository;
    }
    /**
     * Alias for make() - resolves a service from the container.
     */
    make(token) {
        // Check the container first — explicit bindings always win
        if (this.container.has(token)) {
            return this.container.make(token);
        }
        // If it's a config token (e.g. 'config.app.name'),
        // fall back to the configuration repository
        if (typeof token === 'string' && token.startsWith('config.')) {
            const configKey = token.substring(7);
            const value = this.configRepository.get(configKey);
            if (value !== undefined) {
                return value;
            }
        }
        // Final attempt — let the container throw if not found
        return this.container.make(token);
    }
    /**
     * Bind a transient service.
     */
    bind(token, factory) {
        this.container.bind(token, factory);
    }
    /**
     * Bind a singleton service.
     */
    singleton(token, factory) {
        this.container.singleton(token, factory);
    }
    /**
     * Register an existing instance.
     */
    instance(token, value) {
        this.container.instance(token, value);
    }
    /**
     * Register an alias for a token.
     */
    alias(token, alias) {
        this.container.alias(token, alias);
    }
    /**
     * Tag a token.
     */
    tag(token, tag) {
        this.container.tag(token, tag);
    }
    /**
     * Resolve all tagged services.
     */
    tagged(tag) {
        return this.container.tagged(tag);
    }
    /**
     * Extend a service.
     */
    extend(token, callback) {
        this.container.extend(token, callback);
    }
    /**
     * Register a service provider.
     * Can accept a class (will be instantiated) or an instance.
     */
    register(provider) {
        if (this.booted) {
            throw new Error('Cannot register providers after the application has been booted.');
        }
        const providerInstance = provider instanceof ServiceProvider_1.ServiceProvider
            ? provider
            : new provider(this);
        this.providers.push(providerInstance);
    }
    /**
     * Boot all registered service providers.
     * First runs all register() methods, then all boot() methods.
     */
    async boot() {
        if (this.booted) {
            return;
        }
        // Phase 1: Register all providers
        for (const provider of this.providers) {
            await provider.register();
        }
        // Phase 2: Boot all providers
        for (const provider of this.providers) {
            await provider.boot();
        }
        // Phase 3: Apply runtime configuration
        // Re-apply timezone in case it was changed after construction
        const timezone = this.configRepository.get('app.timezone');
        if (timezone) {
            process.env.TZ = timezone;
        }
        // Validate APP_KEY (unless in debug/local mode)
        if (!this.configRepository.get('app.key') && process.env.NODE_ENV === 'production') {
            throw new Error('Application key (APP_KEY) is not set. Please run "arika key:generate" to generate one.');
        }
        // Mark config as booted (read-only)
        this.configRepository.markAsBooted();
        this.booted = true;
    }
    /**
     * Run the application.
     */
    async run() {
        if (!this.booted) {
            await this.boot();
        }
    }
    /**
     * Gracefully terminate the application.
     */
    async terminate() {
        for (const callback of this.terminatingCallbacks) {
            await callback();
        }
    }
    /**
     * Register a callback to be run when the application is terminating.
     */
    onTerminate(callback) {
        this.terminatingCallbacks.push(callback);
    }
    /**
     * Alias for make() - resolves a service from the container.
     */
    resolve(token) {
        return this.make(token);
    }
    /**
     * Check if a service is registered in the container.
     */
    has(token) {
        return this.container.has(token);
    }
    /**
     * Check if the application has been booted.
     */
    isBooted() {
        return this.booted;
    }
}
exports.Application = Application;
Application.instance = null;
