// @arikajs/storage/dist/Disk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Disk = void 0;
const middleware_1 = require("@arikajs/middleware");
class Disk {
    constructor(driver) {
        this.middlewares = [];
        this.driver = driver;
    }
    /**
     * Register middleware for this disk.
     */
    middleware(middleware) {
        this.middlewares.push(middleware);
        return this;
    }
    /**
     * Create a pipeline for an operation.
     */
    async runPipeline(operation, params, callback) {
        if (this.middlewares.length === 0) {
            return await callback(params);
        }
        return await new middleware_1.Pipeline()
            .pipe(this.middlewares)
            .handle({ operation, ...params }, async (data) => {
            return await callback(data);
        });
    }
    /**
     * Write contents to a file.
     */
    async put(path, contents) {
        return await this.runPipeline('put', { path, contents }, async (data) => {
            return await this.driver.put(data.path, data.contents);
        });
    }
    /**
     * Write a stream to a file.
     */
    async putStream(path, stream) {
        return await this.runPipeline('putStream', { path, stream }, async (data) => {
            return await this.driver.putStream(data.path, data.stream);
        });
    }
    /**
     * Read file contents.
     */
    async get(path) {
        return await this.runPipeline('get', { path }, async (data) => {
            return await this.driver.get(data.path);
        });
    }
    /**
     * Read file as a stream.
     */
    readStream(path) {
        // Note: readStream is tricky with async middleware.
        // For now, we return the driver stream directly or 
        // we'd need a more complex implementation for streaming middleware.
        return this.driver.readStream(path);
    }
    /**
     * Check if a file exists.
     */
    async exists(path) {
        return await this.runPipeline('exists', { path }, async (data) => {
            return await this.driver.exists(data.path);
        });
    }
    /**
     * Delete a file.
     */
    async delete(path) {
        return await this.runPipeline('delete', { path }, async (data) => {
            return await this.driver.delete(data.path);
        });
    }
    /**
     * Get the URL for a file.
     */
    url(path) {
        return this.driver.url(path);
    }
    /**
     * Get a temporary URL for a file.
     */
    async temporaryUrl(path, expiresAt) {
        return await this.runPipeline('temporaryUrl', { path, expiresAt }, async (data) => {
            return await this.driver.temporaryUrl(data.path, data.expiresAt);
        });
    }
    /**
     * Get the file size in bytes.
     */
    async size(path) {
        return await this.runPipeline('size', { path }, async (data) => {
            return await this.driver.size(data.path);
        });
    }
    /**
     * Get the last modified time of the file.
     */
    async lastModified(path) {
        return await this.runPipeline('lastModified', { path }, async (data) => {
            return await this.driver.lastModified(data.path);
        });
    }
    /**
     * Get the mime type of the file.
     */
    async mimeType(path) {
        return await this.runPipeline('mimeType', { path }, async (data) => {
            return await this.driver.mimeType(data.path);
        });
    }
    /**
     * Copy a file to a new location.
     */
    async copy(path, newPath) {
        return await this.runPipeline('copy', { path, newPath }, async (data) => {
            if (typeof this.driver.copy === 'function') {
                return await this.driver.copy(data.path, data.newPath);
            }
            // Fallback
            const content = await this.driver.get(data.path);
            await this.driver.put(data.newPath, content);
        });
    }
    /**
     * Move a file to a new location.
     */
    async move(path, newPath) {
        return await this.runPipeline('move', { path, newPath }, async (data) => {
            if (typeof this.driver.move === 'function') {
                return await this.driver.move(data.path, data.newPath);
            }
            // Fallback
            const content = await this.driver.get(data.path);
            await this.driver.put(data.newPath, content);
            await this.driver.delete(data.path);
        });
    }
    /**
     * Append to a file.
     */
    async append(path, contents) {
        return await this.runPipeline('append', { path, contents }, async (data) => {
            if (typeof this.driver.append === 'function') {
                return await this.driver.append(data.path, data.contents);
            }
            // Fallback
            let existing = '';
            try {
                existing = await this.driver.get(data.path);
            }
            catch (e) { } // Ignore if it doesn't exist
            const separator = existing.length > 0 ? '\n' : '';
            const newContent = Buffer.isBuffer(existing)
                ? Buffer.concat([existing, Buffer.from(separator), Buffer.isBuffer(data.contents) ? data.contents : Buffer.from(data.contents)])
                : existing + separator + data.contents;
            return await this.driver.put(data.path, newContent);
        });
    }
    /**
     * Prepend to a file.
     */
    async prepend(path, contents) {
        return await this.runPipeline('prepend', { path, contents }, async (data) => {
            if (typeof this.driver.prepend === 'function') {
                return await this.driver.prepend(data.path, data.contents);
            }
            // Fallback
            let existing = '';
            try {
                existing = await this.driver.get(data.path);
            }
            catch (e) { } // Ignore if it doesn't exist
            const separator = existing.length > 0 ? '\n' : '';
            const newContent = Buffer.isBuffer(existing)
                ? Buffer.concat([Buffer.isBuffer(data.contents) ? data.contents : Buffer.from(data.contents), Buffer.from(separator), existing])
                : data.contents + separator + existing;
            return await this.driver.put(data.path, newContent);
        });
    }
}
exports.Disk = Disk;
