// @arikajs/authorization/dist/src/Gate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Gate = void 0;
const PolicyResolver_1 = require("./PolicyResolver");
const AuthorizationException_1 = require("./Exceptions/AuthorizationException");
const AuthResponse_1 = require("./AuthResponse");
class Gate {
    // ── Gate Definitions ────────────────────────────────────────────
    /**
     * Define a new ability.
     */
    static define(ability, callback) {
        this.abilities.set(ability, callback);
    }
    /**
     * Register a policy for a model.
     */
    static policy(model, policy) {
        this.policyResolver.register(model, policy);
    }
    // ── Before / After Hooks ────────────────────────────────────────
    /**
     * Register a callback to run before all gate checks.
     * Return `true` to allow immediately (super admin bypass).
     * Return `false` to deny immediately.
     * Return `null`/`undefined` to continue to the actual gate check.
     */
    static before(callback) {
        this.beforeCallbacks.push(callback);
    }
    /**
     * Register a callback to run after all gate checks.
     * Useful for logging, auditing decisions, etc.
     */
    static after(callback) {
        this.afterCallbacks.push(callback);
    }
    // ── User Binding ────────────────────────────────────────────────
    /**
     * Set the current user for authorization checks.
     */
    static forUser(user) {
        this.currentUser = user;
        return this;
    }
    // ── Single Ability Checks ───────────────────────────────────────
    /**
     * Determine if the user is authorized for an ability.
     * Returns the AuthResponse for rich deny messages.
     */
    static async inspect(ability, ...args) {
        // 1. Run before hooks
        for (const cb of this.beforeCallbacks) {
            const beforeResult = await cb(this.currentUser, ability, ...args);
            if (beforeResult === true) {
                return AuthResponse_1.AuthResponse.allow();
            }
            if (beforeResult === false) {
                return AuthResponse_1.AuthResponse.deny();
            }
            // null/undefined = continue
        }
        // 2. Check policy before() method
        if (args.length > 0) {
            const resource = args[0];
            const policy = this.policyResolver.resolvePolicy(resource);
            if (policy) {
                const instance = this.policyResolver.getInstance(policy);
                if (typeof instance.before === 'function') {
                    const policyBefore = await instance.before(this.currentUser, ability, ...args);
                    if (policyBefore === true)
                        return AuthResponse_1.AuthResponse.allow();
                    if (policyBefore === false)
                        return AuthResponse_1.AuthResponse.deny();
                }
                const method = this.policyResolver.getPolicyMethod(policy, ability);
                if (method) {
                    const result = await method(this.currentUser, ...args);
                    const response = this.normalizeResult(result);
                    await this.runAfterCallbacks(ability, response.allowed(), ...args);
                    return response;
                }
            }
        }
        // 3. Check direct gate definition
        if (this.abilities.has(ability)) {
            const callback = this.abilities.get(ability);
            const result = await callback(this.currentUser, ...args);
            const response = this.normalizeResult(result);
            await this.runAfterCallbacks(ability, response.allowed(), ...args);
            return response;
        }
        // 4. Default deny
        const response = AuthResponse_1.AuthResponse.deny();
        await this.runAfterCallbacks(ability, false, ...args);
        return response;
    }
    /**
     * Determine if the user is authorized to perform an ability.
     */
    static async allows(ability, ...args) {
        const response = await this.inspect(ability, ...args);
        return response.allowed();
    }
    /**
     * Determine if the user is NOT authorized.
     */
    static async denies(ability, ...args) {
        return !(await this.allows(ability, ...args));
    }
    /**
     * Authorize or throw exception with optional custom message.
     */
    static async authorize(ability, ...args) {
        const response = await this.inspect(ability, ...args);
        if (response.denied()) {
            throw new AuthorizationException_1.AuthorizationException(response.message() || 'This action is unauthorized.', response.code());
        }
    }
    /**
     * Check authorization (alias for allows).
     */
    static async check(ability, ...args) {
        return await this.allows(ability, ...args);
    }
    // ── Bulk Ability Checks ─────────────────────────────────────────
    /**
     * Check if the user can perform ANY of the given abilities.
     */
    static async any(abilities, ...args) {
        for (const ability of abilities) {
            if (await this.allows(ability, ...args)) {
                return true;
            }
        }
        return false;
    }
    /**
     * Check if the user can perform ALL of the given abilities.
     */
    static async every(abilities, ...args) {
        for (const ability of abilities) {
            if (!(await this.allows(ability, ...args))) {
                return false;
            }
        }
        return true;
    }
    /**
     * Check if the user can perform NONE of the given abilities.
     */
    static async none(abilities, ...args) {
        for (const ability of abilities) {
            if (await this.allows(ability, ...args)) {
                return false;
            }
        }
        return true;
    }
    // ── Internals ───────────────────────────────────────────────────
    static normalizeResult(result) {
        if (result instanceof AuthResponse_1.AuthResponse) {
            return result;
        }
        return result ? AuthResponse_1.AuthResponse.allow() : AuthResponse_1.AuthResponse.deny();
    }
    static async runAfterCallbacks(ability, result, ...args) {
        for (const cb of this.afterCallbacks) {
            await cb(this.currentUser, ability, result, ...args);
        }
    }
    /**
     * Reset all gates, policies, and hooks (useful for testing).
     */
    static reset() {
        this.abilities.clear();
        this.policyResolver = new PolicyResolver_1.PolicyResolver();
        this.beforeCallbacks = [];
        this.afterCallbacks = [];
        this.currentUser = null;
    }
}
exports.Gate = Gate;
Gate.abilities = new Map();
Gate.policyResolver = new PolicyResolver_1.PolicyResolver();
Gate.beforeCallbacks = [];
Gate.afterCallbacks = [];
Gate.currentUser = null;
