// @arikajs/authorization/dist/src/RolePermission.js
"use strict";
/**
 * Contracts for entities that have roles and permissions.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RolePermissionMixin = void 0;
/**
 * Mixin helper to add Role & Permission checking to any user object.
 * Works with both string arrays and object arrays ({name: string}).
 */
class RolePermissionMixin {
    /**
     * Normalize roles/permissions to string arrays.
     */
    static normalize(items) {
        if (!items)
            return [];
        return items.map((item) => typeof item === 'string' ? item : item.name);
    }
    static hasRole(user, role) {
        const userRoles = this.normalize(user.roles);
        if (Array.isArray(role)) {
            return role.every(r => userRoles.includes(r));
        }
        return userRoles.includes(role);
    }
    static hasAnyRole(user, roles) {
        const userRoles = this.normalize(user.roles);
        return roles.some(r => userRoles.includes(r));
    }
    static hasAllRoles(user, roles) {
        const userRoles = this.normalize(user.roles);
        return roles.every(r => userRoles.includes(r));
    }
    static hasPermission(user, permission) {
        // Check direct permissions
        const userPerms = this.normalize(user.permissions);
        if (userPerms.includes(permission))
            return true;
        // Check role-based permissions via a role→permissions map if user has rolePermissions
        if (user.rolePermissions && typeof user.rolePermissions === 'object') {
            const userRoles = this.normalize(user.roles);
            for (const role of userRoles) {
                const rolePerms = user.rolePermissions[role] || [];
                if (rolePerms.includes(permission))
                    return true;
            }
        }
        return false;
    }
    static hasAnyPermission(user, permissions) {
        return permissions.some(p => this.hasPermission(user, p));
    }
    static hasAllPermissions(user, permissions) {
        return permissions.every(p => this.hasPermission(user, p));
    }
}
exports.RolePermissionMixin = RolePermissionMixin;
