// @arikajs/database/dist/Model/GlobalScope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallbackGlobalScope = exports.GlobalScopeRegistry = void 0;
/**
 * Registry for global scopes on model classes
 */
class GlobalScopeRegistry {
    /**
     * Register a named global scope for a model class
     */
    static register(modelName, scopeName, scope) {
        if (!this.scopes.has(modelName)) {
            this.scopes.set(modelName, new Map());
        }
        this.scopes.get(modelName).set(scopeName, scope);
    }
    /**
     * Retrieve all global scopes for a model class
     */
    static getScopes(modelName) {
        return this.scopes.get(modelName) ?? new Map();
    }
    /**
     * Remove a specific named scope from a model
     */
    static remove(modelName, scopeName) {
        this.scopes.get(modelName)?.delete(scopeName);
    }
    /**
     * Apply all registered global scopes to a query builder
     */
    static applyAll(modelName, builder, model) {
        const scopes = this.getScopes(modelName);
        for (const scope of scopes.values()) {
            scope.apply(builder, model);
        }
    }
    /**
     * Clear all scopes (useful in tests)
     */
    static flush(modelName) {
        if (modelName) {
            this.scopes.delete(modelName);
        }
        else {
            this.scopes.clear();
        }
    }
}
exports.GlobalScopeRegistry = GlobalScopeRegistry;
GlobalScopeRegistry.scopes = new Map();
/**
 * A functional scope — can be registered inline without a class
 */
class CallbackGlobalScope {
    constructor(callback) {
        this.callback = callback;
    }
    apply(builder, model) {
        this.callback(builder, model);
    }
}
exports.CallbackGlobalScope = CallbackGlobalScope;
