// @arikajs/database/dist/Model/Relations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MorphTo = exports.MorphMany = exports.MorphOne = exports.HasManyThrough = exports.HasOneThrough = exports.BelongsToMany = exports.BelongsTo = exports.HasMany = exports.HasOne = exports.Relation = void 0;
const Database_1 = require("../Database");
// ─── Base ──────────────────────────────────────────────────────────────────
/**
 * Base class for all relationship types
 */
class Relation {
    constructor(related, parent) {
        this.related = related;
        this.parent = parent;
    }
    /**
     * Compare if a related model instance is the same as another
     */
    is(model) {
        if (!model)
            return false;
        const instance = new this.related();
        const pk = instance['getPrimaryKey']();
        return this.parent[pk] === model[pk];
    }
    isNot(model) {
        return !this.is(model);
    }
}
exports.Relation = Relation;
// ─── Has One ───────────────────────────────────────────────────────────────
class HasOne extends Relation {
    constructor(related, parent, foreignKey, localKey = 'id') {
        super(related, parent);
        this.foreignKey = foreignKey;
        this.localKey = localKey;
    }
    async get() {
        const parentKeyValue = this.parent[this.localKey];
        if (!parentKeyValue)
            return null;
        return await this.related
            .where(this.foreignKey, parentKeyValue)
            .first();
    }
    query() {
        const parentKeyValue = this.parent[this.localKey];
        return this.related.query().where(this.foreignKey, parentKeyValue);
    }
    getRelationCountQuery(outerTable, outerKey = this.localKey) {
        const relatedTable = this.related.table;
        return {
            sql: `(SELECT COUNT(*) FROM ${relatedTable} WHERE ${relatedTable}.${this.foreignKey} = ${outerTable}.${outerKey})`,
            bindings: []
        };
    }
    /**
     * Create a related record and associate it with the parent
     */
    async create(data) {
        const parentKeyValue = this.parent[this.localKey];
        return this.related.query().create({
            ...data,
            [this.foreignKey]: parentKeyValue,
        });
    }
    /**
     * Save a related model and associate it with the parent
     */
    async save(model) {
        const parentKeyValue = this.parent[this.localKey];
        model.setAttribute(this.foreignKey, parentKeyValue);
        await model.save();
        return model;
    }
}
exports.HasOne = HasOne;
// ─── Has Many ──────────────────────────────────────────────────────────────
class HasMany extends Relation {
    constructor(related, parent, foreignKey, localKey = 'id') {
        super(related, parent);
        this.foreignKey = foreignKey;
        this.localKey = localKey;
    }
    async get() {
        const parentKeyValue = this.parent[this.localKey];
        if (!parentKeyValue)
            return [];
        return await this.related
            .where(this.foreignKey, parentKeyValue)
            .get();
    }
    query() {
        const parentKeyValue = this.parent[this.localKey];
        return this.related.query().where(this.foreignKey, parentKeyValue);
    }
    getRelationCountQuery(outerTable, outerKey = this.localKey) {
        const relatedTable = this.related.table;
        return {
            sql: `(SELECT COUNT(*) FROM ${relatedTable} WHERE ${relatedTable}.${this.foreignKey} = ${outerTable}.${outerKey})`,
            bindings: []
        };
    }
    /**
     * Create a related record
     */
    async create(data) {
        const parentKeyValue = this.parent[this.localKey];
        return this.related.query().create({
            ...data,
            [this.foreignKey]: parentKeyValue,
        });
    }
    /**
     * Bulk-create multiple related records
     */
    async createMany(records) {
        return Promise.all(records.map(data => this.create(data)));
    }
    /**
     * Save a related model and associate it with the parent
     */
    async save(model) {
        const parentKeyValue = this.parent[this.localKey];
        model.setAttribute(this.foreignKey, parentKeyValue);
        await model.save();
        return model;
    }
    /**
     * Save multiple related models
     */
    async saveMany(models) {
        return Promise.all(models.map(m => this.save(m)));
    }
}
exports.HasMany = HasMany;
// ─── Belongs To ────────────────────────────────────────────────────────────
class BelongsTo extends Relation {
    constructor(related, parent, foreignKey, ownerKey = 'id') {
        super(related, parent);
        this.foreignKey = foreignKey;
        this.ownerKey = ownerKey;
    }
    async get() {
        const foreignKeyValue = this.parent[this.foreignKey];
        if (!foreignKeyValue)
            return null;
        return await this.related
            .where(this.ownerKey, foreignKeyValue)
            .first();
    }
    query() {
        const foreignKeyValue = this.parent[this.foreignKey];
        return this.related.query().where(this.ownerKey, foreignKeyValue);
    }
    getRelationCountQuery(outerTable, outerKey = this.foreignKey) {
        const relatedTable = this.related.table;
        return {
            sql: `(SELECT COUNT(*) FROM ${relatedTable} WHERE ${relatedTable}.${this.ownerKey} = ${outerTable}.${outerKey})`,
            bindings: []
        };
    }
    /**
     * Associate this model with the given related model (sets the FK on parent)
     */
    associate(model) {
        this.parent.setAttribute(this.foreignKey, model[this.ownerKey]);
        return this;
    }
    /**
     * Dissociate — nullify the foreign key on the parent
     */
    dissociate() {
        this.parent.setAttribute(this.foreignKey, null);
        return this;
    }
}
exports.BelongsTo = BelongsTo;
class BelongsToMany extends Relation {
    constructor(related, parent, pivotTable, foreignPivotKey, relatedPivotKey, parentKey = 'id', relatedKey = 'id') {
        super(related, parent);
        this.pivotTable = pivotTable;
        this.foreignPivotKey = foreignPivotKey;
        this.relatedPivotKey = relatedPivotKey;
        this.parentKey = parentKey;
        this.relatedKey = relatedKey;
        this.pivotColumns = [];
        this.pivotWheres = [];
    }
    /**
     * Include extra pivot columns in the result
     */
    withPivot(...columns) {
        this.pivotColumns.push(...columns);
        return this;
    }
    /**
     * Filter results by a pivot column value
     */
    wherePivot(column, value) {
        this.pivotWheres.push({ column, value });
        return this;
    }
    async get() {
        const parentKeyValue = this.parent[this.parentKey];
        if (!parentKeyValue)
            return [];
        const relatedTable = this.related.table;
        const connection = await Database_1.Database.connection(this.related.connection);
        // Build the pivot extra columns selector
        const pivotCols = this.pivotColumns
            .map(c => `${this.pivotTable}.${c} AS __pivot_${c}`)
            .join(', ');
        const selectExtra = pivotCols ? `, ${pivotCols}` : '';
        // Build pivot where clauses
        const pivotWhereSQL = this.pivotWheres
            .map(w => `AND ${this.pivotTable}.${w.column} = ?`)
            .join(' ');
        const pivotWhereBindings = this.pivotWheres.map(w => w.value);
        const sql = `
            SELECT ${relatedTable}.*${selectExtra}
            FROM ${relatedTable}
            INNER JOIN ${this.pivotTable}
                ON ${relatedTable}.${this.relatedKey} = ${this.pivotTable}.${this.relatedPivotKey}
            WHERE ${this.pivotTable}.${this.foreignPivotKey} = ?
            ${pivotWhereSQL}
        `;
        const results = await connection.query(sql, [parentKeyValue, ...pivotWhereBindings]);
        return results.map((data) => {
            const instance = new this.related();
            // Separate pivot columns from model columns
            const pivotData = {};
            const modelData = {};
            for (const [key, val] of Object.entries(data)) {
                if (key.startsWith('__pivot_')) {
                    pivotData[key.replace('__pivot_', '')] = val;
                }
                else {
                    modelData[key] = val;
                }
            }
            instance.fill(modelData);
            instance.setExists(true);
            // Attach pivot data as a property on the model
            instance.pivot = pivotData;
            return instance;
        });
    }
    getRelationCountQuery(outerTable, outerKey = this.parentKey) {
        const relatedTable = this.related.table;
        const pivotWhereSQL = this.pivotWheres
            .map(w => `AND ${this.pivotTable}.${w.column} = ?`)
            .join(' ');
        const pivotWhereBindings = this.pivotWheres.map(w => w.value);
        const sql = `(SELECT COUNT(*) FROM ${relatedTable} INNER JOIN ${this.pivotTable} ON ${relatedTable}.${this.relatedKey} = ${this.pivotTable}.${this.relatedPivotKey} WHERE ${this.pivotTable}.${this.foreignPivotKey} = ${outerTable}.${outerKey} ${pivotWhereSQL})`;
        return { sql, bindings: pivotWhereBindings };
    }
    /**
     * Attach one or many related models to the pivot table
     */
    async attach(ids, pivotData = {}) {
        const parentKeyValue = this.parent[this.parentKey];
        const connection = await Database_1.Database.connection(this.related.connection);
        const idList = Array.isArray(ids) ? ids : [ids];
        const extraCols = Object.keys(pivotData);
        const extraVals = Object.values(pivotData);
        for (const relatedId of idList) {
            const cols = [this.foreignPivotKey, this.relatedPivotKey, ...extraCols].join(', ');
            const placeholders = [parentKeyValue, relatedId, ...extraVals].map(() => '?').join(', ');
            const sql = `INSERT INTO ${this.pivotTable} (${cols}) VALUES (${placeholders})`;
            await connection.query(sql, [parentKeyValue, relatedId, ...extraVals]);
        }
    }
    /**
     * Detach one, many, or all related models from the pivot table
     */
    async detach(ids) {
        const parentKeyValue = this.parent[this.parentKey];
        const connection = await Database_1.Database.connection(this.related.connection);
        if (ids === undefined) {
            // Detach all
            await connection.query(`DELETE FROM ${this.pivotTable} WHERE ${this.foreignPivotKey} = ?`, [parentKeyValue]);
        }
        else {
            const idList = Array.isArray(ids) ? ids : [ids];
            const placeholders = idList.map(() => '?').join(', ');
            await connection.query(`DELETE FROM ${this.pivotTable} WHERE ${this.foreignPivotKey} = ? AND ${this.relatedPivotKey} IN (${placeholders})`, [parentKeyValue, ...idList]);
        }
    }
    /**
     * Sync the pivot table — attach missing, detach extras
     */
    async sync(ids, pivotData = {}) {
        const parentKeyValue = this.parent[this.parentKey];
        const connection = await Database_1.Database.connection(this.related.connection);
        // Get current pivot IDs
        const existing = await connection.query(`SELECT ${this.relatedPivotKey} FROM ${this.pivotTable} WHERE ${this.foreignPivotKey} = ?`, [parentKeyValue]);
        const existingIds = existing.map((row) => row[this.relatedPivotKey]);
        const toAttach = ids.filter(id => !existingIds.includes(id));
        const toDetach = existingIds.filter((id) => !ids.includes(id));
        if (toAttach.length > 0)
            await this.attach(toAttach, pivotData);
        if (toDetach.length > 0)
            await this.detach(toDetach);
    }
    /**
     * Toggle — attach if not present, detach if already present
     */
    async toggle(ids) {
        const parentKeyValue = this.parent[this.parentKey];
        const connection = await Database_1.Database.connection(this.related.connection);
        const idList = Array.isArray(ids) ? ids : [ids];
        const existing = await connection.query(`SELECT ${this.relatedPivotKey} FROM ${this.pivotTable} WHERE ${this.foreignPivotKey} = ?`, [parentKeyValue]);
        const existingIds = existing.map((row) => row[this.relatedPivotKey]);
        const toAttach = idList.filter((id) => !existingIds.includes(id));
        const toDetach = idList.filter((id) => existingIds.includes(id));
        if (toAttach.length > 0)
            await this.attach(toAttach);
        if (toDetach.length > 0)
            await this.detach(toDetach);
    }
}
exports.BelongsToMany = BelongsToMany;
// ─── Has One Through ───────────────────────────────────────────────────────
/**
 * HasOneThrough — relate models through an intermediate table.
 *
 * Example: Country → hasOneThrough(Capital, User)
 *   countries → through users → capitals
 */
class HasOneThrough extends Relation {
    constructor(related, parent, through, firstKey, // FK on through table pointing to parent
    secondKey, // FK on related table pointing to through
    localKey = 'id', // PK of parent
    secondLocalKey = 'id' // PK of through model
    ) {
        super(related, parent);
        this.through = through;
        this.firstKey = firstKey;
        this.secondKey = secondKey;
        this.localKey = localKey;
        this.secondLocalKey = secondLocalKey;
    }
    async get() {
        const parentKeyValue = this.parent[this.localKey];
        if (!parentKeyValue)
            return null;
        const throughTable = this.through.table;
        const relatedTable = this.related.table;
        const connection = await Database_1.Database.connection(this.related.connection);
        const sql = `
            SELECT ${relatedTable}.*
            FROM ${relatedTable}
            INNER JOIN ${throughTable}
                ON ${relatedTable}.${this.secondKey} = ${throughTable}.${this.secondLocalKey}
            WHERE ${throughTable}.${this.firstKey} = ?
            LIMIT 1
        `;
        const results = await connection.query(sql, [parentKeyValue]);
        if (!results.length)
            return null;
        const instance = new this.related();
        instance.fill(results[0]);
        instance.setExists(true);
        return instance;
    }
    getRelationCountQuery(outerTable, outerKey = this.localKey) {
        const throughTable = this.through.table;
        const relatedTable = this.related.table;
        const sql = `(SELECT COUNT(*) FROM ${relatedTable} INNER JOIN ${throughTable} ON ${relatedTable}.${this.secondKey} = ${throughTable}.${this.secondLocalKey} WHERE ${throughTable}.${this.firstKey} = ${outerTable}.${outerKey})`;
        return { sql, bindings: [] };
    }
}
exports.HasOneThrough = HasOneThrough;
// ─── Has Many Through ──────────────────────────────────────────────────────
/**
 * HasManyThrough — relate models through an intermediate table.
 *
 * Example: Country → hasManyThrough(Post, User)
 *   countries → through users → posts
 */
class HasManyThrough extends Relation {
    constructor(related, parent, through, firstKey, secondKey, localKey = 'id', secondLocalKey = 'id') {
        super(related, parent);
        this.through = through;
        this.firstKey = firstKey;
        this.secondKey = secondKey;
        this.localKey = localKey;
        this.secondLocalKey = secondLocalKey;
    }
    async get() {
        const parentKeyValue = this.parent[this.localKey];
        if (!parentKeyValue)
            return [];
        const throughTable = this.through.table;
        const relatedTable = this.related.table;
        const connection = await Database_1.Database.connection(this.related.connection);
        const sql = `
            SELECT ${relatedTable}.*
            FROM ${relatedTable}
            INNER JOIN ${throughTable}
                ON ${relatedTable}.${this.secondKey} = ${throughTable}.${this.secondLocalKey}
            WHERE ${throughTable}.${this.firstKey} = ?
        `;
        const results = await connection.query(sql, [parentKeyValue]);
        return results.map((data) => {
            const instance = new this.related();
            instance.fill(data);
            instance.setExists(true);
            return instance;
        });
    }
    query() {
        const parentKeyValue = this.parent[this.localKey];
        const throughTable = this.through.table;
        const relatedTable = this.related.table;
        return Database_1.Database.table(relatedTable)
            .whereRaw(`${relatedTable}.${this.secondKey} IN (SELECT ${this.secondLocalKey} FROM ${throughTable} WHERE ${this.firstKey} = ?)`, [parentKeyValue]);
    }
    getRelationCountQuery(outerTable, outerKey = this.localKey) {
        const throughTable = this.through.table;
        const relatedTable = this.related.table;
        const sql = `(SELECT COUNT(*) FROM ${relatedTable} INNER JOIN ${throughTable} ON ${relatedTable}.${this.secondKey} = ${throughTable}.${this.secondLocalKey} WHERE ${throughTable}.${this.firstKey} = ${outerTable}.${outerKey})`;
        return { sql, bindings: [] };
    }
}
exports.HasManyThrough = HasManyThrough;
// ─── Polymorphic Relationships ─────────────────────────────────────────────
/**
 * MorphOne — polymorphic has-one
 *
 * Example: Post → morphOne(Image, 'imageable')
 *   images table has: imageable_id, imageable_type
 */
class MorphOne extends Relation {
    constructor(related, parent, morphName, // e.g. 'imageable'
    morphType, // column: imageable_type
    morphId, // column: imageable_id
    localKey = 'id') {
        super(related, parent);
        this.morphName = morphName;
        this.morphType = morphType;
        this.morphId = morphId;
        this.localKey = localKey;
    }
    getMorphType() {
        return this.parent.constructor.name;
    }
    async get() {
        const parentKeyValue = this.parent[this.localKey];
        if (!parentKeyValue)
            return null;
        return await this.related
            .where(this.morphType, this.getMorphType())
            .where(this.morphId, parentKeyValue)
            .first();
    }
    query() {
        const parentKeyValue = this.parent[this.localKey];
        return this.related.query()
            .where(this.morphType, this.getMorphType())
            .where(this.morphId, parentKeyValue);
    }
    async create(data) {
        const parentKeyValue = this.parent[this.localKey];
        return this.related.query().create({
            ...data,
            [this.morphType]: this.getMorphType(),
            [this.morphId]: parentKeyValue,
        });
    }
    getRelationCountQuery(outerTable, outerKey = this.localKey) {
        const relatedTable = this.related.table;
        const sql = `(SELECT COUNT(*) FROM ${relatedTable} WHERE ${relatedTable}.${this.morphId} = ${outerTable}.${outerKey} AND ${relatedTable}.${this.morphType} = ?)`;
        return { sql, bindings: [this.getMorphType()] };
    }
}
exports.MorphOne = MorphOne;
/**
 * MorphMany — polymorphic has-many
 *
 * Example: Post → morphMany(Comment, 'commentable')
 */
class MorphMany extends Relation {
    constructor(related, parent, morphName, morphType, morphId, localKey = 'id') {
        super(related, parent);
        this.morphName = morphName;
        this.morphType = morphType;
        this.morphId = morphId;
        this.localKey = localKey;
    }
    getMorphType() {
        return this.parent.constructor.name;
    }
    async get() {
        const parentKeyValue = this.parent[this.localKey];
        if (!parentKeyValue)
            return [];
        return await this.related
            .where(this.morphType, this.getMorphType())
            .where(this.morphId, parentKeyValue)
            .get();
    }
    query() {
        const parentKeyValue = this.parent[this.localKey];
        return this.related.query()
            .where(this.morphType, this.getMorphType())
            .where(this.morphId, parentKeyValue);
    }
    async create(data) {
        const parentKeyValue = this.parent[this.localKey];
        return this.related.query().create({
            ...data,
            [this.morphType]: this.getMorphType(),
            [this.morphId]: parentKeyValue,
        });
    }
    async createMany(records) {
        return Promise.all(records.map(d => this.create(d)));
    }
    getRelationCountQuery(outerTable, outerKey = this.localKey) {
        const relatedTable = this.related.table;
        const sql = `(SELECT COUNT(*) FROM ${relatedTable} WHERE ${relatedTable}.${this.morphId} = ${outerTable}.${outerKey} AND ${relatedTable}.${this.morphType} = ?)`;
        return { sql, bindings: [this.getMorphType()] };
    }
}
exports.MorphMany = MorphMany;
/**
 * MorphTo — inverse polymorphic relationship
 *
 * Example: Comment → morphTo('commentable')
 *   reads commentable_type + commentable_id to find the owner
 */
class MorphTo extends Relation {
    constructor(parent, morphName, morphType, // column: commentable_type
    morphId, // column: commentable_id
    morphMap = {}) {
        // Pass parent as related too (placeholder — resolved dynamically)
        super(parent.constructor, parent);
        this.morphName = morphName;
        this.morphType = morphType;
        this.morphId = morphId;
        this.morphMap = new Map(Object.entries(morphMap));
    }
    /**
     * Register a morph type → Model class mapping
     */
    map(morphMap) {
        for (const [key, val] of Object.entries(morphMap)) {
            this.morphMap.set(key, val);
        }
        return this;
    }
    async get() {
        const morphType = this.parent[this.morphType];
        const morphId = this.parent[this.morphId];
        if (!morphType || !morphId)
            return null;
        const relatedClass = this.morphMap.get(morphType);
        if (!relatedClass) {
            throw new Error(`No morph map entry for type "${morphType}". ` +
                `Call .map({ '${morphType}': YourModel }) on this MorphTo instance.`);
        }
        return await relatedClass.find(morphId);
    }
    getRelationCountQuery(outerTable, outerKey = this.morphId) {
        throw new Error('Calling withCount on a MorphTo relationship is not supported yet natively. Please count inside the mapped relations.');
    }
}
exports.MorphTo = MorphTo;
