// @arikajs/database/dist/Model/SoftDeletes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withSoftDeletes = withSoftDeletes;
/**
 * Mixin to add soft delete functionality to models
 *
 * Usage:
 * ```typescript
 * class User extends withSoftDeletes(Model) {
 *   static table = 'users';
 * }
 *
 * await User.delete(1); // Soft deletes
 * await User.withTrashed().get(); // Include soft deleted
 * await User.restore(1); // Restore soft deleted
 * await User.forceDelete(1); // Permanently delete
 * ```
 */
function withSoftDeletes(Base) {
    class SoftDeleteModel extends Base {
        constructor(...args) {
            super(...args);
        }
        /**
         * Override delete to use soft deletes
         */
        static async delete(id) {
            if (this.useSoftDeletes) {
                const primaryKey = this.getPrimaryKeyName();
                return await this.query()
                    .where(primaryKey, id)
                    .update({ [this.deletedAtColumn]: new Date() });
            }
            // Call the parent delete method
            return await this.query().delete(id);
        }
        /**
         * Include trashed (soft deleted) records in query
         */
        static withTrashed() {
            // Create a fresh query without soft delete filtering
            const connection = this.getConnectionName();
            const table = this.getTableName();
            const { Database } = require('../Database');
            const queryBuilder = Database.table(table, connection);
            const { ModelQueryBuilder } = require('./Model');
            return new ModelQueryBuilder(queryBuilder, this);
        }
        /**
         * Only get trashed (soft deleted) records
         */
        static onlyTrashed() {
            return this.withTrashed().whereNotNull(this.deletedAtColumn);
        }
        /**
         * Restore a soft deleted record
         */
        static async restore(id) {
            const primaryKey = this.getPrimaryKeyName();
            return await this.withTrashed()
                .where(primaryKey, id)
                .update({ [this.deletedAtColumn]: null });
        }
        /**
         * Force delete a record (permanently delete)
         */
        static async forceDelete(id) {
            const primaryKey = this.getPrimaryKeyName();
            return await this.withTrashed()
                .where(primaryKey, id)
                .delete();
        }
        /**
         * Override query to automatically exclude soft deleted records
         */
        static query() {
            const baseQuery = super.query();
            const queryBuilder = baseQuery;
            if (this.useSoftDeletes) {
                queryBuilder.whereNull(this.deletedAtColumn);
            }
            return queryBuilder;
        }
    }
    /**
     * The name of the "deleted at" column
     */
    SoftDeleteModel.deletedAtColumn = 'deleted_at';
    /**
     * Indicates if the model should use soft deletes
     */
    SoftDeleteModel.useSoftDeletes = true;
    return SoftDeleteModel;
}
