// @arikajs/view/dist/src/Compiler/CodeGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeGenerator = void 0;
const AST_1 = require("./AST");
class CodeGenerator {
    constructor(registry) {
        this.registry = registry;
    }
    generate(root) {
        let jsCode = 'let _output = "";\n';
        jsCode += 'const _escape = (val) => String(val ?? "").replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/\\\'/g, "&#39;");\n';
        jsCode += this.generateNodes(root.children);
        jsCode += 'return _output;';
        return jsCode;
    }
    generateNodes(nodes) {
        return nodes.map(node => this.generateNode(node)).join('\n');
    }
    generateNode(node) {
        switch (node.type) {
            case AST_1.NodeType.Text:
                const text = node.content
                    .replace(/\\/g, '\\\\')
                    .replace(/`/g, '\\`')
                    .replace(/\${/g, '\\${');
                return `_output += \`${text}\`;`;
            case AST_1.NodeType.Expression:
                return `_output += _escape(${this.cleanExpression(node.content)});`;
            case AST_1.NodeType.RawExpression:
                return `_output += (${this.cleanExpression(node.content)});`;
            case AST_1.NodeType.Directive:
                const dir = node;
                const childrenCode = dir.children ? this.generateNodes(dir.children) : '';
                const result = this.registry.handle(dir.name, dir.expression, childrenCode);
                if (result === null) {
                    // Fallback for unknown directives (e.g., CSS @media, @keyframes)
                    const originalText = `@${dir.name}${dir.expression ? '(' + dir.expression + ')' : ''}`;
                    const sanitized = originalText.replace(/`/g, '\\`').replace(/\${/g, '\\${');
                    return `_output += \`${sanitized}\`;\n${childrenCode}`;
                }
                return result;
            default:
                return '';
        }
    }
    cleanExpression(exp) {
        // Strip leading $ from variables to support Laravel-style {{ $var }}
        // This regex finds $ followed by an identifier and replaces it with just the identifier
        return exp.replace(/(^|[^a-zA-Z0-9_])\$([a-zA-Z_][a-zA-Z0-9_]*)/g, '$1$2');
    }
}
exports.CodeGenerator = CodeGenerator;
