// @arikajs/view/dist/src/Compiler/Parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Parser = void 0;
const Lexer_1 = require("./Lexer");
const AST_1 = require("./AST");
class Parser {
    constructor(tokens) {
        this.position = 0;
        this.tokens = tokens;
    }
    parse() {
        const root = {
            type: AST_1.NodeType.Root,
            children: []
        };
        while (this.position < this.tokens.length) {
            const lastPos = this.position;
            const node = this.parseNode();
            if (node) {
                root.children.push(node);
            }
            if (this.position <= lastPos && this.position < this.tokens.length) {
                this.position++; // Safety advance
            }
        }
        return root;
    }
    parseNode() {
        const token = this.tokens[this.position];
        if (token.type === Lexer_1.TokenType.Text) {
            this.position++;
            return { type: AST_1.NodeType.Text, content: token.value };
        }
        if (token.type === Lexer_1.TokenType.Expression) {
            this.position++;
            return { type: AST_1.NodeType.Expression, content: token.value };
        }
        if (token.type === Lexer_1.TokenType.RawExpression) {
            this.position++;
            return { type: AST_1.NodeType.RawExpression, content: token.value };
        }
        if (token.type === Lexer_1.TokenType.Directive) {
            return this.parseDirective();
        }
        if (token.type === Lexer_1.TokenType.ComponentStart) {
            return this.parseComponent();
        }
        this.position++;
        return null;
    }
    parseDirective() {
        const token = this.tokens[this.position];
        this.position++;
        const match = token.value.match(/@(\w+)(?:\s*\((.*)\))?/s);
        const name = match ? match[1] : token.value.substring(1);
        const expression = match && match[2] ? match[2].trim() : null;
        const node = {
            type: AST_1.NodeType.Directive,
            name,
            expression,
            line: token.line
        };
        // Check if this is a block directive (if, for, each, etc.)
        if (this.isBlockDirective(name)) {
            node.children = [];
            const endName = this.getEndDirectiveName(name);
            while (this.position < this.tokens.length) {
                const nextToken = this.tokens[this.position];
                if (nextToken.type === Lexer_1.TokenType.Directive) {
                    const nextMatch = nextToken.value.match(/@(\w+)/);
                    const nextName = nextMatch ? nextMatch[1] : '';
                    if (nextName === endName) {
                        this.position++; // Consume end directive
                        break;
                    }
                    // Handle else/elseif/case/default which are part of the same block but start new sub-blocks?
                    // For now, let's keep it simple and just nest them as children.
                }
                const child = this.parseNode();
                if (child) {
                    node.children.push(child);
                }
            }
        }
        return node;
    }
    parseComponent() {
        const token = this.tokens[this.position];
        this.position++;
        // Basic parsing of <x-component-name attr="val">
        const match = token.value.match(/<x-([\w\.-]+)\s*(.*)>/s);
        const name = match ? match[1] : '';
        const attributesRaw = match ? match[2] : '';
        const node = {
            type: AST_1.NodeType.Directive,
            name: 'component',
            expression: this.parseComponentAttributes(name, attributesRaw),
            line: token.line,
            children: []
        };
        // Components always have a closing tag </x-component-name>
        const endTag = `</x-${name}>`;
        if (endTag) {
            while (this.position < this.tokens.length) {
                const nextToken = this.tokens[this.position];
                if (nextToken.type === Lexer_1.TokenType.ComponentEnd && nextToken.value.trim() === endTag) {
                    this.position++;
                    break;
                }
                const child = this.parseNode();
                if (child && node.children) {
                    node.children.push(child);
                }
            }
        }
        return node;
    }
    parseComponentAttributes(name, raw) {
        // Convert attributes to a JS object string
        // type="danger" :user="user"
        const attrs = {};
        const regex = /([:@]?[\w\.-]+)(?:="([^"]*)")?/g;
        let match;
        while ((match = regex.exec(raw)) !== null) {
            const key = match[1];
            const value = match[2] || 'true';
            if (key.startsWith(':')) {
                // Binding
                attrs[key.substring(1)] = value;
            }
            else {
                // Literal
                attrs[key] = `'${value}'`;
            }
        }
        const attrsStr = Object.entries(attrs).map(([k, v]) => `${k}: ${v}`).join(', ');
        return `'${name}', { ${attrsStr} }`;
    }
    isBlockDirective(name) {
        const blockDirectives = [
            // Core
            'if', 'for', 'each', 'unless', 'empty', 'section', 'push', 'prepend',
            'component', 'slot', 'auth', 'guest', 'switch', 'once', 'verbatim', 'fragment', 'await',
            // New
            'forelse', 'isset', 'unset', 'env', 'production', 'can', 'cannot', 'canany', 'error', 'style'
        ];
        return blockDirectives.includes(name);
    }
    getEndDirectiveName(name) {
        if (name === 'section')
            return 'endsection';
        if (name === 'push')
            return 'endpush';
        if (name === 'prepend')
            return 'endprepend';
        if (name === 'component')
            return 'endcomponent';
        if (name === 'slot')
            return 'endslot';
        if (name === 'verbatim')
            return 'endverbatim';
        if (name === 'fragment')
            return 'endfragment';
        if (name === 'forelse')
            return 'endforelse';
        if (name === 'isset')
            return 'endisset';
        if (name === 'unset')
            return 'endunset';
        if (name === 'env')
            return 'endenv';
        if (name === 'production')
            return 'endproduction';
        if (name === 'can')
            return 'endcan';
        if (name === 'cannot')
            return 'endcannot';
        if (name === 'canany')
            return 'endcanany';
        if (name === 'error')
            return 'enderror';
        if (name === 'style')
            return 'endstyle';
        return `end${name}`;
    }
}
exports.Parser = Parser;
