// @arikajs/middleware/dist/Pipeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pipeline = void 0;
/**
 * Pipeline executes a stack of middleware in an onion-style model.
 */
class Pipeline {
    /**
     * Create a new Pipeline instance.
     */
    constructor(container) {
        this.container = container;
        /**
         * The stack of middleware handlers.
         */
        this.handlers = [];
        this.middlewareGroups = {};
        this.aliases = {};
    }
    /**
     * Set the middleware groups.
     */
    setMiddlewareGroups(groups) {
        this.middlewareGroups = groups;
        return this;
    }
    /**
     * Set the middleware aliases.
     */
    setAliases(aliases) {
        this.aliases = aliases;
        return this;
    }
    /**
     * Add middleware to the pipeline.
     */
    pipe(middleware) {
        if (Array.isArray(middleware)) {
            this.handlers.push(...middleware);
        }
        else {
            this.handlers.push(middleware);
        }
        return this;
    }
    /**
     * Run the pipeline through the given destination.
     */
    async handle(request, destination, response) {
        const flattened = this.flattenHandlers(this.handlers);
        const invoke = async (index, req) => {
            if (index >= flattened.length) {
                return destination(req, response);
            }
            const { handler: rawHandler, args } = this.parseHandler(flattened[index]);
            const handler = this.resolve(rawHandler);
            if (typeof handler === 'function') {
                return handler(req, (nextReq) => invoke(index + 1, nextReq), response, ...args);
            }
            if (typeof handler === 'object' && 'handle' in handler && typeof handler.handle === 'function') {
                return handler.handle(req, (nextReq) => invoke(index + 1, nextReq), response, ...args);
            }
            throw new Error(`Invalid middleware handler: ${typeof handler}`);
        };
        return invoke(0, request);
    }
    /**
     * Resolve the middleware handler.
     */
    resolve(handler) {
        // If it's a string, try resolving from aliases first, then container
        if (typeof handler === 'string') {
            if (this.aliases[handler]) {
                return this.resolve(this.aliases[handler]);
            }
            if (this.container && this.container.has(handler)) {
                return this.container.make(handler);
            }
            return handler;
        }
        // If it's a class/constructor (has handle on prototype), instantiate it
        if (typeof handler === 'function') {
            let isClass = Pipeline.isClassCache.get(handler);
            if (isClass === undefined) {
                const check = /^\s*class\s+/.test(handler.toString()) ||
                    (handler.prototype && typeof handler.prototype.handle === 'function');
                isClass = !!check;
                Pipeline.isClassCache.set(handler, isClass);
            }
            if (isClass) {
                if (this.container) {
                    return this.container.make(handler);
                }
                return new handler();
            }
        }
        return handler;
    }
    /**
     * Flatten handlers by resolving groups and aliases.
     */
    flattenHandlers(handlers) {
        let flattened = [];
        for (const handler of handlers) {
            if (typeof handler === 'string') {
                const [name, args] = handler.split(':');
                if (this.middlewareGroups[name]) {
                    flattened.push(...this.flattenHandlers(this.middlewareGroups[name]));
                    continue;
                }
                if (this.aliases[name]) {
                    const resolved = this.aliases[name];
                    // If it's an alias but we have args, we keep it as a string to parse later in handle()
                    if (args) {
                        flattened.push(handler);
                    }
                    else if (Array.isArray(resolved)) {
                        flattened.push(...this.flattenHandlers(resolved));
                    }
                    else {
                        flattened.push(...this.flattenHandlers([resolved]));
                    }
                    continue;
                }
            }
            flattened.push(handler);
        }
        return flattened;
    }
    /**
     * Parse handler string to extract arguments.
     */
    parseHandler(handler) {
        if (typeof handler !== 'string') {
            return { handler, args: [] };
        }
        const [name, ...args] = handler.split(':');
        return {
            handler: name,
            args: args.length > 0 ? args[0].split(',') : []
        };
    }
}
exports.Pipeline = Pipeline;
Pipeline.isClassCache = new WeakMap();
