// @arikajs/storage/dist/Drivers/AzureDriver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureDriver = void 0;
const storage_blob_1 = require("@azure/storage-blob");
const mime = __importStar(require("mime-types"));
class AzureDriver {
    constructor(config) {
        this.client = storage_blob_1.BlobServiceClient.fromConnectionString(config.connectionString);
        this.containerName = config.container;
        this.container = this.client.getContainerClient(config.container);
        this.urlEndpoint = config.url;
    }
    async put(path, contents) {
        const blockBlobClient = this.container.getBlockBlobClient(path);
        const buffer = Buffer.isBuffer(contents) ? contents : Buffer.from(contents);
        await blockBlobClient.uploadData(buffer, {
            blobHTTPHeaders: {
                blobContentType: mime.lookup(path) || 'application/octet-stream'
            }
        });
    }
    async putStream(path, stream) {
        const blockBlobClient = this.container.getBlockBlobClient(path);
        await blockBlobClient.uploadStream(stream, 4 * 1024 * 1024, 20, {
            blobHTTPHeaders: {
                blobContentType: mime.lookup(path) || 'application/octet-stream'
            }
        });
    }
    async get(path) {
        try {
            const blockBlobClient = this.container.getBlockBlobClient(path);
            return await blockBlobClient.downloadToBuffer();
        }
        catch (error) {
            if (error.statusCode === 404) {
                throw new Error(`File not found: ${path}`);
            }
            throw error;
        }
    }
    readStream(path) {
        const blockBlobClient = this.container.getBlockBlobClient(path);
        const { PassThrough } = require('stream');
        const passThrough = new PassThrough();
        blockBlobClient.download().then((response) => {
            if (response.readableStreamBody) {
                response.readableStreamBody.pipe(passThrough);
            }
        }).catch((err) => {
            passThrough.emit('error', err);
        });
        return passThrough;
    }
    async exists(path) {
        const blockBlobClient = this.container.getBlockBlobClient(path);
        return await blockBlobClient.exists();
    }
    async delete(path) {
        try {
            const blockBlobClient = this.container.getBlockBlobClient(path);
            await blockBlobClient.delete();
        }
        catch (error) {
            if (error.statusCode !== 404)
                throw error;
        }
    }
    url(path) {
        if (this.urlEndpoint) {
            return `${this.urlEndpoint}/${path}`;
        }
        const blockBlobClient = this.container.getBlockBlobClient(path);
        return blockBlobClient.url;
    }
    async temporaryUrl(path, expiresAt) {
        try {
            const blockBlobClient = this.container.getBlockBlobClient(path);
            const permissions = new storage_blob_1.BlobSASPermissions();
            permissions.read = true;
            const sasUrl = await blockBlobClient.generateSasUrl({
                permissions,
                expiresOn: expiresAt
            });
            return sasUrl;
        }
        catch (e) {
            if (e.message?.includes('generateSasUrl') || e.message?.includes('StorageSharedKeyCredential')) {
                throw new Error("Temporary URL requires Shared Key Credential which must be parsed from Connection String. Alternatively configure SAS manually.");
            }
            throw e;
        }
    }
    async size(path) {
        const blockBlobClient = this.container.getBlockBlobClient(path);
        const properties = await blockBlobClient.getProperties();
        return properties.contentLength || 0;
    }
    async lastModified(path) {
        const blockBlobClient = this.container.getBlockBlobClient(path);
        const properties = await blockBlobClient.getProperties();
        return properties.lastModified?.getTime() || 0;
    }
    async mimeType(path) {
        const blockBlobClient = this.container.getBlockBlobClient(path);
        const properties = await blockBlobClient.getProperties();
        return properties.contentType || mime.lookup(path) || 'application/octet-stream';
    }
}
exports.AzureDriver = AzureDriver;
