// @arikajs/storage/dist/Drivers/GCSDriver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GCSDriver = void 0;
const storage_1 = require("@google-cloud/storage");
const mime = __importStar(require("mime-types"));
class GCSDriver {
    constructor(config) {
        const gcsConfig = {};
        if (config.projectId)
            gcsConfig.projectId = config.projectId;
        if (config.keyFilename)
            gcsConfig.keyFilename = config.keyFilename;
        if (config.credentials)
            gcsConfig.credentials = config.credentials;
        this.client = new storage_1.Storage(gcsConfig);
        this.bucketName = config.bucket;
        this.bucket = this.client.bucket(config.bucket);
        this.urlEndpoint = config.url;
    }
    getFile(path) {
        return this.bucket.file(path);
    }
    async put(path, contents) {
        const file = this.getFile(path);
        await file.save(contents, {
            contentType: mime.lookup(path) || 'application/octet-stream',
            resumable: false
        });
    }
    async putStream(path, stream) {
        return new Promise((resolve, reject) => {
            const file = this.getFile(path);
            const writeStream = file.createWriteStream({
                contentType: mime.lookup(path) || 'application/octet-stream',
                resumable: false
            });
            stream.pipe(writeStream)
                .on('error', reject)
                .on('finish', resolve);
        });
    }
    async get(path) {
        try {
            const [content] = await this.getFile(path).download();
            return content;
        }
        catch (error) {
            if (error.code === 404) {
                throw new Error(`File not found: ${path}`);
            }
            throw error;
        }
    }
    readStream(path) {
        return this.getFile(path).createReadStream();
    }
    async exists(path) {
        const [exists] = await this.getFile(path).exists();
        return exists;
    }
    async delete(path) {
        try {
            await this.getFile(path).delete();
        }
        catch (error) {
            // Ignore if file doesn't exist
            if (error.code !== 404)
                throw error;
        }
    }
    url(path) {
        if (this.urlEndpoint) {
            return `${this.urlEndpoint}/${path}`;
        }
        return `https://storage.googleapis.com/${this.bucketName}/${path}`;
    }
    async temporaryUrl(path, expiresAt) {
        const [url] = await this.getFile(path).getSignedUrl({
            version: 'v4',
            action: 'read',
            expires: expiresAt.getTime(),
        });
        return url;
    }
    async size(path) {
        const [metadata] = await this.getFile(path).getMetadata();
        return parseInt(metadata.size || '0', 10);
    }
    async lastModified(path) {
        const [metadata] = await this.getFile(path).getMetadata();
        return metadata.updated ? new Date(metadata.updated).getTime() : 0;
    }
    async mimeType(path) {
        const [metadata] = await this.getFile(path).getMetadata();
        return metadata.contentType || mime.lookup(path) || 'application/octet-stream';
    }
    async copy(path, newPath) {
        await this.getFile(path).copy(this.getFile(newPath));
    }
    async move(path, newPath) {
        await this.getFile(path).move(this.getFile(newPath));
    }
}
exports.GCSDriver = GCSDriver;
