// @arikajs/storage/dist/Drivers/LocalDriver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalDriver = void 0;
const fs = __importStar(require("fs/promises"));
const fs_1 = require("fs");
const path = __importStar(require("path"));
const promises_1 = require("stream/promises");
const mime = __importStar(require("mime-types"));
const FileNotFoundException_1 = require("../Exceptions/FileNotFoundException");
class LocalDriver {
    constructor(config) {
        this.root = path.resolve(config.root);
        this.urlPrefix = config.url || '';
    }
    /**
     * Get the full path for a file.
     */
    getFullPath(filePath) {
        return path.join(this.root, filePath);
    }
    /**
     * Ensure directory exists.
     */
    async ensureDirectory(filePath) {
        const dir = path.dirname(filePath);
        await fs.mkdir(dir, { recursive: true });
    }
    /**
     * Write contents to a file.
     */
    async put(filePath, contents) {
        const fullPath = this.getFullPath(filePath);
        await this.ensureDirectory(fullPath);
        await fs.writeFile(fullPath, contents);
    }
    /**
     * Write a stream to a file.
     */
    async putStream(filePath, stream) {
        const fullPath = this.getFullPath(filePath);
        await this.ensureDirectory(fullPath);
        const writeStream = (0, fs_1.createWriteStream)(fullPath);
        await (0, promises_1.pipeline)(stream, writeStream);
    }
    /**
     * Read file contents.
     */
    async get(filePath) {
        const fullPath = this.getFullPath(filePath);
        if (!(await this.exists(filePath))) {
            throw new FileNotFoundException_1.FileNotFoundException(filePath);
        }
        return await fs.readFile(fullPath);
    }
    /**
     * Read file as a stream.
     */
    readStream(filePath) {
        const fullPath = this.getFullPath(filePath);
        return (0, fs_1.createReadStream)(fullPath);
    }
    /**
     * Check if a file exists.
     */
    async exists(filePath) {
        const fullPath = this.getFullPath(filePath);
        try {
            await fs.access(fullPath);
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * Delete a file.
     */
    async delete(filePath) {
        const fullPath = this.getFullPath(filePath);
        if (await this.exists(filePath)) {
            await fs.unlink(fullPath);
        }
    }
    /**
     * Get the URL for a file.
     */
    url(filePath) {
        if (!this.urlPrefix) {
            return filePath;
        }
        return `${this.urlPrefix}/${filePath}`.replace(/\/+/g, '/').replace(':/', '://');
    }
    /**
     * Get a temporary URL for a file.
     * Note: For local driver, this just returns the normal URL as local files don't support signing by default
     * without a custom server implementation.
     */
    async temporaryUrl(filePath, expiresAt) {
        return this.url(filePath);
    }
    /**
     * Get the file size in bytes.
     */
    async size(filePath) {
        const stats = await fs.stat(this.getFullPath(filePath));
        return stats.size;
    }
    /**
     * Get the last modified time of the file.
     */
    async lastModified(filePath) {
        const stats = await fs.stat(this.getFullPath(filePath));
        return stats.mtimeMs;
    }
    /**
     * Get the mime type of the file.
     */
    async mimeType(filePath) {
        return mime.lookup(filePath) || 'application/octet-stream';
    }
}
exports.LocalDriver = LocalDriver;
