// @arikajs/validation/dist/tests/AdvancedValidation.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = require("node:test");
const node_assert_1 = __importDefault(require("node:assert"));
const src_1 = require("../src");
(0, node_test_1.describe)('Advanced Validation Features', () => {
    (0, node_test_1.it)('handles nullable rule', async () => {
        const data = { email: null };
        const rules = { email: 'nullable|email' };
        const validator = new src_1.Validator(data, rules);
        const passed = await validator.validate();
        node_assert_1.default.strictEqual(passed, true);
        node_assert_1.default.deepStrictEqual(validator.validated(), { email: null });
    });
    (0, node_test_1.it)('handles bail rule', async () => {
        const data = { email: 'invalid-email' };
        const rules = { email: 'bail|email|min:10' };
        const validator = new src_1.Validator(data, rules);
        await validator.validate();
        const errors = validator.errors();
        // Should only have one error because of bail
        node_assert_1.default.strictEqual(errors.email.length, 1);
        node_assert_1.default.strictEqual(errors.email[0], 'The email must be a valid email address.');
    });
    (0, node_test_1.it)('validates nested objects with dot notation', async () => {
        const data = {
            user: {
                profile: {
                    age: 15
                }
            }
        };
        const rules = {
            'user.profile.age': 'number|min:18'
        };
        const validator = new src_1.Validator(data, rules);
        const passed = await validator.validate();
        node_assert_1.default.strictEqual(passed, false);
        node_assert_1.default.ok(validator.errors()['user.profile.age']);
    });
    (0, node_test_1.it)('validates arrays with wildcards', async () => {
        const data = {
            users: [
                { email: 'valid@example.com' },
                { email: 'invalid' }
            ]
        };
        const rules = {
            'users.*.email': 'required|email'
        };
        const validator = new src_1.Validator(data, rules);
        const passed = await validator.validate();
        node_assert_1.default.strictEqual(passed, false);
        node_assert_1.default.ok(validator.errors()['users.1.email']);
        node_assert_1.default.ok(!validator.errors()['users.0.email']);
    });
    (0, node_test_1.it)('handles conditional validation (required_if)', async () => {
        const data = {
            provider: 'email',
            password: ''
        };
        const rules = {
            password: 'required_if:provider,email|min:8'
        };
        const validator = new src_1.Validator(data, rules);
        const passed = await validator.validate();
        node_assert_1.default.strictEqual(passed, false);
        node_assert_1.default.ok(validator.errors().password);
        // Test fallback if condition not met
        const data2 = { provider: 'google', password: '' };
        const validator2 = new src_1.Validator(data2, rules);
        node_assert_1.default.strictEqual(await validator2.validate(), true);
    });
    (0, node_test_1.it)('returns only validated data with validated()', async () => {
        const data = {
            name: 'Prakash',
            email: 'prakash@example.com',
            extra: 'ignore me'
        };
        const rules = {
            name: 'required',
            email: 'required|email'
        };
        const validator = new src_1.Validator(data, rules);
        await validator.validate();
        const validated = validator.validated();
        node_assert_1.default.deepStrictEqual(validated, {
            name: 'Prakash',
            email: 'prakash@example.com'
        });
        node_assert_1.default.ok(!('extra' in validated));
    });
});
