// @arikajs/validation/dist/tests/Validator.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = require("node:test");
const assert = __importStar(require("node:assert"));
const src_1 = require("../src");
(0, node_test_1.describe)('Arika Validation', () => {
    (0, node_test_1.it)('validates required fields', async () => {
        const data = { name: '' };
        const rules = { name: 'required' };
        const validator = new src_1.Validator(data, rules);
        assert.strictEqual(await validator.fails(), true);
        assert.deepStrictEqual(validator.errors(), {
            name: ['The name field is required.']
        });
        const validData = { name: 'Arika' };
        const validValidator = new src_1.Validator(validData, rules);
        assert.strictEqual(await validValidator.fails(), false);
    });
    (0, node_test_1.it)('validates email fields', async () => {
        const data = { email: 'invalid-email' };
        const rules = { email: 'email' };
        const validator = new src_1.Validator(data, rules);
        assert.strictEqual(await validator.fails(), true);
        assert.deepStrictEqual(validator.errors(), {
            email: ['The email must be a valid email address.']
        });
        const validData = { email: 'test@example.com' };
        const validValidator = new src_1.Validator(validData, rules);
        assert.strictEqual(await validValidator.fails(), false);
    });
    (0, node_test_1.it)('validates min rule', async () => {
        const data = { password: 'short', age: 10 };
        const rules = { password: 'min:8', age: 'min:18' };
        const validator = new src_1.Validator(data, rules);
        assert.strictEqual(await validator.fails(), true);
        const errors = validator.errors();
        assert.ok(errors.password);
        assert.ok(errors.age);
    });
    (0, node_test_1.it)('validates max rule', async () => {
        const data = { username: 'reallylongusername', score: 101 };
        const rules = { username: 'max:10', score: 'max:100' };
        const validator = new src_1.Validator(data, rules);
        assert.strictEqual(await validator.fails(), true);
        const errors = validator.errors();
        assert.ok(errors.username);
        assert.ok(errors.score);
    });
    (0, node_test_1.it)('validates string rule', async () => {
        const data = { name: 'Arika', age: 25 };
        const rules = { name: 'string', age: 'string' };
        const validator = new src_1.Validator(data, rules);
        assert.strictEqual(await validator.fails(), true);
        const errors = validator.errors();
        assert.ok(errors.age);
        assert.strictEqual(errors.age[0], 'The age must be a string.');
        assert.ok(!errors.name);
    });
    (0, node_test_1.it)('validates number rule', async () => {
        const data = { age: 25, name: 'Arika' };
        const rules = { age: 'number', name: 'number' };
        const validator = new src_1.Validator(data, rules);
        assert.strictEqual(await validator.fails(), true);
        const errors = validator.errors();
        assert.ok(errors.name);
        assert.strictEqual(errors.name[0], 'The name must be a number.');
        assert.ok(!errors.age);
    });
    (0, node_test_1.it)('supports custom rule classes', async () => {
        class Uppercase {
            validate(value) {
                return value === value.toUpperCase();
            }
            message() {
                return 'Must be uppercase';
            }
        }
        const data = { code: 'abc' };
        const validator = new src_1.Validator(data);
        validator.extend('uppercase', new Uppercase());
        // Directly test validate method as extension happens after constructor
        const rules = { code: 'uppercase' };
        const v = new src_1.Validator(data, rules);
        v.extend('uppercase', new Uppercase());
        assert.strictEqual(await v.fails(), true);
        assert.deepStrictEqual(v.errors(), {
            code: ['Must be uppercase']
        });
    });
    (0, node_test_1.it)('supports pipe syntax for multiple rules', async () => {
        const data = { email: 'invalid' };
        const rules = { email: 'required|email|min:10' };
        const validator = new src_1.Validator(data, rules);
        assert.strictEqual(await validator.fails(), true);
        const errors = validator.errors();
        // Should have at least one error (email format)
        assert.strictEqual(errors.email.includes('The email must be a valid email address.'), true);
    });
});
