// @arikajs/cache/dist/src/Drivers/FileDriver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileDriver = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class FileDriver {
    constructor(directory, prefix = '') {
        this.prefix = prefix;
        this.directory = directory;
        this.ensureDirectoryExists();
    }
    async get(key) {
        const path = this.getPath(key);
        if (!fs_1.default.existsSync(path)) {
            return null;
        }
        try {
            const content = fs_1.default.readFileSync(path, 'utf8');
            const data = JSON.parse(content);
            if (data.expiration && Date.now() / 1000 > data.expiration) {
                await this.forget(key);
                return null;
            }
            return data.value;
        }
        catch (e) {
            return null;
        }
    }
    async put(key, value, seconds) {
        const path = this.getPath(key);
        const expiration = seconds > 0 ? (Date.now() / 1000) + seconds : 0;
        const data = {
            value,
            expiration
        };
        try {
            this.ensureDirectoryExists(path);
            fs_1.default.writeFileSync(path, JSON.stringify(data), 'utf8');
        }
        catch (e) {
            // Silently fail or log
        }
    }
    async increment(key, value = 1) {
        const current = await this.get(key);
        const newValue = (typeof current === 'number' ? current : 0) + value;
        await this.put(key, newValue, 0);
        return newValue;
    }
    async decrement(key, value = 1) {
        return this.increment(key, -value);
    }
    async forever(key, value) {
        await this.put(key, value, 0);
    }
    async forget(key) {
        const path = this.getPath(key);
        if (fs_1.default.existsSync(path)) {
            fs_1.default.unlinkSync(path);
        }
    }
    async flush() {
        if (fs_1.default.existsSync(this.directory)) {
            this.deleteFolderRecursive(this.directory);
            this.ensureDirectoryExists();
        }
    }
    getPrefix() {
        return this.prefix;
    }
    getPath(key) {
        // Simple hash-based partitioning like Laravel: 2 levels deep
        const hash = this.sha1(this.prefix + key);
        const folder1 = hash.substring(0, 2);
        const folder2 = hash.substring(2, 4);
        return path_1.default.join(this.directory, folder1, folder2, hash);
    }
    ensureDirectoryExists(filePath) {
        const dir = filePath ? path_1.default.dirname(filePath) : this.directory;
        if (!fs_1.default.existsSync(dir)) {
            fs_1.default.mkdirSync(dir, { recursive: true });
        }
    }
    deleteFolderRecursive(dir) {
        if (fs_1.default.existsSync(dir)) {
            fs_1.default.readdirSync(dir).forEach((file) => {
                const curPath = path_1.default.join(dir, file);
                if (fs_1.default.lstatSync(curPath).isDirectory()) {
                    this.deleteFolderRecursive(curPath);
                }
                else {
                    fs_1.default.unlinkSync(curPath);
                }
            });
            // Don't delete the root directory itself if it's the cache root
            if (dir !== this.directory) {
                fs_1.default.rmdirSync(dir);
            }
        }
    }
    sha1(str) {
        const crypto = require('crypto');
        return crypto.createHash('sha1').update(str).digest('hex');
    }
}
exports.FileDriver = FileDriver;
