// @arikajs/cache/dist/src/Drivers/MemoryDriver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryDriver = void 0;
class MemoryDriver {
    constructor() {
        this.storage = new Map();
    }
    async get(key) {
        const item = this.storage.get(key);
        if (!item) {
            return null;
        }
        if (item.expiresAt !== null && item.expiresAt < Date.now()) {
            this.forget(key);
            return null;
        }
        return item.value;
    }
    async put(key, value, seconds) {
        const expiresAt = Date.now() + (seconds * 1000);
        this.storage.set(key, { value, expiresAt });
    }
    async increment(key, value = 1) {
        const current = await this.get(key) || 0;
        const newValue = current + value;
        // We set it forever if it was already existing or start now.
        // Usually increment doesn't change TTL if exists, or sets a default.
        // For simplicity, we just update the value.
        const item = this.storage.get(key);
        this.storage.set(key, {
            value: newValue,
            expiresAt: item ? item.expiresAt : null
        });
        return newValue;
    }
    async decrement(key, value = 1) {
        return this.increment(key, -value);
    }
    async forever(key, value) {
        this.storage.set(key, { value, expiresAt: null });
    }
    async forget(key) {
        this.storage.delete(key);
    }
    async flush() {
        this.storage.clear();
    }
    async add(key, value, seconds) {
        const item = this.storage.get(key);
        if (item && (item.expiresAt === null || item.expiresAt > Date.now())) {
            return false;
        }
        await this.put(key, value, seconds);
        return true;
    }
    async getMultiple(keys) {
        const results = {};
        for (const key of keys) {
            results[key] = await this.get(key);
        }
        return results;
    }
    async putMultiple(values, seconds) {
        for (const [key, value] of Object.entries(values)) {
            await this.put(key, value, seconds);
        }
    }
    async forgetMultiple(keys) {
        for (const key of keys) {
            this.storage.delete(key);
        }
    }
    getPrefix() {
        return '';
    }
}
exports.MemoryDriver = MemoryDriver;
