// @arikajs/cache/dist/src/Drivers/RedisDriver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisDriver = void 0;
const ioredis_1 = __importDefault(require("ioredis"));
class RedisDriver {
    constructor(config, prefix = '') {
        this.prefix = prefix;
        const mode = config.mode || 'standalone';
        if (mode === 'cluster') {
            this.redis = new ioredis_1.default.Cluster(config.nodes, {
                redisOptions: {
                    password: config.password,
                    ...config.options
                }
            });
        }
        else if (mode === 'sentinel') {
            this.redis = new ioredis_1.default({
                sentinels: config.nodes,
                name: config.name,
                password: config.password,
                db: config.database,
                ...config.options
            });
        }
        else {
            this.redis = new ioredis_1.default({
                host: config.host,
                port: config.port,
                password: config.password,
                db: config.database,
                ...config.options
            });
        }
    }
    async get(key) {
        const value = await this.redis.get(this.prefix + key);
        return value ? JSON.parse(value) : null;
    }
    async put(key, value, seconds) {
        const serializedValue = JSON.stringify(value);
        if (seconds > 0) {
            await this.redis.set(this.prefix + key, serializedValue, 'EX', seconds);
        }
        else {
            await this.forever(key, value);
        }
    }
    async increment(key, value = 1) {
        return await this.redis.incrby(this.prefix + key, value);
    }
    async decrement(key, value = 1) {
        return await this.redis.decrby(this.prefix + key, value);
    }
    async forever(key, value) {
        await this.redis.set(this.prefix + key, JSON.stringify(value));
    }
    async forget(key) {
        await this.redis.del(this.prefix + key);
    }
    async flush() {
        await this.redis.flushdb();
    }
    async add(key, value, seconds) {
        const serializedValue = JSON.stringify(value);
        let result;
        if (seconds > 0) {
            result = await this.redis.set(this.prefix + key, serializedValue, 'EX', seconds, 'NX');
        }
        else {
            result = await this.redis.set(this.prefix + key, serializedValue, 'NX');
        }
        return result === 'OK';
    }
    async getMultiple(keys) {
        if (keys.length === 0)
            return {};
        const prefixedKeys = keys.map(k => this.prefix + k);
        const values = await this.redis.mget(...prefixedKeys);
        const result = {};
        for (let i = 0; i < keys.length; i++) {
            result[keys[i]] = values[i] ? JSON.parse(values[i]) : null;
        }
        return result;
    }
    async putMultiple(values, seconds) {
        if (Object.keys(values).length === 0)
            return;
        if (seconds > 0) {
            const pipeline = this.redis.pipeline();
            for (const [key, value] of Object.entries(values)) {
                pipeline.set(this.prefix + key, JSON.stringify(value), 'EX', seconds);
            }
            await pipeline.exec();
        }
        else {
            const multiValues = {};
            for (const [key, value] of Object.entries(values)) {
                multiValues[this.prefix + key] = JSON.stringify(value);
            }
            await this.redis.mset(multiValues);
        }
    }
    async forgetMultiple(keys) {
        if (keys.length === 0)
            return;
        const prefixedKeys = keys.map(k => this.prefix + k);
        await this.redis.del(...prefixedKeys);
    }
    getPrefix() {
        return this.prefix;
    }
    /**
     * Get the underlying Redis instance.
     */
    getRedis() {
        return this.redis;
    }
}
exports.RedisDriver = RedisDriver;
