// @arikajs/validation/dist/ErrorBag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorBag = void 0;
class ErrorBag {
    constructor() {
        this.errors = {};
    }
    /**
     * Add an error message for a specific field.
     */
    add(field, message) {
        if (!this.errors[field]) {
            this.errors[field] = [];
        }
        this.errors[field].push(message);
    }
    /**
     * Determine if the error bag is empty.
     */
    isEmpty() {
        return Object.keys(this.errors).length === 0;
    }
    /**
     * Determine if there are any errors.
     */
    isNotEmpty() {
        return !this.isEmpty();
    }
    /**
     * Get all error messages.
     */
    all() {
        return this.errors;
    }
    /**
     * Get the first error message for a specific field.
     */
    first(field) {
        return this.errors[field] ? this.errors[field][0] : null;
    }
    /**
     * Determine if a field has an error.
     */
    has(field) {
        return !!this.errors[field];
    }
}
exports.ErrorBag = ErrorBag;
