// @arikajs/validation/dist/ValidationError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationError = void 0;
class ValidationError extends Error {
    constructor(errors) {
        super('The given data was invalid.');
        this.errors = errors;
        this.status = 422;
        this.name = 'ValidationError';
    }
    /**
     * Render the error into an HTTP response.
     */
    render(request, response) {
        if (request.expectsJson()) {
            return response.status(this.status).json({
                message: this.message,
                errors: this.errors
            });
        }
        // For web requests, we redirect back with errors and old input
        // Note: ArikaJS should have a session.flashErrors or similar. 
        // For now, we'll redirect back. If sessions are enabled, they will handle the flashing.
        if (typeof response.back === 'function') {
            return response.back(request);
        }
        return response.status(this.status).json({
            message: this.message,
            errors: this.errors
        });
    }
}
exports.ValidationError = ValidationError;
