// @arikajs/validation/dist/Validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Validator = void 0;
const ErrorBag_1 = require("./ErrorBag");
const Required_1 = require("./Rules/Required");
const Email_1 = require("./Rules/Email");
const Min_1 = require("./Rules/Min");
const Max_1 = require("./Rules/Max");
const String_1 = require("./Rules/String");
const Number_1 = require("./Rules/Number");
const RequiredIf_1 = require("./Rules/RequiredIf");
const In_1 = require("./Rules/In");
const NotIn_1 = require("./Rules/NotIn");
const Alpha_1 = require("./Rules/Alpha");
const AlphaNum_1 = require("./Rules/AlphaNum");
const Url_1 = require("./Rules/Url");
const Boolean_1 = require("./Rules/Boolean");
const IsArray_1 = require("./Rules/IsArray");
const Confirmed_1 = require("./Rules/Confirmed");
class Validator {
    constructor(data, rules = {}, messages = {}) {
        this.messages = {};
        this.customRules = new Map();
        this.validatedData = {};
        this.data = data;
        this.initialRules = rules;
        this.messages = messages;
        this.errorBag = new ErrorBag_1.ErrorBag();
        // Register default rules
        this.customRules.set('required', new Required_1.Required());
        this.customRules.set('email', new Email_1.Email());
        this.customRules.set('string', new String_1.IsString());
        this.customRules.set('number', new Number_1.IsNumber());
        this.customRules.set('alpha', new Alpha_1.Alpha());
        this.customRules.set('alpha_num', new AlphaNum_1.AlphaNum());
        this.customRules.set('url', new Url_1.Url());
        this.customRules.set('boolean', new Boolean_1.IsBoolean());
        this.customRules.set('array', new IsArray_1.IsArray());
    }
    /**
     * Determine if the data passes the validation rules.
     */
    async validate() {
        this.errorBag = new ErrorBag_1.ErrorBag();
        this.validatedData = {};
        const expandedRules = this.expandRules(this.initialRules);
        for (const [attribute, rules] of Object.entries(expandedRules)) {
            const ruleSet = this.parseRules(rules);
            const value = this.getValue(attribute);
            const isNullable = ruleSet.some(r => r.name === 'nullable');
            const shouldBail = ruleSet.some(r => r.name === 'bail');
            // Handle nullable
            if (isNullable && (value === null || value === undefined)) {
                this.setValidatedValue(attribute, value);
                continue;
            }
            let failed = false;
            for (const item of ruleSet) {
                const { rule, name: ruleName } = item;
                // Skip pseudo-rules
                if (ruleName === 'nullable' || ruleName === 'bail') {
                    continue;
                }
                // If value is null/undefined/empty and rule is NOT 'required' (or similar), skip validation
                if (this.isEmpty(value) && !this.isRequirementRule(rule)) {
                    continue;
                }
                const passes = await rule.validate(value, attribute);
                if (!passes) {
                    const defaultMessage = rule.message();
                    const message = this.formatMessage(defaultMessage, attribute, ruleName);
                    this.errorBag.add(attribute, message);
                    failed = true;
                    if (shouldBail) {
                        break;
                    }
                }
            }
            if (!failed) {
                this.setValidatedValue(attribute, value);
            }
        }
        return this.errorBag.isEmpty();
    }
    isRequirementRule(rule) {
        return rule instanceof Required_1.Required || rule instanceof RequiredIf_1.RequiredIf;
    }
    async fails() {
        return !(await this.validate());
    }
    errors() {
        return this.errorBag.all();
    }
    validated() {
        return this.validatedData;
    }
    /**
     * Register a custom validation rule.
     */
    extend(name, rule) {
        this.customRules.set(name, rule);
    }
    parseRules(rules) {
        let parsed;
        if (typeof rules === 'string') {
            parsed = rules.split('|');
        }
        else if (Array.isArray(rules)) {
            parsed = rules;
        }
        else {
            parsed = [rules];
        }
        return parsed.flatMap(r => {
            if (typeof r === 'string') {
                return r.split('|').map(sub => this.createRuleFromString(sub));
            }
            return { rule: r, name: r.constructor.name.toLowerCase() };
        });
    }
    createRuleFromString(ruleString) {
        const [fullRule, ...rest] = ruleString.split(':');
        const name = fullRule.trim();
        const params = rest.length > 0 ? rest.join(':') : '';
        const parameters = params ? params.split(',') : [];
        let rule;
        switch (name) {
            case 'min':
                rule = new Min_1.Min(Number(parameters[0]));
                break;
            case 'max':
                rule = new Max_1.Max(Number(parameters[0]));
                break;
            case 'in':
                rule = new In_1.In(parameters);
                break;
            case 'not_in':
                rule = new NotIn_1.NotIn(parameters);
                break;
            case 'confirmed':
                rule = new Confirmed_1.Confirmed(this.data);
                break;
            case 'required_if':
                rule = new RequiredIf_1.RequiredIf(parameters[0], parameters[1], this.data);
                break;
            case 'nullable':
            case 'bail':
                rule = {
                    validate: async () => true,
                    message: () => ''
                };
                break;
            default:
                const registered = this.customRules.get(name);
                if (!registered) {
                    throw new Error(`Validation rule '${name}' not found.`);
                }
                rule = registered;
        }
        return { rule, name };
    }
    getValue(key) {
        return key.split('.').reduce((o, i) => (o ? o[i] : undefined), this.data);
    }
    setValidatedValue(path, value) {
        const keys = path.split('.');
        let current = this.validatedData;
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            if (i === keys.length - 1) {
                current[key] = value;
            }
            else {
                current[key] = current[key] || {};
                current = current[key];
            }
        }
    }
    isEmpty(value) {
        return value === null || value === undefined || value === '';
    }
    formatMessage(defaultMessage, attribute, ruleName) {
        // Look for exact attribute.rule custom message
        let template = this.messages[`${attribute}.${ruleName}`];
        // Look for general rule custom message
        if (!template) {
            template = this.messages[ruleName];
        }
        // Fallback to default message
        if (!template) {
            template = defaultMessage;
        }
        return template.replace(/:attribute/g, attribute);
    }
    expandRules(rules) {
        const expanded = {};
        for (const [key, rule] of Object.entries(rules)) {
            if (key.includes('*')) {
                this.expandWildcards(key, rule, this.data, expanded);
            }
            else {
                expanded[key] = rule;
            }
        }
        return expanded;
    }
    expandWildcards(path, rule, data, expanded) {
        const parts = path.split('.');
        const starIndex = parts.indexOf('*');
        if (starIndex === -1) {
            expanded[path] = rule;
            return;
        }
        const before = parts.slice(0, starIndex).join('.');
        const after = parts.slice(starIndex + 1).join('.');
        const collection = this.getValue(before);
        if (Array.isArray(collection)) {
            for (let i = 0; i < collection.length; i++) {
                const subPath = after ? `${before}.${i}.${after}` : `${before}.${i}`;
                if (subPath.includes('*')) {
                    this.expandWildcards(subPath, rule, data, expanded);
                }
                else {
                    expanded[subPath] = rule;
                }
            }
        }
    }
}
exports.Validator = Validator;
