// @arikajs/cache/dist/CacheManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheManager = void 0;
const Repository_1 = require("./Repository");
const MemoryDriver_1 = require("./Drivers/MemoryDriver");
const DatabaseDriver_1 = require("./Drivers/DatabaseDriver");
const RedisDriver_1 = require("./Drivers/RedisDriver");
const FileDriver_1 = require("./Drivers/FileDriver");
class CacheManager {
    constructor(config, database, basePath = process.cwd()) {
        this.config = config;
        this.stores = new Map();
        this.customCreators = new Map();
        this.database = database;
        this.basePath = basePath;
    }
    store(name) {
        const storeName = name || this.config.default;
        if (!this.stores.has(storeName)) {
            this.stores.set(storeName, this.resolve(storeName));
        }
        return this.stores.get(storeName);
    }
    resolve(name) {
        const config = this.config.stores[name];
        if (!config) {
            throw new Error(`Cache store [${name}] is not defined.`);
        }
        if (this.customCreators.has(config.driver)) {
            return new Repository_1.Repository(this.customCreators.get(config.driver)(config));
        }
        const driverMethod = `create${config.driver.charAt(0).toUpperCase() + config.driver.slice(1)}Driver`;
        if (typeof this[driverMethod] === 'function') {
            return new Repository_1.Repository(this[driverMethod](config));
        }
        throw new Error(`Driver [${config.driver}] is not supported.`);
    }
    createMemoryDriver() {
        return new MemoryDriver_1.MemoryDriver();
    }
    createDatabaseDriver(config) {
        return new DatabaseDriver_1.DatabaseDriver(this.database, config.table, this.config.prefix || '', config.connection);
    }
    createRedisDriver(config) {
        return new RedisDriver_1.RedisDriver(config, this.config.prefix || '');
    }
    createFileDriver(config) {
        const path = require('path');
        const directory = path.isAbsolute(config.path)
            ? config.path
            : path.resolve(this.basePath, config.path);
        return new FileDriver_1.FileDriver(directory, this.config.prefix || '');
    }
    extend(driver, callback) {
        this.customCreators.set(driver, callback);
        return this;
    }
    // Proxy common methods to default store
    async get(key, defaultValue = null) { return this.store().get(key, defaultValue); }
    async put(key, value, seconds) { return this.store().put(key, value, seconds); }
    async has(key) { return this.store().has(key); }
    async forever(key, value) { return this.store().forever(key, value); }
    async forget(key) { return this.store().forget(key); }
    async flush() { return this.store().flush(); }
    async increment(key, value = 1) { return this.store().increment(key, value); }
    async decrement(key, value = 1) { return this.store().decrement(key, value); }
    async add(key, value, seconds) { return this.store().add(key, value, seconds); }
    async pull(key, defaultValue = null) { return this.store().pull(key, defaultValue); }
    async remember(key, seconds, callback) { return this.store().remember(key, seconds, callback); }
    async rememberForever(key, callback) { return this.store().rememberForever(key, callback); }
    // Enterprise methods
    async getMultiple(keys) { return this.store().getMultiple(keys); }
    async putMultiple(values, seconds) { return this.store().putMultiple(values, seconds); }
    async forgetMultiple(keys) { return this.store().forgetMultiple(keys); }
    lock(name, seconds = 0, owner = null) { return this.store().lock(name, seconds, owner); }
    tags(names) { return this.store().tags(names); }
}
exports.CacheManager = CacheManager;
