// @arikajs/cache/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatabaseDriver = exports.MemoryDriver = exports.Repository = exports.CacheManager = exports.Cache = void 0;
const CacheManager_1 = require("./CacheManager");
Object.defineProperty(exports, "CacheManager", { enumerable: true, get: function () { return CacheManager_1.CacheManager; } });
let cacheManager;
class Cache {
    static setManager(manager) {
        cacheManager = manager;
    }
    static getManager() {
        if (!cacheManager) {
            throw new Error('Cache system not configured. Please use Cache.setManager() to configure.');
        }
        return cacheManager;
    }
    static store(name) {
        return this.getManager().store(name);
    }
    static async get(key, defaultValue = null) { return this.getManager().get(key, defaultValue); }
    static async put(key, value, seconds) { return this.getManager().put(key, value, seconds); }
    static async has(key) { return this.getManager().has(key); }
    static async forever(key, value) { return this.getManager().forever(key, value); }
    static async forget(key) { return this.getManager().forget(key); }
    static async flush() { return this.getManager().flush(); }
    static async increment(key, value = 1) { return this.getManager().increment(key, value); }
    static async decrement(key, value = 1) { return this.getManager().decrement(key, value); }
    static async add(key, value, seconds) { return this.getManager().add(key, value, seconds); }
    static async pull(key, defaultValue = null) { return this.getManager().pull(key, defaultValue); }
    static async remember(key, seconds, callback) { return this.getManager().remember(key, seconds, callback); }
    static async rememberForever(key, callback) { return this.getManager().rememberForever(key, callback); }
    // Enterprise methods
    static async getMultiple(keys) { return this.getManager().getMultiple(keys); }
    static async putMultiple(values, seconds) { return this.getManager().putMultiple(values, seconds); }
    static async forgetMultiple(keys) { return this.getManager().forgetMultiple(keys); }
    static lock(name, seconds = 0, owner = null) { return this.getManager().lock(name, seconds, owner); }
    static tags(names) { return this.getManager().tags(names); }
}
exports.Cache = Cache;
var Repository_1 = require("./Repository");
Object.defineProperty(exports, "Repository", { enumerable: true, get: function () { return Repository_1.Repository; } });
var MemoryDriver_1 = require("./Drivers/MemoryDriver");
Object.defineProperty(exports, "MemoryDriver", { enumerable: true, get: function () { return MemoryDriver_1.MemoryDriver; } });
var DatabaseDriver_1 = require("./Drivers/DatabaseDriver");
Object.defineProperty(exports, "DatabaseDriver", { enumerable: true, get: function () { return DatabaseDriver_1.DatabaseDriver; } });
