// @arikajs/cache/dist/TagSet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TagSet = void 0;
class TagSet {
    constructor(store, names) {
        this.store = store;
        this.names = names;
    }
    /**
     * Get a combined namespace prefix for all tags.
     */
    async getNamespace() {
        const ids = await Promise.all(this.names.map(name => this.tagId(name)));
        return ids.join('|') + ':';
    }
    /**
     * Get or create entirely a unique identifier for a tag.
     */
    async tagId(name) {
        const key = `tag:${name}:key`;
        let id = await this.store.get(key);
        if (!id) {
            id = this.generateId();
            await this.store.forever(key, id);
        }
        return String(id);
    }
    /**
     * Reset all assigned tags by generating new IDs.
     * This effectively orphans all previous cache entries.
     */
    async reset() {
        await Promise.all(this.names.map(async (name) => {
            const key = `tag:${name}:key`;
            await this.store.forever(key, this.generateId());
        }));
    }
    /**
     * Generate a unique tag ID.
     */
    generateId() {
        return Date.now().toString(36) + Math.random().toString(36).substring(2, 15);
    }
}
exports.TagSet = TagSet;
