// @arikajs/http/dist/HttpKernel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpKernel = void 0;
const Request_1 = require("./Request");
const Response_1 = require("./Response");
const Pipeline_1 = require("./Pipeline");
const Router_1 = require("./Router");
class HttpKernel {
    constructor(app) {
        this.app = app;
        this.middleware = [];
        this.router = new Router_1.Router();
    }
    /**
     * Get the router instance.
     */
    getRouter() {
        return this.router;
    }
    /**
     * Set the router instance.
     */
    setRouter(router) {
        this.router = router;
    }
    /**
     * Add global middleware.
     */
    use(middleware) {
        this.middleware.push(middleware);
        return this;
    }
    /**
     * Handle an incoming HTTP request.
     */
    async handle(req, res) {
        try {
            const request = new Request_1.Request(this.app, req);
            const response = new Response_1.Response(res);
            const result = await this.sendRequestThroughRouter(request, response);
            result.terminate();
        }
        catch (error) {
            this.handleException(error, res);
        }
    }
    /**
     * Send the given request through the middleware / router.
     */
    async sendRequestThroughRouter(request, response) {
        return (new Pipeline_1.Pipeline(this.app.getContainer()))
            .pipe(this.middleware)
            .handle(request, (req) => this.dispatchToRouter(req, response), response);
    }
    /**
     * Dispatch the request to the router.
     */
    async dispatchToRouter(request, response) {
        return this.router.dispatch(request, response);
    }
    /**
     * Handle an exception during request execution.
     */
    handleException(error, res) {
        const response = new Response_1.Response(res);
        const status = error.statusCode || error.status || 500;
        response.status(status).json({
            message: error.message || 'Internal Server Error',
            error: process.env.NODE_ENV === 'development' ? error.stack : undefined
        });
        response.terminate();
    }
}
exports.HttpKernel = HttpKernel;
