// @arikajs/http/dist/ObjectPool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPool = void 0;
/**
 * A high-performance generic object pool.
 * Prevents repeated memory allocation/GC by recycling objects across requests.
 *
 * Strategy: maintain a fixed-size free list (stack). On acquire(), pop a
 * pre-warmed object and reset its state. On release(), push it back.
 */
class ObjectPool {
    constructor(factory, reset, poolSize = 512) {
        this.pool = [];
        this.factory = factory;
        this.reset = reset;
        this.maxSize = poolSize;
        // Pre-warm the pool — allocate & shape objects at startup so V8
        // can immediately JIT-compile them into a stable hidden class.
        for (let i = 0; i < poolSize; i++) {
            this.pool.push(factory());
        }
    }
    /**
     * Acquire an object from the pool (or allocate a new one if the pool is empty).
     */
    acquire() {
        if (this.pool.length > 0) {
            return this.pool.pop();
        }
        return this.factory();
    }
    /**
     * Return an object to the pool after resetting its state.
     */
    release(obj) {
        if (this.pool.length < this.maxSize) {
            this.reset(obj);
            this.pool.push(obj);
        }
        // If pool is full, let the object be GC'd — this is rare at high concurrency.
    }
    /**
     * Current number of available objects in the pool.
     */
    get available() {
        return this.pool.length;
    }
}
exports.ObjectPool = ObjectPool;
