// @arikajs/http/dist/Pipeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pipeline = void 0;
/**
 * Pipeline executes a stack of middleware for HTTP requests.
 */
class Pipeline {
    constructor(container) {
        this.container = container;
        this.handlers = [];
    }
    /**
     * Add middleware to the pipeline.
     */
    pipe(middleware) {
        if (Array.isArray(middleware)) {
            this.handlers.push(...middleware);
        }
        else {
            this.handlers.push(middleware);
        }
        return this;
    }
    /**
     * Run the pipeline through the given destination.
     */
    async handle(request, destination, response) {
        const invoke = async (index, req) => {
            if (index >= this.handlers.length) {
                return destination(req, response);
            }
            const handler = this.resolve(this.handlers[index]);
            if (typeof handler === 'function') {
                return handler(req, (nextReq) => invoke(index + 1, nextReq), response);
            }
            if (typeof handler === 'object' && 'handle' in handler && typeof handler.handle === 'function') {
                return handler.handle(req, (nextReq) => invoke(index + 1, nextReq), response);
            }
            throw new Error(`Invalid middleware handler at index ${index}`);
        };
        return invoke(0, request);
    }
    /**
     * Resolve the middleware handler.
     */
    resolve(handler) {
        if (typeof handler === 'string' && this.container) {
            return this.container.make(handler);
        }
        if (typeof handler === 'function') {
            const isClass = /^\s*class\s+/.test(handler.toString()) ||
                (handler.prototype && typeof handler.prototype.handle === 'function');
            if (isClass) {
                return this.container ? this.container.make(handler) : new handler();
            }
        }
        return handler;
    }
}
exports.Pipeline = Pipeline;
