// @arikajs/http/dist/Request.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Request = void 0;
const cookie = __importStar(require("cookie"));
const validation_1 = require("@arikajs/validation");
class Request {
    constructor(app, req) {
        this._session = null;
        this._cookies = null;
        this._body = null;
        this._params = {};
        this._auth = null;
        this._view = null;
        this._baseUrl = null;
        this.app = app;
        if (req) {
            this.reset(req);
        }
    }
    reset(req) {
        if (!req) {
            this.req = null;
            return;
        }
        this.req = req;
        this.searchParams = null;
        this._cookies = null;
        this._body = null;
        this._params = {};
        this._baseUrl = null;
        this._session = null;
        this._auth = null;
        this._view = null;
    }
    /**
     * Get the view engine instance.
     */
    get view() {
        if (!this._view) {
            try {
                this._view = this.app.make('view');
            }
            catch (e) {
                return null;
            }
        }
        return this._view;
    }
    set view(value) {
        this._view = value;
    }
    /**
     * Get the authentication context for the request.
     */
    get auth() {
        if (!this._auth) {
            try {
                const authManager = this.app.make('auth');
                this._auth = authManager.createContext(this);
            }
            catch (e) {
                // Auth manager not registered
                return null;
            }
        }
        return this._auth;
    }
    set auth(value) {
        this._auth = value;
    }
    ensureSearchParams() {
        if (!this.searchParams && this.req) {
            const url = this.req.url || '/';
            const queryIndex = url.indexOf('?');
            if (queryIndex === -1) {
                this.searchParams = new URLSearchParams();
            }
            else {
                this.searchParams = new URLSearchParams(url.slice(queryIndex + 1));
            }
        }
    }
    /**
     * Get the base URL (scheme + host) of the request.
     * Falls back to app.url config if host header is missing.
     */
    baseUrl() {
        if (this._baseUrl)
            return this._baseUrl;
        const trustProxy = this.app.config().get('http.trustProxy', false);
        const protocol = trustProxy ? (this.header('x-forwarded-proto') || 'http') : 'http';
        const host = trustProxy ? (this.header('x-forwarded-host') || this.req.headers.host) : this.req.headers.host;
        if (host) {
            this._baseUrl = `${protocol}://${host}`;
            return this._baseUrl;
        }
        this._baseUrl = this.app.config().get('app.url', 'http://localhost');
        return this._baseUrl;
    }
    /**
     * Get the application instance.
     */
    getApplication() {
        return this.app;
    }
    /**
     * Get the underlying Node request.
     */
    getRaw() {
        return this.req;
    }
    /**
     * Get the HTTP method.
     */
    method() {
        return this.req.method || 'GET';
    }
    /**
     * Get the request path.
     */
    path() {
        const url = this.req.url || '/';
        const queryIndex = url.indexOf('?');
        return queryIndex === -1 ? url : url.slice(0, queryIndex);
    }
    /**
     * Get all headers.
     */
    headers() {
        return this.req.headers;
    }
    /**
     * Get the client IP address.
     */
    ip() {
        const trustProxy = this.app.config().get('http.trustProxy', false);
        if (trustProxy) {
            const forwardedFor = this.header('x-forwarded-for');
            if (typeof forwardedFor === 'string') {
                return forwardedFor.split(',')[0].trim();
            }
        }
        return this.req.socket.remoteAddress;
    }
    /**
     * Get a specific header.
     */
    header(name) {
        return this.req.headers[name.toLowerCase()];
    }
    /**
     * Get all cookies.
     */
    cookies() {
        if (this._cookies === null) {
            const cookieHeader = this.header('cookie');
            this._cookies = typeof cookieHeader === 'string' ? cookie.parse(cookieHeader) : {};
        }
        return this._cookies;
    }
    /**
     * Get a specific cookie.
     */
    cookie(name) {
        return this.cookies()[name];
    }
    /**
     * Get a query parameter.
     */
    query(key) {
        this.ensureSearchParams();
        return this.searchParams.get(key);
    }
    /**
     * Set the route parameters.
     */
    setParams(params) {
        this._params = params;
    }
    /**
     * Get all route parameters.
     */
    params() {
        return this._params;
    }
    /**
     * Get a specific route parameter.
     */
    param(key, defaultValue = null) {
        return this._params[key] ?? defaultValue;
    }
    /**
     * Set the parsed body (usually set by middleware).
     */
    setBody(body) {
        this._body = body;
    }
    /**
     * Get an input value from route params, body, or query.
     */
    input(key, defaultValue = null) {
        // 1. Check route parameters
        if (this._params[key] !== undefined) {
            return this._params[key];
        }
        // 2. Check body
        if (this._body && typeof this._body === 'object' && key in this._body) {
            return this._body[key];
        }
        // 3. Check query string
        return this.query(key) ?? defaultValue;
    }
    /**
     * Get the parsed body.
     */
    body() {
        return this._body;
    }
    /**
     * Get all input (query + body).
     */
    all() {
        this.ensureSearchParams();
        const query = Object.fromEntries(this.searchParams.entries());
        const body = typeof this._body === 'object' && this._body !== null ? this._body : {};
        return { ...query, ...body };
    }
    /**
     * Get a subset of the input data.
     */
    only(keys) {
        const all = this.all();
        const result = {};
        keys.forEach(key => {
            if (key in all) {
                result[key] = all[key];
            }
        });
        return result;
    }
    get session() {
        if (!this._session) {
            const store = {};
            this._session = {
                get: (key) => store[key] ?? null,
                put: (key, value) => { store[key] = value; },
                forget: (key) => { delete store[key]; },
                has: (key) => key in store,
                all: () => ({ ...store }),
            };
        }
        return this._session;
    }
    set session(value) {
        this._session = value;
    }
    /**
     * Get all input data except for a specified array of keys.
     */
    except(keys) {
        const all = this.all();
        const result = { ...all };
        keys.forEach(key => {
            delete result[key];
        });
        return result;
    }
    /**
     * Get the original incoming message.
     */
    getIncomingMessage() {
        return this.req;
    }
    /**
     * Determine if the current request is asking for JSON.
     */
    wantsJson() {
        const acceptable = this.header('accept');
        return typeof acceptable === 'string' &&
            (acceptable.includes('application/json') || acceptable.includes('+json'));
    }
    /**
     * Determine if the current request is an AJAX request.
     */
    ajax() {
        return this.header('x-requested-with') === 'XMLHttpRequest';
    }
    /**
     * Determine if the current request expects a JSON response.
     */
    expectsJson() {
        return this.ajax() || this.wantsJson();
    }
    /**
     * Validate the request with the given rules.
     */
    async validate(rules, messages = {}) {
        const validator = new validation_1.Validator(this.all(), rules, messages);
        if (await validator.fails()) {
            throw new validation_1.ValidationError(validator.errors());
        }
        return validator.validated();
    }
}
exports.Request = Request;
