// @arikajs/http/dist/Response.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Response = void 0;
const cookie = __importStar(require("cookie"));
class Response {
    constructor(res) {
        this._status = 200;
        this._headers = {};
        this._cookies = [];
        this._content = null;
        if (res)
            this.reset(res);
    }
    reset(res) {
        this.res = res;
        this._status = 200;
        this._headers = {};
        this._cookies = [];
        this._content = null;
    }
    /**
     * Set the HTTP status code.
     */
    status(code) {
        this._status = code;
        return this;
    }
    /**
     * Helper: 200 OK.
     */
    ok() {
        return this.status(200);
    }
    /**
     * Helper: 201 Created.
     */
    created() {
        return this.status(201);
    }
    /**
     * Helper: 204 No Content.
     */
    noContent() {
        return this.status(204);
    }
    /**
     * Helper: 403 Forbidden.
     */
    forbidden() {
        return this.status(403);
    }
    /**
     * Helper: 404 Not Found.
     */
    notFound() {
        return this.status(404);
    }
    /**
     * Set a header.
     */
    header(name, value) {
        this._headers[name] = value;
        return this;
    }
    /**
     * Set a cookie.
     */
    cookie(name, value, options = {}) {
        this._cookies.push(cookie.serialize(name, value, {
            path: '/',
            ...options
        }));
        return this;
    }
    /**
     * Set the response content as JSON.
     */
    json(data, status) {
        if (status) {
            this.status(status);
        }
        this._content = JSON.stringify(data);
        this.header('Content-Type', 'application/json');
        return this;
    }
    /**
     * Set the response content as plain text or HTML.
     */
    send(content, status) {
        if (status) {
            this.status(status);
        }
        this._content = content;
        if (typeof content === 'string' && !this._headers['Content-Type']) {
            this.header('Content-Type', content.startsWith('<') ? 'text/html' : 'text/plain');
        }
        return this;
    }
    /**
     * Redirect to a specific URL.
     */
    redirect(url, status = 302) {
        this.status(status);
        this.header('Location', url);
        this._content = `Redirecting to ${url}`;
        return this;
    }
    /**
     * Redirect to the previous URL.
     */
    back(request, fallback = '/', status = 302) {
        const url = request.header('referer') || request.header('referrer') || fallback;
        return this.redirect(url, status);
    }
    /**
     * Set the response content.
     */
    setContent(content) {
        this._content = content;
        return this;
    }
    /**
     * Get the current content.
     */
    getContent() {
        return this._content;
    }
    /**
     * Get the current status code.
     */
    getStatusCode() {
        return this._status;
    }
    /**
     * Get all currently set headers.
     */
    getHeaders() {
        return this._headers;
    }
    /**
     * Actually terminate the request and send to client.
     * This is called by the Kernel.
     */
    terminate() {
        if (this.res.writableEnded) {
            return;
        }
        // 1. Set Status
        this.res.statusCode = this._status;
        // 2. Set Headers
        const headerKeys = Object.keys(this._headers);
        for (let i = 0; i < headerKeys.length; i++) {
            const key = headerKeys[i];
            this.res.setHeader(key, this._headers[key]);
        }
        // 3. Set Cookies
        if (this._cookies.length > 0) {
            this.res.setHeader('Set-Cookie', this._cookies);
        }
        // 4. Send body and end
        this.res.end(this._content ?? '');
    }
    /**
     * Get the underlying Node response.
     */
    getOriginalResponse() {
        return this.res;
    }
    /**
     * Get the underlying Node response (alias for getOriginalResponse)
     */
    get raw() {
        return this.res;
    }
}
exports.Response = Response;
