// @arikajs/http/dist/Router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Router = void 0;
const Response_1 = require("./Response");
const Pipeline_1 = require("./Pipeline");
class Router {
    constructor() {
        this.routes = [];
        this.groupStack = [];
        this.middlewareGroups = {};
    }
    group(attributes, callback) {
        this.groupStack.push(attributes);
        callback(this);
        this.groupStack.pop();
    }
    get(path, action) {
        return this.addRoute('GET', path, action);
    }
    post(path, action) {
        return this.addRoute('POST', path, action);
    }
    put(path, action) {
        return this.addRoute('PUT', path, action);
    }
    delete(path, action) {
        return this.addRoute('DELETE', path, action);
    }
    patch(path, action) {
        return this.addRoute('PATCH', path, action);
    }
    options(path, action) {
        return this.addRoute('OPTIONS', path, action);
    }
    /**
     * Add middleware to the last registered route.
     */
    middleware(middleware) {
        const lastRoute = this.routes[this.routes.length - 1];
        if (lastRoute) {
            if (Array.isArray(middleware)) {
                lastRoute.middleware.push(...middleware);
            }
            else {
                lastRoute.middleware.push(middleware);
            }
        }
        return this;
    }
    addRoute(method, path, action) {
        const attributes = this.mergeGroupAttributes();
        const fullPath = this.prefixPath(attributes.prefix, path);
        const middleware = [...(attributes.middleware || [])];
        this.routes.push({
            method,
            path: fullPath,
            action,
            middleware
        });
        return this;
    }
    mergeGroupAttributes() {
        const attributes = { prefix: '', middleware: [] };
        for (const group of this.groupStack) {
            if (group.prefix) {
                attributes.prefix = this.prefixPath(attributes.prefix, group.prefix);
            }
            if (group.middleware) {
                attributes.middleware = [...attributes.middleware, ...group.middleware];
            }
        }
        return attributes;
    }
    prefixPath(prefix, path) {
        const cleanedPrefix = prefix.replace(/\/+$/, '');
        const cleanedPath = path.replace(/^\/+/, '');
        if (!cleanedPrefix)
            return cleanedPath.startsWith('/') ? cleanedPath : '/' + cleanedPath;
        return cleanedPrefix + '/' + cleanedPath;
    }
    async dispatch(request, response) {
        const route = this.findRoute(request);
        if (!route) {
            return response.status(404).json({ error: 'Route not found' });
        }
        const pipeline = new Pipeline_1.Pipeline(request.getApplication().getContainer());
        return pipeline
            .pipe(route.middleware)
            .handle(request, (req) => this.runRoute(route, req, response), response);
    }
    findRoute(request) {
        const path = request.path();
        const method = request.method().toUpperCase();
        for (const route of this.routes) {
            if (route.method === method && this.matchPath(route.path, path, request)) {
                return route;
            }
        }
        return null;
    }
    matchPath(routePath, requestPath, request) {
        // Simple parameter matching: /users/:id
        const routeParts = routePath.split('/').filter(Boolean);
        const requestParts = requestPath.split('/').filter(Boolean);
        if (routeParts.length !== requestParts.length) {
            return false;
        }
        const params = {};
        for (let i = 0; i < routeParts.length; i++) {
            if (routeParts[i].startsWith(':')) {
                const paramName = routeParts[i].substring(1);
                params[paramName] = requestParts[i];
                continue;
            }
            if (routeParts[i] !== requestParts[i]) {
                return false;
            }
        }
        request.setParams(params);
        return true;
    }
    async runRoute(route, request, response) {
        const action = route.action;
        if (typeof action === 'function') {
            const result = await action(request, ...Object.values(request.params()));
            return this.toResponse(result, response);
        }
        if (Array.isArray(action)) {
            const [controllerClass, methodName] = action;
            const controller = request.getApplication().make(controllerClass);
            const result = await controller[methodName](request, ...Object.values(request.params()));
            return this.toResponse(result, response);
        }
        throw new Error('Invalid route action');
    }
    toResponse(result, response) {
        if (result instanceof Response_1.Response) {
            return result;
        }
        if (typeof result === 'object') {
            return response.json(result);
        }
        if (typeof result === 'string') {
            return response.send(result);
        }
        return response;
    }
}
exports.Router = Router;
