// @arikajs/queue/dist/Drivers/DatabaseDriver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatabaseDriver = void 0;
class DatabaseDriver {
    constructor(database, config) {
        this.database = database;
        this.config = config;
    }
    async push(job, options = {}) {
        const payload = this.createPayload(job);
        const queue = options.queue || job.queue || 'default';
        let availableAt = Math.floor(Date.now() / 1000);
        if (options.delay) {
            if (options.delay instanceof Date) {
                availableAt = Math.floor(options.delay.getTime() / 1000);
            }
            else {
                availableAt += options.delay;
            }
        }
        await this.database.table(this.config.table, this.config.connection).insert({
            queue: queue,
            payload: JSON.stringify(payload),
            attempts: 0,
            available_at: availableAt,
            created_at: Math.floor(Date.now() / 1000),
        });
    }
    async pop(queue = 'default') {
        const now = Math.floor(Date.now() / 1000);
        // Find the next available job
        const job = await this.database.table(this.config.table, this.config.connection)
            .where('queue', queue)
            .where('available_at', '<=', now)
            .whereNull('reserved_at')
            .orderBy('available_at', 'asc')
            .first();
        if (!job) {
            return null;
        }
        // Reserve the job
        await this.database.table(this.config.table, this.config.connection)
            .where('id', job.id)
            .update({
            reserved_at: now,
            attempts: job.attempts + 1
        });
        return {
            id: job.id,
            payload: JSON.parse(job.payload),
            attempts: job.attempts + 1
        };
    }
    async acknowledge(jobData) {
        await this.database.table(this.config.table, this.config.connection)
            .where('id', jobData.id)
            .delete();
    }
    async fail(jobData, error) {
        const failedTable = 'failed_jobs';
        await this.database.table(failedTable, this.config.connection).insert({
            uuid: Math.random().toString(36).substring(2),
            connection: this.config.connection || 'default',
            queue: jobData.queue || 'default',
            payload: JSON.stringify(jobData.payload),
            exception: error.stack || error.message,
            failed_at: new Date()
        });
        await this.acknowledge(jobData);
    }
    createPayload(job) {
        return {
            displayName: job.constructor.name,
            job: job.constructor.name,
            data: this.getJobData(job),
        };
    }
    getJobData(job) {
        // Simple serialization of job properties
        const data = {};
        for (const key of Object.keys(job)) {
            // Avoid serializing functions or complex objects
            if (typeof job[key] !== 'function') {
                data[key] = job[key];
            }
        }
        return data;
    }
}
exports.DatabaseDriver = DatabaseDriver;
