// @arikajs/queue/dist/Drivers/RedisDriver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisDriver = void 0;
const ioredis_1 = require("ioredis");
class RedisDriver {
    constructor(config) {
        this.config = config;
        this.queueName = config.queue || 'default';
        if (config.client) {
            this.redis = config.client;
        }
        else if (typeof config.redisConfig === 'string') {
            this.redis = new ioredis_1.Redis(config.redisConfig);
        }
        else if (typeof config.redisConfig === 'object') {
            this.redis = new ioredis_1.Redis(config.redisConfig);
        }
        else {
            this.redis = new ioredis_1.Redis(); // defaults to localhost:6379
        }
    }
    async push(job, options = {}) {
        const payload = this.createPayload(job);
        const queue = options.queue || job.queue || this.queueName;
        const json = JSON.stringify(payload);
        if (options.delay) {
            let availableAt = Math.floor(Date.now() / 1000);
            if (options.delay instanceof Date) {
                availableAt = Math.floor(options.delay.getTime() / 1000);
            }
            else {
                availableAt += options.delay;
            }
            await this.redis.zadd(`queues:${queue}:delayed`, availableAt, json);
        }
        else {
            await this.redis.rpush(`queues:${queue}`, json);
        }
    }
    createPayload(job) {
        return {
            displayName: job.constructor.name,
            job: job.constructor.name,
            data: this.getJobData(job),
            id: this.generateJobId(),
            attempts: 0,
            createdAt: new Date().toISOString()
        };
    }
    getJobData(job) {
        const data = {};
        for (const key of Object.keys(job)) {
            data[key] = job[key];
        }
        return data;
    }
    generateJobId() {
        return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
    }
    // Helper method useful for fetching jobs if Worker implements polling later
    async pop(queue) {
        const queueName = queue || this.queueName;
        // First try to move delayed jobs that are now available
        await this.migrateDelayedJobs(queueName);
        const result = await this.redis.lpop(`queues:${queueName}`);
        return result ? JSON.parse(result) : null;
    }
    async migrateDelayedJobs(queue) {
        const now = Math.floor(Date.now() / 1000);
        const delayedKey = `queues:${queue}:delayed`;
        const jobs = await this.redis.zrangebyscore(delayedKey, 0, now);
        if (jobs.length > 0) {
            await this.redis.zremrangebyscore(delayedKey, 0, now);
            for (const job of jobs) {
                await this.redis.rpush(`queues:${queue}`, job);
            }
        }
    }
    async acknowledge(jobData) {
        // Redis lpop already removed the job, but we could implement ack for RPOPPLPUSH if needed
    }
    async fail(jobData, error) {
        const failedKey = `queues:failed`;
        await this.redis.rpush(failedKey, JSON.stringify({
            ...jobData,
            exception: error.stack || error.message,
            failedAt: new Date().toISOString()
        }));
    }
    // Helper method to clear for tests
    async flush() {
        await this.redis.del(`queues:${this.queueName}`);
    }
    async close() {
        await this.redis.quit();
    }
}
exports.RedisDriver = RedisDriver;
