// @arikajs/foundation/dist/tests/application.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = require("node:test");
const node_assert_1 = __importDefault(require("node:assert"));
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const os = __importStar(require("os"));
const Application_1 = require("../src/application/Application");
const ServiceProvider_1 = require("../src/providers/ServiceProvider");
(0, node_test_1.test)('Application: constructor accepts basePath', () => {
    const basePath = __dirname;
    const app = new Application_1.Application(basePath);
    node_assert_1.default.strictEqual(app.getBasePath(), path.resolve(basePath));
});
(0, node_test_1.test)('Application: exposes container', () => {
    const app = new Application_1.Application(__dirname);
    const container = app.getContainer();
    node_assert_1.default.ok(container);
});
(0, node_test_1.test)('Application: exposes config repository', () => {
    const app = new Application_1.Application(__dirname);
    const config = app.config();
    node_assert_1.default.ok(config);
});
(0, node_test_1.test)('Application: can bind services', () => {
    const app = new Application_1.Application(__dirname);
    app.bind('test', () => 'value');
    const value = app.resolve('test');
    node_assert_1.default.strictEqual(value, 'value');
});
(0, node_test_1.test)('Application: can register singleton', () => {
    const app = new Application_1.Application(__dirname);
    let callCount = 0;
    app.singleton('counter', () => {
        callCount++;
        return callCount;
    });
    node_assert_1.default.strictEqual(app.resolve('counter'), 1);
    node_assert_1.default.strictEqual(app.resolve('counter'), 1);
    node_assert_1.default.strictEqual(callCount, 1);
});
(0, node_test_1.test)('Application: can register instance', () => {
    const app = new Application_1.Application(__dirname);
    const instance = { id: 123 };
    app.instance('instance', instance);
    node_assert_1.default.strictEqual(app.resolve('instance'), instance);
});
(0, node_test_1.test)('Application: register() accepts ServiceProvider instance', async () => {
    const app = new Application_1.Application(__dirname);
    let registered = false;
    let booted = false;
    class TestProvider extends ServiceProvider_1.ServiceProvider {
        register() {
            registered = true;
            this.app.instance('test', 'value');
        }
        boot() {
            booted = true;
        }
    }
    app.register(new TestProvider(app));
    app.config().set('app.key', 'base64:sm957Y1wUYo8Uj8yL1fD7vX+X6y8gG+E6XpXnJz+I=');
    await app.boot();
    node_assert_1.default.strictEqual(registered, true);
    node_assert_1.default.strictEqual(booted, true);
    node_assert_1.default.strictEqual(app.resolve('test'), 'value');
});
(0, node_test_1.test)('Application: register() accepts ServiceProvider class', async () => {
    const app = new Application_1.Application(__dirname);
    let registered = false;
    class TestProvider extends ServiceProvider_1.ServiceProvider {
        register() {
            registered = true;
        }
    }
    app.register(TestProvider);
    app.config().set('app.key', 'base64:sm957Y1wUYo8Uj8yL1fD7vX+X6y8gG+E6XpXnJz+I=');
    await app.boot();
    node_assert_1.default.strictEqual(registered, true);
});
(0, node_test_1.test)('Application: boot() runs all register() then all boot()', async () => {
    const app = new Application_1.Application(__dirname);
    const order = [];
    class Provider1 extends ServiceProvider_1.ServiceProvider {
        register() {
            order.push('register1');
        }
        boot() {
            order.push('boot1');
        }
    }
    class Provider2 extends ServiceProvider_1.ServiceProvider {
        register() {
            order.push('register2');
        }
        boot() {
            order.push('boot2');
        }
    }
    app.register(Provider1);
    app.register(Provider2);
    app.config().set('app.key', 'base64:sm957Y1wUYo8Uj8yL1fD7vX+X6y8gG+E6XpXnJz+I=');
    await app.boot();
    node_assert_1.default.deepStrictEqual(order, ['register1', 'register2', 'boot1', 'boot2']);
});
(0, node_test_1.test)('Application: boot() is idempotent', async () => {
    const app = new Application_1.Application(__dirname);
    let bootCount = 0;
    class TestProvider extends ServiceProvider_1.ServiceProvider {
        register() { }
        boot() {
            bootCount++;
        }
    }
    app.register(TestProvider);
    app.config().set('app.key', 'base64:sm957Y1wUYo8Uj8yL1fD7vX+X6y8gG+E6XpXnJz+I=');
    await app.boot();
    await app.boot();
    await app.boot();
    node_assert_1.default.strictEqual(bootCount, 1);
    node_assert_1.default.strictEqual(app.isBooted(), true);
});
(0, node_test_1.test)('Application: cannot register providers after boot', async () => {
    const app = new Application_1.Application(__dirname);
    class TestProvider extends ServiceProvider_1.ServiceProvider {
        register() { }
    }
    app.config().set('app.key', 'base64:sm957Y1wUYo8Uj8yL1fD7vX+X6y8gG+E6XpXnJz+I=');
    await app.boot();
    node_assert_1.default.throws(() => {
        app.register(TestProvider);
    }, /Cannot register providers after the application has been booted/);
});
(0, node_test_1.test)('Application: run() calls boot() if not booted', async () => {
    const app = new Application_1.Application(__dirname);
    let booted = false;
    class TestProvider extends ServiceProvider_1.ServiceProvider {
        register() { }
        boot() {
            booted = true;
        }
    }
    app.register(TestProvider);
    app.config().set('app.key', 'base64:sm957Y1wUYo8Uj8yL1fD7vX+X6y8gG+E6XpXnJz+I=');
    await app.run();
    node_assert_1.default.strictEqual(booted, true);
    node_assert_1.default.strictEqual(app.isBooted(), true);
});
(0, node_test_1.test)('Application: config loads from config directory', async () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'arika-test-'));
    const configDir = path.join(tmpDir, 'config');
    fs.mkdirSync(configDir, { recursive: true });
    // Create a test config file
    const configContent = `module.exports = {
  default: {
    name: 'Test App',
    version: '1.0.0'
  }
};`;
    fs.writeFileSync(path.join(configDir, 'app.js'), configContent);
    const app = new Application_1.Application(tmpDir);
    app.config().set('app.key', 'base64:sm957Y1wUYo8Uj8yL1fD7vX+X6y8gG+E6XpXnJz+I=');
    await app.boot();
    const name = app.config().get('app.name');
    node_assert_1.default.strictEqual(name, 'Test App');
    // Cleanup
    fs.rmSync(tmpDir, { recursive: true, force: true });
});
(0, node_test_1.test)('Application: config is read-only after boot', async () => {
    const app = new Application_1.Application(__dirname);
    app.config().set('app.key', 'base64:sm957Y1wUYo8Uj8yL1fD7vX+X6y8gG+E6XpXnJz+I=');
    await app.boot();
    node_assert_1.default.throws(() => {
        app.config().loadConfigDirectory(__dirname);
    }, /Configuration cannot be modified after boot/);
});
