// arikajs/dist/http/Middleware/VerifyCsrfToken.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifyCsrfToken = void 0;
const http_1 = require("@arikajs/http");
const node_crypto_1 = __importDefault(require("node:crypto"));
class VerifyCsrfToken {
    constructor() {
        /**
         * The URIs that should be excluded from CSRF verification.
         */
        this.except = [];
    }
    /**
     * Handle the incoming request.
     */
    async handle(request, next) {
        if (this.isReading(request) || this.runningUnitTests() || await this.inExceptArray(request)) {
            return await this.addCookieToResponse(request, await next(request));
        }
        if (await this.tokensMatch(request)) {
            return await this.addCookieToResponse(request, await next(request));
        }
        throw new http_1.HttpException(419, 'CSRF token mismatch.');
    }
    /**
     * Determine if the request has a URI that should pass through CSRF verification.
     */
    async inExceptArray(request) {
        const path = request.path();
        for (const except of this.except) {
            if (except === '/') {
                if (path === '/')
                    return true;
                continue;
            }
            if (path.startsWith(except)) {
                return true;
            }
        }
        return false;
    }
    /**
     * Determine if the session and input tokens match.
     */
    async tokensMatch(request) {
        const token = await this.getTokenFromRequest(request);
        const sessionToken = await request.session.get('_token');
        if (!token || !sessionToken || token.length !== sessionToken.length) {
            return false;
        }
        return node_crypto_1.default.timingSafeEqual(Buffer.from(token), Buffer.from(sessionToken));
    }
    /**
     * Get the CSRF token from the request.
     */
    async getTokenFromRequest(request) {
        const token = request.input('_token') || request.header('X-CSRF-TOKEN') || request.header('X-XSRF-TOKEN');
        if (!token) {
            return null;
        }
        return token;
    }
    /**
     * Determine if the HTTP request uses a read-only method.
     */
    isReading(request) {
        return ['GET', 'HEAD', 'OPTIONS'].includes(request.method());
    }
    /**
     * Determine if the application is running unit tests.
     */
    runningUnitTests() {
        return process.env.NODE_ENV === 'test';
    }
    /**
     * Add the CSRF token to the response cookies.
     */
    async addCookieToResponse(request, response) {
        const token = await request.session.get('_token');
        if (token) {
            const config = request.app.config();
            response.cookie('XSRF-TOKEN', token, {
                path: '/',
                httpOnly: false,
                sameSite: 'lax',
                secure: config.get('session.secure', false)
            });
        }
        return response;
    }
}
exports.VerifyCsrfToken = VerifyCsrfToken;
