// @arikajs/cache/dist/src/Repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Repository = void 0;
const CacheLock_1 = require("./CacheLock");
const TagSet_1 = require("./TagSet");
const TaggedStore_1 = require("./TaggedStore");
class Repository {
    constructor(store) {
        this.store = store;
    }
    async get(key, defaultValue = null) {
        const value = await this.store.get(key);
        return value !== null ? value : defaultValue;
    }
    async has(key) {
        return (await this.get(key)) !== null;
    }
    async put(key, value, seconds) {
        await this.store.put(key, value, seconds);
    }
    async add(key, value, seconds) {
        if (await this.has(key)) {
            return false;
        }
        await this.put(key, value, seconds);
        return true;
    }
    async increment(key, value = 1) {
        return this.store.increment(key, value);
    }
    async decrement(key, value = 1) {
        return this.store.decrement(key, value);
    }
    async forever(key, value) {
        await this.store.forever(key, value);
    }
    async remember(key, seconds, callback) {
        let value = await this.get(key);
        if (value !== null) {
            return value;
        }
        value = await callback();
        await this.put(key, value, seconds);
        return value;
    }
    async rememberForever(key, callback) {
        let value = await this.get(key);
        if (value !== null) {
            return value;
        }
        value = await callback();
        await this.forever(key, value);
        return value;
    }
    async pull(key, defaultValue = null) {
        const value = await this.get(key, defaultValue);
        await this.forget(key);
        return value;
    }
    async forget(key) {
        await this.store.forget(key);
    }
    async flush() {
        await this.store.flush();
    }
    async getMultiple(keys) {
        if (keys.length === 0)
            return {};
        if (this.store.getMultiple) {
            return await this.store.getMultiple(keys);
        }
        const results = {};
        for (const key of keys) {
            results[key] = await this.get(key);
        }
        return results;
    }
    async putMultiple(values, seconds) {
        if (Object.keys(values).length === 0)
            return;
        if (this.store.putMultiple) {
            return await this.store.putMultiple(values, seconds);
        }
        for (const [key, value] of Object.entries(values)) {
            await this.put(key, value, seconds);
        }
    }
    async forgetMultiple(keys) {
        if (keys.length === 0)
            return;
        if (this.store.forgetMultiple) {
            return await this.store.forgetMultiple(keys);
        }
        for (const key of keys) {
            await this.forget(key);
        }
    }
    lock(name, seconds = 0, owner = null) {
        return new CacheLock_1.CacheLock(this.store, name, seconds, owner);
    }
    tags(names) {
        const tagNames = Array.isArray(names) ? names : [names];
        return new Repository(new TaggedStore_1.TaggedStore(this.store, new TagSet_1.TagSet(this.store, tagNames)));
    }
    getStore() {
        return this.store;
    }
}
exports.Repository = Repository;
