// @arikajs/cache/dist/src/TaggedStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaggedStore = void 0;
class TaggedStore {
    constructor(store, tags) {
        this.store = store;
        this.tags = tags;
    }
    async prefixKey(key) {
        return (await this.tags.getNamespace()) + key;
    }
    async get(key) {
        return await this.store.get(await this.prefixKey(key));
    }
    async put(key, value, seconds) {
        await this.store.put(await this.prefixKey(key), value, seconds);
    }
    async add(key, value, seconds) {
        if (this.store.add) {
            return await this.store.add(await this.prefixKey(key), value, seconds);
        }
        if (await this.get(key) !== null)
            return false;
        await this.put(key, value, seconds);
        return true;
    }
    async increment(key, value = 1) {
        return await this.store.increment(await this.prefixKey(key), value);
    }
    async decrement(key, value = 1) {
        return await this.store.decrement(await this.prefixKey(key), value);
    }
    async forever(key, value) {
        await this.store.forever(await this.prefixKey(key), value);
    }
    async forget(key) {
        await this.store.forget(await this.prefixKey(key));
    }
    async flush() {
        await this.tags.reset();
    }
    getPrefix() {
        return this.store.getPrefix();
    }
    async getMultiple(keys) {
        const namespace = await this.tags.getNamespace();
        const prefixedKeys = keys.map(k => namespace + k);
        if (this.store.getMultiple) {
            const results = await this.store.getMultiple(prefixedKeys);
            const mapped = {};
            for (const [key, value] of Object.entries(results)) {
                // If the underlying store returns prefixed keys (e.g. database driver), it might not include namespace.
                // We just rely on mapping index-to-index since we pass exactly prefixedKeys.
            }
            // A safer approach: fetch everything and rebuild map
            const finalMap = {};
            for (let i = 0; i < keys.length; i++) {
                finalMap[keys[i]] = results[prefixedKeys[i]];
            }
            return finalMap;
        }
        // Fallback for custom drivers
        const fallback = {};
        for (const key of keys) {
            fallback[key] = await this.get(key);
        }
        return fallback;
    }
    async putMultiple(values, seconds) {
        const namespace = await this.tags.getNamespace();
        const prefixedValues = {};
        for (const [key, value] of Object.entries(values)) {
            prefixedValues[namespace + key] = value;
        }
        if (this.store.putMultiple) {
            await this.store.putMultiple(prefixedValues, seconds);
        }
        else {
            // Fallback
            for (const [k, v] of Object.entries(prefixedValues)) {
                await this.store.put(k, v, seconds);
            }
        }
    }
    async forgetMultiple(keys) {
        const namespace = await this.tags.getNamespace();
        const prefixedKeys = keys.map(k => namespace + k);
        if (this.store.forgetMultiple) {
            await this.store.forgetMultiple(prefixedKeys);
        }
        else {
            // Fallback
            for (const k of prefixedKeys) {
                await this.store.forget(k);
            }
        }
    }
}
exports.TaggedStore = TaggedStore;
